diff --git a/Makefile.in b/Makefile.in
index d4b308d..4c4b04b 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -924,7 +924,7 @@ c-common.o : c-common.c $(CONFIG_H) $(TREE_H) c-tree.h c-lex.h flags.h
 
 gcc.o: gcc.c $(CONFIG_H) gvarargs.h config.status
 	$(CC) $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
-  -DSTANDARD_STARTFILE_PREFIX=\"$(libdir)/\" \
+  -DSTANDARD_STARTFILE_PREFIX=\"$(tooldir)/lib/\" \
   -DSTANDARD_EXEC_PREFIX=\"$(libdir)/gcc-lib/\" \
   -DDEFAULT_TARGET_MACHINE=\"$(target)\" \
   -DTOOLDIR_BASE_PREFIX=\"$(exec_prefix)/\" \
diff --git a/cccp.c b/cccp.c
index 80e3e74..65b352e 100644
--- a/cccp.c
+++ b/cccp.c
@@ -39,10 +39,6 @@ typedef unsigned char U_CHAR;
 #include "config.h"
 #endif /* not EMACS */
 
-#ifndef STANDARD_INCLUDE_DIR
-#define STANDARD_INCLUDE_DIR "/usr/include"
-#endif
-
 #ifndef LOCAL_INCLUDE_DIR
 #define LOCAL_INCLUDE_DIR "/usr/local/include"
 #endif
@@ -589,7 +585,6 @@ static struct default_include { char *fname; int cplusplus; } include_defaults_a
 #ifdef SYSTEM_INCLUDE_DIR
     { SYSTEM_INCLUDE_DIR, 0},
 #endif
-    { STANDARD_INCLUDE_DIR, 0},
 #endif /* not CROSS_COMPILE */
     { 0, 0}
     };
diff --git a/g++.c b/g++.c
index aee9edd..068103a 100644
--- a/g++.c
+++ b/g++.c
@@ -58,7 +58,7 @@ the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */
    Makefile should compile this file with the proper name
    (e.g., "i386-aout-gcc").  */
 #ifndef GCC_NAME
-#define GCC_NAME "gcc-2.5.8"
+#define GCC_NAME "i586-legacy-linux-gnulibc1-gcc-2.5.8"
 #endif
 
 /* This bit is set if we saw a `-xfoo' language specification.  */
diff --git a/protoize.c b/protoize.c
index f3b1307..9d138db 100644
--- a/protoize.c
+++ b/protoize.c
@@ -262,13 +262,6 @@ static const unexpansion unexpansions[] = {
 
 static const int hash_mask = (HASH_TABLE_SIZE - 1);
 
-/* Make a table of default system include directories
-   just as it is done in cccp.c.  */
-
-#ifndef STANDARD_INCLUDE_DIR
-#define STANDARD_INCLUDE_DIR "/usr/include"
-#endif
-
 #ifndef LOCAL_INCLUDE_DIR
 #define LOCAL_INCLUDE_DIR "/usr/local/include"
 #endif
@@ -303,7 +296,6 @@ struct default_include { const char *fname; int cplusplus; } include_defaults[]
 #ifdef SYSTEM_INCLUDE_DIR
     { SYSTEM_INCLUDE_DIR, 0},
 #endif
-    { STANDARD_INCLUDE_DIR, 0},
 #endif /* not CROSS_COMPILE */
     { 0, 0}
     };
