diff --git a/stlcomp.h b/stlcomp.h
index bb08744..316acb3 100644
--- a/stlcomp.h
+++ b/stlcomp.h
@@ -327,7 +327,9 @@
 // g++ 2.7.x and above 
 # if defined (__GNUC__ )
 #   define __STL_BOOL_KEYWORD 1
-#   define __STL_EXPLICIT 1
+#   if __GNUC__ > 2 || (__GNUC__ == 2 && __GNUC_MINOR__ >= 7)
+#     define __STL_EXPLICIT 1
+#   endif
 #   define __STL_LONG_LONG  1
 #   define __STL_WCHAR_T  1
 #   define __STL_MUTABLE  1
