diff --git a/src/dll_main.cpp b/src/dll_main.cpp
index b9588b3..bbc8379 100644
--- a/src/dll_main.cpp
+++ b/src/dll_main.cpp
@@ -153,9 +153,11 @@ template struct _STLP_CLASS_DECLSPEC __stl_debug_engine<bool>;
 
 template class _STLP_CLASS_DECLSPEC __node_alloc<false,0>;
 template class _STLP_CLASS_DECLSPEC __node_alloc<true,0>;
+# if defined (_STLP_DEBUG)
 template class _STLP_CLASS_DECLSPEC __debug_alloc< __node_alloc<true,0> >;
 template class _STLP_CLASS_DECLSPEC __debug_alloc< __node_alloc<false,0> >;
 template class _STLP_CLASS_DECLSPEC __debug_alloc<__new_alloc>;
+# endif
 template class _STLP_CLASS_DECLSPEC __malloc_alloc<0>;
 
 # if defined (_STLP_THREADS) && ! defined ( _STLP_ATOMIC_EXCHANGE ) && (defined(_STLP_PTHREADS) || defined (_STLP_UITHREADS)  || defined (_STLP_OS2THREADS))
diff --git a/src/gcc-linux.mak b/src/gcc-linux.mak
index fc128b1..4d08f8d 100644
--- a/src/gcc-linux.mak
+++ b/src/gcc-linux.mak
@@ -36,9 +36,7 @@ include common_macros.mak
 
 WARNING_FLAGS= -W -Wno-sign-compare -Wno-unused -Wno-uninitialized -D_STLP_USE_GLIBC
 
-# boris : real locale implementation may not really work
-CXXFLAGS_COMMON = -I${STLPORT_DIR} ${WARNING_FLAGS} -D_STLP_REAL_LOCALE_IMPLEMENTED -D_GNU_SOURCE
-# CXXFLAGS_COMMON = -I${STLPORT_DIR} ${WARNING_FLAGS}
+CXXFLAGS_COMMON = -I${STLPORT_DIR} ${WARNING_FLAGS} -D_GNU_SOURCE
 
 CXXFLAGS_RELEASE_static = $(CXXFLAGS_COMMON) -O2
 CXXFLAGS_RELEASE_dynamic = $(CXXFLAGS_COMMON) -O2 -fPIC
