diff --git a/gcc/config.gcc b/gcc/config.gcc
index 2e5cf8ee..78a9049e 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -1679,7 +1679,7 @@ mips*-*-netbsd*)			# NetBSD/mips, either endian.
 mips64*-*-linux*)
 	tm_file="dbxelf.h elfos.h svr4.h linux.h ${tm_file} mips/linux.h mips/linux64.h"
 	tmake_file="${tmake_file} mips/t-linux64"
-	tm_defines="${tm_defines} MIPS_ABI_DEFAULT=ABI_N32"
+	tm_defines="${tm_defines} MIPS_ABI_DEFAULT=ABI_64"
 	gnu_ld=yes
 	gas=yes
 	test x$with_llsc != x || with_llsc=yes
diff --git a/gcc/config/mips/linux64.h b/gcc/config/mips/linux64.h
index 36b67d6c..6b4466ac 100644
--- a/gcc/config/mips/linux64.h
+++ b/gcc/config/mips/linux64.h
@@ -24,7 +24,7 @@ along with GCC; see the file COPYING3.  If not see
 #define DRIVER_SELF_SPECS \
 NO_SHARED_SPECS \
 " %{!EB:%{!EL:%(endian_spec)}}" \
-" %{!mabi=*: -mabi=n32}"
+" %{!mabi=*: -mabi=64}"
 
 #undef SUBTARGET_ASM_SPEC
 #define SUBTARGET_ASM_SPEC "\
