diff --git a/gcc/cprop.c b/gcc/cprop.c
index c9fb2fc51..8a5a3c844 100644
--- a/gcc/cprop.c
+++ b/gcc/cprop.c
@@ -1620,9 +1620,12 @@ bypass_block (basic_block bb, rtx_insn *setcc, rtx_insn *jump)
 	    {
 	      dest = BLOCK_FOR_INSN (XEXP (new_rtx, 0));
 	      /* Don't bypass edges containing instructions.  */
-	      edest = find_edge (bb, dest);
-	      if (edest && edest->insns.r)
-		dest = NULL;
+	      if (dest)
+		{
+		  edest = find_edge (bb, dest);
+		  if (edest && edest->insns.r)
+		    dest = NULL;
+		}
 	    }
 	  else
 	    dest = NULL;
diff --git a/gcc/testsuite/gcc.dg/pr106421.c b/gcc/testsuite/gcc.dg/pr106421.c
new file mode 100644
index 000000000..73e522afc
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/pr106421.c
@@ -0,0 +1,13 @@
+/* { dg-do compile } */
+/* { dg-options "-O2" } */
+
+int main(int argc, char **argv)
+{
+	__label__ loop, end;
+	void jmp(int c) { goto *(c ? &&loop : &&end); }
+loop:
+	jmp(argc < 0);
+end:
+	return 0;
+}
+
