diff --git a/gcc/cp/typeck.c b/gcc/cp/typeck.c
index 0dc15899b..b4e8bb4de 100644
--- a/gcc/cp/typeck.c
+++ b/gcc/cp/typeck.c
@@ -4483,7 +4483,7 @@ cp_build_binary_op (location_t location,
 	      tree pfn0, delta0, e1, e2;
 
 	      if (TREE_SIDE_EFFECTS (op0))
-		op0 = save_expr (op0);
+		op0 = cp_save_expr (op0);
 
 	      pfn0 = pfn_from_ptrmemfunc (op0);
 	      delta0 = delta_from_ptrmemfunc (op0);
diff --git a/gcc/testsuite/g++.dg/cpp0x/pr88120.C b/gcc/testsuite/g++.dg/cpp0x/pr88120.C
new file mode 100644
index 000000000..7004e03f4
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/pr88120.C
@@ -0,0 +1,23 @@
+// PR c++/88120
+// { dg-do compile { target c++11 } }
+
+typedef int a;
+enum b : a;
+class c {
+  enum f { d };
+  c(f);
+  friend c operator&(c, c);
+  typedef void (c::*e)();
+  operator e();
+};
+class g {
+  template <typename, typename> b h();
+  struct k {
+    c i;
+  };
+};
+template <typename, typename> b g::h() {
+  k j;
+  &j || j.i &c::d;
+  return b();
+}
