diff --git a/gcc/testsuite/gcc.c-torture/compile/pr87647.c b/gcc/testsuite/gcc.c-torture/compile/pr87647.c
new file mode 100644
index 000000000..67f9d8734
--- /dev/null
+++ b/gcc/testsuite/gcc.c-torture/compile/pr87647.c
@@ -0,0 +1,15 @@
+/* PR middle-end/87647 */
+
+struct A {};
+struct A *const b = &(struct A) {};
+struct B { char *s; struct A *t; };
+void bar (struct B *);
+
+void
+foo (void)
+{
+  struct B a[] = { "", b, "", b, "", b, "", b, "", b, "", b, "", b, "", b,
+		   "", b, "", b, "", b, "", b, "", b, "", b, "", b, "", b,
+		   "", b };
+  bar (a);
+}
diff --git a/gcc/varasm.c b/gcc/varasm.c
index 246d6dd29..43462de63 100644
--- a/gcc/varasm.c
+++ b/gcc/varasm.c
@@ -2900,6 +2900,11 @@ decode_addr_const (tree exp, struct addr_const *value)
       x = output_constant_def (target, 1);
       break;
 
+    case COMPOUND_LITERAL_EXPR:
+      gcc_assert (COMPOUND_LITERAL_EXPR_DECL (target));
+      x = DECL_RTL (COMPOUND_LITERAL_EXPR_DECL (target));
+      break;
+
     default:
       gcc_unreachable ();
     }
