diff --git a/gcc/cp/constexpr.c b/gcc/cp/constexpr.c
index 8f9089d2c..8c501ed6f 100644
--- a/gcc/cp/constexpr.c
+++ b/gcc/cp/constexpr.c
@@ -389,6 +389,9 @@ build_data_member_initialization (tree t, vec<constructor_elt, va_gc> **vec)
   if (TREE_CODE (member) == COMPONENT_REF)
     {
       tree aggr = TREE_OPERAND (member, 0);
+      if (TREE_CODE (aggr) == VAR_DECL)
+	/* Initializing a local variable, don't add anything.  */
+	return true;
       if (TREE_CODE (aggr) != COMPONENT_REF)
 	/* Normal member initialization.  */
 	member = TREE_OPERAND (member, 1);
diff --git a/gcc/testsuite/g++.dg/cpp1z/constexpr-lambda15.C b/gcc/testsuite/g++.dg/cpp1z/constexpr-lambda15.C
new file mode 100644
index 000000000..fc363cc65
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp1z/constexpr-lambda15.C
@@ -0,0 +1,9 @@
+// PR c++/79461
+// { dg-options -std=c++1z }
+
+struct S {
+  constexpr S(int i) {
+    auto f = [i]{};
+  }
+};
+int main() {}
