diff --git a/gcc/cp/call.c b/gcc/cp/call.c
index efa61ef70..881460555 100644
--- a/gcc/cp/call.c
+++ b/gcc/cp/call.c
@@ -7541,6 +7541,9 @@ build_over_call (struct z_candidate *cand, int flags, tsubst_flags_t complain)
 	{
 	  arg = cp_build_indirect_ref (arg, RO_NULL, complain);
 	  val = build2 (MODIFY_EXPR, TREE_TYPE (to), to, arg);
+	  if (cxx_dialect >= cxx14)
+	    /* Handle NSDMI that refer to the object being initialized.  */
+	    replace_placeholders (arg, to);
 	}
       else
 	{
diff --git a/gcc/testsuite/g++.dg/cpp0x/nsdmi13.C b/gcc/testsuite/g++.dg/cpp0x/nsdmi13.C
new file mode 100644
index 000000000..2751da3b5
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/nsdmi13.C
@@ -0,0 +1,13 @@
+// PR c++/79796
+// { dg-do compile { target c++11 } }
+
+struct A
+{
+  A* p = this;
+};
+
+void foo()
+{
+  A a;
+  a = A({});
+}
