diff --git a/gcc/testsuite/c-c++-common/cpp/pr67046.c b/gcc/testsuite/c-c++-common/cpp/pr67046.c
new file mode 100644
index 000000000..f37f20c62
--- /dev/null
+++ b/gcc/testsuite/c-c++-common/cpp/pr67046.c
@@ -0,0 +1,10 @@
+/* { dg-do preprocess } */
+
+_Pragma(
+"message(\"msg\")"
+)
+
+_Pragma(
+"message(\"msg\")"
+#
+)
diff --git a/libcpp/directives.c b/libcpp/directives.c
index 2a824e6ed..07bcedc82 100644
--- a/libcpp/directives.c
+++ b/libcpp/directives.c
@@ -1862,7 +1862,12 @@ destringize_and_run (cpp_reader *pfile, const cpp_string *in)
 int
 _cpp_do__Pragma (cpp_reader *pfile)
 {
+  /* Make sure we don't invalidate the string token, if the closing parenthesis
+   ended up on a different line.  */
+  ++pfile->keep_tokens;
   const cpp_token *string = get__Pragma_string (pfile);
+  --pfile->keep_tokens;
+
   pfile->directive_result.type = CPP_PADDING;
 
   if (string)
