diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 0cb5ad88b..79aea17b7 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -7921,6 +7921,11 @@ grokfndecl (tree ctype,
   DECL_EXTERNAL (decl) = 1;
   if (TREE_CODE (type) == FUNCTION_TYPE)
     {
+      if (quals || rqual)
+	TREE_TYPE (decl) = apply_memfn_quals (TREE_TYPE (decl),
+					      TYPE_UNQUALIFIED,
+					      REF_QUAL_NONE);
+
       if (quals)
 	{
 	  error (ctype
diff --git a/gcc/testsuite/g++.dg/other/cv_func4.C b/gcc/testsuite/g++.dg/other/cv_func4.C
new file mode 100644
index 000000000..df4f19eb5
--- /dev/null
+++ b/gcc/testsuite/g++.dg/other/cv_func4.C
@@ -0,0 +1,6 @@
+// PR c++/67845
+
+typedef void F () const;
+
+F foo;  // { dg-error "cv-qualifier" }
+void foo ();
