diff --git a/gcc/targhooks.c b/gcc/targhooks.c
index 003e9ef97..870afefa5 100644
--- a/gcc/targhooks.c
+++ b/gcc/targhooks.c
@@ -1040,7 +1040,10 @@ tree default_mangle_decl_assembler_name (tree decl ATTRIBUTE_UNUSED,
 HOST_WIDE_INT
 default_vector_alignment (const_tree type)
 {
-  return tree_to_shwi (TYPE_SIZE (type));
+  HOST_WIDE_INT align = tree_to_shwi (TYPE_SIZE (type));
+  if (align > MAX_OFILE_ALIGNMENT)
+    align = MAX_OFILE_ALIGNMENT;
+  return align;
 }
 
 bool
diff --git a/gcc/testsuite/gcc.dg/pr69973.c b/gcc/testsuite/gcc.dg/pr69973.c
new file mode 100644
index 000000000..648bc5679
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/pr69973.c
@@ -0,0 +1,2 @@
+/* { dg-do compile } */
+typedef int v4si __attribute__ ((vector_size (1 << 29)));
