diff --git a/gcc/cp/lambda.c b/gcc/cp/lambda.c
index 8161c24fd..ba9bfe0c0 100644
--- a/gcc/cp/lambda.c
+++ b/gcc/cp/lambda.c
@@ -503,6 +503,8 @@ add_capture (tree lambda, tree id, tree orig_init, bool by_reference_p,
   else
     {
       type = lambda_capture_field_type (initializer, explicit_init_p);
+      if (type == error_mark_node)
+	return error_mark_node;
       if (by_reference_p)
 	{
 	  type = build_reference_type (type);
diff --git a/gcc/testsuite/g++.dg/cpp1y/lambda-ice1.C b/gcc/testsuite/g++.dg/cpp1y/lambda-ice1.C
new file mode 100644
index 000000000..86fb88aad
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp1y/lambda-ice1.C
@@ -0,0 +1,7 @@
+// PR c++/72800
+// { dg-do compile { target c++14 } }
+
+void foo ()
+{
+  [n {}] {};  // { dg-error "one element|deducing" }
+}
