diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index 74e85583c..4062f7620 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -9404,10 +9404,11 @@ cp_parser_lambda_declarator_opt (cp_parser* parser, tree lambda_expr)
 
       /* Default arguments shall not be specified in the
 	 parameter-declaration-clause of a lambda-declarator.  */
-      for (tree t = param_list; t; t = TREE_CHAIN (t))
-	if (TREE_PURPOSE (t) && cxx_dialect < cxx14)
-	  pedwarn (DECL_SOURCE_LOCATION (TREE_VALUE (t)), OPT_Wpedantic,
-		   "default argument specified for lambda parameter");
+      if (cxx_dialect < cxx14)
+	for (tree t = param_list; t; t = TREE_CHAIN (t))
+	  if (TREE_PURPOSE (t) && DECL_P (TREE_VALUE (t)))
+	    pedwarn (DECL_SOURCE_LOCATION (TREE_VALUE (t)), OPT_Wpedantic,
+		     "default argument specified for lambda parameter");
 
       cp_parser_require (parser, CPP_CLOSE_PAREN, RT_CLOSE_PAREN);
 
diff --git a/gcc/testsuite/g++.dg/cpp0x/lambda/lambda-77791.C b/gcc/testsuite/g++.dg/cpp0x/lambda/lambda-77791.C
new file mode 100644
index 000000000..0475d2786
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/lambda/lambda-77791.C
@@ -0,0 +1,4 @@
+// PR c++/77791
+// { dg-do compile { target c++11 } }
+
+auto a = [] (int i, int i = 0) {};	// { dg-error "redefinition of" }
