diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 5b86ee76e..1b213216c 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -10282,9 +10282,9 @@ grokdeclarator (const cp_declarator *declarator,
       else if (TREE_CODE (type) == FUNCTION_TYPE)
 	{
 	  if (current_class_type
-	      && (!friendp || funcdef_flag))
+	      && (!friendp || funcdef_flag || initialized))
 	    {
-	      error (funcdef_flag
+	      error (funcdef_flag || initialized
 		     ? G_("cannot define member function %<%T::%s%> "
 			  "within %<%T%>")
 		     : G_("cannot declare member function %<%T::%s%> "
diff --git a/gcc/testsuite/g++.dg/cpp0x/deleted13.C b/gcc/testsuite/g++.dg/cpp0x/deleted13.C
new file mode 100644
index 000000000..f10551f96
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/deleted13.C
@@ -0,0 +1,12 @@
+// PR c++/79519
+// { dg-do compile { target c++11 } }
+
+struct A
+{
+  template<typename> void foo();
+};
+
+struct B
+{
+  template<typename> friend void A::foo() = delete; // { dg-error "" }
+};
