diff --git a/gcc/cp/rtti.c b/gcc/cp/rtti.c
index 6efe57984..9558ae639 100644
--- a/gcc/cp/rtti.c
+++ b/gcc/cp/rtti.c
@@ -1611,7 +1611,8 @@ emit_tinfo_decl (tree decl)
       /* Avoid targets optionally bumping up the alignment to improve
 	 vector instruction accesses, tinfo are never accessed this way.  */
 #ifdef DATA_ABI_ALIGNMENT
-      DECL_ALIGN (decl) = DATA_ABI_ALIGNMENT (decl, TYPE_ALIGN (TREE_TYPE (decl)));
+      DECL_ALIGN (decl) = DATA_ABI_ALIGNMENT (TREE_TYPE (decl),
+						TYPE_ALIGN (TREE_TYPE (decl)));
       DECL_USER_ALIGN (decl) = true;
 #endif
       return true;
