diff --git a/Makefile.in b/Makefile.in
index 2e25dfb..1f78429 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -128,7 +128,7 @@ RANLIB_FOR_TARGET = $(TARGET_TOOLPREFIX)ranlib
 RANLIB_TEST_FOR_TARGET = [ -f $(TARGET_TOOLPREFIX)ranlib ]
 
 # Dir to search for system headers.  Overridden by cross-make.
-SYSTEM_HEADER_DIR = /usr/include
+SYSTEM_HEADER_DIR = $(tooldir)/include
 
 # Control whether to run fixproto.
 STMP_FIXPROTO = stmp-fixproto
diff --git a/gbl-ctors.h b/gbl-ctors.h
index 3e2ab83..e1a7e43 100644
--- a/gbl-ctors.h
+++ b/gbl-ctors.h
@@ -32,11 +32,11 @@ the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */
 */
 
 #ifdef HAVE_ATEXIT
-extern void atexit (void (*) (void));
+extern int atexit (void (*) (void));
 #define ON_EXIT(FUNC,ARG) atexit ((FUNC))
 #else
 #ifdef sun
-extern void on_exit (void*, void*);
+extern int on_exit (void*, void*);
 #define ON_EXIT(FUNC,ARG) on_exit ((FUNC), (ARG))
 #endif
 #endif
diff --git a/libgcc2.c b/libgcc2.c
index 63a7114..4478c57 100644
--- a/libgcc2.c
+++ b/libgcc2.c
@@ -1431,11 +1431,11 @@ BLOCK_PROFILER_CODE
 char *ctime ();
 
 #ifdef HAVE_ATEXIT
-extern void atexit (void (*) (void));
+extern int atexit (void (*) (void));
 #define ON_EXIT(FUNC,ARG) atexit ((FUNC))
 #else
 #ifdef sun
-extern void on_exit (void*, void*);
+extern int on_exit (void*, void*);
 #define ON_EXIT(FUNC,ARG) on_exit ((FUNC), (ARG))
 #endif
 #endif
