diff --git a/bi-arity.c b/bi-arity.c
index c47971e..4a6e666 100644
--- a/bi-arity.c
+++ b/bi-arity.c
@@ -56,7 +56,7 @@ main ()
       }
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
@@ -72,7 +72,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-lexer.c b/bi-lexer.c
index 0766b5f..c7b38c8 100644
--- a/bi-lexer.c
+++ b/bi-lexer.c
@@ -38,7 +38,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
@@ -60,7 +60,7 @@ xrealloc (block, nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't reallocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-opcode.c b/bi-opcode.c
index 0325fa4..7c7bd18 100644
--- a/bi-opcode.c
+++ b/bi-opcode.c
@@ -54,7 +54,7 @@ main(argc, argv)
     fprintf (stderr, "(Number of opcodes is %d)\n", i);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
@@ -70,7 +70,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-opname.c b/bi-opname.c
index 02dfabc..1b7fadd 100644
--- a/bi-opname.c
+++ b/bi-opname.c
@@ -35,7 +35,7 @@ main()
       printf("\"%s%s\",\n", d->basename, v->name);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
@@ -51,7 +51,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-parser.c b/bi-parser.c
index e167ce8..b54c08c 100644
--- a/bi-parser.c
+++ b/bi-parser.c
@@ -958,5 +958,5 @@ yyerror (s)
      char *s;
 {
   fprintf (stderr, "syntax error in input\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
diff --git a/bi-parser.y b/bi-parser.y
index 5b3be52..9fd009f 100644
--- a/bi-parser.y
+++ b/bi-parser.y
@@ -164,5 +164,5 @@ yyerror (s)
      char *s;
 {
   fprintf (stderr, "syntax error in input\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
diff --git a/cccp.c b/cccp.c
index b6eef80..9abd2c5 100644
--- a/cccp.c
+++ b/cccp.c
@@ -2153,8 +2153,8 @@ main (argc, argv)
     fatal ("I/O error on output");
 
   if (errors)
-    exit (FAILURE_EXIT_CODE);
-  exit (SUCCESS_EXIT_CODE);
+    _exit (FAILURE_EXIT_CODE);
+  _exit (SUCCESS_EXIT_CODE);
 
  perror:
   pfatal_with_name (in_fname);
@@ -9554,7 +9554,7 @@ fatal (str, arg)
   fprintf (stderr, "%s: ", progname);
   fprintf (stderr, str, arg);
   fprintf (stderr, "\n");
-  exit (FAILURE_EXIT_CODE);
+  _exit (FAILURE_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -9581,9 +9581,9 @@ pfatal_with_name (name)
 {
   perror_with_name (name);
 #ifdef VMS
-  exit (vaxc$errno);
+  _exit (vaxc$errno);
 #else
-  exit (FAILURE_EXIT_CODE);
+  _exit (FAILURE_EXIT_CODE);
 #endif
 }
 
diff --git a/collect2.c b/collect2.c
index 356bad0..55425d9 100644
--- a/collect2.c
+++ b/collect2.c
@@ -338,7 +338,7 @@ my_exit (status)
   if (status != 0 && output_file != 0 && output_file[0])
     maybe_unlink (output_file);
 
-  exit (status);
+  _exit (status);
 }
 
 
diff --git a/cp/edsel.c b/cp/edsel.c
index 78b2637..34b4b67 100644
--- a/cp/edsel.c
+++ b/cp/edsel.c
@@ -205,7 +205,7 @@ init_cadillac ()
 
       default:
 	perror ("Editor to kernel connection");
-	exit (0);
+	_exit (0);
       }
 }
 
diff --git a/cp/g++.c b/cp/g++.c
index 372ba27..61afbee 100644
--- a/cp/g++.c
+++ b/cp/g++.c
@@ -118,7 +118,7 @@ fatal VPROTO((char *format, ...))
   /* XXX Not needed for g++ driver.  */
   delete_temp_files ();
 #endif
-  exit (1);
+  _exit (1);
 }
 
 static void
@@ -162,7 +162,7 @@ fatal (msg, arg1, arg2)
   /* XXX Not needed for g++ driver.  */
   delete_temp_files ();
 #endif
-  exit (1);
+  _exit (1);
 }
 
 #endif /* not HAVE_VPRINTF */
diff --git a/cp/typeck2.c b/cp/typeck2.c
index 5d3322d..5ee3b90 100644
--- a/cp/typeck2.c
+++ b/cp/typeck2.c
@@ -355,7 +355,7 @@ my_friendly_abort (i)
       else
 	error ("confused by earlier errors, bailing out");
       
-      exit (34);
+      _exit (34);
     }
   ++abortcount;
 
diff --git a/dbxout.c b/dbxout.c
index f09d52c..d6129f6 100644
--- a/dbxout.c
+++ b/dbxout.c
@@ -359,7 +359,7 @@ abspath (rel_filename)
 
               	  fprintf (stderr, "%s: invalid file name: %s\n",
 			   pname, rel_filename);
-              	  exit (1);
+              	  _exit (1);
               	}
                     *++outp = '\0';
                     continue;
diff --git a/enquire.c b/enquire.c
index 763b1a4..91520d7 100644
--- a/enquire.c
+++ b/enquire.c
@@ -600,13 +600,13 @@ int setmode(s) char *s; {
 #endif
 
 Procedure farewell(bugs) int bugs; {
-	if (bugs == 0) exit(0);
+	if (bugs == 0) _exit(0);
 	printf("\n%sFor hints on dealing with the ", co);
 	if (bugs == 1) printf("problem");
 	else printf("%d problems", bugs);
 	printf(" above\n   see the section 'TROUBLESHOOTING' in the file ");
 	printf("%s%s\n", FILENAME, oc);
-	exit(bugs);
+	_exit(bugs);
 }
 
 /* The program has received a signal where it wasn't expecting one */
@@ -620,7 +620,7 @@ char *xmalloc(size) unsigned size; {
 	char *value = (char *)malloc(size);
 	if (value == 0) {
 		fprintf(stderr, "Virtual memory exceeded\n");
-		exit(bugs+1);
+		_exit(bugs+1);
 	}
 	return value;
 }
@@ -682,7 +682,7 @@ int main(argc, argv) int argc; char *argv[]; {
 		fprintf(stderr,
 			"Usage: %s [-vlf]\n  v=Verbose l=Limits.h f=Float.h\n",
 			argv[0]);
-		exit(1);
+		_exit(1);
 	}
 	if (L || F) {
 		co[0]= '/'; oc[0]= ' ';
diff --git a/fix-header.c b/fix-header.c
index c21acf0..92db044 100644
--- a/fix-header.c
+++ b/fix-header.c
@@ -353,7 +353,7 @@ read_scan_file (scan_file)
     {
       if (verbose)
 	fprintf (stderr, "%s: OK, nothing needs to be done.\n", inc_filename);
-      exit (0);
+      _exit (0);
     }
   if (!verbose)
     fprintf (stderr, "%s: fixing %s\n", progname, inc_filename);
@@ -695,7 +695,7 @@ main (argc, argv)
     {
       fprintf (stderr, "%s: Usage: foo.h infile.h outfile.h req_funcs <scan-file-name\n",
 	       progname);
-      exit (-1);
+      _exit (-1);
     }
 
   inc_filename = argv[1];
@@ -741,13 +741,13 @@ main (argc, argv)
       fprintf (stderr, "%s: Cannot open '%s' for reading -",
 	       progname, argv[2]);
       perror (NULL);
-      exit (-1);
+      _exit (-1);
     }
   if (fstat (inf_fd, &sbuf) < 0)
     {
       fprintf (stderr, "%s: Cannot get size of '%s' -", progname, argv[2]);
       perror (NULL);
-      exit (-1);
+      _exit (-1);
     }
   inf_size = sbuf.st_size;
   inf_buffer = (char*) xmalloc (inf_size + 2);
@@ -764,7 +764,7 @@ main (argc, argv)
 	{
 	  fprintf (stderr, "%s: Failed to read '%s' -", progname, argv[2]);
 	  perror (NULL);
-	  exit (-1);
+	  _exit (-1);
 	}
       if (i == 0)
 	{
@@ -787,7 +787,7 @@ main (argc, argv)
       fprintf (stderr, "%s: Cannot open '%s' for writing -",
 	       progname, argv[3]);
       perror (NULL);
-      exit (-1);
+      _exit (-1);
     }
 
   lineno = 1;
diff --git a/gcc.c b/gcc.c
index 75c1d7c..84eb3dd 100644
--- a/gcc.c
+++ b/gcc.c
@@ -2068,7 +2068,7 @@ pexecute (search_flag, program, argv, not_last)
       /* Exec the program.  */
       (*func) (program, argv);
       perror_exec (program);
-      exit (-1);
+      _exit (-1);
       /* NOTREACHED */
       return 0;
 
@@ -2497,17 +2497,17 @@ process_command (argc, argv)
 	  printf ("*cross_compile:\n%d\n\n", cross_compile);
 	  printf ("*multilib:\n%s\n\n", multilib_select);
 
-	  exit (0);
+	  _exit (0);
 	}
       else if (! strcmp (argv[i], "-dumpversion"))
 	{
 	  printf ("%s\n", version_string);
-	  exit (0);
+	  _exit (0);
 	}
       else if (! strcmp (argv[i], "-dumpmachine"))
 	{
 	  printf ("%s\n", spec_machine);
-	  exit  (0);
+	  _exit (0);
 	}
       else if (! strcmp (argv[i], "-print-libgcc-file-name"))
 	  print_file_name = "libgcc.a";
@@ -4324,20 +4324,20 @@ main (argc, argv)
   if (print_file_name)
     {
       printf ("%s\n", find_file (print_file_name));
-      exit (0);
+      _exit (0);
     }
 
   if (print_prog_name)
     {
       char *newname = find_a_file (&exec_prefixes, print_prog_name, X_OK);
       printf ("%s\n", (newname ? newname : print_prog_name));
-      exit (0);
+      _exit (0);
     }
 
   if (print_multi_lib)
     {
       print_multilib_info ();
-      exit (0);
+      _exit (0);
     }
 
   if (print_multi_directory)
@@ -4346,14 +4346,14 @@ main (argc, argv)
 	printf (".\n");
       else
 	printf ("%s\n", multilib_dir);
-      exit (0);
+      _exit (0);
     }
 
   if (verbose_flag)
     {
       fprintf (stderr, "gcc version %s\n", version_string);
       if (n_infiles == 0)
-	exit (0);
+	_exit (0);
     }
 
   if (n_infiles == 0)
@@ -4518,7 +4518,7 @@ main (argc, argv)
     delete_failure_queue ();
   delete_temp_files ();
 
-  exit (error_count > 0 ? (signal_count ? 2 : 1) : 0);
+  _exit (error_count > 0 ? (signal_count ? 2 : 1) : 0);
   /* NOTREACHED */
   return 0;
 }
@@ -4737,7 +4737,7 @@ fatal VPROTO((char *format, ...))
   va_end (ap);
   fprintf (stderr, "\n");
   delete_temp_files ();
-  exit (1);
+  _exit (1);
 }
 
 static void
@@ -4769,7 +4769,7 @@ fatal (msg, arg1, arg2)
 {
   error (msg, arg1, arg2);
   delete_temp_files ();
-  exit (1);
+  _exit (1);
 }
 
 static void
diff --git a/genattr.c b/genattr.c
index 119e010..8289e78 100644
--- a/genattr.c
+++ b/genattr.c
@@ -219,7 +219,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genattr: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -263,7 +263,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -425,7 +425,7 @@ from the machine description file `md'.  */\n\n");
   printf("#define ATTR_FLAG_very_unlikely\t0x20\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genattrtab.c b/genattrtab.c
index 0d42cac..4e97ddb 100644
--- a/genattrtab.c
+++ b/genattrtab.c
@@ -5471,7 +5471,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genattrtab: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -5558,7 +5558,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -5704,7 +5704,7 @@ from the machine description file `md'.  */\n\n");
   write_const_num_delay_slots ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/gencodes.c b/gencodes.c
index 255e2aa..0c4cd09 100644
--- a/gencodes.c
+++ b/gencodes.c
@@ -81,7 +81,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "gencodes: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -111,7 +111,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -153,7 +153,7 @@ from the machine description file `md'.  */\n\n");
   printf ("#endif /* MAX_INSN_CODE */\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genconfig.c b/genconfig.c
index f1585c6..1297aa9 100644
--- a/genconfig.c
+++ b/genconfig.c
@@ -265,7 +265,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genconfig: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -295,7 +295,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -357,7 +357,7 @@ from the machine description file `md'.  */\n\n");
     printf ("#define HAVE_lo_sum\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genemit.c b/genemit.c
index 5b60f46..18ac701 100644
--- a/genemit.c
+++ b/genemit.c
@@ -695,7 +695,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genemit: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -725,7 +725,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -792,7 +792,7 @@ from the machine description file `md'.  */\n\n");
   output_init_mov_optab ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genextract.c b/genextract.c
index 9a608cb..67e3982 100644
--- a/genextract.c
+++ b/genextract.c
@@ -367,7 +367,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genextract: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -423,7 +423,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -549,7 +549,7 @@ from the machine description file `md'.  */\n\n");
   printf ("    }\n}\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genflags.c b/genflags.c
index 5642e5f..0a1cae6 100644
--- a/genflags.c
+++ b/genflags.c
@@ -196,7 +196,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genflags: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -232,7 +232,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -285,7 +285,7 @@ from the machine description file `md'.  */\n\n");
   printf ("#endif  /* NO_MD_PROTOTYPES */\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genopinit.c b/genopinit.c
index cd8f7c0..11c5b65 100644
--- a/genopinit.c
+++ b/genopinit.c
@@ -308,7 +308,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genopinit: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -340,7 +340,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -377,7 +377,7 @@ from the machine description file `md'.  */\n\n");
   printf ("}\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genoutput.c b/genoutput.c
index 6636fc4..0512699 100644
--- a/genoutput.c
+++ b/genoutput.c
@@ -905,7 +905,7 @@ fatal (s, a1, a2, a3, a4)
   fprintf (stderr, "genoutput: ");
   fprintf (stderr, s, a1, a2, a3, a4);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -946,7 +946,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -980,7 +980,7 @@ main (argc, argv)
   output_epilogue ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 || have_error
+  _exit (ferror (stdout) != 0 || have_error
 	? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
 
   /* NOTREACHED */
diff --git a/genpeep.c b/genpeep.c
index 1e11a68..b7b849a 100644
--- a/genpeep.c
+++ b/genpeep.c
@@ -413,7 +413,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genpeep: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -445,7 +445,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -502,7 +502,7 @@ from the machine description file `md'.  */\n\n");
   printf ("rtx peep_operand[%d];\n", max_opno + 1);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genrecog.c b/genrecog.c
index c0d31d2..761f7a9 100644
--- a/genrecog.c
+++ b/genrecog.c
@@ -1682,7 +1682,7 @@ fatal (s)
   fprintf (stderr, s);
   fprintf (stderr, "\n");
   fprintf (stderr, "after %d definitions\n", next_index);
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -1715,7 +1715,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -1798,7 +1798,7 @@ from the machine description file `md'.  */\n\n");
   write_subroutine (split_tree.first, SPLIT);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/getopt.c b/getopt.c
index 43c0a6a..f4d1cec 100644
--- a/getopt.c
+++ b/getopt.c
@@ -742,7 +742,7 @@ main (argc, argv)
       printf ("\n");
     }
 
-  exit (0);
+  _exit (0);
 }
 
 #endif /* TEST */
diff --git a/getopt1.c b/getopt1.c
index 4580211..7c36887 100644
--- a/getopt1.c
+++ b/getopt1.c
@@ -174,7 +174,7 @@ main (argc, argv)
       printf ("\n");
     }
 
-  exit (0);
+  _exit (0);
 }
 
 #endif /* TEST */
diff --git a/libgcc2.c b/libgcc2.c
index 63a7114..7c67f89 100644
--- a/libgcc2.c
+++ b/libgcc2.c
@@ -1875,7 +1875,7 @@ __enable_execute_stack (addr)
   if (r != KERN_SUCCESS)
     {
       mach_error("vm_protect VM_PROT_ALL", r);
-      exit(1);
+      _exit(1);
     }
 
   /* We inline the i-cache invalidation for speed */
diff --git a/objc-act.c b/objc-act.c
index 4f89d5a..ce868aa 100644
--- a/objc-act.c
+++ b/objc-act.c
@@ -2205,7 +2205,7 @@ build_ivar_chain (interface, copy)
 	  error ("Cannot find interface declaration for `%s', superclass of `%s'",
 		 IDENTIFIER_POINTER (super_name),
 		 IDENTIFIER_POINTER (my_name));
-	  exit (34);
+	  _exit (34);
         }
       if (super_interface == interface)
         {
@@ -5739,7 +5739,7 @@ start_class (code, class_name, super_name, protocol_list)
 	{
 	  error ("Cannot find interface declaration for `%s'",
 		 IDENTIFIER_POINTER (class_name));
-	  exit (1);
+	  _exit (1);
 	}
       else
         add_category (class_category_is_assoc_with, class);
@@ -5771,7 +5771,7 @@ start_class (code, class_name, super_name, protocol_list)
         {
 	  error ("Cannot find interface declaration for `%s'",
 		 IDENTIFIER_POINTER (class_name));
-	  exit (1);
+	  _exit (1);
         }
     }
   return class;
diff --git a/protoize.c b/protoize.c
index f4c92f3..0b22593 100644
--- a/protoize.c
+++ b/protoize.c
@@ -165,7 +165,7 @@ extern int fprintf (FILE *, const char *, ...);
 extern int printf (const char *, ...);
 extern int open (const char *, int, ...);
 #endif /* 0 */
-extern void exit ();
+extern void _exit ();
 extern void free ();
 extern int read ();
 extern int write ();
@@ -623,7 +623,7 @@ xmalloc (byte_count)
   if (rv == NULL)
     {
       fprintf (stderr, "\n%s: virtual memory exceeded\n", pname);
-      exit (1);
+      _exit (1);
       return 0;		/* avoid warnings */
     }
   else
@@ -643,7 +643,7 @@ xrealloc (old_space, byte_count)
   if (rv == NULL)
     {
       fprintf (stderr, "\n%s: virtual memory exceeded\n", pname);
-      exit (1);
+      _exit (1);
       return 0;		/* avoid warnings */
     }
   else
@@ -696,7 +696,7 @@ void
 fancy_abort ()
 {
   fprintf (stderr, "%s: internal abort\n", pname);
-  exit (1);
+  _exit (1);
 }
 
 /* Make a duplicate of the first N bytes of a given string in a newly
@@ -834,7 +834,7 @@ usage ()
   fprintf (stderr, "%s: usage '%s [ -VqfnkNlgC ] [ -B <diname> ] [ filename ... ]'\n",
 	   pname, pname);
 #endif /* !defined (UNPROTOIZE) */
-  exit (1);
+  _exit (1);
 }
 
 /* Return true if the given filename (assumed to be an absolute filename)
@@ -1328,7 +1328,7 @@ abspath (cwd, rel_filename)
 
               	  fprintf (stderr, "%s: invalid file name: %s\n",
 			   pname, rel_filename);
-              	  exit (1);
+              	  _exit (1);
               	}
                     *++outp = '\0';
                     continue;
@@ -1498,7 +1498,7 @@ aux_info_corrupted ()
 {
   fprintf (stderr, "\n%s: fatal error: aux info file corrupted at line %d\n",
 	   pname, current_aux_info_lineno);
-  exit (1);
+  _exit (1);
 }
 
 /* ??? This comment is vague.  Say what the condition is for.  */
@@ -1824,7 +1824,7 @@ save_def_or_dec (l, is_syscalls)
 			 def_dec_p->file->hash_entry->symbol,
 			 def_dec_p->line,
 			 def_dec_p->hash_entry->symbol);
-                exit (1);
+                _exit (1);
               }
             free_def_dec (def_dec_p);
             return;
@@ -4554,7 +4554,7 @@ main (argc, argv)
     {
       fprintf (stderr, "%s: cannot get working directory: %s\n",
 	       pname, sys_errlist[errno]);
-      exit (1);
+      _exit (1);
     }
 
   /* By default, convert the files in the current directory.  */
@@ -4680,8 +4680,8 @@ main (argc, argv)
       do_processing ();
     }
   if (errors)
-    exit (1);
+    _exit (1);
   else
-    exit (0);
+    _exit (0);
   return 1;
 }
diff --git a/toplev.c b/toplev.c
index 28114c8..140fe0b 100644
--- a/toplev.c
+++ b/toplev.c
@@ -939,7 +939,7 @@ pfatal_with_name (name)
 {
   fprintf (stderr, "%s: ", progname);
   perror (name);
-  exit (35);
+  _exit (35);
 }
 
 void
@@ -947,7 +947,7 @@ fatal_io_error (name)
      char *name;
 {
   fprintf (stderr, "%s: %s: I/O error\n", progname, name);
-  exit (35);
+  _exit (35);
 }
 
 /* Called to give a better error message for a bad insn rather than
@@ -1398,7 +1398,7 @@ vfatal (s, ap)
      va_list ap;
 {
   verror (s, ap);
-  exit (34);
+  _exit (34);
 }
 
 void
@@ -3968,10 +3968,10 @@ You Lose!  You must define PREFERRED_DEBUGGING_TYPE!
 #endif /* not OS2 and not VMS and not WINNT */
 
   if (errorcount)
-    exit (FATAL_EXIT_CODE);
+    _exit (FATAL_EXIT_CODE);
   if (sorrycount)
-    exit (FATAL_EXIT_CODE);
-  exit (SUCCESS_EXIT_CODE);
+    _exit (FATAL_EXIT_CODE);
+  _exit (SUCCESS_EXIT_CODE);
   return 34;
 }
 
