diff --git a/configure b/configure
index 3fd699d..d4a514f 100755
--- a/configure
+++ b/configure
@@ -289,7 +289,7 @@ do
 	--norecursion | --no*)
 		norecursion=yes
 		;;
-	--ol*)
+	--oldincludedir | --ol*)
 		oldincludedir=$optarg
 		diroptions="$diroptions --oldincludedir=$optarg"
 		;;
diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 602d54e..8e68670 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -1331,7 +1331,7 @@ c-common.o : c-common.c $(CONFIG_H) system.h $(TREE_H) c-tree.h c-lex.h \
 # Language-independent files.
 
 DRIVER_DEFINES = \
-  -DSTANDARD_STARTFILE_PREFIX=\"$(libdir)/\" \
+  -DSTANDARD_STARTFILE_PREFIX=\"$(tooldir)/lib/\" \
   -DSTANDARD_EXEC_PREFIX=\"$(libdir)/gcc-lib/\" \
   -DDEFAULT_TARGET_VERSION=\"$(version)\" \
   -DDEFAULT_TARGET_MACHINE=\"$(target_alias)\" \
diff --git a/gcc/cccp.c b/gcc/cccp.c
index 2e87c36..b747d1b 100644
--- a/gcc/cccp.c
+++ b/gcc/cccp.c
@@ -49,10 +49,6 @@ typedef unsigned char U_CHAR;
 #define GET_ENVIRONMENT(ENV_VALUE,ENV_NAME) ENV_VALUE = getenv (ENV_NAME)
 #endif
 
-#ifndef STANDARD_INCLUDE_DIR
-# define STANDARD_INCLUDE_DIR "/usr/include"
-#endif
-
 /* By default, colon separates directories in a path.  */
 #ifndef PATH_SEPARATOR
 # define PATH_SEPARATOR ':'
@@ -454,7 +450,6 @@ static struct default_include {
 #ifndef STANDARD_INCLUDE_COMPONENT
 #define STANDARD_INCLUDE_COMPONENT 0
 #endif
-    { STANDARD_INCLUDE_DIR, STANDARD_INCLUDE_COMPONENT, 0, 0 },
 #endif /* not CROSS_COMPILE */
     { 0, 0, 0, 0 }
     };
diff --git a/gcc/cpplib.c b/gcc/cpplib.c
index 5b10349..553dac8 100644
--- a/gcc/cpplib.c
+++ b/gcc/cpplib.c
@@ -90,9 +90,6 @@ extern char *update_path PARAMS ((char *, char *));
 #define PATH_SEPARATOR ':'
 #endif
 
-#ifndef STANDARD_INCLUDE_DIR
-#define STANDARD_INCLUDE_DIR "/usr/include"
-#endif
 #ifndef INCLUDE_LEN_FUDGE
 #define INCLUDE_LEN_FUDGE 0
 #endif
@@ -335,7 +332,6 @@ static struct default_include {
 #ifndef STANDARD_INCLUDE_COMPONENT
 #define STANDARD_INCLUDE_COMPONENT 0
 #endif
-    { STANDARD_INCLUDE_DIR, STANDARD_INCLUDE_COMPONENT, 0, 0 },
 #endif /* not CROSS_COMPILE */
     { 0, 0, 0, 0 }
     };
diff --git a/gcc/eh-common.h b/gcc/eh-common.h
index 143ddff..c81a5b5 100644
--- a/gcc/eh-common.h
+++ b/gcc/eh-common.h
@@ -25,6 +25,14 @@
    the location of 'dynamic_handler_chain'. If its location is changed, 
    that routine must be modified as well. */
 
+#ifndef PARAMS
+# if defined __GNUC__ || (defined __STDC__ && __STDC__)
+#  define PARAMS(args) args
+# else
+#  define PARAMS(args) ()
+# endif  /* GCC.  */
+#endif  /* Not PARAMS.  */
+
 struct eh_context
 {
   void *handler_label;
diff --git a/gcc/machmode.h b/gcc/machmode.h
index 5f8459d..6234ef7 100644
--- a/gcc/machmode.h
+++ b/gcc/machmode.h
@@ -23,6 +23,14 @@ Boston, MA 02111-1307, USA.  */
 
 #include "gansidecl.h"
 
+#ifndef PARAMS
+# if defined __GNUC__ || (defined __STDC__ && __STDC__)
+#  define PARAMS(args) args
+# else
+#  define PARAMS(args) ()
+# endif  /* GCC.  */
+#endif  /* Not PARAMS.  */
+
 /* Strictly speaking, this isn't the proper place to include these definitions,
    but this file is included by every GCC file.
 
diff --git a/gcc/protoize.c b/gcc/protoize.c
index 48908e5..531dc81 100644
--- a/gcc/protoize.c
+++ b/gcc/protoize.c
@@ -258,13 +258,6 @@ static const unexpansion unexpansions[] = {
 
 static const int hash_mask = (HASH_TABLE_SIZE - 1);
 
-/* Make a table of default system include directories
-   just as it is done in cccp.c.  */
-
-#ifndef STANDARD_INCLUDE_DIR
-#define STANDARD_INCLUDE_DIR "/usr/include"
-#endif
-
 #ifndef LOCAL_INCLUDE_DIR
 #define LOCAL_INCLUDE_DIR "/usr/local/include"
 #endif
@@ -301,7 +294,6 @@ struct default_include { const char *fname;
 #ifdef SYSTEM_INCLUDE_DIR
     { SYSTEM_INCLUDE_DIR, 0, 0, 0 },
 #endif
-    { STANDARD_INCLUDE_DIR, 0, 0, 0},
 #endif /* not CROSS_COMPILE */
     { 0, 0, 0, 0}
     };
