diff --git a/gcc/testsuite/g++.dg/other/default1.C b/gcc/testsuite/g++.dg/other/default1.C
new file mode 100644
index 0000000..a6d9681
--- /dev/null
+++ b/gcc/testsuite/g++.dg/other/default1.C
@@ -0,0 +1,22 @@
+// PR c++/24103
+// ICE in simple_cst_equal
+// Origin: Alexander Stepanov <astepanov@softminecorp.com>
+// { dg-do compile }
+// { dg-options "" }
+
+struct S
+{
+  int i;
+};
+
+struct A
+{
+  A(S = (S){0});
+};
+
+struct B
+{
+  B(S = (S){0});
+};
+
+B::B(S) {}
diff --git a/gcc/tree.c b/gcc/tree.c
index 5c93372..60a0962 100644
--- a/gcc/tree.c
+++ b/gcc/tree.c
@@ -3593,10 +3593,8 @@ simple_cst_equal (t1, t2)
 			 TREE_STRING_LENGTH (t1)));
 
     case CONSTRUCTOR:
-      if (CONSTRUCTOR_ELTS (t1) == CONSTRUCTOR_ELTS (t2))
-	return 1;
-      else
-	abort ();
+      return simple_cst_list_equal (CONSTRUCTOR_ELTS (t1),
+				    CONSTRUCTOR_ELTS (t2));
 
     case SAVE_EXPR:
       return simple_cst_equal (TREE_OPERAND (t1, 0), TREE_OPERAND (t2, 0));
