diff --git a/gcc/c-semantics.c b/gcc/c-semantics.c
index eaa4452..af8fae7 100644
--- a/gcc/c-semantics.c
+++ b/gcc/c-semantics.c
@@ -673,8 +673,7 @@ genrtl_case_label (case_label)
   if (cleanup)
     {
       static int explained = 0;
-      warning_with_decl (TREE_PURPOSE (cleanup), 
-			 "destructor needed for `%#D'");
+      warning ("destructor needed for `%#D'", TREE_PURPOSE (cleanup));
       warning ("where case label appears here");
       if (!explained)
 	{
diff --git a/gcc/testsuite/g++.dg/parse/jumptocaselab1.C b/gcc/testsuite/g++.dg/parse/jumptocaselab1.C
new file mode 100644
index 0000000..7f028d2
--- /dev/null
+++ b/gcc/testsuite/g++.dg/parse/jumptocaselab1.C
@@ -0,0 +1,19 @@
+// PR c++/7566
+// { dg-do compile }
+
+struct bar { bar (); ~bar ();};
+
+void foo (int c)
+{
+  switch (c)
+    {
+    case 0:
+    case 1:
+      bar x;	// { dg-error "crosses initialization" }
+      break;
+    default:	// { dg-error "jump to case label" }
+      break;
+    }
+}
+
+// { dg-warning "destructor needed|where case label|enclose actions" "" { target *-*-* } 14 }
