diff --git a/libstdc++-v3/testsuite/22_locale/collate_members_wchar_t.cc b/libstdc++-v3/testsuite/22_locale/collate_members_wchar_t.cc
index c59ab32..049f470 100644
--- a/libstdc++-v3/testsuite/22_locale/collate_members_wchar_t.cc
+++ b/libstdc++-v3/testsuite/22_locale/collate_members_wchar_t.cc
@@ -24,7 +24,7 @@
 #include <testsuite_hooks.h>
 
 // XXX This may not work for non-glibc locale models.
-// { dg-do run { xfail *-*-* } }
+// { dg-do run { xfail *-*-linux-gnu } }
 #ifdef _GLIBCPP_USE_WCHAR_T
 void test01()
 {
@@ -92,8 +92,8 @@ void test01()
   VERIFY ( i1 * i2 > 0 );
 
   // Check German "de_DE" locale.
-  const wchar_t* strlit3 = L"Äuglein Augment"; // "C" == "Augment Äuglein"
-  const wchar_t* strlit4 = L"Base baß Baß Bast"; // "C" == "Base baß Baß Bast"
+  const wchar_t* strlit3 = L"?uglein Augment"; // "C" == "Augment ?uglein"
+  const wchar_t* strlit4 = L"Base ba? Ba? Bast"; // "C" == "Base ba? Ba? Bast"
 
   int size3 = char_traits<wchar_t>::length(strlit3) - 1;
   i1 = coll_de.compare(strlit3, strlit3 + size3, strlit3, strlit3 + 7);
diff --git a/libstdc++-v3/testsuite/22_locale/ctype_is_wchar_t.cc b/libstdc++-v3/testsuite/22_locale/ctype_is_wchar_t.cc
index 5ca171b..927889c 100644
--- a/libstdc++-v3/testsuite/22_locale/ctype_is_wchar_t.cc
+++ b/libstdc++-v3/testsuite/22_locale/ctype_is_wchar_t.cc
@@ -32,7 +32,7 @@
 #include <testsuite_hooks.h>
 
 // XXX This test (test02) is not working for non-glibc locale models.
-// { dg-do run { xfail *-*-* } }
+// { dg-do run { xfail *-*-linux-gnu } }
 
 #ifdef _GLIBCPP_USE_WCHAR_T
 typedef wchar_t char_type;
diff --git a/libstdc++-v3/testsuite/22_locale/money_get_members_wchar_t.cc b/libstdc++-v3/testsuite/22_locale/money_get_members_wchar_t.cc
index 3ac25d0..f6531fc 100644
--- a/libstdc++-v3/testsuite/22_locale/money_get_members_wchar_t.cc
+++ b/libstdc++-v3/testsuite/22_locale/money_get_members_wchar_t.cc
@@ -25,7 +25,7 @@
 #include <testsuite_hooks.h>
 
 // XXX This test is not working for non-glibc locale models.
-// { dg-do run { xfail *-*-* } }
+// { dg-do run { xfail *-*-linux-gnu } }
 
 #ifdef _GLIBCPP_USE_WCHAR_T
 // test string version
diff --git a/libstdc++-v3/testsuite/22_locale/money_put_members_wchar_t.cc b/libstdc++-v3/testsuite/22_locale/money_put_members_wchar_t.cc
index 1bfe86b..2f752e8 100644
--- a/libstdc++-v3/testsuite/22_locale/money_put_members_wchar_t.cc
+++ b/libstdc++-v3/testsuite/22_locale/money_put_members_wchar_t.cc
@@ -25,7 +25,7 @@
 #include <testsuite_hooks.h>
 
 // XXX This test is not working for non-glibc locale models.
-// { dg-do run { xfail *-*-* } }
+// { dg-do run { xfail *-*-linux-gnu } }
 
 #ifdef _GLIBCPP_USE_WCHAR_T
 // test string version
diff --git a/libstdc++-v3/testsuite/22_locale/moneypunct_members_wchar_t.cc b/libstdc++-v3/testsuite/22_locale/moneypunct_members_wchar_t.cc
index a81fe0a..e004df8 100644
--- a/libstdc++-v3/testsuite/22_locale/moneypunct_members_wchar_t.cc
+++ b/libstdc++-v3/testsuite/22_locale/moneypunct_members_wchar_t.cc
@@ -24,7 +24,7 @@
 #include <testsuite_hooks.h>
 
 // XXX This test is not working for non-glibc locale models.
-// { dg-do run { xfail *-*-* } }
+// { dg-do run { xfail *-*-linux-gnu } }
 
 #ifdef _GLIBCPP_USE_WCHAR_T
 void test01()
diff --git a/libstdc++-v3/testsuite/22_locale/num_get_members_wchar_t.cc b/libstdc++-v3/testsuite/22_locale/num_get_members_wchar_t.cc
index 088d4f8..e043a22 100644
--- a/libstdc++-v3/testsuite/22_locale/num_get_members_wchar_t.cc
+++ b/libstdc++-v3/testsuite/22_locale/num_get_members_wchar_t.cc
@@ -25,7 +25,7 @@
 #include <testsuite_hooks.h>
 
 // XXX This test is not working for non-glibc locale models.
-// { dg-do run { xfail *-*-* } }
+// { dg-do run { xfail *-*-linux-gnu } }
 
 #ifdef _GLIBCPP_USE_WCHAR_T
 void test01()
diff --git a/libstdc++-v3/testsuite/22_locale/num_put_members_wchar_t.cc b/libstdc++-v3/testsuite/22_locale/num_put_members_wchar_t.cc
index d3060dd..e260432 100644
--- a/libstdc++-v3/testsuite/22_locale/num_put_members_wchar_t.cc
+++ b/libstdc++-v3/testsuite/22_locale/num_put_members_wchar_t.cc
@@ -25,7 +25,7 @@
 #include <testsuite_hooks.h>
 
 // XXX This test is not working for non-glibc locale models.
-// { dg-do run { xfail *-*-* } }
+// { dg-do run { xfail *-*-linux-gnu } }
 
 #ifdef _GLIBCPP_USE_WCHAR_T
 void test01()
diff --git a/libstdc++-v3/testsuite/22_locale/numpunct_members_wchar_t.cc b/libstdc++-v3/testsuite/22_locale/numpunct_members_wchar_t.cc
index 14e9321..8319313 100644
--- a/libstdc++-v3/testsuite/22_locale/numpunct_members_wchar_t.cc
+++ b/libstdc++-v3/testsuite/22_locale/numpunct_members_wchar_t.cc
@@ -24,7 +24,7 @@
 #include <testsuite_hooks.h>
 
 // XXX This test is not working for non-glibc locale models.
-// { dg-do run { xfail *-*-* } }
+// { dg-do run { xfail *-*-linux-gnu } }
 
 #ifdef _GLIBCPP_USE_WCHAR_T
 void test01()
diff --git a/libstdc++-v3/testsuite/22_locale/time_get_members_wchar_t.cc b/libstdc++-v3/testsuite/22_locale/time_get_members_wchar_t.cc
index 89efdb9..223a6fd 100644
--- a/libstdc++-v3/testsuite/22_locale/time_get_members_wchar_t.cc
+++ b/libstdc++-v3/testsuite/22_locale/time_get_members_wchar_t.cc
@@ -25,7 +25,7 @@
 #include <testsuite_hooks.h>
 
 // XXX This test is not working for non-glibc locale models.
-// { dg-do run { xfail *-*-* } }
+// { dg-do run { xfail *-*-linux-gnu } }
 
 #ifdef _GLIBCPP_USE_WCHAR_T
 void test01()
diff --git a/libstdc++-v3/testsuite/22_locale/time_put_members_wchar_t.cc b/libstdc++-v3/testsuite/22_locale/time_put_members_wchar_t.cc
index 347f41f..b937644 100644
--- a/libstdc++-v3/testsuite/22_locale/time_put_members_wchar_t.cc
+++ b/libstdc++-v3/testsuite/22_locale/time_put_members_wchar_t.cc
@@ -25,7 +25,7 @@
 #include <testsuite_hooks.h>
 
 // XXX This test is not working for non-glibc locale models.
-// { dg-do run { xfail *-*-* } }
+// { dg-do run { xfail *-*-linux-gnu } }
 
 #ifdef _GLIBCPP_USE_WCHAR_T
 void test01()
