diff --git a/gcc/testsuite/g++.old-deja/g++.other/init5.C b/gcc/testsuite/g++.old-deja/g++.other/init5.C
index 92b1d32..566ae11 100644
--- a/gcc/testsuite/g++.old-deja/g++.other/init5.C
+++ b/gcc/testsuite/g++.old-deja/g++.other/init5.C
@@ -1,6 +1,6 @@
 // Objects must be destructed in decreasing cnt order
 // Original test attributed to James Kanze <jkanze@otelo.ibmmail.com>
-// execution test - XFAIL *-*-*
+// execution test - XFAIL *-*-gnulibc1
 
 extern "C" void abort ();
 
diff --git a/gcc/testsuite/gcc.c-torture/execute/20020227-1.x b/gcc/testsuite/gcc.c-torture/execute/20020227-1.x
index 67afb82..4e71947 100644
--- a/gcc/testsuite/gcc.c-torture/execute/20020227-1.x
+++ b/gcc/testsuite/gcc.c-torture/execute/20020227-1.x
@@ -24,6 +24,8 @@ if { [istarget "sparc64-*-*"] || [istarget "sparcv9-*-*"] } {
             { "-O0" "-O1" }
         }
     }
+} elseif { [istarget "powerpc64-*-*"] || [istarget "x86_64-*-*"] } {
+    # PowerPC-64 and x86_64 do not fail at any optimization level.
 } elseif { [istarget "*64*-*-*"] || [istarget "alpha*-*-*"]
         || [istarget "mmix-*-*"] || [istarget "mips*-*-irix6*"] } {
     # Other 64-bit targets fail at all optimization levels.
