diff --git a/libstdc++/std/new.h b/libstdc++/std/new.h
index cfdfbeb..71a5968 100644
--- a/libstdc++/std/new.h
+++ b/libstdc++/std/new.h
@@ -27,8 +27,8 @@ void operator delete (void *);
 void operator delete[] (void *);
 
 // default placement versions of operator new
-inline void *operator new(size_t, void *place) { return place; }
-inline void *operator new[](size_t, void *place) { return place; }
+static void *operator new(size_t, void *place) { return place; }
+static void *operator new[](size_t, void *place) { return place; }
 } // extern "C++"
 
 #endif
