diff --git a/gcc/config/host-linux.c b/gcc/config/host-linux.c
index d5fea9d..ac8d453 100644
--- a/gcc/config/host-linux.c
+++ b/gcc/config/host-linux.c
@@ -89,6 +89,10 @@
 # define TRY_EMPTY_VM_SPACE	0
 #endif
 
+#ifndef MAP_FAILED
+#define MAP_FAILED -1
+#endif
+
 /* Determine a location where we might be able to reliably allocate SIZE
    bytes.  FD is the PCH file, though we should return with the file 
    unmapped.  */
@@ -205,7 +209,7 @@ linux_gt_pch_use_address (void *base, size_t size, int fd, size_t offset)
     {
       ssize_t nbytes;
 
-      nbytes = read (fd, base, MIN (size, SSIZE_MAX));
+      nbytes = read (fd, base, MIN (size, INT_MAX));
       if (nbytes <= 0)
         return -1;
       base = (char *) base + nbytes;
diff --git a/include/libiberty.h b/include/libiberty.h
index 761b2cf..fa64dbe 100644
--- a/include/libiberty.h
+++ b/include/libiberty.h
@@ -285,14 +285,6 @@ extern int pwait PARAMS ((int, int *, int));
 extern int asprintf PARAMS ((char **, const char *, ...)) ATTRIBUTE_PRINTF_2;
 #endif
 
-#if !HAVE_DECL_VASPRINTF
-/* Like vsprintf but provides a pointer to malloc'd storage, which
-   must be freed by the caller.  */
-
-extern int vasprintf PARAMS ((char **, const char *, va_list))
-  ATTRIBUTE_PRINTF(2,0);
-#endif
-
 #define ARRAY_SIZE(a) (sizeof (a) / sizeof ((a)[0]))
 
 /* Drastically simplified alloca configurator.  If we're using GCC,
