diff --git a/Makefile.in b/Makefile.in
index b09bea9..e6478d0 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -174,7 +174,7 @@ build_xm_file= ... `configure' substitutes actual build xm- file name here.
 host_xm_file= ... `configure' substitutes actual host xm- file name here.
 lang_specs_files= ... `configure' substitutes actual lang spec file names here.
 lang_options_files= ... `configure' puts actual lang options file names here.
-version=`sed -e 's/.*\"\([^ \"]*\)[ \"].*/\1/' < $(srcdir)/version.c`
+version=`sed -e 's/.*\"\([0-9]*\.[0-9]*\.[0-9]*\).*/\1/' < $(srcdir)/version.c`
 mainversion=`sed -e 's/.*\"\([0-9]*\.[0-9]*\).*/\1/' < $(srcdir)/version.c`
 
 # Directory where sources are, from where we are.
@@ -190,27 +190,27 @@ local_prefix = /usr/local
 # Directory in which to put host dependent programs and libraries
 exec_prefix = $(prefix)
 # Directory in which to put the executable for the command `gcc'
-bindir = $(exec_prefix)/bin
+bindir = $(exec_prefix)/$(target)/gcc-bin/$(version)
 # Directory in which to put the directories used by the compiler.
 libdir = $(exec_prefix)/lib
 # Directory in which the compiler finds executables, libraries, etc.
 libsubdir = $(libdir)/gcc-lib/$(target)/$(version)
-# Directory in which the compiler finds g++ includes.
-gxx_include_dir= $(libdir)/g++-include
 # Directory to search for site-specific includes.
-includedir = $(local_prefix)/include
+includedir = $(libsubdir)/include
+# Directory in which the compiler finds g++ includes.
+gxx_include_dir= $(includedir)/g++
 # assertdir is overridden in cross-make.
 # (But this currently agrees with what is in cross-make.)
 assertdir = $(tooldir)/include
 # where the info files go
-infodir = $(prefix)/info
+infodir = $(prefix)/share/gcc-data/$(target)/$(version)/info
 # Extension (if any) to put in installed man-page filename.
 manext = .1
 objext = .o
 exeext =
 
 # Directory in which to put man pages.
-mandir = $(prefix)/man/man1
+mandir = $(prefix)/share/gcc-data/$(target)/$(version)/man/man1
 # Directory in which to find other cross-compilation tools and headers.
 # Used in install-cross.
 tooldir = $(exec_prefix)/$(target)
diff --git a/configure b/configure
index 9c2b12e..e6c4604 100755
--- a/configure
+++ b/configure
@@ -2548,7 +2548,7 @@ if [ x$host_truncate_target != x ]; then
 fi
 
 # Get the version number from the toplevel
-version=`sed -e 's/.*\"\([^ \"]*\)[ \"].*/\1/' < ${srcdir}/version.c`
+version=`sed -e 's/.*\"\([0-9]*\.[0-9]*\.[0-9]*\).*/\1/' < ${srcdir}/version.c`
 
 # For the current directory and all of the language subdirectories,
 # do the rest of the script ...
@@ -2569,7 +2569,7 @@ done
 case $canon_host in
 *linux*)	# All GNU/Linux systems use gcc as the native compiler.
 	prefix=$native_prefix
-	gxx_include_dir=$prefix/include/g++
+	gxx_include_dir=$prefix/lib/gcc-lib/$target/$version/include/g++
 	;;
 esac
 
