diff --git a/Makefile.in b/Makefile.in
index 100e869..5b82763 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -148,7 +148,7 @@ RANLIB_FOR_TARGET = $(TARGET_TOOLPREFIX)ranlib
 RANLIB_TEST_FOR_TARGET = [ -f $(TARGET_TOOLPREFIX)ranlib ]
 
 # Dir to search for system headers.  Overridden by cross-make.
-SYSTEM_HEADER_DIR = /usr/include
+SYSTEM_HEADER_DIR = $(tooldir)/include
 
 # Control whether to run fixproto.
 STMP_FIXPROTO = stmp-fixproto
diff --git a/gbl-ctors.h b/gbl-ctors.h
index 537df50..848bb67 100644
--- a/gbl-ctors.h
+++ b/gbl-ctors.h
@@ -29,17 +29,20 @@ Boston, MA 02111-1307, USA.  */
 
 	Note that this file should only be compiled with GCC.
 */
+#ifdef NEED_ATEXIT
+#ifndef HAVE_ATEXIT
+#define HAVE_ATEXIT	1	/* Take it from libgcc2.c */
+#endif
+#endif
 
 #ifdef HAVE_ATEXIT
 #ifdef WINNT
 extern int atexit (void (*) (void));
-#else
-extern void atexit (void (*) (void));
 #endif
 #define ON_EXIT(FUNC,ARG) atexit ((FUNC))
 #else
 #ifdef sun
-extern void on_exit (void*, void*);
+extern int on_exit (void*, void*);
 #define ON_EXIT(FUNC,ARG) on_exit ((FUNC), (ARG))
 #endif
 #endif
diff --git a/libgcc2.c b/libgcc2.c
index 1e01e91..63bcf31 100644
--- a/libgcc2.c
+++ b/libgcc2.c
@@ -1465,19 +1465,7 @@ BLOCK_PROFILER_CODE
 #include <stdio.h>
 char *ctime ();
 
-#ifdef HAVE_ATEXIT
-#ifdef WINNT
-extern int atexit (void (*) (void));
-#else
-extern void atexit (void (*) (void));
-#endif
-#define ON_EXIT(FUNC,ARG) atexit ((FUNC))
-#else
-#ifdef sun
-extern void on_exit (void*, void*);
-#define ON_EXIT(FUNC,ARG) on_exit ((FUNC), (ARG))
-#endif
-#endif
+#include "gbl-ctors.h"
 
 static struct bb *bb_head;
 
