diff --git a/bi-arity.c b/bi-arity.c
index da3607f..7a063af 100644
--- a/bi-arity.c
+++ b/bi-arity.c
@@ -57,7 +57,7 @@ main ()
       }
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
@@ -73,7 +73,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-lexer.c b/bi-lexer.c
index 6601c52..3d866b5 100644
--- a/bi-lexer.c
+++ b/bi-lexer.c
@@ -35,7 +35,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
@@ -57,7 +57,7 @@ xrealloc (block, nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't reallocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-opcode.c b/bi-opcode.c
index 6b1ebe4..8a1bfa5 100644
--- a/bi-opcode.c
+++ b/bi-opcode.c
@@ -55,7 +55,7 @@ main(argc, argv)
     fprintf (stderr, "(Number of opcodes is %d)\n", i);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
@@ -71,7 +71,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-opname.c b/bi-opname.c
index 1862e71..67116a8 100644
--- a/bi-opname.c
+++ b/bi-opname.c
@@ -36,7 +36,7 @@ main()
       printf("\"%s%s\",\n", d->basename, v->name);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
@@ -52,7 +52,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-parser.c b/bi-parser.c
index 1f18170..e9bc8a1 100644
--- a/bi-parser.c
+++ b/bi-parser.c
@@ -964,5 +964,5 @@ yyerror (s)
      char *s;
 {
   fprintf (stderr, "syntax error in input\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
diff --git a/bi-parser.y b/bi-parser.y
index 0a03d0f..aa80d48 100644
--- a/bi-parser.y
+++ b/bi-parser.y
@@ -165,5 +165,5 @@ yyerror (s)
      char *s;
 {
   fprintf (stderr, "syntax error in input\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
diff --git a/cccp.c b/cccp.c
index 39f1a5a..8d3f551 100644
--- a/cccp.c
+++ b/cccp.c
@@ -2331,8 +2331,8 @@ main (argc, argv)
     fatal ("I/O error on output");
 
   if (errors)
-    exit (FATAL_EXIT_CODE);
-  exit (SUCCESS_EXIT_CODE);
+    _exit (FATAL_EXIT_CODE);
+  _exit (SUCCESS_EXIT_CODE);
 
  perror:
   pfatal_with_name (in_fname);
@@ -9959,7 +9959,7 @@ fatal (PRINTF_ALIST (msg))
   vfprintf (stderr, msg, args);
   va_end (args);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -9986,9 +9986,9 @@ pfatal_with_name (name)
 {
   perror_with_name (name);
 #ifdef VMS
-  exit (vaxc$errno);
+  _exit (vaxc$errno);
 #else
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 #endif
 }
 
diff --git a/collect2.c b/collect2.c
index 30fb49d..b11889f 100644
--- a/collect2.c
+++ b/collect2.c
@@ -391,7 +391,7 @@ collect_exit (status)
   if (status != 0 && output_file != 0 && output_file[0])
     maybe_unlink (output_file);
 
-  exit (status);
+  _exit (status);
 }
 
 
diff --git a/cp/g++.c b/cp/g++.c
index f694898..6b708cd 100644
--- a/cp/g++.c
+++ b/cp/g++.c
@@ -153,7 +153,7 @@ fatal VPROTO((char *format, ...))
   /* XXX Not needed for g++ driver.  */
   delete_temp_files ();
 #endif
-  exit (1);
+  _exit (1);
 }
 
 static void
@@ -197,7 +197,7 @@ fatal (msg, arg1, arg2)
   /* XXX Not needed for g++ driver.  */
   delete_temp_files ();
 #endif
-  exit (1);
+  _exit (1);
 }
 
 #endif /* not HAVE_VPRINTF */
diff --git a/cp/typeck2.c b/cp/typeck2.c
index 653d6a5..674438b 100644
--- a/cp/typeck2.c
+++ b/cp/typeck2.c
@@ -356,7 +356,7 @@ my_friendly_abort (i)
       else
 	error ("confused by earlier errors, bailing out");
       
-      exit (34);
+      _exit (34);
     }
   ++abortcount;
 
diff --git a/cplus-dem.c b/cplus-dem.c
index 3eb27cc..8c98ce8 100644
--- a/cplus-dem.c
+++ b/cplus-dem.c
@@ -2836,7 +2836,7 @@ Usage: %s [-_] [-n] [-s {gnu,lucid,arm}] [--strip-underscores]\n\
        [--no-strip-underscores] [--format={gnu,lucid,arm}]\n\
        [--help] [--version] [arg...]\n",
 	   program_name);
-  exit (status);
+  _exit (status);
 }
 
 #define MBUF_SIZE 512
@@ -2882,7 +2882,7 @@ main (argc, argv)
 	    break;
 	  case 'v':
 	    printf ("GNU %s version %s\n", program_name, program_version);
-	    exit (0);
+	    _exit (0);
 	  case '_':
 	    strip_underscore = 1;
 	    break;
@@ -2964,7 +2964,7 @@ main (argc, argv)
 	}
     }
 
-  exit (0);
+  _exit (0);
 }
 
 static void
@@ -2972,7 +2972,7 @@ fatal (str)
      char *str;
 {
   fprintf (stderr, "%s: %s\n", program_name, str);
-  exit (1);
+  _exit (1);
 }
 
 char * malloc ();
diff --git a/cpperror.c b/cpperror.c
index 5ba3b78..6a7fe5d 100644
--- a/cpperror.c
+++ b/cpperror.c
@@ -110,7 +110,7 @@ fatal (str, arg)
   fprintf (stderr, "%s: ", progname);
   fprintf (stderr, str, arg);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 
@@ -121,8 +121,8 @@ cpp_pfatal_with_name (pfile, name)
 {
   cpp_perror_with_name (pfile, name);
 #ifdef VMS
-  exit (vaxc$errno);
+  _exit (vaxc$errno);
 #else
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 #endif
 }
diff --git a/cppmain.c b/cppmain.c
index 506ce05..8371b31 100644
--- a/cppmain.c
+++ b/cppmain.c
@@ -95,6 +95,6 @@ main (argc, argv)
   cpp_finish (&parse_in);
 
   if (parse_in.errors)
-    exit (FATAL_EXIT_CODE);
-  exit (SUCCESS_EXIT_CODE);
+    _exit (FATAL_EXIT_CODE);
+  _exit (SUCCESS_EXIT_CODE);
 }
diff --git a/enquire.c b/enquire.c
index 763b1a4..8f29d98 100644
--- a/enquire.c
+++ b/enquire.c
@@ -600,13 +600,13 @@ int setmode(s) char *s; {
 #endif
 
 Procedure farewell(bugs) int bugs; {
-	if (bugs == 0) exit(0);
+	if (bugs == 0) _exit(0);
 	printf("\n%sFor hints on dealing with the ", co);
 	if (bugs == 1) printf("problem");
 	else printf("%d problems", bugs);
 	printf(" above\n   see the section 'TROUBLESHOOTING' in the file ");
 	printf("%s%s\n", FILENAME, oc);
-	exit(bugs);
+	_exit(bugs);
 }
 
 /* The program has received a signal where it wasn't expecting one */
@@ -620,7 +620,7 @@ char *xmalloc(size) unsigned size; {
 	char *value = (char *)malloc(size);
 	if (value == 0) {
 		fprintf(stderr, "Virtual memory exceeded\n");
-		exit(bugs+1);
+		_exit(bugs+1);
 	}
 	return value;
 }
diff --git a/fix-header.c b/fix-header.c
index 7059d12..453f2f5 100644
--- a/fix-header.c
+++ b/fix-header.c
@@ -550,7 +550,7 @@ read_scan_file (in_fname, argc, argv)
     fprintf (stderr, "(%s: %d errors and %d warnings from cpp)\n",
 	     inc_filename, scan_in.errors, warnings);
   if (scan_in.errors)
-    exit (0);
+    _exit (0);
 
   /* Traditionally, getc and putc are defined in terms of _filbuf and _flsbuf.
      If so, those functions are also required. */
@@ -622,7 +622,7 @@ read_scan_file (in_fname, argc, argv)
     {
       if (verbose)
 	fprintf (stderr, "%s: OK, nothing needs to be done.\n", inc_filename);
-      exit (0);
+      _exit (0);
     }
   if (!verbose)
     fprintf (stderr, "%s: fixing %s\n", progname, inc_filename);
@@ -966,7 +966,7 @@ main (argc, argv)
     {
       fprintf (stderr, "%s: Usage: foo.h infile.h outfile.h options\n",
 	       progname);
-      exit (-1);
+      _exit (-1);
     }
 
   inc_filename = argv[1];
@@ -984,7 +984,7 @@ main (argc, argv)
 	    {
 	      if (verbose)
 		fprintf (stderr, "%s: ignoring %s\n", progname, inc_filename);
-	      exit (0);
+	      _exit (0);
 	    }
 	}
 	  
@@ -1025,13 +1025,13 @@ main (argc, argv)
       fprintf (stderr, "%s: Cannot open '%s' for reading -",
 	       progname, argv[2]);
       perror (NULL);
-      exit (-1);
+      _exit (-1);
     }
   if (fstat (inf_fd, &sbuf) < 0)
     {
       fprintf (stderr, "%s: Cannot get size of '%s' -", progname, argv[2]);
       perror (NULL);
-      exit (-1);
+      _exit (-1);
     }
   inf_size = sbuf.st_size;
   inf_buffer = (char*) xmalloc (inf_size + 2);
@@ -1048,7 +1048,7 @@ main (argc, argv)
 	{
 	  fprintf (stderr, "%s: Failed to read '%s' -", progname, argv[2]);
 	  perror (NULL);
-	  exit (-1);
+	  _exit (-1);
 	}
       if (i == 0)
 	{
@@ -1071,7 +1071,7 @@ main (argc, argv)
       fprintf (stderr, "%s: Cannot open '%s' for writing -",
 	       progname, argv[3]);
       perror (NULL);
-      exit (-1);
+      _exit (-1);
     }
 
   lineno = 1;
@@ -1200,7 +1200,7 @@ fatal (str, arg)
   fprintf (stderr, "%s: %s: ", progname, inc_filename);
   fprintf (stderr, str, arg);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 void
@@ -1209,5 +1209,5 @@ cpp_pfatal_with_name (pfile, name)
      char *name;
 {
   cpp_perror_with_name (pfile, name);
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
diff --git a/gcc.c b/gcc.c
index fe7de5b..92d9ecf 100644
--- a/gcc.c
+++ b/gcc.c
@@ -2095,7 +2095,7 @@ pexecute (search_flag, program, argv, not_last)
       /* Exec the program.  */
       (*func) (program, argv);
       perror_exec (program);
-      exit (-1);
+      _exit (-1);
       /* NOTREACHED */
       return 0;
 
@@ -2496,12 +2496,12 @@ process_command (argc, argv)
 	  printf ("*cross_compile:\n%d\n\n", cross_compile);
 	  printf ("*multilib:\n%s\n\n", multilib_select);
 
-	  exit (0);
+	  _exit (0);
 	}
       else if (! strcmp (argv[i], "-dumpversion"))
 	{
 	  printf ("%s\n", version_string);
-	  exit (0);
+	  _exit (0);
 	}
       else if (! strcmp (argv[i], "-dumpmachine"))
 	{
@@ -4379,26 +4379,26 @@ main (argc, argv)
       printf ("install: %s%s\n", standard_exec_prefix, machine_suffix);
       printf ("programs: %s\n", build_search_list (&exec_prefixes, "", 0));
       printf ("libraries: %s\n", build_search_list (&startfile_prefixes, "", 0));
-      exit (0);
+      _exit (0);
     }
 
   if (print_file_name)
     {
       printf ("%s\n", find_file (print_file_name));
-      exit (0);
+      _exit (0);
     }
 
   if (print_prog_name)
     {
       char *newname = find_a_file (&exec_prefixes, print_prog_name, X_OK);
       printf ("%s\n", (newname ? newname : print_prog_name));
-      exit (0);
+      _exit (0);
     }
 
   if (print_multi_lib)
     {
       print_multilib_info ();
-      exit (0);
+      _exit (0);
     }
 
   if (print_multi_directory)
@@ -4407,7 +4407,7 @@ main (argc, argv)
 	printf (".\n");
       else
 	printf ("%s\n", multilib_dir);
-      exit (0);
+      _exit (0);
     }
 
   if (verbose_flag)
@@ -4584,7 +4584,7 @@ main (argc, argv)
     delete_failure_queue ();
   delete_temp_files ();
 
-  exit (error_count > 0 ? (signal_count ? 2 : 1) : 0);
+  _exit (error_count > 0 ? (signal_count ? 2 : 1) : 0);
   /* NOTREACHED */
   return 0;
 }
@@ -4800,7 +4800,7 @@ fatal VPROTO((char *format, ...))
   va_end (ap);
   fprintf (stderr, "\n");
   delete_temp_files ();
-  exit (1);
+  _exit (1);
 }
 
 static void
@@ -4832,7 +4832,7 @@ fatal (msg, arg1, arg2)
 {
   error (msg, arg1, arg2);
   delete_temp_files ();
-  exit (1);
+  _exit (1);
 }
 
 static void
diff --git a/gen-protos.c b/gen-protos.c
index 44c900a..175387f 100644
--- a/gen-protos.c
+++ b/gen-protos.c
@@ -175,5 +175,5 @@ fatal (s)
      char *s;
 {
   fprintf (stderr, "%s: %s\n", "gen-protos", s);
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
diff --git a/genattr.c b/genattr.c
index f87ce06..ebacbf1 100644
--- a/genattr.c
+++ b/genattr.c
@@ -220,7 +220,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genattr: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -264,7 +264,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -426,7 +426,7 @@ from the machine description file `md'.  */\n\n");
   printf("#define ATTR_FLAG_very_unlikely\t0x20\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genattrtab.c b/genattrtab.c
index c764871..c85ab27 100644
--- a/genattrtab.c
+++ b/genattrtab.c
@@ -5476,7 +5476,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genattrtab: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -5563,7 +5563,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -5709,7 +5709,7 @@ from the machine description file `md'.  */\n\n");
   write_const_num_delay_slots ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/gencodes.c b/gencodes.c
index 8baf3b8..f943cc9 100644
--- a/gencodes.c
+++ b/gencodes.c
@@ -83,7 +83,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "gencodes: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -113,7 +113,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -155,7 +155,7 @@ from the machine description file `md'.  */\n\n");
   printf ("#endif /* MAX_INSN_CODE */\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genconfig.c b/genconfig.c
index 404534e..c523d30 100644
--- a/genconfig.c
+++ b/genconfig.c
@@ -266,7 +266,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genconfig: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -296,7 +296,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -350,7 +350,7 @@ from the machine description file `md'.  */\n\n");
     printf ("#define HAVE_lo_sum\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genemit.c b/genemit.c
index ebed4f3..61f39cc 100644
--- a/genemit.c
+++ b/genemit.c
@@ -697,7 +697,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genemit: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -727,7 +727,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -794,7 +794,7 @@ from the machine description file `md'.  */\n\n");
   output_init_mov_optab ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genextract.c b/genextract.c
index 3861bd5..7fe6598 100644
--- a/genextract.c
+++ b/genextract.c
@@ -368,7 +368,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genextract: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -424,7 +424,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -550,7 +550,7 @@ from the machine description file `md'.  */\n\n");
   printf ("    }\n}\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genflags.c b/genflags.c
index e91b45d..2443061 100644
--- a/genflags.c
+++ b/genflags.c
@@ -199,7 +199,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genflags: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -235,7 +235,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -288,7 +288,7 @@ from the machine description file `md'.  */\n\n");
   printf ("#endif  /* NO_MD_PROTOTYPES */\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genopinit.c b/genopinit.c
index 3939126..d978cc6 100644
--- a/genopinit.c
+++ b/genopinit.c
@@ -310,7 +310,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genopinit: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -342,7 +342,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -379,7 +379,7 @@ from the machine description file `md'.  */\n\n");
   printf ("}\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genoutput.c b/genoutput.c
index bbf6200..a0cd6ff 100644
--- a/genoutput.c
+++ b/genoutput.c
@@ -906,7 +906,7 @@ fatal (s, a1, a2, a3, a4)
   fprintf (stderr, "genoutput: ");
   fprintf (stderr, s, a1, a2, a3, a4);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -947,7 +947,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -981,7 +981,7 @@ main (argc, argv)
   output_epilogue ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 || have_error
+  _exit (ferror (stdout) != 0 || have_error
 	? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
 
   /* NOTREACHED */
diff --git a/genpeep.c b/genpeep.c
index ba2328d..46aeaab 100644
--- a/genpeep.c
+++ b/genpeep.c
@@ -414,7 +414,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genpeep: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -446,7 +446,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -503,7 +503,7 @@ from the machine description file `md'.  */\n\n");
   printf ("rtx peep_operand[%d];\n", max_opno + 1);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genrecog.c b/genrecog.c
index 21f1c06..3e75ec5 100644
--- a/genrecog.c
+++ b/genrecog.c
@@ -1692,7 +1692,7 @@ fatal (s)
   fprintf (stderr, s);
   fprintf (stderr, "\n");
   fprintf (stderr, "after %d definitions\n", next_index);
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -1725,7 +1725,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -1808,7 +1808,7 @@ from the machine description file `md'.  */\n\n");
   write_subroutine (split_tree.first, SPLIT);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/getopt.c b/getopt.c
index beb7450..d7e7ed3 100644
--- a/getopt.c
+++ b/getopt.c
@@ -759,7 +759,7 @@ main (argc, argv)
       printf ("\n");
     }
 
-  exit (0);
+  _exit (0);
 }
 
 #endif /* TEST */
diff --git a/getopt1.c b/getopt1.c
index 4580211..7c36887 100644
--- a/getopt1.c
+++ b/getopt1.c
@@ -174,7 +174,7 @@ main (argc, argv)
       printf ("\n");
     }
 
-  exit (0);
+  _exit (0);
 }
 
 #endif /* TEST */
diff --git a/libgcc2.c b/libgcc2.c
index 1e01e91..b39d2de 100644
--- a/libgcc2.c
+++ b/libgcc2.c
@@ -1979,7 +1979,7 @@ __enable_execute_stack (addr)
   if (r != KERN_SUCCESS)
     {
       mach_error("vm_protect VM_PROT_ALL", r);
-      exit(1);
+      _exit(1);
     }
 
   /* We inline the i-cache invalidation for speed */
diff --git a/objc-act.c b/objc-act.c
index 951f469..8a4dd9c 100644
--- a/objc-act.c
+++ b/objc-act.c
@@ -2431,7 +2431,7 @@ build_ivar_chain (interface, copy)
 	  error ("Cannot find interface declaration for `%s', superclass of `%s'",
 		 IDENTIFIER_POINTER (super_name),
 		 IDENTIFIER_POINTER (my_name));
-	  exit (FATAL_EXIT_CODE);
+	  _exit (FATAL_EXIT_CODE);
         }
 
       if (super_interface == interface)
@@ -6087,7 +6087,7 @@ start_class (code, class_name, super_name, protocol_list)
 	{
 	  error ("Cannot find interface declaration for `%s'",
 		 IDENTIFIER_POINTER (class_name));
-	  exit (FATAL_EXIT_CODE);
+	  _exit (FATAL_EXIT_CODE);
 	}
       else
         add_category (class_category_is_assoc_with, class);
@@ -6123,7 +6123,7 @@ start_class (code, class_name, super_name, protocol_list)
         {
 	  error ("Cannot find interface declaration for `%s'",
 		 IDENTIFIER_POINTER (class_name));
-	  exit (FATAL_EXIT_CODE);
+	  _exit (FATAL_EXIT_CODE);
         }
     }
   return class;
diff --git a/protoize.c b/protoize.c
index 9728bf2..173ba5c 100644
--- a/protoize.c
+++ b/protoize.c
@@ -179,7 +179,7 @@ extern int fprintf (FILE *, const char *, ...);
 extern int printf (const char *, ...);
 extern int open (const char *, int, ...);
 #endif /* 0 */
-extern void exit ();
+extern void _exit ();
 extern void free ();
 extern int read ();
 extern int write ();
@@ -659,7 +659,7 @@ xmalloc (byte_count)
   if (rv == NULL)
     {
       fprintf (stderr, "\n%s: virtual memory exceeded\n", pname);
-      exit (1);
+      _exit (1);
       return 0;		/* avoid warnings */
     }
   else
@@ -679,7 +679,7 @@ xrealloc (old_space, byte_count)
   if (rv == NULL)
     {
       fprintf (stderr, "\n%s: virtual memory exceeded\n", pname);
-      exit (1);
+      _exit (1);
       return 0;		/* avoid warnings */
     }
   else
@@ -732,7 +732,7 @@ void
 fancy_abort ()
 {
   fprintf (stderr, "%s: internal abort\n", pname);
-  exit (1);
+  _exit (1);
 }
 
 /* Make a duplicate of the first N bytes of a given string in a newly
@@ -870,7 +870,7 @@ usage ()
   fprintf (stderr, "%s: usage '%s [ -VqfnkNlgC ] [ -B <diname> ] [ filename ... ]'\n",
 	   pname, pname);
 #endif /* !defined (UNPROTOIZE) */
-  exit (1);
+  _exit (1);
 }
 
 /* Return true if the given filename (assumed to be an absolute filename)
@@ -1364,7 +1364,7 @@ abspath (cwd, rel_filename)
 
               	  fprintf (stderr, "%s: invalid file name: %s\n",
 			   pname, rel_filename);
-              	  exit (1);
+              	  _exit (1);
               	}
                     *++outp = '\0';
                     continue;
@@ -1534,7 +1534,7 @@ aux_info_corrupted ()
 {
   fprintf (stderr, "\n%s: fatal error: aux info file corrupted at line %d\n",
 	   pname, current_aux_info_lineno);
-  exit (1);
+  _exit (1);
 }
 
 /* ??? This comment is vague.  Say what the condition is for.  */
@@ -1860,7 +1860,7 @@ save_def_or_dec (l, is_syscalls)
 			 def_dec_p->file->hash_entry->symbol,
 			 def_dec_p->line,
 			 def_dec_p->hash_entry->symbol);
-                exit (1);
+                _exit (1);
               }
             free_def_dec (def_dec_p);
             return;
@@ -4599,7 +4599,7 @@ main (argc, argv)
     {
       fprintf (stderr, "%s: cannot get working directory: %s\n",
 	       pname, my_strerror(errno));
-      exit (1);
+      _exit (1);
     }
 
   /* By default, convert the files in the current directory.  */
@@ -4725,8 +4725,8 @@ main (argc, argv)
       do_processing ();
     }
   if (errors)
-    exit (1);
+    _exit (1);
   else
-    exit (0);
+    _exit (0);
   return 1;
 }
diff --git a/toplev.c b/toplev.c
index 1b6fccb..f760228 100644
--- a/toplev.c
+++ b/toplev.c
@@ -878,7 +878,7 @@ pfatal_with_name (name)
 {
   fprintf (stderr, "%s: ", progname);
   perror (name);
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 void
@@ -886,7 +886,7 @@ fatal_io_error (name)
      char *name;
 {
   fprintf (stderr, "%s: %s: I/O error\n", progname, name);
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* Called to give a better error message for a bad insn rather than
@@ -1355,7 +1355,7 @@ vfatal (s, ap)
      va_list ap;
 {
   verror (s, ap);
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 void
@@ -3925,10 +3925,10 @@ You Lose!  You must define PREFERRED_DEBUGGING_TYPE!
 #endif /* not OS2 and not VMS and not _WIN32 */
 
   if (errorcount)
-    exit (FATAL_EXIT_CODE);
+    _exit (FATAL_EXIT_CODE);
   if (sorrycount)
-    exit (FATAL_EXIT_CODE);
-  exit (SUCCESS_EXIT_CODE);
+    _exit (FATAL_EXIT_CODE);
+  _exit (SUCCESS_EXIT_CODE);
   return 0;
 }
 
