diff --git a/usr/include/sys/features.h b/usr/include/sys/features.h
index 2188071..a1b99f0 100644
--- a/usr/include/sys/features.h
+++ b/usr/include/sys/features.h
@@ -262,7 +262,7 @@ extern "C" {
 #endif
 
 #if defined(_ISOC11_SOURCE) || \
-  (__STDC_VERSION__ - 0) >= 201112L || (__cplusplus - 0) >= 201103L
+  (__STDC_VERSION__ - 0) >= 201112L || ((__cplusplus - 0) >= 201103L && __GNUC_PREREQ (4, 8))
 #define	__ISO_C_VISIBLE		2011
 #elif defined(_ISOC99_SOURCE) || (_POSIX_C_SOURCE - 0) >= 200112L || \
   (__STDC_VERSION__ - 0) >= 199901L || defined(__cplusplus)
