diff --git a/usr/include/stdlib.h b/usr/include/stdlib.h
index 15b3494..d0e1d5c 100644
--- a/usr/include/stdlib.h
+++ b/usr/include/stdlib.h
@@ -131,7 +131,12 @@ int	mkstemp (char *);
 int	mkstemps (char *, int);
 #endif
 #if __BSD_VISIBLE || (__XSI_VISIBLE >= 4 && __POSIX_VISIBLE < 200112)
-char *	mktemp (char *) _ATTRIBUTE ((__deprecated__("the use of `mktemp' is dangerous; use `mkstemp' instead")));
+char *	mktemp (char *)
+#if __GNUC_PREREQ (4, 5)
+_ATTRIBUTE ((__deprecated__("the use of `mktemp' is dangerous; use `mkstemp' instead")));
+#else
+__attribute__ ((__deprecated__));
+#endif
 #endif
 #endif /* !_REENT_ONLY */
 char *	_mkdtemp_r (struct _reent *, char *);
@@ -139,7 +144,12 @@ int	_mkostemp_r (struct _reent *, char *, int);
 int	_mkostemps_r (struct _reent *, char *, int, int);
 int	_mkstemp_r (struct _reent *, char *);
 int	_mkstemps_r (struct _reent *, char *, int);
-char *	_mktemp_r (struct _reent *, char *) _ATTRIBUTE ((__deprecated__("the use of `mktemp' is dangerous; use `mkstemp' instead")));
+char *	_mktemp_r (struct _reent *, char *)
+#if __GNUC_PREREQ (4, 5)
+_ATTRIBUTE ((__deprecated__("the use of `mktemp' is dangerous; use `mkstemp' instead")));
+#else
+__attribute__ ((__deprecated__));
+#endif
 void	qsort (void *__base, size_t __nmemb, size_t __size, __compar_fn_t _compar);
 int	rand (void);
 void	*realloc(void *, size_t) __result_use_check __alloc_size(2) _NOTHROW;
