diff --git a/gcc/fold-const.c b/gcc/fold-const.c
index 7271af18..80459ded 100644
--- a/gcc/fold-const.c
+++ b/gcc/fold-const.c
@@ -9786,7 +9786,8 @@ fold (tree expr)
           && integer_zerop (TREE_OPERAND (t, 2))
           && (tem = sign_bit_p (TREE_OPERAND (arg0, 0), arg1)))
         return fold_convert (type, fold (build2 (BIT_AND_EXPR,
-						 TREE_TYPE (tem), tem, arg1)));
+						TREE_TYPE (tem), tem, 
+						fold_convert (TREE_TYPE (tem), arg1))));
 
       /* (A >> N) & 1 ? (1 << N) : 0 is simply A & (1 << N).  A & 1 was
 	 already handled above.  */
diff --git a/gcc/testsuite/gcc.c-torture/compile/pr26622.c b/gcc/testsuite/gcc.c-torture/compile/pr26622.c
new file mode 100644
index 00000000..aea991ec
--- /dev/null
+++ b/gcc/testsuite/gcc.c-torture/compile/pr26622.c
@@ -0,0 +1,11 @@
+/* PR middle-end/26622
+   fold_ternary used to create a tree with mismatching types, causing
+   (const_int 128) to appear in QImode rtx.  */
+
+unsigned char g;
+
+unsigned long long
+foo (void)
+{
+  return ((long long) ((g & 0x80) != 0)) << 7;
+}
