diff --git a/gcc/cfgexpand.c b/gcc/cfgexpand.c
index 9cd34987..ae8e9e2c 100644
--- a/gcc/cfgexpand.c
+++ b/gcc/cfgexpand.c
@@ -56,6 +56,8 @@ add_reg_br_prob_note (FILE *dump_file, rtx last, int probability)
 	if (!any_condjump_p (last)
 	    || GET_CODE (NEXT_INSN (last)) != JUMP_INSN
 	    || !simplejump_p (NEXT_INSN (last))
+	    || !NEXT_INSN (NEXT_INSN (last))
+	    || !NEXT_INSN (NEXT_INSN (NEXT_INSN (last)))
 	    || GET_CODE (NEXT_INSN (NEXT_INSN (last))) != BARRIER
 	    || GET_CODE (NEXT_INSN (NEXT_INSN (NEXT_INSN (last)))) != CODE_LABEL
 	    || NEXT_INSN (NEXT_INSN (NEXT_INSN (NEXT_INSN (last)))))
diff --git a/gcc/testsuite/gcc.c-torture/compile/20060823-1.c b/gcc/testsuite/gcc.c-torture/compile/20060823-1.c
new file mode 100644
index 00000000..fbce7fe9
--- /dev/null
+++ b/gcc/testsuite/gcc.c-torture/compile/20060823-1.c
@@ -0,0 +1,18 @@
+/* PR middle-end/28683 */
+
+extern void foo (int *);
+
+struct A
+{
+  int f;
+};
+
+struct A *
+test (struct A *r)
+{
+  int *f = &r->f;
+  static int i = 0;
+  if (!i && !((void *) f == (void *) r))
+    foo (&i);
+  return r;
+}
