diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index c6912b21..df692579 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -4513,7 +4513,7 @@ cp_parser_postfix_dot_deref_expression (cp_parser *parser,
 	    }
 	  if (scope && name && BASELINK_P (name))
 	    adjust_result_of_qualified_name_lookup
-	      (name, BINFO_TYPE (BASELINK_BINFO (name)), scope);
+	      (name, BINFO_TYPE (BASELINK_ACCESS_BINFO (name)), scope);
 	  postfix_expression
 	    = finish_class_member_access_expr (postfix_expression, name);
 	}
diff --git a/gcc/testsuite/g++.dg/template/member7.C b/gcc/testsuite/g++.dg/template/member7.C
new file mode 100644
index 00000000..1c75e495
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/member7.C
@@ -0,0 +1,15 @@
+// PR c++/29080
+
+struct Base {
+  template<class C> void method() { }
+};
+
+struct Left : public Base { };
+struct Right : public Base { };
+struct Join : public Left, public Right { };
+
+void function()
+{
+  Join join;
+  join.Left::method<int>();
+}
