diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 5445ab5f..5c94e49f 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -11128,7 +11128,11 @@ revert_static_member_fn (tree decl)
   tmp = build_exception_variant (tmp,
 				 TYPE_RAISES_EXCEPTIONS (function));
   TREE_TYPE (decl) = tmp;
-  if (DECL_ARGUMENTS (decl))
+  if (DECL_ARGUMENTS (decl)
+      /* revert_static_member_fn might be called before grokclassfn
+         had time to add the "this" argument.  */
+      && DECL_ARTIFICIAL (DECL_ARGUMENTS (decl))
+      && DECL_NAME (DECL_ARGUMENTS (decl)) == this_identifier)
     DECL_ARGUMENTS (decl) = TREE_CHAIN (DECL_ARGUMENTS (decl));
   DECL_STATIC_FUNCTION_P (decl) = 1;
 }
diff --git a/gcc/testsuite/g++.dg/template/friend49.C b/gcc/testsuite/g++.dg/template/friend49.C
new file mode 100644
index 00000000..b4d717f5
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/friend49.C
@@ -0,0 +1,18 @@
+// PR c++/29054
+// { dg-do compile }
+
+struct A
+{
+  template <typename T, typename U> static void create (U) {}
+};
+
+struct B
+{
+  friend void A::create <B, const char *> (const char *);
+};
+
+int
+main ()
+{
+  A::create<B>("test");
+}
