diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index df692579..dd10d0fd 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -11103,7 +11103,7 @@ cp_parser_declarator (cp_parser* parser,
 						member_p);
     }
 
-  if (attributes && declarator != cp_error_declarator)
+  if (attributes && declarator && declarator != cp_error_declarator)
     declarator->attributes = attributes;
 
   return declarator;
diff --git a/gcc/testsuite/g++.dg/ext/attrib22.C b/gcc/testsuite/g++.dg/ext/attrib22.C
new file mode 100644
index 00000000..5304a35d
--- /dev/null
+++ b/gcc/testsuite/g++.dg/ext/attrib22.C
@@ -0,0 +1,6 @@
+// PR c++/27648
+
+void f()
+{
+  static_cast<float *__attribute((unused))>(0); // { dg-error "expected" }
+}
