diff --git a/gcc/testsuite/g++.dg/opt/pr27826.C b/gcc/testsuite/g++.dg/opt/pr27826.C
new file mode 100644
index 00000000..5e40f174
--- /dev/null
+++ b/gcc/testsuite/g++.dg/opt/pr27826.C
@@ -0,0 +1,14 @@
+/* { dg-do compile } */
+/* { dg-options "-O3" } */
+
+struct Geometry
+{
+        int type:16;
+};
+struct Geometry get() {};
+int f()
+{
+        struct Geometry test;
+        return get().type == test.type;
+}
+
diff --git a/gcc/tree.c b/gcc/tree.c
index 17399352..e8f22bd7 100644
--- a/gcc/tree.c
+++ b/gcc/tree.c
@@ -4923,12 +4923,10 @@ get_narrower (tree op, int *unsignedp_ptr)
 	{
 	  if (first)
 	    uns = DECL_UNSIGNED (TREE_OPERAND (op, 1));
-	  win = build3 (COMPONENT_REF, type, TREE_OPERAND (op, 0),
-			TREE_OPERAND (op, 1), NULL_TREE);
-	  TREE_SIDE_EFFECTS (win) = TREE_SIDE_EFFECTS (op);
-	  TREE_THIS_VOLATILE (win) = TREE_THIS_VOLATILE (op);
+	  win = fold_convert (type, op);
 	}
     }
+
   *unsignedp_ptr = uns;
   return win;
 }
