diff --git a/gcc/testsuite/g++.dg/init/ptrmem3.C b/gcc/testsuite/g++.dg/init/ptrmem3.C
new file mode 100644
index 00000000..95b60379
--- /dev/null
+++ b/gcc/testsuite/g++.dg/init/ptrmem3.C
@@ -0,0 +1,8 @@
+// PR c++/28148
+
+struct foo {
+public:
+  virtual int bar(int);
+};
+
+void (foo::*__Virtual__foo__Var1)() = (void (foo::*)())(&foo::bar);
diff --git a/gcc/varasm.c b/gcc/varasm.c
index 368eb9ed..fc4458f9 100644
--- a/gcc/varasm.c
+++ b/gcc/varasm.c
@@ -3765,6 +3765,9 @@ output_constant (tree exp, unsigned HOST_WIDE_INT size, unsigned int align)
   code = TREE_CODE (TREE_TYPE (exp));
   thissize = int_size_in_bytes (TREE_TYPE (exp));
 
+  /* Give the front end another chance to expand constants.  */
+  exp = lang_hooks.expand_constant (exp);
+
   /* Allow a constructor with no elements for any data type.
      This means to fill the space with zeros.  */
   if (TREE_CODE (exp) == CONSTRUCTOR && CONSTRUCTOR_ELTS (exp) == 0)
