diff --git a/gcc/dwarf2out.c b/gcc/dwarf2out.c
index 6b1427fd..b67abcef 100644
--- a/gcc/dwarf2out.c
+++ b/gcc/dwarf2out.c
@@ -6082,7 +6082,8 @@ is_symbol_die (dw_die_ref c)
 {
   return (is_type_die (c)
 	  || (get_AT (c, DW_AT_declaration)
-	      && !get_AT (c, DW_AT_specification)));
+	      && !get_AT (c, DW_AT_specification))
+	  || c->die_tag == DW_TAG_namespace);
 }
 
 static char *
diff --git a/gcc/testsuite/g++.dg/debug/dwarf2-2.C b/gcc/testsuite/g++.dg/debug/dwarf2-2.C
new file mode 100644
index 00000000..9e6dbd21
--- /dev/null
+++ b/gcc/testsuite/g++.dg/debug/dwarf2-2.C
@@ -0,0 +1,17 @@
+// PR debug/27057
+// { dg-do compile }
+// { dg-options "-g -feliminate-dwarf2-dups" }
+
+namespace N
+{
+}
+
+struct A
+{
+  void foo ();
+};
+
+void A::foo ()
+{
+  using namespace N;
+}
