diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 5c94e49f..9616d326 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -3565,7 +3565,8 @@ shadow_tag (cp_decl_specifier_seq *declspecs)
 
     }
 
-  maybe_process_partial_specialization (t);
+  if (maybe_process_partial_specialization (t) == error_mark_node)
+    return NULL_TREE;
 
   /* This is where the variables in an anonymous union are
      declared.  An anonymous union declaration looks like:
diff --git a/gcc/testsuite/g++.dg/parse/crash9.C b/gcc/testsuite/g++.dg/parse/crash9.C
index 44ce57b1..d121d5dd 100644
--- a/gcc/testsuite/g++.dg/parse/crash9.C
+++ b/gcc/testsuite/g++.dg/parse/crash9.C
@@ -6,4 +6,4 @@
 // contains error.
 
 template <typename> struct A {};
-template <typename> struct A<INVALID> : A<int> { }; // { dg-error "not declared|invalid" }
+template <typename> struct A<INVALID> : A<int> { }; // { dg-error "not declared|invalid|token|extra" }
diff --git a/gcc/testsuite/g++.dg/template/spec32.C b/gcc/testsuite/g++.dg/template/spec32.C
new file mode 100644
index 00000000..6ef172b5
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/spec32.C
@@ -0,0 +1,6 @@
+//PR c++/28861
+
+struct A
+{
+  template<template<int> class B> struct B<0>;  // { dg-error "name of class shadows" }
+};
