diff --git a/gcc/cp/call.c b/gcc/cp/call.c
index 1c0ad7d6..b7ca6bd0 100644
--- a/gcc/cp/call.c
+++ b/gcc/cp/call.c
@@ -1960,6 +1960,12 @@ add_builtin_candidate (struct z_candidate **candidates, enum tree_code code,
 	 types are TYPE2.  */
       break;
 
+    case REALPART_EXPR:
+    case IMAGPART_EXPR:
+      if (ARITHMETIC_TYPE_P (type1))
+	break;
+      return;
+ 
     default:
       gcc_unreachable ();
     }
diff --git a/gcc/testsuite/g++.dg/ext/complex9.C b/gcc/testsuite/g++.dg/ext/complex9.C
new file mode 100644
index 00000000..5bb1625c
--- /dev/null
+++ b/gcc/testsuite/g++.dg/ext/complex9.C
@@ -0,0 +1,8 @@
+// PR c++/28501
+
+struct A
+{
+  operator int();
+};
+
+int i = __real__ A();
