diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 38226469..24c0d2b7 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -6977,7 +6977,7 @@ grokdeclarator (const cp_declarator *declarator,
   if (dependant_name && !friendp)
     {
       error ("%<%T::%D%> is not a valid declarator", ctype, dependant_name);
-      return void_type_node;
+      return error_mark_node;
     }
 
   /* Issue errors about use of storage classes for parameters.  */
@@ -7218,7 +7218,7 @@ grokdeclarator (const cp_declarator *declarator,
 			if (! member_function_or_else (ctype,
 						       current_class_type,
 						       flags))
-			  return void_type_node;
+			  return error_mark_node;
 		      }
 		  }
 		else            /* It's a constructor.  */
@@ -7246,10 +7246,10 @@ grokdeclarator (const cp_declarator *declarator,
 			if (! member_function_or_else (ctype,
 						       current_class_type,
 						       flags))
-			  return void_type_node;
+			  return error_mark_node;
 			TYPE_HAS_CONSTRUCTOR (ctype) = 1;
 			if (sfk != sfk_constructor)
-			  return NULL_TREE;
+			  return error_mark_node;
 		      }
 		  }
 		if (decl_context == FIELD)
@@ -7512,7 +7512,7 @@ grokdeclarator (const cp_declarator *declarator,
 	    {
 	      error ("cannot declare member %<%T::%s%> within %qT",
 		     ctype, name, current_class_type);
-	      return void_type_node;
+	      return error_mark_node;
 	    }
 	}
       else
@@ -7901,7 +7901,7 @@ grokdeclarator (const cp_declarator *declarator,
 		  {
 		    error ("can't make %qD into a method -- not in a class",
 			   unqualified_id);
-		    return void_type_node;
+		    return error_mark_node;
 		  }
 
 		/* ``A union may [ ... ] not [ have ] virtual functions.''
@@ -7910,7 +7910,7 @@ grokdeclarator (const cp_declarator *declarator,
 		  {
 		    error ("function %qD declared virtual inside a union",
 			   unqualified_id);
-		    return void_type_node;
+		    return error_mark_node;
 		  }
 
 		if (NEW_DELETE_OPNAME_P (unqualified_id))
@@ -8014,7 +8014,7 @@ grokdeclarator (const cp_declarator *declarator,
 			       funcdef_flag, template_count, in_namespace, 
 			       attrlist); 
 	    if (decl == NULL_TREE)
-	      return NULL_TREE;
+	      return error_mark_node;
 	  }
 	else if (!staticp && !dependent_type_p (type)
 		 && !COMPLETE_TYPE_P (complete_type (type))
@@ -8070,7 +8070,7 @@ grokdeclarator (const cp_declarator *declarator,
 		return decl;
 	      }
 	    else
-	      return void_type_node;
+	      return error_mark_node;
 	  }
 
 	/* Structure field.  It may not be a function, except for C++.  */
@@ -8109,7 +8109,7 @@ grokdeclarator (const cp_declarator *declarator,
 		     void_type_node, as if this was a friend
 		     declaration, to cause callers to completely
 		     ignore this declaration.  */
-		  return void_type_node;
+		  return error_mark_node;
 	      }
 
 	    if (staticp)
@@ -8156,7 +8156,7 @@ grokdeclarator (const cp_declarator *declarator,
 	int publicp = 0;
 
 	if (!unqualified_id)
-	  return NULL_TREE;
+	  return error_mark_node;
 
 	if (TREE_CODE (unqualified_id) == TEMPLATE_ID_EXPR)
 	  original_name = dname;
@@ -8211,7 +8211,7 @@ grokdeclarator (const cp_declarator *declarator,
 			   publicp, inlinep, sfk, funcdef_flag,
 			   template_count, in_namespace, attrlist);
 	if (decl == NULL_TREE)
-	  return NULL_TREE;
+	  return error_mark_node;
 
 	if (staticp == 1)
 	  {
diff --git a/gcc/testsuite/g++.dg/template/crash50.C b/gcc/testsuite/g++.dg/template/crash50.C
new file mode 100644
index 00000000..00605612
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/crash50.C
@@ -0,0 +1,7 @@
+// PR c++/27398
+// { dg-do compile }
+
+struct A
+{
+  template<int> void* foo(; // { dg-error "primary-expression|initialization|static" }
+};
