diff --git a/gcc/c-common.c b/gcc/c-common.c
index bd7402f9..89fbb583 100644
--- a/gcc/c-common.c
+++ b/gcc/c-common.c
@@ -4971,6 +4971,8 @@ handle_vector_size_attribute (tree *node, tree name, tree args,
   orig_mode = TYPE_MODE (type);
 
   if (TREE_CODE (type) == RECORD_TYPE
+      || TREE_CODE (type) == UNION_TYPE
+      || TREE_CODE (type) == VECTOR_TYPE
       || (GET_MODE_CLASS (orig_mode) != MODE_FLOAT
 	  && GET_MODE_CLASS (orig_mode) != MODE_INT)
       || !host_integerp (TYPE_SIZE_UNIT (type), 1)
diff --git a/gcc/testsuite/gcc.dg/pr29736.c b/gcc/testsuite/gcc.dg/pr29736.c
new file mode 100644
index 00000000..54eb9aaa
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/pr29736.c
@@ -0,0 +1,13 @@
+/* PR c/29736 */
+
+int __attribute__ ((vector_size (8), vector_size (8))) v; /* { dg-error "invalid vector type" } */
+
+extern int __attribute__ ((vector_size (8))) w;
+int __attribute__ ((vector_size (8))) w;
+
+void
+foo ()
+{
+  v = v + v;
+  w = w + w;
+}
