diff --git a/gcc/config/ia64/linux.h b/gcc/config/ia64/linux.h
index c8a60d65..3cc7b001 100644
--- a/gcc/config/ia64/linux.h
+++ b/gcc/config/ia64/linux.h
@@ -48,6 +48,7 @@ do {						\
       %{!dynamic-linker:-dynamic-linker /lib/ld-linux-ia64.so.2}} \
       %{static:-static}}"
 
+#define CPP_SPEC "%{posix:-D_POSIX_SOURCE} %{pthread:-D_REENTRANT}"
 
 #define JMP_BUF_SIZE  76
 
diff --git a/gcc/config/s390/linux.h b/gcc/config/s390/linux.h
index 369b8c1a..28c1e81f 100644
--- a/gcc/config/s390/linux.h
+++ b/gcc/config/s390/linux.h
@@ -89,6 +89,7 @@ Software Foundation, 59 Temple Place - Suite 330, Boston, MA
           %{m31:-dynamic-linker /lib/ld.so.1} \
           %{m64:-dynamic-linker /lib/ld64.so.1}}}}"
 
+#define CPP_SPEC "%{posix:-D_POSIX_SOURCE} %{pthread:-D_REENTRANT}"
 
 #define TARGET_ASM_FILE_END file_end_indicate_exec_stack
 
diff --git a/gcc/gthr-posix.h b/gcc/gthr-posix.h
index 938b4f15..cdcecf62 100644
--- a/gcc/gthr-posix.h
+++ b/gcc/gthr-posix.h
@@ -36,7 +36,7 @@ Software Foundation, 59 Temple Place - Suite 330, Boston, MA
 #define __GTHREADS 1
 
 /* Some implementations of <pthread.h> require this to be defined.  */
-#ifndef _REENTRANT
+#if !defined(_REENTRANT) && defined(__osf__)
 #define _REENTRANT 1
 #endif
 
