diff --git a/gcc/stmt.c b/gcc/stmt.c
index 5dd5e82d..db19fa6f 100644
--- a/gcc/stmt.c
+++ b/gcc/stmt.c
@@ -868,7 +868,7 @@ expand_asm_operands (tree string, tree outputs, tree inputs,
 
       if (asm_operand_ok (op, constraint) <= 0)
 	{
-	  if (allows_reg)
+	  if (allows_reg && TYPE_MODE (type) != BLKmode)
 	    op = force_reg (TYPE_MODE (type), op);
 	  else if (!allows_mem)
 	    warning ("asm operand %d probably doesn%'t match constraints",
diff --git a/gcc/testsuite/gcc.dg/pr8788-1.c b/gcc/testsuite/gcc.dg/pr8788-1.c
new file mode 100644
index 00000000..74e16940
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/pr8788-1.c
@@ -0,0 +1,20 @@
+/* PR inline-asm/8788 */
+/* { dg-do compile } */
+/* { dg-options "-O2" } */
+
+typedef struct {
+    long x[6];
+} myjmp_buf;
+
+typedef struct {
+    myjmp_buf regs;
+} my_stack;
+
+void switch_to_stack (my_stack *stack){
+    asm (  /* { dg-error "impossible constraint" } */
+/* { dg-warning "asm operand 1" "asm operand 1" { target *-*-* } 14 } */
+        "\n"
+        : "+r" (stack->regs)
+    );
+}
+
