diff --git a/gcc/common.opt b/gcc/common.opt
index 566bdcb7..4584c7ae 100644
--- a/gcc/common.opt
+++ b/gcc/common.opt
@@ -540,7 +540,7 @@ Common RejectNegative Report Var(flag_mudflap)
 Add mudflap bounds-checking instrumentation for single-threaded program.
 
 fmudflapth
-Common RejectNegative Report Var(flag_mudflap_threads)
+Common RejectNegative Report VarExists Var(flag_mudflap,2)
 Add mudflap bounds-checking instrumentation for multi-threaded program.
 
 fmudflapir
diff --git a/gcc/tree-mudflap.c b/gcc/tree-mudflap.c
index b704d049..dc86bfb3 100644
--- a/gcc/tree-mudflap.c
+++ b/gcc/tree-mudflap.c
@@ -48,6 +48,10 @@ Software Foundation, 59 Temple Place - Suite 330, Boston, MA
 
 /* Internal function decls */
 
+
+/* Options.  */
+#define flag_mudflap_threads (flag_mudflap == 2)
+
 /* Helpers.  */
 static tree mf_build_string (const char *string);
 static tree mf_varname_tree (tree);
