diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index dbb8c079..062c4024 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -10687,7 +10687,8 @@ finish_function (int flags)
   /* If this function can't throw any exceptions, remember that.  */
   if (!processing_template_decl
       && !cp_function_chain->can_throw
-      && !flag_non_call_exceptions)
+      && !flag_non_call_exceptions
+      && targetm.binds_local_p (fndecl))
     TREE_NOTHROW (fndecl) = 1;
 
   /* This must come after expand_function_end because cleanups might
diff --git a/gcc/except.c b/gcc/except.c
index 73caf732..7fe5c666 100644
--- a/gcc/except.c
+++ b/gcc/except.c
@@ -2507,6 +2507,9 @@ set_nothrow_function_flags (void)
 {
   rtx insn;
 
+  if (!targetm.binds_local_p (current_function_decl))
+    return;
+
   TREE_NOTHROW (current_function_decl) = 1;
 
   /* Assume cfun->all_throwers_are_sibcalls until we encounter
