diff --git a/gcc/cp/semantics.c b/gcc/cp/semantics.c
index 1885740a..f9f8494c 100644
--- a/gcc/cp/semantics.c
+++ b/gcc/cp/semantics.c
@@ -2255,7 +2255,7 @@ note_decl_for_pch (tree decl)
   if (TREE_CODE (decl) == FUNCTION_DECL
       && TREE_PUBLIC (decl)
       && DECL_DECLARED_INLINE_P (decl)
-      && !DECL_IMPLICIT_INSTANTIATION (decl))
+      && !DECL_TEMPLATE_INFO (decl))
     {
       comdat_linkage (decl);
       DECL_INTERFACE_KNOWN (decl) = 1;
