diff --git a/Makefile.in b/Makefile.in
index 83dd5fc..792caf1 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -150,22 +150,22 @@ local_prefix = /usr/local
 # Directory in which to put host dependent programs and libraries
 exec_prefix = $(prefix)
 # Directory in which to put the executable for the command `gcc'
-bindir = $(exec_prefix)/bin
+bindir = $(exec_prefix)/$(target)/gcc-bin/$(version)
 # Directory in which to put the directories used by the compiler.
 libdir = $(exec_prefix)/lib
 # Directory in which the compiler finds executables, libraries, etc.
 libsubdir = $(libdir)/gcc-lib/$(target)/$(version)
 # Directory to search for site-specific includes.
-includedir = $(local_prefix)/include
+includedir = $(libsubdir)/include
 # assertdir is overridden in cross-make.
 # (But this currently agrees with what is in cross-make.)
 assertdir = $(tooldir)/include
 # where the info files go
-infodir = $(prefix)/info
+infodir = $(prefix)/share/gcc-data/$(target)/$(version)/info
 # Extension (if any) to put in installed man-page filename.
 manext = .1
 # Directory in which to put man pages.
-mandir = $(prefix)/man/man1
+mandir = $(prefix)/share/gcc-data/$(target)/$(version)/man/man1
 # Directory in which to find other cross-compilation tools and headers.
 # Used in install-cross.
 tooldir = $(exec_prefix)/$(target)
