diff --git a/bi-arity.c b/bi-arity.c
index 0b72897..54c12ed 100644
--- a/bi-arity.c
+++ b/bi-arity.c
@@ -56,7 +56,7 @@ main ()
       }
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
@@ -72,7 +72,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-lexer.c b/bi-lexer.c
index 0766b5f..c7b38c8 100644
--- a/bi-lexer.c
+++ b/bi-lexer.c
@@ -38,7 +38,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
@@ -60,7 +60,7 @@ xrealloc (block, nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't reallocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-opcode.c b/bi-opcode.c
index 3320845..0f3f8f1 100644
--- a/bi-opcode.c
+++ b/bi-opcode.c
@@ -51,7 +51,7 @@ enum bytecode_opcode\n{");
     fprintf (stderr, "(Number of opcodes is %d)\n", i);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
@@ -67,7 +67,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-opname.c b/bi-opname.c
index 02dfabc..1b7fadd 100644
--- a/bi-opname.c
+++ b/bi-opname.c
@@ -35,7 +35,7 @@ main()
       printf("\"%s%s\",\n", d->basename, v->name);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
@@ -51,7 +51,7 @@ xmalloc (nbytes)
   if (!tmp)
     {
       fprintf (stderr, "can't allocate %d bytes (out of virtual memory)\n", nbytes);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   return tmp;
diff --git a/bi-parser.c b/bi-parser.c
index e167ce8..b54c08c 100644
--- a/bi-parser.c
+++ b/bi-parser.c
@@ -958,5 +958,5 @@ yyerror (s)
      char *s;
 {
   fprintf (stderr, "syntax error in input\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
diff --git a/bi-parser.y b/bi-parser.y
index 5b3be52..9fd009f 100644
--- a/bi-parser.y
+++ b/bi-parser.y
@@ -164,5 +164,5 @@ yyerror (s)
      char *s;
 {
   fprintf (stderr, "syntax error in input\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
diff --git a/cccp.c b/cccp.c
index 97651a2..7ce9247 100644
--- a/cccp.c
+++ b/cccp.c
@@ -2046,8 +2046,8 @@ main (argc, argv)
     fatal ("I/O error on output");
 
   if (errors)
-    exit (FAILURE_EXIT_CODE);
-  exit (SUCCESS_EXIT_CODE);
+    _exit (FAILURE_EXIT_CODE);
+  _exit (SUCCESS_EXIT_CODE);
 
  perror:
   pfatal_with_name (in_fname);
@@ -6139,7 +6139,7 @@ do_error (buf, limit, op, keyword)
   copy[length] = 0;
   SKIP_WHITE_SPACE (copy);
   error ("#error %s", copy);
-  exit (FAILURE_EXIT_CODE);
+  _exit (FAILURE_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
@@ -9024,7 +9024,7 @@ fatal (str, arg)
   fprintf (stderr, "%s: ", progname);
   fprintf (stderr, str, arg);
   fprintf (stderr, "\n");
-  exit (FAILURE_EXIT_CODE);
+  _exit (FAILURE_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -9051,9 +9051,9 @@ pfatal_with_name (name)
 {
   perror_with_name (name);
 #ifdef VMS
-  exit (vaxc$errno);
+  _exit (vaxc$errno);
 #else
-  exit (FAILURE_EXIT_CODE);
+  _exit (FAILURE_EXIT_CODE);
 #endif
 }
 
diff --git a/collect2.c b/collect2.c
index 4adae08..f7983db 100644
--- a/collect2.c
+++ b/collect2.c
@@ -323,7 +323,7 @@ my_exit (status)
   if (status != 0 && output_file != 0 && output_file[0])
     maybe_unlink (output_file);
 
-  exit (status);
+  _exit (status);
 }
 
 
diff --git a/cp-type2.c b/cp-type2.c
index efdc496..8e22af4 100644
--- a/cp-type2.c
+++ b/cp-type2.c
@@ -323,7 +323,7 @@ my_friendly_abort (i)
       else
 	error ("confused by earlier errors, bailing out");
       
-      exit (34);
+      _exit (34);
     }
   ++abortcount;
 
diff --git a/dbxout.c b/dbxout.c
index 08881fc..a2acf15 100644
--- a/dbxout.c
+++ b/dbxout.c
@@ -354,7 +354,7 @@ abspath (rel_filename)
 
               	  fprintf (stderr, "%s: invalid file name: %s\n",
 			   pname, rel_filename);
-              	  exit (1);
+              	  _exit (1);
               	}
                     *++outp = '\0';
                     continue;
diff --git a/enquire.c b/enquire.c
index 029fbd2..ae6b58e 100644
--- a/enquire.c
+++ b/enquire.c
@@ -598,13 +598,13 @@ int setmode(s) char *s; {
 #endif
 
 Procedure farewell(bugs) int bugs; {
-	if (bugs == 0) exit(0);
+	if (bugs == 0) _exit(0);
 	printf("\n%sFor hints on dealing with the ", co);
 	if (bugs == 1) printf("problem");
 	else printf("%d problems", bugs);
 	printf(" above\n   see the section 'TROUBLESHOOTING' in the file ");
 	printf("%s%s\n", FILENAME, oc);
-	exit(bugs);
+	_exit(bugs);
 }
 
 /* The program has received a signal where it wasn't expecting one */
diff --git a/fix-header.c b/fix-header.c
index b6de982..24617b3 100644
--- a/fix-header.c
+++ b/fix-header.c
@@ -340,7 +340,7 @@ read_scan_file (scan_file)
     {
       if (verbose)
 	fprintf (stderr, "%s: OK, nothing needs to be done.\n", inc_filename);
-      exit (0);
+      _exit (0);
     }
   if (!verbose)
     fprintf (stderr, "%s: fixing %s\n", progname, inc_filename);
@@ -667,7 +667,7 @@ main (argc, argv)
     {
       fprintf (stderr, "%s: Usage: foo.h infile.h outfile.h req_funcs <scan-file-name\n",
 	       progname);
-      exit (-1);
+      _exit (-1);
     }
 
   inc_filename = argv[1];
@@ -713,13 +713,13 @@ main (argc, argv)
       fprintf (stderr, "%s: Cannot open '%s' for reading -",
 	       progname, argv[2]);
       perror (NULL);
-      exit (-1);
+      _exit (-1);
     }
   if (fstat (inf_fd, &sbuf) < 0)
     {
       fprintf (stderr, "%s: Cannot get size of '%s' -", progname, argv[2]);
       perror (NULL);
-      exit (-1);
+      _exit (-1);
     }
   inf_size = sbuf.st_size;
   inf_buffer = (char*) xmalloc (inf_size + 2);
@@ -736,7 +736,7 @@ main (argc, argv)
 	{
 	  fprintf (stderr, "%s: Failed to read '%s' -", progname, argv[2]);
 	  perror (NULL);
-	  exit (-1);
+	  _exit (-1);
 	}
       if (i == 0)
 	{
@@ -759,7 +759,7 @@ main (argc, argv)
       fprintf (stderr, "%s: Cannot open '%s' for writing -",
 	       progname, argv[3]);
       perror (NULL);
-      exit (-1);
+      _exit (-1);
     }
 
   lineno = 1;
diff --git a/g++.c b/g++.c
index 7db2ae2..31a2cf9 100644
--- a/g++.c
+++ b/g++.c
@@ -119,7 +119,7 @@ fatal VPROTO((char *format, ...))
   /* XXX Not needed for g++ driver.  */
   delete_temp_files ();
 #endif
-  exit (1);
+  _exit (1);
 }
 
 static void
@@ -163,7 +163,7 @@ fatal (msg, arg1, arg2)
   /* XXX Not needed for g++ driver.  */
   delete_temp_files ();
 #endif
-  exit (1);
+  _exit (1);
 }
 
 #endif /* not HAVE_VPRINTF */
diff --git a/gcc.c b/gcc.c
index 9b36eb2..f986e93 100644
--- a/gcc.c
+++ b/gcc.c
@@ -1904,7 +1904,7 @@ pexecute (search_flag, program, argv, not_last)
       /* Exec the program.  */
       (*func) (program, argv);
       perror_exec (program);
-      exit (-1);
+      _exit (-1);
       /* NOTREACHED */
       return 0;
 
@@ -2275,12 +2275,12 @@ process_command (argc, argv)
 	  printf ("*predefines:\n%s\n\n", cpp_predefines);
 	  printf ("*cross_compile:\n%d\n\n", cross_compile);
 
-	  exit (0);
+	  _exit (0);
 	}
       else if (! strcmp (argv[i], "-dumpversion"))
 	{
 	  printf ("%s\n", version_string);
-	  exit (0);
+	  _exit (0);
 	}
       else if (! strcmp (argv[i], "-print-libgcc-file-name"))
 	{
@@ -3830,7 +3830,7 @@ main (argc, argv)
   if (print_libgcc_file_name)
     {
       printf ("%s\n", find_file ("libgcc.a"));
-      exit (0);
+      _exit (0);
     }
 
   /* Obey some of the options.  */
@@ -4004,7 +4004,7 @@ main (argc, argv)
     delete_failure_queue ();
   delete_temp_files ();
 
-  exit (error_count > 0 ? (signal_count ? 2 : 1) : 0);
+  _exit (error_count > 0 ? (signal_count ? 2 : 1) : 0);
   /* NOTREACHED */
   return 0;
 }
@@ -4194,7 +4194,7 @@ fatal VPROTO((char *format, ...))
   va_end (ap);
   fprintf (stderr, "\n");
   delete_temp_files ();
-  exit (1);
+  _exit (1);
 }
 
 static void
@@ -4226,7 +4226,7 @@ fatal (msg, arg1, arg2)
 {
   error (msg, arg1, arg2);
   delete_temp_files ();
-  exit (1);
+  _exit (1);
 }
 
 static void
diff --git a/genattr.c b/genattr.c
index 6137851..5714d00 100644
--- a/genattr.c
+++ b/genattr.c
@@ -220,7 +220,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genattr: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -264,7 +264,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -427,7 +427,7 @@ from the machine description file `md'.  */\n\n");
   printf("#define ATTR_FLAG_very_unlikely\t0x20\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genattrtab.c b/genattrtab.c
index b5e1d43..6fb2249 100644
--- a/genattrtab.c
+++ b/genattrtab.c
@@ -5441,7 +5441,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genattrtab: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -5528,7 +5528,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -5674,7 +5674,7 @@ from the machine description file `md'.  */\n\n");
   write_const_num_delay_slots ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/gencodes.c b/gencodes.c
index 255e2aa..0c4cd09 100644
--- a/gencodes.c
+++ b/gencodes.c
@@ -81,7 +81,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "gencodes: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -111,7 +111,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -153,7 +153,7 @@ from the machine description file `md'.  */\n\n");
   printf ("#endif /* MAX_INSN_CODE */\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genconfig.c b/genconfig.c
index da230f1..cda9d35 100644
--- a/genconfig.c
+++ b/genconfig.c
@@ -265,7 +265,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genconfig: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -295,7 +295,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -349,7 +349,7 @@ from the machine description file `md'.  */\n\n");
     printf ("#define HAVE_lo_sum\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genemit.c b/genemit.c
index 6feb972..e7cbb26 100644
--- a/genemit.c
+++ b/genemit.c
@@ -695,7 +695,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genemit: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -725,7 +725,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -792,7 +792,7 @@ from the machine description file `md'.  */\n\n");
   output_init_mov_optab ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genextract.c b/genextract.c
index 3234d4e..c809612 100644
--- a/genextract.c
+++ b/genextract.c
@@ -367,7 +367,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genextract: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -423,7 +423,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -551,7 +551,7 @@ from the machine description file `md'.  */\n\n");
   printf ("    }\n}\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genflags.c b/genflags.c
index 5642e5f..0a1cae6 100644
--- a/genflags.c
+++ b/genflags.c
@@ -196,7 +196,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genflags: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -232,7 +232,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -285,7 +285,7 @@ from the machine description file `md'.  */\n\n");
   printf ("#endif  /* NO_MD_PROTOTYPES */\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genopinit.c b/genopinit.c
index ee65de4..5963f17 100644
--- a/genopinit.c
+++ b/genopinit.c
@@ -307,7 +307,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genopinit: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -339,7 +339,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -376,7 +376,7 @@ from the machine description file `md'.  */\n\n");
   printf ("}\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genoutput.c b/genoutput.c
index eea252a..3ce3114 100644
--- a/genoutput.c
+++ b/genoutput.c
@@ -905,7 +905,7 @@ fatal (s, a1, a2, a3, a4)
   fprintf (stderr, "genoutput: ");
   fprintf (stderr, s, a1, a2, a3, a4);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -946,7 +946,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -980,7 +980,7 @@ main (argc, argv)
   output_epilogue ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 || have_error
+  _exit (ferror (stdout) != 0 || have_error
 	? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
 
   /* NOTREACHED */
diff --git a/genpeep.c b/genpeep.c
index 1e11a68..b7b849a 100644
--- a/genpeep.c
+++ b/genpeep.c
@@ -413,7 +413,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genpeep: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -445,7 +445,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -502,7 +502,7 @@ from the machine description file `md'.  */\n\n");
   printf ("rtx peep_operand[%d];\n", max_opno + 1);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genrecog.c b/genrecog.c
index fb7a3a1..a182929 100644
--- a/genrecog.c
+++ b/genrecog.c
@@ -1682,7 +1682,7 @@ fatal (s)
   fprintf (stderr, s);
   fprintf (stderr, "\n");
   fprintf (stderr, "after %d definitions\n", next_index);
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -1715,7 +1715,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -1798,7 +1798,7 @@ from the machine description file `md'.  */\n\n");
   write_subroutine (split_tree.first, SPLIT);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/getopt.c b/getopt.c
index 7a4673b..7b4a05f 100755
--- a/getopt.c
+++ b/getopt.c
@@ -751,7 +751,7 @@ main (argc, argv)
       printf ("\n");
     }
 
-  exit (0);
+  _exit (0);
 }
 
 #endif /* TEST */
diff --git a/getopt1.c b/getopt1.c
index f784b57..3ee4cb4 100755
--- a/getopt1.c
+++ b/getopt1.c
@@ -181,7 +181,7 @@ main (argc, argv)
       printf ("\n");
     }
 
-  exit (0);
+  _exit (0);
 }
 
 #endif /* TEST */
diff --git a/libgcc2.c b/libgcc2.c
index 6b58e0f..566cd16 100644
--- a/libgcc2.c
+++ b/libgcc2.c
@@ -1801,7 +1801,7 @@ __enable_execute_stack (addr)
   if (r != KERN_SUCCESS)
     {
       mach_error("vm_protect VM_PROT_ALL", r);
-      exit(1);
+      _exit(1);
     }
 
   /* We inline the i-cache invalidation for speed */
diff --git a/objc-act.c b/objc-act.c
index 3431f19..8991d65 100644
--- a/objc-act.c
+++ b/objc-act.c
@@ -2102,7 +2102,7 @@ build_ivar_chain (interface, copy)
 	  error ("Cannot find interface declaration for `%s', superclass of `%s'",
 		 IDENTIFIER_POINTER (super_name),
 		 IDENTIFIER_POINTER (my_name));
-	  exit (34);
+	  _exit (34);
         }
       if (super_interface == interface)
         {
@@ -5588,7 +5588,7 @@ start_class (code, class_name, super_name, protocol_list)
 	{
 	  error ("Cannot find interface declaration for `%s'",
 		 IDENTIFIER_POINTER (class_name));
-	  exit (1);
+	  _exit (1);
 	}
       else
         add_category (class_category_is_assoc_with, class);
@@ -5620,7 +5620,7 @@ start_class (code, class_name, super_name, protocol_list)
         {
 	  error ("Cannot find interface declaration for `%s'",
 		 IDENTIFIER_POINTER (class_name));
-	  exit (1);
+	  _exit (1);
         }
     }
   return class;
diff --git a/protoize.c b/protoize.c
index f3b1307..f653551 100644
--- a/protoize.c
+++ b/protoize.c
@@ -147,7 +147,7 @@ extern int fprintf (FILE *, const char *, ...);
 extern int printf (const char *, ...);
 extern int open (const char *, int, ...);
 #endif /* 0 */
-extern void exit ();
+extern void _exit ();
 extern void free ();
 extern int read ();
 extern int write ();
@@ -615,7 +615,7 @@ xmalloc (byte_count)
   if (rv == NULL)
     {
       fprintf (stderr, "\n%s: virtual memory exceeded\n", pname);
-      exit (1);
+      _exit (1);
       return 0;		/* avoid warnings */
     }
   else
@@ -635,7 +635,7 @@ xrealloc (old_space, byte_count)
   if (rv == NULL)
     {
       fprintf (stderr, "\n%s: virtual memory exceeded\n", pname);
-      exit (1);
+      _exit (1);
       return 0;		/* avoid warnings */
     }
   else
@@ -688,7 +688,7 @@ void
 fancy_abort ()
 {
   fprintf (stderr, "%s: internal abort\n", pname);
-  exit (1);
+  _exit (1);
 }
 
 /* Make a duplicate of the first N bytes of a given string in a newly
@@ -826,7 +826,7 @@ usage ()
   fprintf (stderr, "%s: usage '%s [ -VqfnkNlgC ] [ -B <diname> ] [ filename ... ]'\n",
 	   pname, pname);
 #endif /* !defined (UNPROTOIZE) */
-  exit (1);
+  _exit (1);
 }
 
 /* Return true if the given filename (assumed to be an absolute filename)
@@ -1320,7 +1320,7 @@ abspath (cwd, rel_filename)
 
               	  fprintf (stderr, "%s: invalid file name: %s\n",
 			   pname, rel_filename);
-              	  exit (1);
+              	  _exit (1);
               	}
                     *++outp = '\0';
                     continue;
@@ -1490,7 +1490,7 @@ aux_info_corrupted ()
 {
   fprintf (stderr, "\n%s: fatal error: aux info file corrupted at line %d\n",
 	   pname, current_aux_info_lineno);
-  exit (1);
+  _exit (1);
 }
 
 /* ??? This comment is vague.  Say what the condition is for.  */
@@ -1816,7 +1816,7 @@ save_def_or_dec (l, is_syscalls)
 			 def_dec_p->file->hash_entry->symbol,
 			 def_dec_p->line,
 			 def_dec_p->hash_entry->symbol);
-                exit (1);
+                _exit (1);
               }
             free_def_dec (def_dec_p);
             return;
@@ -4545,7 +4545,7 @@ main (argc, argv)
     {
       fprintf (stderr, "%s: cannot get working directory: %s\n",
 	       pname, sys_errlist[errno]);
-      exit (1);
+      _exit (1);
     }
 
   /* By default, convert the files in the current directory.  */
@@ -4671,8 +4671,8 @@ main (argc, argv)
       do_processing ();
     }
   if (errors)
-    exit (1);
+    _exit (1);
   else
-    exit (0);
+    _exit (0);
   return 1;
 }
diff --git a/toplev.c b/toplev.c
index 7acabf6..5570560 100644
--- a/toplev.c
+++ b/toplev.c
@@ -874,7 +874,7 @@ pfatal_with_name (name)
 {
   fprintf (stderr, "%s: ", progname);
   perror (name);
-  exit (35);
+  _exit (35);
 }
 
 void
@@ -882,7 +882,7 @@ fatal_io_error (name)
      char *name;
 {
   fprintf (stderr, "%s: %s: I/O error\n", progname, name);
-  exit (35);
+  _exit (35);
 }
 
 /* Called to give a better error message when we don't have an insn to match
@@ -1131,7 +1131,7 @@ fatal (s, arglist)
      arglist_dcl
 {
   error (s, arglist);
-  exit (34);
+  _exit (34);
 }
 
 /* Report a warning at line LINE.
@@ -3525,10 +3525,10 @@ You Lose!  You must define PREFERRED_DEBUGGING_TYPE!
 #endif /* not OS2 */
 
   if (errorcount)
-    exit (FATAL_EXIT_CODE);
+    _exit (FATAL_EXIT_CODE);
   if (sorrycount)
-    exit (FATAL_EXIT_CODE);
-  exit (SUCCESS_EXIT_CODE);
+    _exit (FATAL_EXIT_CODE);
+  _exit (SUCCESS_EXIT_CODE);
   return 34;
 }
 
