diff --git a/src/stdio_streambuf.cpp b/src/stdio_streambuf.cpp
index 45aad79..abcf967 100644
--- a/src/stdio_streambuf.cpp
+++ b/src/stdio_streambuf.cpp
@@ -138,7 +138,8 @@ stdio_streambuf_base::seekpos(pos_type pos, ios_base::openmode /* mode */) {
   FPOS_T p(pos);
 #endif
 
-  return FSETPOS(_M_file, &p) == 0 ? pos : pos_type(-1);
+  int r = FSETPOS(_M_file, &p);
+  return (r == 0) ? pos : pos_type(-1);
 }
 
 int stdio_streambuf_base::sync() {
