diff --git a/Makefile.in b/Makefile.in
index 4c4b04b..9c12182 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -241,7 +241,7 @@ EXTRA_HEADERS =
 
 # Target to use whe installing assert.h.  Some systems may
 # want to set this empty.
-INSTALL_ASSERT_H = install-assert-h
+INSTALL_ASSERT_H =
 
 # Set this to `ld' to enable use of collect2.
 # USE_COLLECT2 =
diff --git a/g++.c b/g++.c
index 2329701..8529c3c 100644
--- a/g++.c
+++ b/g++.c
@@ -58,7 +58,7 @@ the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */
    Makefile should compile this file with the proper name
    (e.g., "i386-aout-gcc").  */
 #ifndef GCC_NAME
-#define GCC_NAME "i586-legacy-linux-gnulibc1-gcc-2.5.8"
+#define GCC_NAME "i486-legacy-linuxaout-gcc-2.5.8"
 #endif
 
 /* This bit is set if we saw a `-xfoo' language specification.  */
