diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index c19d690ea..902837f79 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -218,6 +218,11 @@ libgcov-merge-tool.o-warn = -Wno-error
 gimple-match.o-warn = -Wno-unused
 generic-match.o-warn = -Wno-unused
 dfp.o-warn = -Wno-strict-aliasing
+build/genhooks.o-warn = -Wno-error=unused-result
+gcc.o-warn = -Wno-error=unused-result
+ipa-inline.o-warn = -Wno-error=strict-overflow
+lto-wrapper.o-warn = -Wno-error=unused-result
+lto/lto.o-warn = -Wno-error=unused-result
 
 # All warnings have to be shut off in stage1 if the compiler used then
 # isn't gcc; configure determines that.  WARN_CFLAGS will be either
