diff --git a/gcc/combine.c b/gcc/combine.c
index 0015c1a8c..be5b67306 100644
--- a/gcc/combine.c
+++ b/gcc/combine.c
@@ -10029,7 +10029,8 @@ reg_nonzero_bits_for_combine (const_rtx x, machine_mode mode,
   rsp = &reg_stat[REGNO (x)];
   if (rsp->last_set_value != 0
       && (rsp->last_set_mode == mode
-	  || (GET_MODE_CLASS (rsp->last_set_mode) == MODE_INT
+	  || (REGNO (x) >= FIRST_PSEUDO_REGISTER
+	      && GET_MODE_CLASS (rsp->last_set_mode) == MODE_INT
 	      && GET_MODE_CLASS (mode) == MODE_INT))
       && ((rsp->last_set_label >= label_tick_ebb_start
 	   && rsp->last_set_label < label_tick)
