diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index bffb7721b..f6c4fa267 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -15219,6 +15219,7 @@ tsubst_copy (tree t, tree args, tsubst_flags_t complain, tree in_decl)
     case REAL_CST:
     case STRING_CST:
     case COMPLEX_CST:
+    case VECTOR_CST:
       {
 	/* Instantiate any typedefs in the type.  */
 	tree type = tsubst (TREE_TYPE (t), args, complain, in_decl);
diff --git a/gcc/testsuite/g++.dg/pr103326.C b/gcc/testsuite/g++.dg/pr103326.C
new file mode 100644
index 000000000..260e7da86
--- /dev/null
+++ b/gcc/testsuite/g++.dg/pr103326.C
@@ -0,0 +1,15 @@
+// { dg-do compile }
+// { dg-require-effective-target c++11 }
+
+using x86_64_v16qi [[gnu::__vector_size__ (16)]] = char;
+
+template<typename T>
+void foo()
+{
+  constexpr x86_64_v16qi zero{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+}
+
+void foo2()
+{
+  foo<int>();
+}
