diff --git a/gcc/cp/call.c b/gcc/cp/call.c
index cace30f25..7ad4c787f 100644
--- a/gcc/cp/call.c
+++ b/gcc/cp/call.c
@@ -4809,7 +4809,7 @@ build_conditional_expr_1 (location_t loc, tree arg1, tree arg2, tree arg3,
       if (lvalue_p (arg1))
 	arg2 = arg1 = cp_stabilize_reference (arg1);
       else
-	arg2 = arg1 = save_expr (arg1);
+	arg2 = arg1 = cp_save_expr (arg1);
     }
 
   /* If something has already gone wrong, just pass that fact up the
diff --git a/gcc/testsuite/g++.dg/template/sizeof16.C b/gcc/testsuite/g++.dg/template/sizeof16.C
new file mode 100644
index 000000000..084bb2185
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/sizeof16.C
@@ -0,0 +1,7 @@
+// PR c++/84333
+// { dg-options -Wno-pedantic }
+
+template<typename> int foo()
+{
+  return sizeof(int) > 1 ? : 1;
+}
diff --git a/gcc/testsuite/g++.dg/template/sizeof17.C b/gcc/testsuite/g++.dg/template/sizeof17.C
new file mode 100644
index 000000000..2d5892d2c
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/sizeof17.C
@@ -0,0 +1,7 @@
+// PR c++/84333
+// { dg-options -Wno-pedantic }
+
+template<typename T> int foo()
+{
+  return sizeof(T) > 1 ? : 1;
+}
