diff --git a/gcc/cp/cp-tree.h b/gcc/cp/cp-tree.h
index 905a97ca9..f6f4fdfa1 100644
--- a/gcc/cp/cp-tree.h
+++ b/gcc/cp/cp-tree.h
@@ -4472,7 +4472,8 @@ more_aggr_init_expr_args_p (const aggr_init_expr_arg_iterator *iter)
    entity with its own template parameter list, and which is not a
    full specialization.  */
 #define PROCESSING_REAL_TEMPLATE_DECL_P() \
-  (processing_template_decl > template_class_depth (current_scope ()))
+  (!processing_template_parmlist \
+   && processing_template_decl > template_class_depth (current_scope ()))
 
 /* Nonzero if this VAR_DECL or FUNCTION_DECL has already been
    instantiated, i.e. its definition has been generated from the
diff --git a/gcc/testsuite/g++.dg/template/error58.C b/gcc/testsuite/g++.dg/template/error58.C
new file mode 100644
index 000000000..cede1c948
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/error58.C
@@ -0,0 +1,8 @@
+// PR c++/85242
+
+namespace N
+{
+  struct A {};
+}
+
+template<struct N::A {}> void foo(); // { dg-error "" }
