diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index bc47cf46a..da3dc2417 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -12215,7 +12215,7 @@ grokdeclarator (const cp_declarator *declarator,
 	      {
 		error ("%qE is neither function nor member function; "
 		       "cannot be declared friend", unqualified_id);
-		friendp = 0;
+		return error_mark_node;
 	      }
 	    decl = NULL_TREE;
 	  }
diff --git a/gcc/testsuite/g++.dg/cpp0x/auto50.C b/gcc/testsuite/g++.dg/cpp0x/auto50.C
new file mode 100644
index 000000000..814892d2a
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/auto50.C
@@ -0,0 +1,7 @@
+// PR c++/84348
+// { dg-do compile { target c++11 } }
+
+template<typename> struct A
+{
+  friend auto foo;  // { dg-error "cannot be declared friend" }
+};
diff --git a/gcc/testsuite/g++.dg/parse/friend12.C b/gcc/testsuite/g++.dg/parse/friend12.C
index 84d6e25e7..d4e0ceea9 100644
--- a/gcc/testsuite/g++.dg/parse/friend12.C
+++ b/gcc/testsuite/g++.dg/parse/friend12.C
@@ -3,5 +3,4 @@
 struct A
 {
   friend int i = 0;  // { dg-error "cannot be declared friend" }
-// { dg-error "non-static data member" "" { target { ! c++11 } } .-1 }
 };
