diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 9fac33fc2..49580fd5d 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -14001,7 +14001,10 @@ xref_basetypes (tree ref, tree base_list)
       CLASSTYPE_NON_LAYOUT_POD_P (ref) = true;
 
       if (TREE_CODE (ref) == UNION_TYPE)
-	error ("derived union %qT invalid", ref);
+	{
+	  error ("derived union %qT invalid", ref);
+	  return;
+	}
     }
 
   if (max_bases > 1)
diff --git a/gcc/testsuite/g++.dg/inherit/union3.C b/gcc/testsuite/g++.dg/inherit/union3.C
new file mode 100644
index 000000000..c13b39768
--- /dev/null
+++ b/gcc/testsuite/g++.dg/inherit/union3.C
@@ -0,0 +1,16 @@
+// PR c++/81013
+
+struct A
+{
+  virtual void foo() const;
+};
+
+union B : A  // { dg-error "derived union 'B' invalid" }
+{
+  void foo() const;
+};
+
+void bar(const B& b)
+{
+  b.foo();
+}
