diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index c7b1d3e04..a267ececf 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -1569,6 +1569,33 @@ duplicate_decls (tree newdecl, tree olddecl, bool newdecl_is_friend)
 		   || compparms (TYPE_ARG_TYPES (TREE_TYPE (newdecl)),
 				 TYPE_ARG_TYPES (TREE_TYPE (olddecl))))
 	    {
+	      /* Don't really override olddecl for __* prefixed builtins
+		 except for __[^b]*_chk, the compiler might be using those
+		 explicitly.  */
+	      if (DECL_BUILT_IN (olddecl))
+		{
+		  tree id = DECL_NAME (olddecl);
+		  const char *name = IDENTIFIER_POINTER (id);
+		  size_t len;
+
+		  if (name[0] == '_'
+		      && name[1] == '_'
+		      && (strncmp (name + 2, "builtin_",
+				   strlen ("builtin_")) == 0
+			  || (len = strlen (name)) <= strlen ("___chk")
+			  || memcmp (name + len - strlen ("_chk"),
+				     "_chk", strlen ("_chk") + 1) != 0))
+		    {
+		      if (permerror (DECL_SOURCE_LOCATION (newdecl),
+				     "new declaration %q#D ambiguates built-in"
+				     " declaration %q#D", newdecl, olddecl)
+			  && flag_permissive)
+			inform (DECL_SOURCE_LOCATION (newdecl),
+				"ignoring the %q#D declaration", newdecl);
+		      return olddecl;
+		    }
+		}
+
 	      /* A near match; override the builtin.  */
 
 	      if (TREE_PUBLIC (newdecl))
diff --git a/gcc/testsuite/g++.dg/ext/pr84724-1.C b/gcc/testsuite/g++.dg/ext/pr84724-1.C
new file mode 100644
index 000000000..9f03a292e
--- /dev/null
+++ b/gcc/testsuite/g++.dg/ext/pr84724-1.C
@@ -0,0 +1,14 @@
+// PR c++/84724
+// { dg-do compile }
+// { dg-options "-O3 -fpermissive" }
+
+int __builtin_trap ();		// { dg-warning "ambiguates built-in declaration" }
+				// { dg-message "ignoring the 'int __builtin_trap\\(\\)' declaration" "" { target *-*-* } .-1 }
+
+int
+foo ()
+{
+  int b;
+  int c (&b);			// { dg-warning "invalid conversion from" }
+  return b %= b ? c : 0;
+}
diff --git a/gcc/testsuite/g++.dg/ext/pr84724-2.C b/gcc/testsuite/g++.dg/ext/pr84724-2.C
new file mode 100644
index 000000000..de6942506
--- /dev/null
+++ b/gcc/testsuite/g++.dg/ext/pr84724-2.C
@@ -0,0 +1,14 @@
+// PR c++/84724
+// { dg-do compile }
+// { dg-options "-O3 -fpermissive -w" }
+
+int __builtin_trap ();		// { dg-bogus "ambiguates built-in declaration" }
+				// { dg-bogus "ignoring the 'int __builtin_trap\\(\\)' declaration" "" { target *-*-* } .-1 }
+
+int
+foo ()
+{
+  int b;
+  int c (&b);			// { dg-bogus "invalid conversion from" }
+  return b %= b ? c : 0;
+}
diff --git a/gcc/testsuite/g++.dg/ext/pr84724-3.C b/gcc/testsuite/g++.dg/ext/pr84724-3.C
new file mode 100644
index 000000000..0cd8ccef4
--- /dev/null
+++ b/gcc/testsuite/g++.dg/ext/pr84724-3.C
@@ -0,0 +1,5 @@
+// PR c++/84724
+// { dg-do compile }
+// { dg-options "" }
+
+int __builtin_trap ();		// { dg-error "ambiguates built-in declaration" }
