diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index a267ececf..9c2b6524b 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -3003,6 +3003,7 @@ redeclaration_error_message (tree newdecl, tree olddecl)
 	 specifier is declaration rather than definition (and
 	 deprecated).  */
       if (cxx_dialect >= cxx1z
+	  && VAR_P (olddecl)
 	  && DECL_CLASS_SCOPE_P (olddecl)
 	  && DECL_DECLARED_CONSTEXPR_P (olddecl)
 	  && !DECL_INITIAL (newdecl))
diff --git a/gcc/testsuite/g++.dg/cpp1z/pr84533.C b/gcc/testsuite/g++.dg/cpp1z/pr84533.C
new file mode 100644
index 000000000..0a71fe1b7
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp1z/pr84533.C
@@ -0,0 +1,3 @@
+// { dg-options "-std=c++17" }
+
+enum E { e, e };  // { dg-error "redefinition" }
