diff --git a/Makefile.in b/Makefile.in
index a8ebcb5..7c0fb39 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -1027,7 +1027,7 @@ c-common.o : c-common.c $(CONFIG_H) $(TREE_H) c-tree.h c-lex.h flags.h
 
 gcc.o: gcc.c $(CONFIG_H) multilib.h config.status
 	$(CC) $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
-  -DSTANDARD_STARTFILE_PREFIX=\"$(libdir)/\" \
+  -DSTANDARD_STARTFILE_PREFIX=\"$(tooldir)/lib/\" \
   -DSTANDARD_EXEC_PREFIX=\"$(libdir)/gcc-lib/\" \
   -DDEFAULT_TARGET_VERSION=\"$(version)\" \
   -DDEFAULT_TARGET_MACHINE=\"$(target)\" \
diff --git a/cccp.c b/cccp.c
index e78a764..b1bd6ee 100644
--- a/cccp.c
+++ b/cccp.c
@@ -39,10 +39,6 @@ typedef unsigned char U_CHAR;
 #include "config.h"
 #endif /* not EMACS */
 
-#ifndef STANDARD_INCLUDE_DIR
-#define STANDARD_INCLUDE_DIR "/usr/include"
-#endif
-
 #ifndef LOCAL_INCLUDE_DIR
 #define LOCAL_INCLUDE_DIR "/usr/local/include"
 #endif
@@ -622,7 +618,6 @@ static struct default_include {
 #ifdef SYSTEM_INCLUDE_DIR
     { SYSTEM_INCLUDE_DIR, 0, 0 },
 #endif
-    { STANDARD_INCLUDE_DIR, 0, 0 },
 #endif /* not CROSS_COMPILE */
     { 0, 0, 0 }
     };
diff --git a/cp/g++.c b/cp/g++.c
index bc17020..0153eac 100644
--- a/cp/g++.c
+++ b/cp/g++.c
@@ -46,7 +46,7 @@ the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */
    Makefile should compile this file with the proper name
    (e.g., "i386-aout-gcc").  */
 #ifndef GCC_NAME
-#define GCC_NAME "gcc-2.6.3"
+#define GCC_NAME "i586-legacy-linux-gnulibc1-gcc-2.6.3"
 #endif
 
 /* This bit is set if we saw a `-xfoo' language specification.  */
diff --git a/protoize.c b/protoize.c
index f4c92f3..9029715 100644
--- a/protoize.c
+++ b/protoize.c
@@ -270,13 +270,6 @@ static const unexpansion unexpansions[] = {
 
 static const int hash_mask = (HASH_TABLE_SIZE - 1);
 
-/* Make a table of default system include directories
-   just as it is done in cccp.c.  */
-
-#ifndef STANDARD_INCLUDE_DIR
-#define STANDARD_INCLUDE_DIR "/usr/include"
-#endif
-
 #ifndef LOCAL_INCLUDE_DIR
 #define LOCAL_INCLUDE_DIR "/usr/local/include"
 #endif
@@ -311,7 +304,6 @@ struct default_include { const char *fname; int cplusplus; } include_defaults[]
 #ifdef SYSTEM_INCLUDE_DIR
     { SYSTEM_INCLUDE_DIR, 0},
 #endif
-    { STANDARD_INCLUDE_DIR, 0},
 #endif /* not CROSS_COMPILE */
     { 0, 0}
     };
