diff --git a/Makefile.in b/Makefile.in
index 1d2f2d6..07f7ce1 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -73,7 +73,7 @@ MAKEINFO = `if [ -f $${rootme}/texinfo/C/makeinfo ] ; \
 # C compiler to use to create programs which must be run in the build
 # environment.
 CC_FOR_BUILD = $(CC)
-CXX_FOR_BUILD = gcc
+CXX_FOR_BUILD = $(CXX)
 
 SUBDIRS = libiberty mmalloc glob readline opcodes bfd z8ksim gdb binutils ld gas tgas gcc libg++ newlib deja-gnu
 OTHERS = 
@@ -84,72 +84,6 @@ INSTALL_TARGET = install.all
 ### for debugging
 #GCCVERBOSE=-v
 
-CC_FOR_TARGET = ` \
-  if [ -f $${rootme}/gcc/Makefile ] ; then \
-    echo $${rootme}/gcc/gcc -B$${rootme}/gcc/; \
-  else \
-    if [ "$(host_canonical)" = "$(target_canonical)" ] ; then \
-      echo $(CC); \
-    else \
-      t='$(program_transform_name)'; echo gcc | sed -e '' $$t; \
-    fi; \
-  fi`
-
-CXX_FOR_TARGET = ` \
-  if [ -f $${rootme}/gcc/Makefile ] ; then \
-    echo $${rootme}/gcc/gcc -B$${rootme}/gcc/; \
-  else \
-    if [ "$(host_cpu)-$(host_vendor)-$(host_os)" = "$(target_cpu)-$(target_vendor)-$(target_os)" ] ; then \
-      echo gcc; \
-    else \
-      t='$(program_transform_name)'; echo gcc | sed -e '' $$t; \
-    fi; \
-  fi`
-
-AS_FOR_TARGET = ` \
-  if [ -f $${rootme}/gas/Makefile ] ; then \
-    echo $${rootme}/gas/as.new ; \
-  else \
-    if [ "$(host_cpu)-$(host_vendor)-$(host_os)" = "$(target_cpu)-$(target_vendor)-$(target_os)" ] ; then \
-      echo $(AS); \
-    else \
-       t='$(program_transform_name)'; echo as | sed -e '' $$t ; \
-    fi \
-  fi`
-
-AR_FOR_TARGET = ` \
-  if [ -f $${rootme}/binutils/Makefile ] ; then \
-    echo $${rootme}/binutils/ar ; \
-  else \
-    if [ "$(host_canonical)" = "$(target_canonical)" ] ; then \
-      echo $(AR); \
-    else \
-       t='$(program_transform_name)'; echo ar | sed -e '' $$t ; \
-    fi \
-  fi`
-
-RANLIB_FOR_TARGET = ` \
-  if [ -f $${rootme}/binutils/Makefile ] ; then \
-    echo $${rootme}/binutils/ranlib ; \
-  else \
-    if [ "$(host_canonical)" = "$(target_canonical)" ] ; then \
-      echo $(RANLIB); \
-    else \
-       t='$(program_transform_name)'; echo ranlib | sed -e '' $$t ; \
-    fi \
-  fi`
-
-NM_FOR_TARGET = ` \
-  if [ -f $${rootme}/binutils/Makefile ] ; then \
-    echo $${rootme}/binutils/nm ; \
-  else \
-    if [ "$(host_canonical)" = "$(target_canonical)" ] ; then \
-      echo $(NM); \
-    else \
-       t='$(program_transform_name)'; echo nm | sed -e '' $$t ; \
-    fi \
-  fi`
-
 XTRAFLAGS = ` \
   if [ -f $${rootme}/gcc/Makefile ] ; then \
     if [ -f $${rootme}/newlib/Makefile ] ; then \
@@ -195,16 +129,17 @@ X11_FLAGS_TO_PASS = \
 # Flags to pass down to makes which are built with the target
 # environment (e.g. libg++, xiberty, newlib).  -- keep these in alpha order please
 TARGET_FLAGS_TO_PASS = \
-	"AR=$(AR_FOR_TARGET)" \
+	"AR=$(AR)" \
 	"AR_FLAGS=$(AR_FLAGS)" \
-	"AS=$(AS_FOR_TARGET)" \
+	"AS=$(AS)" \
 	"BISON=$(BISON)" \
-	"CXX=$(CXX_FOR_TARGET)" \
+	"CXX=$(CXX)" \
 	"CXX_FOR_BUILD=$(CXX_FOR_BUILD)" \
 	"CXXFLAGS=$(CXXFLAGS)" \
-	"CC=$(CC_FOR_TARGET)" \
+	"CC=$(CC)" \
 	"CC_FOR_BUILD=$(CC_FOR_BUILD)" \
 	"CFLAGS=$(CFLAGS)" \
+	"DESTDIR=$(DESTDIR)" \
 	"INSTALL=$(INSTALL)" \
 	"INSTALL_DATA=$(INSTALL_DATA)" \
 	"INSTALL_PROGRAM=$(INSTALL_PROGRAM)" \
@@ -212,8 +147,8 @@ TARGET_FLAGS_TO_PASS = \
 	"LEX=$(LEX)" \
 	"LOADLIBES=$(LOADLIBES)" \
 	"MAKEINFO=$(MAKEINFO)" \
-	"NM=$(NM_FOR_TARGET)" \
-	"RANLIB=$(RANLIB_FOR_TARGET)" \
+	"NM=$(NM)" \
+	"RANLIB=$(RANLIB)" \
 	"XTRAFLAGS=$(XTRAFLAGS)" \
 	"exec_prefix=$(exec_prefix)" \
 	"prefix=$(prefix)" \
@@ -235,17 +170,17 @@ clean-info:
 	@rootme='pwd' ; export rootme ; $(MAKE) $(FLAGS_TO_PASS) DO=clean-info "DODIRS=$(SUBDIRS)" subdir_do
 
 install-info: install-info-dirs force
-	-parent=`echo $(infodir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(infodir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(infodir) ] ; then true ; else mkdir $(infodir) ; fi
+	-if [ -d $(DESTDIR)$(infodir) ] ; then true ; else mkdir $(infodir) ; fi
 	@rootme='pwd' ; export rootme ; $(MAKE) $(FLAGS_TO_PASS) DO=install-info "DODIRS=$(SUBDIRS)" subdir_do
 	@rootme='pwd' ; export rootme ; $(MAKE) $(FLAGS_TO_PASS) dir.info install-dir.info
 
 install-dir.info:
-	-parent=`echo $(infodir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(infodir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(infodir) ] ; then true ; else mkdir $(infodir) ; fi
-	$(INSTALL_DATA) dir.info $(infodir)/dir.info
+	-if [ -d $(DESTDIR)$(infodir) ] ; then true ; else mkdir $(DESTDIR)$(infodir) ; fi
+	$(INSTALL_DATA) dir.info $(DESTDIR)$(infodir)/dir.info
 
 all.normal: all-m4 all-autoconf all-libiberty all-mmalloc all-texinfo \
 	all-byacc all-flex all-opcodes all-bfd all-ld all-gas all-tgas \
@@ -279,9 +214,9 @@ clean-stamps:
 	-rm -f all-*
 
 install: $(INSTALL_TARGET) 
-	-parent=`echo $(man1dir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(man1dir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(man1dir) ] ; then true ; else mkdir $(man1dir) ; fi
+	-if [ -d $(DESTDIR)$(man1dir) ] ; then true ; else mkdir $(DESTDIR)$(man1dir) ; fi
 
 install.all: install-no-fixedincludes
 	@if [ -f ./gcc/Makefile ] ; then \
@@ -1603,9 +1538,9 @@ subdir_do:
 # causing Make to quit.
 
 MAKEDIRS= \
-	$(prefix) \
-	$(exec_prefix) \
-	$(tooldir)
+	$(DESTDIR)$(prefix) \
+	$(DESTDIR)$(exec_prefix) \
+	$(DESTDIR)$(tooldir)
 
 #	$(bindir) \
 #	$(libdir) \
diff --git a/etc/Makefile.in b/etc/Makefile.in
index 2de5156..e5b4fb6 100644
--- a/etc/Makefile.in
+++ b/etc/Makefile.in
@@ -41,19 +41,19 @@ INFOFILES = configure.info intro.info standards.info cfg-paper.info
 all:
 
 install:  $(srcdir)/configure.man
-	-parent=`echo $(man1dir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(man1dir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(man1dir) ] ; then true ; else mkdir $(man1dir) ; fi
-	$(INSTALL_DATA) $(srcdir)/configure.man $(man1dir)/configure.1
+	-if [ -d $(DESTDIR)$(man1dir) ] ; then true ; else mkdir $(DESTDIR)$(man1dir) ; fi
+	$(INSTALL_DATA) $(srcdir)/configure.man $(DESTDIR)$(man1dir)/configure.1
 
 info: $(INFOFILES)
 
 install-info:
-	-parent=`echo $(infodir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(infodir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(infodir) ] ; then true ; else mkdir $(infodir) ; fi
+	-if [ -d $(DESTDIR)$(infodir) ] ; then true ; else mkdir $(DESTDIR)$(infodir) ; fi
 	for i in *.info* ; do \
-	  $(INSTALL) $$i $(infodir) ; \
+	  $(INSTALL) $$i $(DESTDIR)$(infodir) ; \
 	done
 
 configure.info: configure.texi
diff --git a/libg++/Makefile.in b/libg++/Makefile.in
index dcf49b1..1b8c3c2 100644
--- a/libg++/Makefile.in
+++ b/libg++/Makefile.in
@@ -282,33 +282,23 @@ libg++.ps: libg++.dvi
 # FIXME - libg++.a, _G_config.h, genclass are all target-dependent,
 # and shoudl be installed in a target-dependent directory.
 install: all
-	-parent=`echo $(INSTALLDIR)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(INSTALLDIR)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(INSTALLDIR) ] ; then true ; else mkdir $(INSTALLDIR) ; fi
-	$(INSTALL_DATA) $(TARGETLIB) $(INSTALLDIR)/$(TARGETLIB)
-	$(RANLIB) $(INSTALLDIR)/$(TARGETLIB)
+	-if [ -d $(DESTDIR)$(INSTALLDIR) ] ; then true ; else mkdir $(DESTDIR)$(INSTALLDIR) ; fi
+	$(INSTALL_DATA) $(TARGETLIB) $(DESTDIR)$(INSTALLDIR)/$(TARGETLIB)
+	$(RANLIB) $(DESTDIR)$(INSTALLDIR)/$(TARGETLIB)
 	-if [ -d $(prefix) ] ; then true ; else mkdir $(prefix) ; fi
-	-parent=`echo $(gxx_includedir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(gxx_includedir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(gxx_includedir) ] ; then true ; else mkdir $(gxx_includedir) ; fi
-	-if [ -d $(gxx_includedir)/sys ] ; then true ; else mkdir $(gxx_includedir)/sys ; fi
-	-if [ -d $(gxx_includedir)/gen ] ; then true ; else mkdir $(gxx_includedir)/gen ; fi
-	rm -f $(gxx_includedir)/_G_config.h ; \
-	  $(INSTALL_DATA) _G_config.h $(gxx_includedir)/_G_config.h || exit 1;
+	-if [ -d $(DESTDIR)$(gxx_includedir) ] ; then true ; else mkdir $(DESTDIR)$(gxx_includedir) ; fi
+	-if [ -d $(DESTDIR)$(gxx_includedir)/sys ] ; then true ; else mkdir $(DESTDIR)$(gxx_includedir)/sys ; fi
+	-if [ -d $(DESTDIR)$(gxx_includedir)/gen ] ; then true ; else mkdir $(DESTDIR)$(gxx_includedir)/gen ; fi
 	cd $(srcdir)/src; \
 	for FILE in *.h gen/*.ccP gen/*.hP; do \
-		rm -f $(gxx_includedir)/$$FILE ; \
-		$(INSTALL_DATA) $$FILE $(gxx_includedir)/$$FILE || exit 1; \
-		chmod a-x $(gxx_includedir)/$$FILE ; \
+		rm -f $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+		$(INSTALL_DATA) $$FILE $(DESTDIR)$(gxx_includedir)/$$FILE || exit 1; \
+		chmod a-x $(DESTDIR)$(gxx_includedir)/$$FILE ; \
 	done
-	if [ "$(HAVE_CPLUS_EXTERN)" = "0" ]; then \
-	    cd $(srcdir)/g++-include; \
-	    for FILE in *.h sys/*.h; do \
-		rm -f $(gxx_includedir)/$$FILE ; \
-		$(INSTALL_DATA) $$FILE $(gxx_includedir)/$$FILE || exit 1; \
-		chmod a-x $(gxx_includedir)/$$FILE ; \
-	    done; \
-	fi
 	cd $(IO_DIR) ; $(MAKE) $(FLAGS_TO_PASS) "gxx_includedir=$(gxx_includedir)" install-include-files
 	if [ "x$(GPERF)" != "x" ]; then \
 	  cd gperf ; $(MAKE) $(FLAGS_TO_PASS) install; \
@@ -318,8 +308,6 @@ install: all
 			(cd $$D; $(MAKE) $(FLAGS_TO_PASS) "gxx_includedir=$(gxx_includedir)" install) ; \
 		fi ; \
 	done
-	cd test-install; $(MAKE) $(TEST_INSTALL) \
-	    CXX=$(bindir)/gcc COMPILE_FLAGS="" LIBS=-lg++
 
 ###**######################################################################
 #
@@ -402,12 +390,6 @@ Makefile: $(srcdir)/Makefile.in $(host_makefile_frag) $(target_makefile_frag)
 	  CC="$(CC)"; export CC; CXX="$(CXX)"; export CXX; \
 	  $(SHELL) ./config.status
 
-_G_config.h: ${srcdir}/utils/gen-params
-	rootme=`pwd`/ ; export rootme; \
-	CC="$(CXX) $(XTRAFLAGS)"; export CC; \
-	CXX="$(CXX) $(XTRAFLAGS) $(NOSTDINC)"; export CXX; \
-	CONFIG_NM="$(NM)"; export CONFIG_NM; \
-	${srcdir}/utils/gen-params LIB_VERSION=$(LIBG++_DIST_VERSION) $(G_CONFIG_ARGS) >tmp-params.h
-	mv tmp-params.h _G_config.h
+_G_config.h:
 
 .NOEXPORT:
diff --git a/libg++/g++-include/ChangeLog b/libg++/g++-include/ChangeLog
deleted file mode 100644
index 19446cf..0000000
--- a/libg++/g++-include/ChangeLog
+++ /dev/null
@@ -1,348 +0,0 @@
-Mon May  3 14:27:47 1993  Per Bothner  (bothner@cygnus.com)
-
-	* time.h:  Fix protection against multiple inclusion.
-	* sys/signal.h:  Check if included implicitly by
-	/usr/include/signal.h (as under Irix 4.0.5).
-	* stdlib.h (bsearch):  Fix return value.
-	* stdlib.h (getopt):  Protect with _G_ARGS (needed for Irix).
-
-Tue Nov 17 21:43:19 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* curses.h:  Added patch for HPUX.
-	* grp.h, memory.h, unistd.h: Added patches for OSF1.
-	* math.h:  Fix definitions of isnan() and isinf() ANSI
-	constants; #include <float.h> to support this.
-
-Thu Nov 12 15:49:17 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* sys/select.h:  New file.
-
-Thu Sep 10 22:46:37 1992  Ian Lance Taylor  (ian@cygnus.com)
-
-	* curses.h: only include the system <curses.h> if the new
-	configuration flag _G_HAVE_CURSES is true.
-
-Wed Aug 12 00:43:48 1992  Per Bothner  (bothner@cygnus.com)
-
-	* sys/dir.h:  As is ./direct.h, remove definition for rewinddir
-	(because it it often macro, which causes problems).
-	* sys/stat.h:  Define S_ISLNK if needed and defined(S_IFLNK).
-	* unistd.h (execv*), signal.h (psignal):  Fix prototypes.
-
-Fri Jun 26 11:34:02 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* curses.h:  Kludges to handle both VMS and SVR4 ...
-
-Tue Jun 23 23:53:15 1992  Per Bothner  (bothner@cygnus.com)
-
-	* direct.h, sys/dir.h:  Kludge to deal with AIX brain-damage.
-
-Fri Jun 19 23:12:15 1992  Per Bothner  (bothner@cygnus.com)
-
-	* unistd.h:  Change return types of read() and write()
-	to _G_ssize_t, following latest Posix standard.
-
-Thu Jun 18 16:22:12 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* curses.h:  More SVR4 kludges (from Fred Fish).
-
-Wed Jun 17 18:40:57 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* curses.h:  Kludges for SVR4 portability.
-	* stdlib.h:  Fix getopt() prototype.
-	* string.h:  Add a comment pointing out the conflicts
-	here are probably due to bad SIZE_TYPE in gcc.
-	* time.h, sys/socket.h:  Protect various functions from
-	prototype mismatch with _G_ARGS.  ALso, fix some typings.
-	* unistd.h:  Lots of little "fixes" for improved portability.
-
-Wed Jun  3 23:20:21 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* sys/wait.h:  If !_G_HAVE_SYS_WAIT, define the Posix
-	macros assuming "traditional" wait() structure.
-	* unistd.h:  Use Posix types where appropriate.
-
-Fri May 29 12:46:57 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* stdio.h:  Typo: _G_sprint_return_type -> _G_sprintf_return_type.
-	* unistd.h:  Hacks so it works on AIX (and elsewhere?).
-
-	* sys/wait.h:  Use _G_pid_t where appropriate.
-	* unistd.h:  Only define F_OK etc if not already defined.
-	* unistd.h:  (Ty) using /usr/include/unistd.h, if available.
-	* dirent.h:  If !_G_HAVE_DIRENT, simulatate missing dirent.h
-	using <sys/dir.h>.
-	* dirent.h:  Don't bother declaring rewinddir - too much of
-	a pain, since it is commonly a macro.
-	* signal.h:  ssignal functions alwasy return int,
-	* stdio.h:  Use new _G_sprint_return_type macro.
-	Use _G_va_list instead of inacurate ... (but protect from
-	clashes using _G_ARGS).
-	* time.h:  Remove use of _G_const.  Use _G_ARGS insted.
-
-Thu May 14 13:24:20 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* sys/socket.h:  Tweak sendto() prototype.
-	* sys/stat.h:  Define Posix macros S_ISDIR etc if they're not defined.
-	* assert.h:  Include stdio.h for the sake of non-ANSI systems
-	(e.g. BSD) whose /usr/include/assert.h need stdio.h.
-	* signal.h:  First arg of kill() is a pid_t.
-	* stdio.h:  setbuf() and setbuffer() return void.
-	Also, add more __hide definitions.
-
-Mon May 11 11:50:03 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* strings.h:  New file.  Just #includes string.h.
-
-Sun May 10 21:49:35 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* curses.h, sys/signal.h:  Add support for "recursive"
-	includes (#include_next gets installed previous version,
-	which then includes from /usr/include) (for those who don't
-	have -nostdinc++).
-
-Wed May  6 01:41:31 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* sys/signal.h:  Add 'hide' for signal().  (Needed on Irix.)
-	Also, remove unused __sys_signal_h_recursive junk.
-
-	* curses.h:  Add prototypes for -lcurses functions.
-
-Tue May  5 15:21:07 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* curses.h:  Give up trying to emulate all possible
-	curses.h implementations.  Just remove everything,
-	and just #include_next <curses.h> instead.
-
-Mon May  4 17:35:04 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* sys/times.h:  Added prototype for times().
-	* time.h:  Removed prototype for times().
-
-Fri May  1 00:07:24 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* time.h:  Remove variable declarations, since these will
-	presumably be in /usr/include/time.h -otherwise they can't
-	be used in C code (without a user declaration) anyway, so
-	we won't bothner.  Same applies to the non-SYSV declaration
-	of the timezone() function.  Thus we remove some conflicts
-	between SYSV and BSD that are impossible to resolve here.
-
-Thu Apr 30 00:39:06 1992  Per Bothner  (bothner@cygnus.com)
-
-	* signal.h:  Comment out sigset, since it seems to be neither
-	Posix nor BSD, and causes a conflict on Irix.
-
-	* memory.h, sys/stat.h, signal.h:  Hide all argument lists with
-	_G_ARGS macro, to avoid conflicts with system header files.
-	* unistd.h: Hide chmod() when doing #next_include.
-	* signal.h:  Use _G_signal_return_type to handle systems
-	that have signals return int instead of void.
-	* sys/resource.h:  Need to #include <_G_config.h>.
-	* math.h:  Add fmod().
-
-Fri Apr 17 14:29:04 1992  Per Bothner  (bothner@cygnus.com)
-
-	* Move all libg++-specific header files (e.g. Integer.h) to
-	../src. leaving just the header files that are wrapper for
-	C header files (e.g. stdio.h).  This allows a system which
-	has C++-raedy C header files (with extern "C" if __cplusplus
-	defined) (such as SVR4 or Linux) to ignore g++-include.
-	* curses.h, time.h:  Replace USG by _G_SYSV, and include
-	_G_config.h (which defines _G_SYSV if needed).
-	* Various files:  Surround parameter lists by _G_ARGS(...)
-	which normally is defined as just (...).  This avoids
-	a lot of conflicts with C header files.
-
-Fri Apr 10 22:14:14 1992  Per Bothner  (bothner@cygnus.com)
-
-	* ctype.h:  Simplify to just #include_next <ctype.h>.
-
-Tue Mar 24 16:13:42 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* Rational.h:  Added Rational::fits_in_float() and
-	fits_in_double() methods.
-
-Mon Mar 23 16:33:23 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* BitSet.h, BitString.h, Fix.h, Integer,h:  Add printon()
-	methods; deprecate *toa functions that use AllocRing.
-	(Note ../src/Changes.)
-
-Tue Mar 10 18:21:34 1992  Per Bothner  (bothner@cygnus.com)
-
-	* sys/wait.h:  Don't include sys/resource.h;
-	use forward struct declaration instead.
-
-Mon Mar  9 13:26:58 1992  Per Bothner  (bothner@cygnus.com)
-
-	* sys/resource.h:  Backed out recent change (until
-	after release) so to again include <sys/time.h>.
-
-Sat Mar  7 18:42:08 1992  Per Bothner  (bothner@rtl.cygnus.com)
-
-	* time.h, sys/time.h, sys/resource.h:  Minor fixes.
-	(They allow tests/test_h.cc to compile on DECstation.)
-
-Fri Mar  6 16:25:44 1992  Per Bothner  (bothner@cygnus.com)
-
-	* signal.h:  Protect signal() from conflicts.
-	* errno.h:  Onlu declare errno if no errno macro.
-	* new.h (operator new):  Don't make static inline, just inline.
-
-Thu Mar  5 23:15:49 1992  Per Bothner  (bothner@cygnus.com)
-
-	Second thoughts about the previous batch of changes to
-	handle '#if defined(__OPTIMIZE__) || defined(USE_LIBGXX_INLINES)'.
-	The problem was that this constrains how users compile their
-	programs:  They have to use -O (or -DUSE_LIBGXX_INLINES).
-	They are likely not to, which will cause them (and us) trouble.
-
-	The new solution:  Just remove the #if, and *always* compile
-	these functions inline.  This is simple, consistent, and
-	understandable.  The disadvantage is that debugging is harder,
-	but the answer is to work on being able to debug inline functions.
-
-	* gen/*.hP (almost): Removed all instances of:
-	#if defined(__OPTIMIZE__) || defined(USE_LIBGXX_INLINES),
-	thus inlining is independent on those flags, as discaussed above,
-
-	* Binomial.h BitSet.h BitString.h ChangeLog Complex.h CursesW.h
-	DiscUnif.h Erlang.h Fix.h Fix16.h Fix24.h Geom.h HypGeom.h
-	Integer.h LogNorm.h MLCG.h NegExp.h Normal.h Obstack.h Poisson.h
-	RNG.h Random.h Rational.h RndInt.h SmplHist.h SmplStat.h String.h
-	Uniform.h Weibull.h builtin.h compare.h ctype.h curses.h math.h
-	minmax.h:  Undo change from earlier today, and make inlining
-	unconditional as above.
-
-Thu Mar  5 16:51:10 1992  Per Bothner  (bothner@cygnus.com)
-
-	* gen/{AVec.ccP,VHSet.ccP,Vec.ccP,Vec.hP}:  Add casts and
-	other minor fixes to get rid of compiler warnings.
-	* sys/param.h:  Put back #undef of setbit, because it
-	would conflict with Integer::Setbit().
-
-	* Binomial.h, BitSet.h, BitString.h, Complex.h, CursesW.h,
-	DiscUnif.h, Erlang.h, Fix.h, Fix16.h, Fix24.h, Geom.h, HypGeom.h,
-	Integer.h, LogNorm.h, MLCG.h, NegExp.h, Normal.h, Obstack.h,
-	Poisson.h, RNG.h, Random.h, Rational.h, RndInt.h, SmplHist.h,
-	SmplStat.h, String.h, Uniform.h, Weibull.h, builtin.h, compare.h,
-	ctype.h, curses.h, math.h, minmax.h:  No longer use
-	USE_LIBGXX_INLINES.  Instead Foo.cc defines _COMPILING_Foo,
-	which is used by Foo.h to turn off inlining, thus
-	forcing a non-inlined version of those inline functions.
-	This allows using libg++ without requiring users to
-	compile without -O or setting funny flags.
-
-Tue Mar  3 17:23:53 1992  Per Bothner  (bothner@cygnus.com)
-
-	* BitSet.h:  Move BitSet::clear() from here to BitSet.cc,
-	and make non-inlined.
-
-Sun Mar  1 16:08:49 1992  Per Bothner  (bothner@cygnus.com)
-
-	* builtin.h:  Error handler functions are now (again)
-	functions returning void, not volatle void, since
-	sometimes these error handlers *do* return.
-	* BitSet.h, BitString.h, Integer.h, Rational.h, String.h:
-	Make error() methods return void, not volatile void.
-	* gen/{PHPQ.ccP,PHPQ.hP,Plex.ccP,Plex.hP,VHBag.ccP,VHBag.hP,
-	VHMap.ccP,VHMap.hP,VHSet.hP,VOHSet.ccP,VOHSet.hP,VQueue.ccP,
-	VQueue.hP,VStack.ccP,VStack.hP,Vec.ccP,Vec.hP}:
-	As above, make error methods return void, not volatile void.
-	Also, update 'delete [size] foo' to 'delete [] foo'.
-	* _G_config.h:  Fix typo in #define of _G_fpos_t.
-
-Wed Feb 26 12:32:08 1992  Per Bothner  (bothner at cygnus.com)
-
-	* _G_config.h:  Add _G_va_list, _G_fpos_t, _G_FOPEN_MAX,
-	and _G_FILENAME_MAX.
-	* builtin.h:  Use _VOLATILE_VOID macro to define
-	function as 'volatile void' only idef __GNUG__.
-	* unistd.h:  Add SEEK_SET and F_OK and their friends.
-	* sys/mman.h:  Fix typo.
-	* gen/SPlayPQ.hP:  Make destructor virtual.
-
-Thu Feb 20 21:47:43 1992  Per Bothner  (bothner at cygnus.com)
-
-	* dirent.h:  Re-write to follow standard pattern
-	(instead of just including sys/dir,h).
-	* osfcn.h:  Also include sys/types.h.
-
-Wed Feb 19 21:56:37 1992  Per Bothner  (bothner at cygnus.com)
-
-	* stdio.h, time.h, sys/time.h: Renamed c_proto_foo to __hide_foo.
-	* _G_config.h:  New file, to contain system-dependent
-	parameters.  (Later, this might be a hist-dependent file.)
-	* stdio.h:  Got rid of lots of old junk.
-	Added a	#define new __hide_new, because AIX uses
-	'new' as a parameter name.
-	* time.h:  Added some missing prototypes.  Use _G_config.h.
-
-Sat Feb 15 15:32:20 1992  Per Bothner  (bothner at cygnus.com)
-
-	* builtin.h: Remove some obsolete functions (that have been
-	moved to ../iostream/stream.h).
-
-Thu Feb  6 16:41:23 1992  Michael Tiemann  (tiemann at cygnus.com)
-
-	* {errno,fcntl,grp,netdb,pwd,setjmp,signal,stdarg,stddef.h}: Make
-	these files work correctly with #include_next.
-	{stdio,time}.h: Ditto
-
-Wed Feb  5 11:55:57 1992  Per Bothner  (bothner at cygnus.com)
-
-	* time.h, sys/times.h, sys/fcntl.h:  Just #include_next
-	the file with the same name; don't include some
-	other file (which could cause a loop).
-
-Fri Jan 31 16:55:19 1992  Per Bothner  (bothner at cygnus.com)
-
-	* gen/CHNode.{h,cc}P, gen/SplayNode.{h,cc}.P:  New files.
-	Pulled shared definition of CHNode and SplayNode
-	classes into their own files, to avoid potential
-	linker problems (from duplicated definitions).
-	* gen/Splay{Bag,Set,PQ}.hP, gen/CH{Bag,Set}.hP:
-	Include {Ch,Splay}Node.hP.
-
-Thu Jan 30 23:50:28 1992  Per Bothner  (bothner at cygnus.com)
-
-	* stdio.h, sys/time.h:  Protect against more function
-	prototype conflicts using macro re-naming.
-	* stdio.h:  Use size_t (actually __SIZE_TYPE__) where
-	appropriate.
-
-Fri Jan 24 15:59:37 1992  Per Bothner  (bothner at cygnus.com)
-
-	* pwd.h, sys/time.h:  Turn #include_next feature on.
-	* stdio.h:  Undefine and then redefine NULL, in case
-	/usr/include/stdio.h defines NULL without checking
-	for an existing definition.
-	* sys/signal.h: Don't just #include <signal.h>;
-	instead include the system (or _next) <sys/signal.h>,
-
-Fri Jan 17 14:12:54 1992  Per Bothner  (bothner at cygnus.com)
-
-	Merge in Doug Lea's latest version; plus other changes.
-	* sys/fcntl.h:  Just include <fcntl.h>.
-	* Most sys/*.h files, grp.h, setjmp.h, stdio.h, signal.h,
-	stdarg.h, stddef.h:  Use new #include_next feature of cpp
-	if using gcc-2.  If so, most of the system-dependent junk
-	in stdio.h goes away.
-	* dir.h, netdb.h, typemacros.h, sys/dir.h, sys/mman.h:  New files.
-	* Various files:  Remove #pragma once (not used by gcc-2).
-	* Various files:  Removed #define KERNEL.
-	* sys/resources.h, sys/socket.h, sys/wait.h, curses.h,
-	fcntl.h:  Add prototypes.
-	* BitString.h, String.h:  (int)->(unsigned int) changes (to fix
-	some gcc-2 warnings).
-	* assert.h, compare.h:  Minor improvements.
-	* Integer.h:  Re-do Integer::operator<< to exploit
-	new iostream facilities.
-	* builtin.h, minmax.h:  Moved min/max from former to latter.
-	* builtin.h: Moved hex/oct/dec to ../old-stream/itoa.cc.
-	* regex.h:  New version (from FSF).
-	* std.h; bstring.h, memory.h, errno.h, string.h, stdlib.h,
-	unistd.h:  Move prototypes into separate include files,
-	mostly split up the way ANSI C does.
-	* abs.h, file.h, malloc.h, max.h, min.h, random.h:  Removed.
diff --git a/libg++/g++-include/assert.h b/libg++/g++-include/assert.h
deleted file mode 100644
index 44a6570..0000000
--- a/libg++/g++-include/assert.h
+++ /dev/null
@@ -1,14 +0,0 @@
-#ifndef __libgxx_assert_h
-
-extern "C" {
-#ifdef __assert_h_recursive
-#include_next <assert.h>
-#else
-/* assert.h on some systems needs stdio.h, in violation of ANSI. */
-#include <stdio.h>
-#include_next <assert.h>
-
-#define __libgxx_assert_h 1
-#endif
-}
-#endif
diff --git a/libg++/g++-include/bstring.h b/libg++/g++-include/bstring.h
deleted file mode 100644
index 3b2f590..0000000
--- a/libg++/g++-include/bstring.h
+++ /dev/null
@@ -1 +0,0 @@
-#include <string.h>
diff --git a/libg++/g++-include/ctype.h b/libg++/g++-include/ctype.h
deleted file mode 100644
index b573e19..0000000
--- a/libg++/g++-include/ctype.h
+++ /dev/null
@@ -1,10 +0,0 @@
-#include <_G_config.h>
-extern "C" {
-#include_next <ctype.h>
-#ifndef toupper
-extern int toupper _G_ARGS((int));
-#endif
-#ifndef tolower
-extern int tolower _G_ARGS((int));
-#endif
-}
diff --git a/libg++/g++-include/curses.h b/libg++/g++-include/curses.h
deleted file mode 100644
index 571a48b..0000000
--- a/libg++/g++-include/curses.h
+++ /dev/null
@@ -1,87 +0,0 @@
-#ifndef _G_curses_h
-
-#include <_G_config.h>
-
-#if _G_HAVE_CURSES
-
-#ifdef __curses_h_recursive
-#include_next <curses.h>
-#else
-#define __curses_h_recursive
-
-extern "C" {
-#include_next <curses.h>
-
-/* Some systems (SVR4 for example) allow the definition of CHTYPE to set the
-   type of some arguments to the curses functions.  It can be set to "char"
-   to save space, or it can be set to something longer to store both a
-   character and some attributes.  By default they do not define CHTYPE,
-   and when CHTYPE is not defined, the default type is "unsigned long" instead
-   of the traditional "char".  However, SVR4 <curses.h> does define
-   _VR3_COMPAT_CODE, so we can use that to detect when we should use the SVR4
-   default if CHTYPE is not defined.  For other systems, just default to the
-   traditional default "char". */
-
-#ifdef CHTYPE
-      typedef CHTYPE _G_chtype;		/* Use specified type. */
-#else
-#ifdef _VR3_COMPAT_CODE
-      typedef unsigned long _G_chtype;	/* SVR4 default is "unsigned long" */
-#elif defined(hpux)
-      typedef unsigned int _G_chtype;
-#else
-      typedef char _G_chtype;		/* Traditional default is "char" */
-#endif
-#endif
-
-/* Some args are conceptually const, but SVR4 (and others?) get it wrong. */
-#define _C_const /* const */
-
-WINDOW * (newwin)(int lines, int cols, int sy, int sx);
-WINDOW * (subwin)(WINDOW *w, int lines, int cols, int sy, int sx);
-WINDOW * (initscr)();
-int      (box) (WINDOW*, _G_chtype, _G_chtype);
-int      (delwin)(WINDOW*);
-int      (getcurx)(WINDOW*);
-int      (getcury)(WINDOW*);
-int      (mvcur)(int, int, int, int);
-int      (overlay)(WINDOW*, WINDOW*);
-int      (overwrite)(WINDOW*, WINDOW*);
-int      (scroll)(WINDOW*);
-int      (touchwin)(WINDOW*);
-int      (waddch)(WINDOW*, _G_chtype);
-int      (waddstr)(WINDOW*, _C_const char*);
-int      (wclear)(WINDOW*);
-int      (wclrtobot)(WINDOW*);
-int      (wclrtoeol)(WINDOW*);
-int      (wdelch)(WINDOW*);
-int      (wdeleteln)(WINDOW*);
-int      (werase)(WINDOW*);
-int      (wgetch)(WINDOW*);
-int      (wgetstr)(WINDOW*, char*);
-int      (winsch)(WINDOW*, _G_chtype);
-int      (winsertln)(WINDOW*);
-int      (wmove)(WINDOW*, int, int);
-int      (wrefresh)(WINDOW*);
-int      (wstandend)(WINDOW*);
-int      (wstandout)(WINDOW*);
-
-// SVR4 rather inanely bundles the format-string parameter with the '...'.
-// This breaks VMS, and I don't want to penalize VMS for being right for once!
-#ifdef _VR3_COMPAT_CODE /* Magic cookie to recognize SVR4 */
-#define _CURSES_FORMAT_ARG /*const char* fmt, */
-#else
-#define _CURSES_FORMAT_ARG const char* fmt,
-#endif
-
-int      (wprintw)(WINDOW*, _CURSES_FORMAT_ARG ...);
-int      (mvwprintw)(WINDOW*, int y, int x, _CURSES_FORMAT_ARG ...);
-int      (wscanw)(WINDOW*, _CURSES_FORMAT_ARG ...);
-int      (mvwscanw)(WINDOW*, int, int, _CURSES_FORMAT_ARG ...);
-int      (endwin)();
-
-}
-#define _G_curses_h
-#endif
-#endif /* _G_HAVE_CURSES */
-#endif /* _G_curses_h */
diff --git a/libg++/g++-include/dir.h b/libg++/g++-include/dir.h
deleted file mode 100644
index 78c5402..0000000
--- a/libg++/g++-include/dir.h
+++ /dev/null
@@ -1 +0,0 @@
-#include <sys/dir.h>
diff --git a/libg++/g++-include/dirent.h b/libg++/g++-include/dirent.h
deleted file mode 100644
index 4c06ea8..0000000
--- a/libg++/g++-include/dirent.h
+++ /dev/null
@@ -1,44 +0,0 @@
-#ifndef __libgxx_dirent_h
-
-#include <_G_config.h>
-
-#if !_G_HAVE_DIRENT
-#define __libgxx_dirent_h
-#define direct dirent
-#include <sys/dir.h>
-#else
-
-extern "C" {
-
-#ifdef __dirent_h_recursive
-#include_next <dirent.h>
-#else
-// Note: sys/dir.h checks __dirent_h_recursive
-#define __dirent_h_recursive
-#define opendir __hide_opendir
-#define closedir __hide_closedir
-#define readdir __hide_readdir
-#define telldir __hide_telldir
-#define seekdir __hide_seekdir
-
-#include_next <dirent.h>
-
-#define __libgxx_dirent_h
-#undef opendir
-#undef closedir
-#undef readdir
-#undef telldir
-#undef seekdir
-
-DIR *opendir(const char *);
-int closedir(DIR *);
-struct dirent *readdir(DIR *);
-long telldir(DIR *);
-void seekdir(DIR *, long);
-// We don't bother with rewinddir (many systems define it as a macro).
-// void rewinddir(DIR *);
-#endif
-}
-
-#endif
-#endif
diff --git a/libg++/g++-include/errno.h b/libg++/g++-include/errno.h
deleted file mode 100644
index f9f86c8..0000000
--- a/libg++/g++-include/errno.h
+++ /dev/null
@@ -1,24 +0,0 @@
-#ifndef errno_h
-
-extern "C" {
-
-#ifdef __errno_h_recursive
-#include_next <errno.h>
-#else
-#define __errno_h_recursive
-#include_next <errno.h>
-
-#define errno_h 1
-
-extern char*    sys_errlist[];
-extern int      sys_nerr;
-#ifndef errno                  
-extern int      errno;
-#endif
-void      perror(const char*);
-char*     strerr(int);
-
-#endif
-}
-
-#endif
diff --git a/libg++/g++-include/fcntl.h b/libg++/g++-include/fcntl.h
deleted file mode 100644
index 48637ef..0000000
--- a/libg++/g++-include/fcntl.h
+++ /dev/null
@@ -1,29 +0,0 @@
-#ifndef fcntl_h
-
-extern "C" {
-
-#ifdef __fcntl_h_recursive
-#include_next <fcntl.h>
-#else
-#define fcntl __hide_fcntl
-#define open  __hide_open
-#define creat __hide_creat
-
-#define __fcntl_h_recursive
-#include <_G_config.h>
-#include_next <fcntl.h>
-
-#undef fcntl
-#undef open
-#undef creat
-
-#define fcntl_h 1
-
-int       fcntl(int, int, ...);
-int	  creat _G_ARGS((const char*, unsigned short int));
-
-int       open _G_ARGS((const char*, int, ...));
-
-#endif
-}
-#endif
diff --git a/libg++/g++-include/grp.h b/libg++/g++-include/grp.h
deleted file mode 100644
index c05c305..0000000
--- a/libg++/g++-include/grp.h
+++ /dev/null
@@ -1,41 +0,0 @@
-#ifndef grp_h
-
-extern "C" {
-
-#ifdef __grp_h_recursive
-#include_next <grp.h>
-#else
-#define __grp_h_recursive
-
-#include <stdio.h>
-
-#define getgrent c_proto_getgrent
-#define getgrgid c_proto_getgrgid
-#define getgrnam c_proto_getgrnam
-#define setgrent c_proto_setgrent
-#define endgrent c_proto_endgrent
-#define fgetgrent c_proto_fgetgrent
-
-#include_next <grp.h>
-
-#define grp_h 1
-
-#undef getgrent
-#undef getgrgid
-#undef getgrnam
-
-extern struct group* getgrent();
-extern struct group* fgetgrent(FILE*);
-extern struct group* getgrgid(int);
-extern struct group* getgrnam(const char*);
-#ifdef __OSF1__
-extern int	     setgrent();
-#else
-extern void          setgrent();
-#endif
-extern void          endgrent();
-
-#endif
-}
-
-#endif
diff --git a/libg++/g++-include/math.h b/libg++/g++-include/math.h
deleted file mode 100644
index 9c0ebfa..0000000
--- a/libg++/g++-include/math.h
+++ /dev/null
@@ -1,219 +0,0 @@
-// This may look like C code, but it is really -*- C++ -*-
-/* 
-Copyright (C) 1988 Free Software Foundation
-    written by Doug Lea (dl@rocky.oswego.edu)
-
-This file is part of the GNU C++ Library.  This library is free
-software; you can redistribute it and/or modify it under the terms of
-the GNU Library General Public License as published by the Free
-Software Foundation; either version 2 of the License, or (at your
-option) any later version.  This library is distributed in the hope
-that it will be useful, but WITHOUT ANY WARRANTY; without even the
-implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
-PURPOSE.  See the GNU Library General Public License for more details.
-You should have received a copy of the GNU Library General Public
-License along with this library; if not, write to the Free Software
-Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
-*/
-
-
-#ifndef _math_h
-#ifdef __GNUG__
-#pragma interface
-#endif
-#define _math_h 1
-
-#if defined(hp300) && defined(__HAVE_FPU__)
-#define __HAVE_68881__ 1
-#endif
-
-#if defined(masscomp)
-#define __HAVE_68881__ 1
-#endif
-
-#ifdef __HAVE_68881__		/* MC68881/2 Floating-Point Coprocessor */
-extern "C" {			/* fill in what we've left out */
-#include <math-68881.h>
-
-double  acosh(double);
-double  asinh(double);
-double  cbrt(double);
-double  copysign(double,double);
-double  erf(double);
-double  erfc(double);
-double  finite(double);
-double  gamma(double);
-double  hypot(double,double);
-double  infnan(int);
-int     isinf(double);
-int     isnan(double);
-double  j0(double);
-double  j1(double);
-double  jn(int, double);
-double  lgamma(double);
-double  y0(double);
-double  y1(double);
-double  yn(int, double);
-
-double aint(double);
-double anint(double);
-int irint(double);
-int nint(double);
-}
-/* Please add inline asm code for other machines here! */
-#else
-extern "C" {
-
-double  acos(double);
-double  acosh(double);
-double  asin(double);
-double  asinh(double);
-double  atan(double);
-double  atan2(double, double);
-double  atanh(double);
-double  cbrt(double);
-double  ceil(double);
-double  copysign(double,double);
-double  cos(double);
-double  cosh(double);
-double  drem(double,double);
-double  erf(double);
-double  erfc(double);
-double  exp(double);
-double  expm1(double);
-double  fabs(double);
-double  finite(double);
-double  floor(double);
-double	fmod(double, double);
-double  frexp(double, int*);
-double  gamma(double);
-double  hypot(double,double);
-double  infnan(int);
-#if !defined(sequent) && !defined(DGUX) &&!defined(sony) && !defined(masscomp) && !defined(hpux)
-/* see below */
-int     isinf(double);
-int     isnan(double);
-#endif
-double  j0(double);
-double  j1(double);
-double  jn(int, double);
-double  ldexp(double, int);
-double  lgamma(double);
-double  log(double);
-double  log10(double);
-double  log1p(double);
-double  logb(double);
-double  modf(double, double*);
-double  pow(double, double);
-double  rint(double);
-double  scalb(double, int);
-double  sin(double);
-double  sinh(double);
-double  sqrt(double);
-double  tan(double);
-double  tanh(double);
-double  y0(double);
-double  y1(double);
-double  yn(int, double);
-
-double aint(double);
-double anint(double);
-int irint(double);
-int nint(double);
-}
-
-#endif
-
-/* libg++ doesn't use this since it is not available on some systems */
-
-/* the following ifdef is just for compiling OOPS */
-
-#ifndef DONT_DECLARE_EXCEPTION
-struct libm_exception
-{
-  int type;
-  char* name;
-  double arg1, arg2, retval;
-};
-
-#define DOMAIN      1
-#define SING        2
-#define OVERFLOW    3
-#define UNDERFLOW   4
-#define TLOSS       5
-#define PLOSS       6
-
-extern "C" int matherr(libm_exception*);
-
-#endif
-
-#include <float.h>
-
-/* On some systems, HUGE ought to be MAXFLOAT or IEEE infinity */
-
-#ifndef HUGE
-#define HUGE    DBL_MAX
-#endif
-#ifndef HUGE_VAL
-#define HUGE_VAL    DBL_MAX
-#endif
-
-
-/* sequents don't supply these. The following should suffice */
-#if defined(sequent) || defined(DGUX) || defined(sony) || defined(masscomp) \
-|| defined(hpux)
-#include <float.h>
-static inline int isnan(double x) { return x != x; }
-static inline int isinf(double x) { return x > DBL_MAX || x < -DBL_MAX; }
-#endif
-
-/* These seem to be sun & sysV names of these constants */
-
-#ifndef M_E
-#define M_E         2.7182818284590452354
-#endif
-#ifndef M_LOG2E
-#define M_LOG2E     1.4426950408889634074
-#endif
-#ifndef M_LOG10E
-#define M_LOG10E    0.43429448190325182765
-#endif
-#ifndef M_LN2
-#define M_LN2       0.69314718055994530942
-#endif
-#ifndef M_LN10
-#define M_LN10      2.30258509299404568402
-#endif
-#ifndef M_PI
-#define M_PI        3.14159265358979323846
-#endif
-#ifndef M_PI_2
-#define M_PI_2      1.57079632679489661923
-#endif
-#ifndef M_1_PI
-#define M_1_PI      0.31830988618379067154
-#endif
-#ifndef M_PI_4
-#define M_PI_4      0.78539816339744830962
-#endif
-#ifndef M_2_PI
-#define M_2_PI      0.63661977236758134308
-#endif
-#ifndef M_2_SQRTPI
-#define M_2_SQRTPI  1.12837916709551257390
-#endif
-#ifndef M_SQRT2
-#define M_SQRT2     1.41421356237309504880
-#endif
-#ifndef M_SQRT1_2
-#define M_SQRT1_2   0.70710678118654752440
-#endif
-
-#ifndef PI                      // as in stroustrup
-#define PI  M_PI
-#endif
-#ifndef PI2
-#define PI2  M_PI_2
-#endif
-
-#endif
diff --git a/libg++/g++-include/memory.h b/libg++/g++-include/memory.h
deleted file mode 100644
index a4ed091..0000000
--- a/libg++/g++-include/memory.h
+++ /dev/null
@@ -1,42 +0,0 @@
-
-#ifndef _memory_h
-#define _memory_h 1
-
-#include "_G_config.h"
-#include <stddef.h>
-
-extern "C" {
-
-void*     memalign _G_ARGS((_G_size_t, _G_size_t));
-void*     memccpy _G_ARGS((void*, const void*, int, _G_size_t));
-void*     memchr _G_ARGS((const void*, int, _G_size_t));
-int       memcmp _G_ARGS((const void*, const void*, _G_size_t));
-void*     memcpy _G_ARGS((void*, const void*, _G_size_t));
-void*     memmove _G_ARGS((void*, const void*, _G_size_t));
-void*     memset _G_ARGS((void*, int, _G_size_t));
-int       ffs _G_ARGS((int));
-#ifdef __OSF1__
-int	  getpagesize _G_ARGS((void));
-#else
-_G_size_t    getpagesize _G_ARGS((void));
-#endif
-void*     valloc _G_ARGS((_G_size_t));
-
-void      bcopy _G_ARGS((const void*, void*, _G_size_t));
-int       bcmp _G_ARGS((const void*, const void*, int));
-void      bzero _G_ARGS((void*, int));
-}
-
-#ifdef __GNUG__
-#ifndef alloca
-#define alloca(x)  __builtin_alloca(x)
-#endif
-#else
-#ifndef IV
-extern "C" void* alloca(_G_size_t);
-#else
-extern "C" void* alloca(unsigned long);
-#endif /* IV */
-#endif
-
-#endif
diff --git a/libg++/g++-include/netdb.h b/libg++/g++-include/netdb.h
deleted file mode 100644
index c718a8e..0000000
--- a/libg++/g++-include/netdb.h
+++ /dev/null
@@ -1,4 +0,0 @@
-extern "C"
-{
-#include_next <netdb.h>
-}
diff --git a/libg++/g++-include/pwd.h b/libg++/g++-include/pwd.h
deleted file mode 100644
index f33d7ee..0000000
--- a/libg++/g++-include/pwd.h
+++ /dev/null
@@ -1,36 +0,0 @@
-#ifndef _pwd_h
-
-// the Interviews-based standard kludge again
-
-extern "C" {
-
-#ifdef __pwd_h_recursive
-#include_next <pwd.h>
-#else
-#define getpwent c_proto_getpwent
-#define getpwuid c_proto_getpwuid
-#define getpwnam c_proto_getpwnam
-#define setpwent c_proto_setpwent
-#define endpwent c_proto_endpwent
-
-#define __pwd_h_recursive
-#include_next <pwd.h>
-
-#define _pwd_h 1
-
-#undef getpwent
-#undef getpwuid
-#undef getpwnam
-#undef setpwent
-#undef endpwent
-
-extern struct passwd* getpwent();
-extern struct passwd* getpwuid(int);
-extern struct passwd* getpwnam(const char*);
-extern int            setpwent();
-extern int            endpwent();
-
-#endif
-}
-
-#endif
diff --git a/libg++/g++-include/setjmp.h b/libg++/g++-include/setjmp.h
deleted file mode 100644
index dc35dea..0000000
--- a/libg++/g++-include/setjmp.h
+++ /dev/null
@@ -1,29 +0,0 @@
-#ifndef _setjmp_h
-
-extern "C" {
-
-#ifdef __setjmp_h_recursive
-#include_next <setjmp.h>
-#else
-#define __setjmp_h_recursive
-#define  setjmp  C_header_setjmp
-#define  longjmp C_header_longjmp
-
-#ifdef VMS
-#include "gnu_cc_include:[000000]setjmp.h"
-#else
-#include_next <setjmp.h>
-#endif
-
-#undef setjmp
-#undef longjmp
-
-#define _setjmp_h 1
-
-extern int setjmp(jmp_buf);
-extern void longjmp(jmp_buf, int);
-
-#endif
-}
-
-#endif
diff --git a/libg++/g++-include/signal.h b/libg++/g++-include/signal.h
deleted file mode 100644
index 942e272..0000000
--- a/libg++/g++-include/signal.h
+++ /dev/null
@@ -1,76 +0,0 @@
-// This may look like C code, but it is really -*- C++ -*-
-/* 
-Copyright (C) 1989 Free Software Foundation
-    written by Doug Lea (dl@rocky.oswego.edu)
-
-This file is part of the GNU C++ Library.  This library is free
-software; you can redistribute it and/or modify it under the terms of
-the GNU Library General Public License as published by the Free
-Software Foundation; either version 2 of the License, or (at your
-option) any later version.  This library is distributed in the hope
-that it will be useful, but WITHOUT ANY WARRANTY; without even the
-implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
-PURPOSE.  See the GNU Library General Public License for more details.
-You should have received a copy of the GNU Library General Public
-License along with this library; if not, write to the Free Software
-Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
-*/
-
-#ifndef _signal_h
-
-#include <_G_config.h>
-
-extern "C" {
-
-#ifdef __signal_h_recursive
-#include_next <signal.h>
-#else
-
-#define __signal_h_recursive
-
-#define signal __hide_signal
-#define psignal __hide_psignal
-#include_next <signal.h>
-#undef signal
-#undef psignal
-
-#define _signal_h 1
-
-// The Interviews folks call this SignalHandler. Might as well conform.
-typedef _G_signal_return_type (*SignalHandler) (...);
-typedef int (*SSignalHandler) (...);
-
-extern SignalHandler signal _G_ARGS((int sig, SignalHandler action));
-//extern SignalHandler sigset _G_ARGS((int sig, SignalHandler action));
-extern SSignalHandler ssignal _G_ARGS((int sig, SSignalHandler action));
-extern int           gsignal  _G_ARGS((int sig));
-extern int           kill  _G_ARGS((_G_pid_t pid, int sig));
-extern int           killpg _G_ARGS((int, int));
-extern int           siginterrupt _G_ARGS((int, int));
-extern void	     psignal _G_ARGS((unsigned, const char*));
-
-#ifndef hpux // Interviews folks claim that hpux doesn't like these
-extern int           sigsetmask _G_ARGS((int mask));
-extern int           sigblock _G_ARGS((int mask));
-extern int           sigpause _G_ARGS((int mask));
-extern int           sigvec _G_ARGS((int, struct sigvec*, struct sigvec*));
-#endif
-
-// The Interviews version also has these ...
-
-#define SignalBad ((SignalHandler)-1)
-#define SignalDefault ((SignalHandler)0)
-#define SignalIgnore ((SignalHandler)1)
-
-#undef BADSIG
-#undef SIG_DFL
-#undef SIG_IGN
-#define BAD_SIG SignalBad
-#define SIG_DFL SignalDefault
-#define SIG_IGN SignalIgnore
-
-#endif
-}
-
-#endif
-
diff --git a/libg++/g++-include/stdarg.h b/libg++/g++-include/stdarg.h
deleted file mode 100644
index 67dc22e..0000000
--- a/libg++/g++-include/stdarg.h
+++ /dev/null
@@ -1,3 +0,0 @@
-extern "C" {
-#include_next <stdarg.h>
-}
diff --git a/libg++/g++-include/stddef.h b/libg++/g++-include/stddef.h
deleted file mode 100644
index 66e78ba..0000000
--- a/libg++/g++-include/stddef.h
+++ /dev/null
@@ -1,12 +0,0 @@
-#ifndef __libgxx_stddef_h
-
-extern "C" {
-#ifdef __stddef_h_recursive
-#include_next <stddef.h>
-#else
-#include_next <stddef.h>
-
-#define __libgxx_stddef_h 1
-#endif
-}
-#endif
diff --git a/libg++/g++-include/stdio.h b/libg++/g++-include/stdio.h
deleted file mode 100644
index 4e8b704..0000000
--- a/libg++/g++-include/stdio.h
+++ /dev/null
@@ -1,174 +0,0 @@
-// This may look like C code, but it is really -*- C++ -*-
-/* 
-Copyright (C) 1988 Free Software Foundation
-    written by Doug Lea (dl@rocky.oswego.edu)
-
-This file is part of the GNU C++ Library.  This library is free
-software; you can redistribute it and/or modify it under the terms of
-the GNU Library General Public License as published by the Free
-Software Foundation; either version 2 of the License, or (at your
-option) any later version.  This library is distributed in the hope
-that it will be useful, but WITHOUT ANY WARRANTY; without even the
-implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
-PURPOSE.  See the GNU Library General Public License for more details.
-You should have received a copy of the GNU Library General Public
-License along with this library; if not, write to the Free Software
-Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
-*/
-
-#ifndef _stdio_h
-#ifdef __GNUG__
-#pragma interface
-#endif
-
-#ifdef __stdio_h_recursive
-#include_next <stdio.h>
-#else
-#define __stdio_h_recursive
-
-// Note:  The #define _stdio_h is at the end of this file,
-// in case #include_next <stdio.h> finds an installed version of this
-// same file -- we want it to continue until it finds the C version.
-
-#include <_G_config.h>
-
-extern "C" {
-
-#undef NULL
-
-#define fdopen __hide_fdopen
-#define fopen __hide_fopen
-#define fprintf __hide_fprintf
-#define fputs __hide_fputs
-#define fread __hide_fread
-#define freopen __hide_freopen
-#define fscanf __hide_fscanf
-#define ftell __hide_ftell
-#define fwrite __hide_fwrite
-#define new __hide_new /* In case 'new' is used as a parameter name. */
-#define perror __hide_perror
-#define popen __hide_popen
-#define printf __hide_printf
-#define puts __hide_puts
-#define putw __hide_putw
-#define rewind __hide_rewind
-#define tempnam __hide_tempnam
-#define scanf __hide_scanf
-#define setbuf __hide_setbuf
-#define setbuffer __hide_setbuffer
-#define setlinebuf __hide_setlinebuf
-#define setvbuf __hide_setvbuf
-#define sprintf __hide_sprintf
-#define sscanf __hide_sscanf
-#define tempnam __hide_tempnam
-#define vfprintf __hide_vfprintf
-#define vprintf __hide_vprintf
-#define vsprintf __hide_vsprintf
-#define _flsbuf __hide__flsbuf
-
-#include_next <stdio.h>
-
-#undef fdopen
-#undef fopen
-#undef fprintf
-#undef fputs
-#undef fread
-#undef freopen
-#undef fscanf
-#undef ftell
-#undef fwrite
-#undef new
-#undef perror
-#undef popen
-#undef printf
-#undef puts
-#undef putw
-#undef rewind
-#undef tempnam
-#undef scanf
-#undef setbuf
-#undef setbuffer
-#undef setlinebuf
-#undef setvbuf
-#undef sprintf
-#undef sscanf
-#undef tempnam
-#undef vprintf
-#undef vfprintf
-#undef vsprintf
-#undef _flsbuf
-
-#ifndef NULL
-#define NULL _G_NULL
-#endif
-
-#ifndef size_t
-#define size_t _G_size_t
-#endif
-}
-
-extern "C" {
-
-int    fclose(FILE*);
-FILE*  fdopen(int, const char*);
-int    fflush(FILE*);
-int    fgetc(FILE*);
-char*  fgets(char*, int, FILE *);
-FILE*  fopen(const char*, const char*);
-int    fprintf(FILE*, const char* ...);
-int    fputc(int, FILE*);
-int    fputs(const char*, FILE*);
-size_t fread(void*, size_t, size_t, FILE*);
-#ifdef VMS
-FILE*  freopen(const char*, const char*, FILE* ...);
-#else
-FILE*  freopen(const char*, const char*, FILE*);
-#endif
-int    fscanf(FILE*, const char* ...);
-int    fseek(FILE*, long, int);
-long   ftell(FILE *);
-size_t fwrite(const void*, size_t, size_t, FILE*);
-char*  gets(char*);
-int    getw(FILE*);
-int    pclose(FILE*);
-void   perror(const char*);
-FILE*  popen(const char*, const char*);
-int    printf(const char* ...);
-int    puts(const char*);
-int    putw(int, FILE*);
-int    rewind(FILE*);
-int    scanf(const char* ...);
-void   setbuf(FILE*, char*);
-void   setbuffer(FILE*, char*, int);
-int    setlinebuf(FILE*);
-int    setvbuf(FILE*, char*, int, size_t);
-int    sscanf(char*, const char* ...);
-FILE*  tmpfile();
-int    ungetc(int, FILE*);
-int    vfprintf _G_ARGS((FILE*, const char*, _G_va_list));
-int    vprintf _G_ARGS((const char*, _G_va_list));
-_G_sprintf_return_type sprintf _G_ARGS((char*, const char* ...));
-_G_sprintf_return_type vsprintf _G_ARGS((char*, const char*, _G_va_list));
-
-extern int _filbuf _G_ARGS((FILE*));
-extern int _flsbuf _G_ARGS((unsigned, FILE*));
-
-}
-
-#ifndef L_ctermid
-#define L_ctermid	9 
-#endif
-#ifndef L_cuserid
-#define L_cuserid	9
-#endif
-#ifndef P_tmpdir
-#define	P_tmpdir    "/tmp/"
-#endif
-#ifndef L_tmpnam
-#define	L_tmpnam    (sizeof(P_tmpdir) + 15)
-#endif
-
-#define _stdio_h 1
-
-#endif
-#endif // _stdio_h
diff --git a/libg++/g++-include/stdlib.h b/libg++/g++-include/stdlib.h
deleted file mode 100644
index 9f4d1fe..0000000
--- a/libg++/g++-include/stdlib.h
+++ /dev/null
@@ -1,81 +0,0 @@
-
-
-#ifndef _stdlib_h
-#define _stdlib_h 1
-
-#include <_G_config.h>
-#include <stddef.h>
-
-extern "C" {
-
-int       abs(int);
-
-#ifdef __GNUG__
-void volatile abort(void);
-#else
-void abort(void);
-#endif
-
-double    atof(const char*);
-int       atoi(const char*);
-long      atol(const char*);
-
-int       atexit(auto void (*p) (void));
-void*     bsearch (const void *, const void *, size_t, 
-                   size_t, auto int (*ptf)(const void*, const void*));
-void*     calloc(size_t, size_t);
-void      cfree(void*);
-
-#ifdef __GNUG__
-void volatile exit(int);
-#else
-void      exit(int);
-#endif
-
-char*     fcvt(double, int, int*, int*);
-void      free(void*);
-char*     getenv(const char*);
-int       getopt _G_ARGS((int, char * const *, const char*));
-int       getpw(int, char*);
-char*     gcvt(double, int, char*);
-char*     ecvt(double, int, int*, int*);
-extern char**   environ;
-
-long      labs(long);
-void*     malloc(size_t);
-size_t    malloc_usable_size(void*);
-int       putenv(const char*);
-extern char*    optarg;
-extern int      opterr;
-extern int      optind;
-void      qsort(void*, size_t, size_t, auto int (*ptf)(void*,void*));
-int       rand(void);
-void*     realloc(void*, size_t);
-int       setkey(const char*);
-int       srand(unsigned int);
-double    strtod(const char*, char**);
-long      strtol(const char*, char**, int);
-unsigned long stroul(const char**, int);
-int       system(const char*);
-
-long      random(void);
-void      srandom(int);
-char*     setstate(char*);
-char*     initstate(unsigned, char*, int);
-
-double    drand48(void);
-void      lcong48(short*);
-long      jrand48(short*);
-long      lrand48(void);
-long      mrand48(void);
-long      nrand48(short*);
-short*    seed48(short*);
-void      srand48(long);
-
-char*     ctermid(char*);
-char*     cuserid(char*);
-char*     tempnam(const char*, const char*);
-char*     tmpnam(char*);
-
-}
-#endif
diff --git a/libg++/g++-include/string.h b/libg++/g++-include/string.h
deleted file mode 100644
index b11ca0a..0000000
--- a/libg++/g++-include/string.h
+++ /dev/null
@@ -1,45 +0,0 @@
-
-#ifndef _string_h
-#define _string_h 1
-
-#include <_G_config.h>
-
-#ifndef size_t
-#define size_t _G_size_t
-#endif
-
-#ifndef NULL
-#define NULL _G_NULL
-#endif
-
-extern "C" {
-
-char*     strcat(char*, const char*);
-char*     strchr(const char*, int);
-int       strcmp(const char*, const char*);
-int       strcoll(const char*, const char*);
-char*     strcpy(char*, const char*);
-size_t    strcspn(const char*, const char*);
-char*     strdup(const char*);
-// NOTE: If you get a error message from g++ that this declaration
-// conflicts with a <built-in> declaration of strlen(), that usually
-// indicates that __SIZE_TYPE__ is incorrectly defined by gcc.
-// Fix or add SIZE_TYPE in the appropriate file in gcc/config/*.h.
-size_t    strlen(const char*);
-char*     strncat(char*, const char*, size_t);
-int       strncmp(const char*, const char*, size_t);
-char*     strncpy(char*, const char*, size_t);
-char*     strpbrk(const char*, const char*);
-char*     strrchr(const char*, int);
-size_t    strspn(const char*, const char*);
-char*     strstr(const char*, const char *);
-char*     strtok(char*, const char*);
-size_t    strxfrm(char*, const char*, size_t);
-
-char*     index(const char*, int);
-char*     rindex(const char*, int);
-}
-
-#include <memory.h>
-
-#endif
diff --git a/libg++/g++-include/strings.h b/libg++/g++-include/strings.h
deleted file mode 100644
index 3b2f590..0000000
--- a/libg++/g++-include/strings.h
+++ /dev/null
@@ -1 +0,0 @@
-#include <string.h>
diff --git a/libg++/g++-include/sys/ChangeLog b/libg++/g++-include/sys/ChangeLog
deleted file mode 100644
index b70b3ae..0000000
--- a/libg++/g++-include/sys/ChangeLog
+++ /dev/null
@@ -1,5 +0,0 @@
-Thu Feb  6 22:19:01 1992  Michael Tiemann  (tiemann at cygnus.com)
-
-	* {dir,fcntl,file,mman,param,resource,signal,socket}.h: Make
-	these files work correctly with #include_next.
-	{stat,time,times,types,wait}.h: Ditto.
diff --git a/libg++/g++-include/sys/dir.h b/libg++/g++-include/sys/dir.h
deleted file mode 100644
index 5868672..0000000
--- a/libg++/g++-include/sys/dir.h
+++ /dev/null
@@ -1,42 +0,0 @@
-#ifndef __libgxx_sys_dir_h
-
-extern "C" {
-
-#ifdef __sys_dir_h_recursive
-#include_next <sys/dir.h>
-#else
-#define __sys_dir_h_recursive
-#define opendir __hide_opendir
-#define closedir __hide_closedir
-#define readdir __hide_readdir
-#define telldir __hide_telldir
-#define seekdir __hide_seekdir
-
-#include_next <sys/dir.h>
-
-#define __libgxx_sys_dir_h
-#undef opendir
-#undef closedir
-#undef readdir
-#undef telldir
-#undef seekdir
-
-DIR *opendir(const char *);
-int closedir(DIR *);
-#ifdef __dirent_h_recursive
-// Some operating systems we won't mention (such as the imitation
-// of Unix marketed by IBM) implement dirent.h by including sys/dir.h,
-// in which case sys/dir.h defines struct dirent, rather than
-// the struct direct originally used by BSD.
-struct dirent *readdir(DIR *);
-#else
-struct direct *readdir(DIR *);
-#endif
-long telldir(DIR *);
-void seekdir(DIR *, long);
-// We don't bother with rewinddir (many systems define it as a macro).
-// void rewinddir(DIR *);
-#endif
-}
-
-#endif
diff --git a/libg++/g++-include/sys/fcntl.h b/libg++/g++-include/sys/fcntl.h
deleted file mode 100644
index d4fdcfe..0000000
--- a/libg++/g++-include/sys/fcntl.h
+++ /dev/null
@@ -1,15 +0,0 @@
-#ifndef __libgxx_sys_fcntl_h
-
-extern "C"
-{
-#ifdef __sys_fcntl_h_recursive
-#include_next <sys/fcntl.h>
-#else
-#define __sys_fcntl_h_recursive
-#include_next <sys/fcntl.h>
-#define __libgxx_sys_fcntl_h 1
-#endif
-}
-
-
-#endif
diff --git a/libg++/g++-include/sys/file.h b/libg++/g++-include/sys/file.h
deleted file mode 100644
index 70b49c2..0000000
--- a/libg++/g++-include/sys/file.h
+++ /dev/null
@@ -1,28 +0,0 @@
-#ifndef __libgxx_sys_file_h
-
-extern "C"
-{
-#ifdef __sys_file_h_recursive
-#include_next <sys/file.h>
-#else
-#define fcntl __hide_fcntl
-#define open  __hide_open
-#define creat __hide_creat
-
-#define __sys_file_h_recursive
-
-#ifdef VMS
-#include "GNU_CC_INCLUDE:[sys]file.h"
-#else
-#include_next <sys/file.h>
-#endif
-
-#undef fcntl
-#undef open
-#undef creat
-
-#define __libgxx_sys_file_h 1
-
-#endif
-}
-#endif
diff --git a/libg++/g++-include/sys/mman.h b/libg++/g++-include/sys/mman.h
deleted file mode 100644
index c4b2e15..0000000
--- a/libg++/g++-include/sys/mman.h
+++ /dev/null
@@ -1,14 +0,0 @@
-#ifndef __libgxx_sys_mman_h
-
-extern "C" {
-#ifdef __sys_mman_h_recursive
-#include_next <sys/mman.h>
-#else
-#define __sys_mman_h_recursive
-#include_next <sys/mman.h>
-
-#define __libgxx_sys_mman_h 1
-#endif
-}
-
-#endif
diff --git a/libg++/g++-include/sys/param.h b/libg++/g++-include/sys/param.h
deleted file mode 100644
index 610a19b..0000000
--- a/libg++/g++-include/sys/param.h
+++ /dev/null
@@ -1,22 +0,0 @@
-#ifndef __libgxx_sys_param_h
-
-extern "C"
-{
-#ifdef __sys_param_h_recursive
-#include_next <sys/param.h>
-#else
-#define __sys_param_h_recursive
-
-#ifdef VMS
-#include "GNU_CC_INCLUDE:[sys]param.h"
-#else
-#include_next <sys/param.h>
-#endif
-
-#undef setbit /* Conflicts with Integer::setbit(). */
-
-#define __libgxx_sys_param_h 1
-#endif
-}
-
-#endif
diff --git a/libg++/g++-include/sys/resource.h b/libg++/g++-include/sys/resource.h
deleted file mode 100644
index 3596c1b..0000000
--- a/libg++/g++-include/sys/resource.h
+++ /dev/null
@@ -1,29 +0,0 @@
-#ifndef __libgxx_sys_resource_h
-
-extern "C"
-{
-#ifdef __sys_resource_h_recursive
-#include_next <sys/resource.h>
-#else
-#include <_G_config.h>
-#define __sys_resource_h_recursive
-#include <sys/time.h>
-
-#ifdef VMS
-#include "GNU_CC_INCLUDE:[sys]resource.h"
-#else
-#include_next <sys/resource.h>
-#endif
-
-#define __libgxx_sys_resource_h 1
-
-int getrusage(int, struct rusage*);
-int getrlimit (int resource, struct rlimit *rlp);
-int setrlimit _G_ARGS((int resource, const struct rlimit *rlp));
-long      ulimit(int, long);
-int       getpriority(int, int);
-int       setpriority(int, int, int);
-#endif
-}
-
-#endif 
diff --git a/libg++/g++-include/sys/select.h b/libg++/g++-include/sys/select.h
deleted file mode 100644
index 1c80ffe..0000000
--- a/libg++/g++-include/sys/select.h
+++ /dev/null
@@ -1,8 +0,0 @@
-/* Needed by Interviews for AIX. */
-#ifndef __libgxx_sys_select_h
-extern "C"
-{
-#include_next <sys/select.h>
-#define __libgxx_sys_select_h 1
-}
-#endif
diff --git a/libg++/g++-include/sys/signal.h b/libg++/g++-include/sys/signal.h
deleted file mode 100644
index 6821205..0000000
--- a/libg++/g++-include/sys/signal.h
+++ /dev/null
@@ -1,37 +0,0 @@
-// This may look like C code, but it is really -*- C++ -*-
-/* 
-Copyright (C) 1989 Free Software Foundation
-    written by Doug Lea (dl@rocky.oswego.edu)
-
-This file is part of the GNU C++ Library.  This library is free
-software; you can redistribute it and/or modify it under the terms of
-the GNU Library General Public License as published by the Free
-Software Foundation; either version 2 of the License, or (at your
-option) any later version.  This library is distributed in the hope
-that it will be useful, but WITHOUT ANY WARRANTY; without even the
-implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
-PURPOSE.  See the GNU Library General Public License for more details.
-You should have received a copy of the GNU Library General Public
-License along with this library; if not, write to the Free Software
-Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
-*/
-
-/* Partly for systems that think signal.h is is sys/ */
-/* But note that some systems that use sys/signal.h to define signal.h. */
-
-#ifndef __libgxx_sys_signal_h
-#if defined(__sys_signal_h_recursive) || defined(__signal_h_recursive)
-#include_next <sys/signal.h>
-#else
-#define __sys_signal_h_recursive
-
-extern "C" {
-#define signal __hide_signal
-#include_next <sys/signal.h>
-#undef signal
-}
-
-#define __libgxx_sys_signal_h 1
-#endif
-#endif
-
diff --git a/libg++/g++-include/sys/socket.h b/libg++/g++-include/sys/socket.h
deleted file mode 100644
index bb00101..0000000
--- a/libg++/g++-include/sys/socket.h
+++ /dev/null
@@ -1,49 +0,0 @@
-#ifndef __libgxx_sys_socket_h
-
-#include <_G_config.h>
-
-extern "C"
-{
-#ifdef __sys_socket_h_recursive
-#include_next <sys/socket.h>
-#else
-#define __sys_socket_h_recursive
-#include <time.h>
-
-#ifdef VMS
-#include "GNU_CC_INCLUDE:[sys]socket.h"
-#else
-#include_next <sys/socket.h>
-#endif
-
-#define __libgxx_sys_socket_h 1
-
-// void* in select, since different systems use int* or fd_set*
-int       select _G_ARGS((int, void*, void*, void*, struct timeval*));
-
-int       accept _G_ARGS((int, struct sockaddr*, int*));
-int       bind _G_ARGS((int, const void*, int));
-int       connect _G_ARGS((int, struct sockaddr*, int));
-int       getsockname _G_ARGS((int, struct sockaddr*, int*));
-int       getpeername _G_ARGS((int, struct sockaddr*, int*));
-int       getsockopt(int, int, int, void*, int*);
-int       listen(int, int);
-#ifndef hpux
-int       rcmd _G_ARGS((char**, int, const char*, const char*, const char*, int*));
-#endif
-int       recv(int, void*, int, int);
-int       recvfrom _G_ARGS((int, void*, int, int, void*, int *));
-int       recvmsg(int, struct msghdr*, int);
-int       rexec(char**, int, const char*, const char*, const char*, int*);
-int       rresvport(int*);
-int       send _G_ARGS((int, const void*, int, int));
-int       sendto _G_ARGS((int, const void*, int, int, void*, int));
-int       sendmsg _G_ARGS((int, const struct msghdr*, int));
-int       setsockopt _G_ARGS((int, int, int, const char*, int));
-int       shutdown(int, int);
-int       socket(int, int, int);
-int       socketpair(int, int, int, int sv[2]);
-#endif
-}
-
-#endif
diff --git a/libg++/g++-include/sys/stat.h b/libg++/g++-include/sys/stat.h
deleted file mode 100644
index cffe2a5..0000000
--- a/libg++/g++-include/sys/stat.h
+++ /dev/null
@@ -1,47 +0,0 @@
-#ifndef __libgxx_sys_stat_h
-
-extern "C"
-{
-#ifdef __sys_stat_h_recursive
-#include_next <sys/stat.h>
-#else
-#define __sys_stat_h_recursive
-#include <_G_config.h>
-#define chmod __hide_chmod
-#ifdef VMS
-#include "GNU_CC_INCLUDE:[sys]stat.h"
-#else
-#include_next <sys/stat.h>
-#endif
-#undef chmod
-
-#define __libgxx_sys_stat_h 1
-
-extern int       chmod  _G_ARGS((const char*, _G_mode_t));
-extern int       stat _G_ARGS((const char *path, struct stat *buf));
-extern int       lstat _G_ARGS((const char *path, struct stat *buf));
-extern int       fstat _G_ARGS((int fd, struct stat *buf));
-
-#ifndef S_ISDIR
-#define S_ISDIR(mode) (((mode) & S_IFMT) == S_IFDIR)
-#endif
-#ifndef S_ISBLK
-#define S_ISBLK(mode) (((mode) & S_IFMT) == S_IFBLK)
-#endif
-#ifndef S_ISCHR
-#define S_ISCHR(mode) (((mode) & S_IFMT) == S_IFCHR)
-#endif
-#ifndef S_ISFIFO
-#define S_ISFIFO(mode) (((mode) & S_IFMT) == S_IFFIFO)
-#endif
-#ifndef S_ISREG
-#define S_ISREG(mode) (((mode) & S_IFMT) == S_IFREG)
-#endif
-#if !defined(S_ISLNK) && defined(S_IFLNK)
-#define S_ISLNK(mode) (((mode) & S_IFMT) == S_IFLNK)
-#endif
-
-#endif
-}
-
-#endif
diff --git a/libg++/g++-include/sys/time.h b/libg++/g++-include/sys/time.h
deleted file mode 100644
index 81419b8..0000000
--- a/libg++/g++-include/sys/time.h
+++ /dev/null
@@ -1,41 +0,0 @@
-#ifndef __libgxx_sys_time_h
-
-extern "C"
-{
-#ifdef __sys_time_h_recursive
-#include_next <sys/time.h>
-#else
-#define __sys_time_h_recursive
-
-// Some system (e.g Ultrix) define these in sys/time.h.
-
-#ifndef __time_h_recursive
-#define time __hide_time
-#define clock __hide_clock
-#define gmtime __hide_gmtime
-#define localtime __hide_localtime
-#define ctime __hide_ctime
-#define asctime __hide_asctime
-#define strftime __hide_strftime
-#endif
-
-#ifdef VMS
-#include "GNU_CC_INCLUDE:[sys]resourcetime.h"
-#else
-#include_next <sys/time.h>
-#endif
-#undef __sys_time_h_recursive
-
-#define __libgxx_sys_time_h 1
-
-#undef clock
-#undef ctime
-#undef gmtime
-#undef localtime
-#undef time
-#undef asctime
-#undef strftime 
-#endif
-}
-
-#endif 
diff --git a/libg++/g++-include/sys/times.h b/libg++/g++-include/sys/times.h
deleted file mode 100644
index d026ca5..0000000
--- a/libg++/g++-include/sys/times.h
+++ /dev/null
@@ -1,20 +0,0 @@
-#ifndef __libgxx_sys_times_h
-
-extern "C"
-{
-#ifdef __sys_times_h_recursive
-#include_next <sys/times.h>
-#else
-#define __sys_times_h_recursive
-#include_next <sys/times.h>
-#define __libgxx_sys_times_h 1
-
-#include <_G_config.h>
-
-extern _G_clock_t times _G_ARGS((struct tms*));
-
-#endif
-}
-
-
-#endif
diff --git a/libg++/g++-include/sys/types.h b/libg++/g++-include/sys/types.h
deleted file mode 100644
index 67f4d31..0000000
--- a/libg++/g++-include/sys/types.h
+++ /dev/null
@@ -1,25 +0,0 @@
-#ifndef __libgxx_sys_types_h
-
-extern "C"
-{
-#ifdef __sys_types_h_recursive
-#include_next <sys/types.h>
-#else
-
-#define __sys_types_h_recursive
-#include <stddef.h>
-
-#ifdef VMS
-#include "GNU_CC_INCLUDE:[sys]types.h"
-#else
-#include_next <sys/types.h>
-#endif
-
-#define __libgxx_sys_types_h 1
-
-#endif
-}
-
-
-#endif
-
diff --git a/libg++/g++-include/sys/wait.h b/libg++/g++-include/sys/wait.h
deleted file mode 100644
index 18323fc..0000000
--- a/libg++/g++-include/sys/wait.h
+++ /dev/null
@@ -1,38 +0,0 @@
-#ifndef __libgxx_sys_wait_h
-
-#include <_G_config.h>
-
-extern "C" {
-#ifdef __sys_wait_h_recursive
-#include_next <sys/wait.h>
-#else
-#define __sys_wait_h_recursive
-
-
-#if _G_HAVE_SYS_WAIT
-#ifdef VMS
-#include "GNU_CC_INCLUDE:[sys]wait.h"
-#else
-#include_next <sys/wait.h>
-#endif
-#else /* !_G_HAVE_SYS_WAIT */
-/* Traditional definitions. */
-#define WEXITSTATUS(status) (((x) >> 8) & 0xFF)
-#define WIFSTOPPED(x) (((x) & 0xFF) == 0177)
-#define WIFEXITED(x) (! WIFSTOPPED(x) && WTERMSIG(x) == 0)
-#define WIFSIGNALED(x) (! WIFSTOPPED(x) && WTERMSIG(x) != 0)
-#define WTERMSIG(status) ((x) & 0x7F)
-#define WSTOPSIG(status) (((x) >> 8) & 0xFF)
-#endif /* !_G_HAVE_SYS_WAIT */
-
-#define __libgxx_sys_wait_h 1
-
-struct rusage;
-extern _G_pid_t wait _G_ARGS((int*));
-extern _G_pid_t waitpid _G_ARGS((_G_pid_t, int*, int));
-extern _G_pid_t wait3 _G_ARGS((int*, int options, struct rusage*));
-extern _G_pid_t wait4 _G_ARGS((int, int*, int, struct rusage*));
-#endif
-}
-
-#endif
diff --git a/libg++/g++-include/time.h b/libg++/g++-include/time.h
deleted file mode 100644
index 2e09e01..0000000
--- a/libg++/g++-include/time.h
+++ /dev/null
@@ -1,103 +0,0 @@
-#ifndef _G_time_h
-#define _G_time_h
-
-extern "C" {
-
-#ifdef __time_h_recursive
-#include_next <time.h>
-#else
-#define __time_h_recursive
-
-#include <_G_config.h>
-
-// A clean way to use and/or define time_t might allow removal of this crud.
-#ifndef __sys_time_h_recursive
-#define time __hide_time
-#define clock __hide_clock
-#define difftime __hide_difftime
-#define gmtime __hide_gmtime
-#define localtime __hide_localtime
-#define ctime __hide_ctime
-#define asctime __hide_asctime
-#define strftime __hide_strftime
-#endif
-#define mktime __hide_mktime
-#define tzset __hide_tzset
-#define tzsetwall __hide_tzsetwall
-#define getitimer __hide_getitimer
-#define setitimer __hide_setitimer
-#define gettimeofday __hide_gettimeofday
-#define settimeofday __hide_settimeofday
-
-#ifdef VMS
-	struct  unix_time
-	{
-		long int	tv_sec;
-		long int	tv_usec;
-	};
-
-	struct rusage
-	{
-		struct unix_time	ru_utime;
-	};
-
-#define RUSAGE_SELF 0		//define it, it will be unused
-#else
-#ifdef hpux
-#define _INCLUDE_POSIX_SOURCE
-#endif
-
-#include_next <time.h>
-#endif
-#undef __time_h_recursive
-
-#define time_h 1
-
-#undef time
-#undef clock
-#undef difftime
-#undef gmtime 
-#undef localtime 
-#undef asctime 
-#undef ctime 
-#undef mktime
-#undef strftime 
-#undef tzset 
-#undef tzsetwall 
-#undef getitimer
-#undef setitimer
-#undef gettimeofday
-#undef settimeofday
-
-extern char* asctime _G_ARGS((const struct tm*));
-extern char* ctime _G_ARGS((const _G_time_t*));
-double difftime _G_ARGS((_G_time_t, _G_time_t));
-extern struct tm* gmtime _G_ARGS((const _G_time_t*));
-extern struct tm* localtime _G_ARGS((const _G_time_t*));
-extern _G_time_t mktime(struct tm*);
-extern _G_size_t strftime _G_ARGS((char*,_G_size_t,const char*,const struct tm*));
-extern void tzset();
-extern void tzsetwall();
-
-extern int getitimer(int, struct itimerval*);
-extern int setitimer _G_ARGS((int, const struct itimerval*,struct itimerval*));
-extern int gettimeofday(struct timeval*, struct timezone*);
-extern int settimeofday _G_ARGS((const struct timeval*,const struct timezone*));
-extern int stime _G_ARGS((const _G_time_t*));
-extern int dysize(int);
-
-#if defined(___AIX__)
-int clock (void);
-#elif defined(hpux)
-unsigned long      clock(void);
-#else
-long      clock(void);
-#endif
-_G_time_t      time(_G_time_t*);
-unsigned  ualarm(unsigned, unsigned);
-unsigned  usleep(unsigned);
-void      profil _G_ARGS((unsigned short*, _G_size_t, unsigned int, unsigned));
-
-#endif
-}
-#endif
diff --git a/libg++/g++-include/unistd.h b/libg++/g++-include/unistd.h
deleted file mode 100644
index 9d9dcf9..0000000
--- a/libg++/g++-include/unistd.h
+++ /dev/null
@@ -1,153 +0,0 @@
-#ifndef _G_unistd_h
-#define _G_unistd_h 1
-
-#include <_G_config.h>
-
-extern "C" {
-
-#if _G_HAVE_UNISTD
-#ifndef _G_USE_PROTOS
-#define chmod __hide_chmod
-#define chown __hide_chown
-#define execl __hide_execl
-#define execlp __hide_execlp
-#define execle __hide_execle
-#define fchown __hide_fchown
-#define ioctl __hide_ioctl
-#define setgid __hide_setgid
-#define setuid __hide_setuid
-#endif
-#ifdef _AIX
-// AIX's unistd.h defines int rename (const char *old, const char *new).
-// This is not legal ANSI.  It causes a C++ syntax error (because of 'new').
-#define new __new
-#endif
-#include_next <unistd.h>
-#ifdef _AIX
-#undef new
-#endif
-#ifndef _G_USE_PROTOS
-#undef chmod
-#undef chown
-#undef execl
-#undef execle
-#undef execlp
-#undef fchown
-#undef ioctl
-#undef setgid
-#undef setuid
-#endif
-#else
-#ifndef SEEK_SET
-#define SEEK_SET        0
-#define SEEK_CUR        1
-#define SEEK_END        2
-#endif
-
-#ifndef F_OK
-#define F_OK            0
-#endif
-#ifndef X_OK
-#define X_OK            1
-#endif
-#ifndef W_OK
-#define W_OK            2
-#endif
-#ifndef R_K
-#define R_OK            4
-#endif
-#endif
-
-#ifdef __GNUG__
-extern void volatile _exit(int);
-#else
-void _exit(int);
-#endif
-
-extern unsigned alarm _G_ARGS((unsigned));
-extern int      brk _G_ARGS((void*));
-extern int      chdir _G_ARGS((const char*));
-extern int      chmod  _G_ARGS((const char*, _G_mode_t));
-extern int      chown (const char*, _G_uid_t, _G_gid_t);
-extern int      close _G_ARGS((int));
-extern char*    crypt _G_ARGS((const char*, const char*));
-extern int      dup _G_ARGS((int));
-extern int      dup2 _G_ARGS((int, int));
-extern void	encrypt _G_ARGS((char*, int));
-extern int      execl (const char*, const char *, ...);
-extern int      execle (const char*, const char *, ...);
-extern int      execlp (const char*, const char*, ...);
-extern int      exect _G_ARGS((const char*,  const char**,  char**));
-extern int      execv _G_ARGS((const char*,  const char * const *));
-extern int      execve _G_ARGS((const char*, const char * const *, const char * const *));
-extern int      execvp _G_ARGS((const char*,  const char * const *));
-extern int      fchown (int, _G_uid_t, _G_gid_t);
-extern _G_pid_t fork _G_ARGS((void));
-extern int      fsync _G_ARGS((int));
-extern int      ftruncate _G_ARGS((int, _G_off_t));
-extern char*    getcwd _G_ARGS((char*, _G_size_t));
-extern int      getdomainname _G_ARGS((char*, int));
-extern int      getdtablesize _G_ARGS((void));
-extern int      getgroups _G_ARGS((int, _G_gid_t*));
-extern _G_uid_t geteuid _G_ARGS((void));
-extern _G_gid_t getegid _G_ARGS((void));
-extern _G_gid_t getgid _G_ARGS((void));
-extern long     gethostid _G_ARGS((void));
-extern int      gethostname _G_ARGS((char*, int));
-extern _G_pid_t getpgrp _G_ARGS((...));
-extern _G_pid_t getpid _G_ARGS((void));
-extern _G_pid_t getppid _G_ARGS((void));
-extern char*    getlogin _G_ARGS((void));
-extern char*    getpass _G_ARGS((const char*));
-extern _G_uid_t getuid _G_ARGS((void));
-extern int      ioctl (int, int, ... );
-extern int      isatty _G_ARGS((int));
-extern int      link _G_ARGS((const char*, const char*));
-extern int      mkstemp _G_ARGS((char*));
-extern char*    mktemp _G_ARGS((char*));
-extern int      nice _G_ARGS((int));
-extern int      pause _G_ARGS((void));
-extern int      pipe _G_ARGS((int*));
-extern int      readlink _G_ARGS((const char*, char*, int));
-extern int      rename _G_ARGS((const char*, const char*));
-extern int      rmdir _G_ARGS((const char*)); 
-#ifdef __OSF1__
-extern char*	sbrk _G_ARGS((int));
-#else
-extern void*    sbrk _G_ARGS((int)); 
-#endif
-extern int      syscall _G_ARGS((int, ...));
-extern int      setgid (_G_gid_t);
-extern int      sethostname _G_ARGS((const char*, int));
-#ifdef _G_SYSV
-extern _G_pid_t setpgrp _G_ARGS((void));
-extern _G_pid_t setsid _G_ARGS((void));
-#else
-extern _G_pid_t setpgrp _G_ARGS((_G_pid_t, _G_pid_t));
-#endif
-extern int      setregid _G_ARGS((int, int));
-extern int      setreuid _G_ARGS((int, int));
-extern int      setuid (_G_uid_t);
-extern unsigned sleep _G_ARGS((unsigned));
-extern void     swab _G_ARGS((void*, void*, int));
-extern int      symlink _G_ARGS((const char*, const char*));
-extern long     sysconf _G_ARGS((int));
-extern int      truncate _G_ARGS((const char*, _G_off_t));
-extern char*    ttyname _G_ARGS((int));
-extern int      ttyslot _G_ARGS((void));
-//extern int   umask _G_ARGS((int)); /* commented out for now; wrong for SunOs4.1 */
-extern int      unlink _G_ARGS((const char*));
-extern _G_pid_t vfork _G_ARGS((void));
-extern int      vadvise _G_ARGS((int));
-extern int      vhangup _G_ARGS((void));
-extern _G_off_t lseek _G_ARGS((int, long, int));
-extern _G_ssize_t read _G_ARGS((int, void*, _G_size_t));
-extern _G_ssize_t write _G_ARGS((int, const void*, _G_size_t));
-extern int      access _G_ARGS((const char*, int));
-#ifndef hpux
-extern int      flock _G_ARGS((int, int));
-#endif
-
-}
-
-#endif
diff --git a/libg++/g++-include/values.h b/libg++/g++-include/values.h
deleted file mode 100644
index 56f8a94..0000000
--- a/libg++/g++-include/values.h
+++ /dev/null
@@ -1,174 +0,0 @@
-// This may look like C code, but it is really -*- C++ -*-
-/* 
-Copyright (C) 1988 Free Software Foundation
-    written by Doug Lea (dl@rocky.oswego.edu)
-
-This file is part of the GNU C++ Library.  This library is free
-software; you can redistribute it and/or modify it under the terms of
-the GNU Library General Public License as published by the Free
-Software Foundation; either version 2 of the License, or (at your
-option) any later version.  This library is distributed in the hope
-that it will be useful, but WITHOUT ANY WARRANTY; without even the
-implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
-PURPOSE.  See the GNU Library General Public License for more details.
-You should have received a copy of the GNU Library General Public
-License along with this library; if not, write to the Free Software
-Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
-*/
-
-
-#ifndef _values_h
-#define _values_h 1
-
-#define BITSPERBYTE 8
-#define BITS(type)  (BITSPERBYTE * (int)sizeof(type))
-
-#define CHARBITS    BITS(char)
-#define SHORTBITS   BITS(short)
-#define INTBITS     BITS(int)
-#define LONGBITS    BITS(long)
-#define PTRBITS     BITS(char*)
-#define DOUBLEBITS  BITS(double)
-#define FLOATBITS   BITS(float)
-
-#define MINSHORT    ((short)(1 << (SHORTBITS - 1)))
-#define MININT      (1 << (INTBITS - 1))
-#define MINLONG     (1L << (LONGBITS - 1))
-
-#define MAXSHORT    ((short)~MINSHORT)
-#define MAXINT      (~MININT)
-#define MAXLONG     (~MINLONG)
-
-#define HIBITS	MINSHORT
-#define HIBITL	MINLONG
-
-#if defined(sun) || defined(hp300) || defined(hpux) || defined(masscomp) || defined(sgi)
-#ifdef masscomp
-#define MAXDOUBLE							\
-({									\
-  double maxdouble_val;							\
-									\
-  __asm ("fmove%.d #0x7fefffffffffffff,%0"	/* Max double */	\
-	 : "=f" (maxdouble_val)						\
-	 : /* no inputs */);						\
-  maxdouble_val;							\
-})
-#define MAXFLOAT ((float) 3.40e+38)
-#else
-#define MAXDOUBLE   1.79769313486231470e+308
-#define MAXFLOAT    ((float)3.40282346638528860e+38)
-#endif
-#define MINDOUBLE   4.94065645841246544e-324
-#define MINFLOAT    ((float)1.40129846432481707e-45)
-#define _IEEE       1
-#define _DEXPLEN    11
-#define _FEXPLEN    8
-#define _HIDDENBIT  1
-#define DMINEXP     (-(DMAXEXP + DSIGNIF - _HIDDENBIT - 3))
-#define FMINEXP     (-(FMAXEXP + FSIGNIF - _HIDDENBIT - 3))
-#define DMAXEXP     ((1 << _DEXPLEN - 1) - 1 + _IEEE)
-#define FMAXEXP     ((1 << _FEXPLEN - 1) - 1 + _IEEE)
-
-#elif defined(sony) 
-#define MAXDOUBLE   1.79769313486231470e+308
-#define MAXFLOAT    ((float)3.40282346638528860e+38)
-#define MINDOUBLE   2.2250738585072010e-308
-#define MINFLOAT    ((float)1.17549435e-38)
-#define _IEEE       1
-#define _DEXPLEN    11
-#define _FEXPLEN    8
-#define _HIDDENBIT  1
-#define DMINEXP     (-(DMAXEXP + DSIGNIF - _HIDDENBIT - 3))
-#define FMINEXP     (-(FMAXEXP + FSIGNIF - _HIDDENBIT - 3))
-#define DMAXEXP     ((1 << _DEXPLEN - 1) - 1 + _IEEE)
-#define FMAXEXP     ((1 << _FEXPLEN - 1) - 1 + _IEEE)
-
-#elif defined(sequent)
-extern double _maxdouble, _mindouble;
-extern float _maxfloat, _minfloat;
-#define MAXDOUBLE	_maxdouble
-#define MAXFLOAT	_maxfloat
-#define MINDOUBLE	_mindouble
-#define MINFLOAT	_minfloat
-#define _IEEE       1
-#define _DEXPLEN    11
-#define _FEXPLEN    8
-#define _HIDDENBIT  1
-#define DMINEXP     (-(DMAXEXP - 3))
-#define FMINEXP     (-(FMAXEXP - 3))
-#define DMAXEXP     ((1 << _DEXPLEN - 1) - 1 + _IEEE)
-#define FMAXEXP     ((1 << _FEXPLEN - 1) - 1 + _IEEE)
-
-#elif defined(i386)
-#define MAXDOUBLE   1.79769313486231570e+308
-#define MAXFLOAT    ((float)3.40282346638528860e+38)
-#define MINDOUBLE   2.22507385850720140e-308
-#define MINFLOAT    ((float)1.17549435082228750e-38)
-#define _IEEE       0
-#define _DEXPLEN    11
-#define _FEXPLEN    8
-#define _HIDDENBIT  1
-#define DMINEXP     (-DMAXEXP)
-#define FMINEXP     (-FMAXEXP)
-#define DMAXEXP     ((1 << _DEXPLEN - 1) - 1 + _IEEE)
-#define FMAXEXP     ((1 << _FEXPLEN - 1) - 1 + _IEEE)
-
-/* from Andrew Klossner <andrew%frip.wv.tek.com@relay.cs.net> */
-#elif defined(m88k)
-	/* These are "good" guesses ...
-	   I'll figure out the true mins and maxes later, at the time I find
-	   out the mins and maxes that the compiler can tokenize. */
-#define MAXDOUBLE   1.79769313486231e+308
-#define MAXFLOAT    ((float)3.40282346638528e+38)
-#define MINDOUBLE   2.22507385850720e-308
-#define MINFLOAT    ((float)1.17549435082228e-38)
-#define _IEEE       1
-#define _DEXPLEN    11
-#define _FEXPLEN    8
-#define _HIDDENBIT  1
-#define DMINEXP     (1-DMAXEXP)
-#define FMINEXP     (1-FMAXEXP)
-#define DMAXEXP     ((1 << _DEXPLEN - 1) - 1 + _IEEE)
-#define FMAXEXP     ((1 << _FEXPLEN - 1) - 1 + _IEEE)
-
-#elif defined(convex)
-#define MAXDOUBLE   8.9884656743115785e+306
-#define MAXFLOAT    ((float) 1.70141173e+38)
-#define MINDOUBLE   5.5626846462680035e-308
-#define MINFLOAT    ((float) 2.93873588e-39)
-#define _IEEE       0
-#define _DEXPLEN    11
-#define _FEXPLEN    8
-#define _HIDDENBIT  1
-#define DMINEXP     (-DMAXEXP)
-#define FMINEXP     (-FMAXEXP)
-#define DMAXEXP     ((1 << _DEXPLEN - 1) - 1 + _IEEE)
-#define FMAXEXP     ((1 << _FEXPLEN - 1) - 1 + _IEEE)
-
-// #elif defined(vax)
-// use vax versions by default -- they seem to be the most conservative
-#else 
-
-#define MAXDOUBLE   1.701411834604692293e+38
-#define MINDOUBLE   (2.938735877055718770e-39)
-
-#define MAXFLOAT    1.7014117331926443e+38
-#define MINFLOAT    2.9387358770557188e-39
-
-#define _IEEE       0
-#define _DEXPLEN    8
-#define _FEXPLEN    8
-#define _HIDDENBIT  1
-#define DMINEXP     (-DMAXEXP)
-#define FMINEXP     (-FMAXEXP)
-#define DMAXEXP     ((1 << _DEXPLEN - 1) - 1 + _IEEE)
-#define FMAXEXP     ((1 << _FEXPLEN - 1) - 1 + _IEEE)
-#endif
-
-#define DSIGNIF     (DOUBLEBITS - _DEXPLEN + _HIDDENBIT - 1)
-#define FSIGNIF     (FLOATBITS  - _FEXPLEN + _HIDDENBIT - 1)
-#define DMAXPOWTWO  ((double)(1L << LONGBITS -2)*(1L << DSIGNIF - LONGBITS +1))
-#define FMAXPOWTWO  ((float)(1L << FSIGNIF - 1))
-
-#endif
-
diff --git a/libg++/genclass/Makefile.in b/libg++/genclass/Makefile.in
index 75d1dee..e15219b 100644
--- a/libg++/genclass/Makefile.in
+++ b/libg++/genclass/Makefile.in
@@ -93,10 +93,10 @@ gentest: $(srcdir)/gentest.sh $(srcdir)/expected.out
 
 .PHONY: install
 install:
-	-parent=`echo $(bindir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(bindir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(bindir) ] ; then true ; else mkdir $(bindir) ; fi
-	$(INSTALL) genclass $(bindir)/genclass
+	-if [ -d $(DESTDIR)$(bindir) ] ; then true ; else mkdir $(DESTDIR)$(bindir) ; fi
+	$(INSTALL) genclass $(DESTDIR)$(bindir)/genclass
 
 .PHONY: mostlyclean clean distclean realclean
 mostlyclean:
diff --git a/libg++/gperf/Makefile.in b/libg++/gperf/Makefile.in
index ed18c73..8e47fc3 100644
--- a/libg++/gperf/Makefile.in
+++ b/libg++/gperf/Makefile.in
@@ -106,10 +106,10 @@ check:
 install:
 	rootme=`pwd` ; export rootme ; cd src ; \
 		$(MAKE) $(FLAGS_TO_PASS) install
-	-parent=`echo $(man1dir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(man1dir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(man1dir) ] ; then true ; else mkdir $(man1dir) ; fi
-	$(INSTALL) $(srcdir)/gperf.1 $(man1dir)/gperf.1
+	-if [ -d $(DESTDIR)$(man1dir) ] ; then true ; else mkdir $(DESTDIR)$(man1dir) ; fi
+	$(INSTALL) $(srcdir)/gperf.1 $(DESTDIR)$(man1dir)/gperf.1
 
 .PHONY: info
 info: gperf.info
@@ -119,12 +119,12 @@ gperf.info: gperf.texinfo
 
 .PHONY: install-info
 install-info: gperf.info
-	-parent=`echo $(infodir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(infodir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(infodir) ] ; then true ; else mkdir $(infodir) ; fi
+	-if [ -d $(DESTDIR)$(infodir) ] ; then true ; else mkdir $(DESTDIR)$(infodir) ; fi
 	for i in *.info* ; do \
-		$(INSTALL_DATA) $$i $(infodir)/$$i ; \
-		chmod a-x $(infodir)/$$i ; \
+		$(INSTALL_DATA) $$i $(DESTDIR)$(infodir)/$$i ; \
+		chmod a-x $(DESTDIR)$(infodir)/$$i ; \
 	done
 
 .PHONY: clean-info
diff --git a/libg++/gperf/src/Makefile.in b/libg++/gperf/src/Makefile.in
index 771692b..fc263b4 100644
--- a/libg++/gperf/src/Makefile.in
+++ b/libg++/gperf/src/Makefile.in
@@ -113,10 +113,10 @@ gperf: $(OBJECTS)
 
 .PHONY: install
 install: $(PROGRAM)
-	-parent=`echo $(bindir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(bindir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(bindir) ] ; then true ; else mkdir $(bindir) ; fi
-	$(INSTALL) $(PROGRAM) $(bindir)/$(PROGRAM)
+	-if [ -d $(DESTDIR)$(bindir) ] ; then true ; else mkdir $(DESTDIR)$(bindir) ; fi
+	$(INSTALL) $(PROGRAM) $(DESTDIR)$(bindir)/$(PROGRAM)
 
 .PHONY: mostlyclean clean distclean realclean
 mostlyclean:
diff --git a/libg++/gperf/src/std-err.cc b/libg++/gperf/src/std-err.cc
index 98365c5..dd1cd67 100644
--- a/libg++/gperf/src/std-err.cc
+++ b/libg++/gperf/src/std-err.cc
@@ -73,10 +73,7 @@ Std_Err::report_error (char *format, ...)
             case 'f' : fprintf (stderr, "%g", va_arg (argp, double)); break;
             case 'n' : fputs (program_name ? program_name : "error", stderr); break;
             case 'p' : 
-              if (errno >= 0 && errno < sys_nerr) 
-                fprintf (stderr, "%s: %s", va_arg (argp, char *), sys_errlist[errno]);
-              else 
-                fprintf (stderr, "<unknown error> %d", errno);
+                fprintf (stderr, "%s: %s", va_arg (argp, char *), strerror(errno));
               break;
             case 's' : fputs (va_arg (argp, char *), stderr); break;
             }
diff --git a/libg++/iostream/Makefile.in b/libg++/iostream/Makefile.in
index 223151d..47d016e 100644
--- a/libg++/iostream/Makefile.in
+++ b/libg++/iostream/Makefile.in
@@ -209,12 +209,12 @@ iostream.info: $(srcdir)/iostream.texi
 
 .PHONY: install-include-files
 install-include-files:
-	[ -d $(gxx_includedir) ] || mkdir $(gxx_includedir)
+	[ -d $(DESTDIR)$(gxx_includedir) ] || mkdir $(DESTDIR)$(gxx_includedir)
 	cd $(srcdir); \
 	for FILE in $(USER_INCLUDES) ; do \
-		rm -f $(gxx_includedir)/$$FILE ; \
-		$(INSTALL_DATA) $$FILE $(gxx_includedir)/$$FILE ; \
-		chmod a-x $(gxx_includedir)/$$FILE ; \
+		rm -f $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+		$(INSTALL_DATA) $$FILE $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+		chmod a-x $(DESTDIR)$(gxx_includedir)/$$FILE ; \
 	done
 
 
diff --git a/libg++/iostream/floatconv.C b/libg++/iostream/floatconv.C
index e973181..0af7816 100644
--- a/libg++/iostream/floatconv.C
+++ b/libg++/iostream/floatconv.C
@@ -1315,10 +1315,7 @@ _Xstrtod
                         if (e1 > DBL_MAX_10_EXP) {
  ovfl:
                                 errno = ERANGE;
-#ifndef HUGE_VAL
-#define HUGE_VAL        1.7976931348623157E+308
-#endif
-                                rv = HUGE_VAL;
+                                rv = DBL_MAX;
                                 goto ret;
                                 }
                         if (e1 >>= 4) {
diff --git a/libg++/iostream/stdio/stdio.h b/libg++/iostream/stdio/stdio.h
index e723304..230752c 100644
--- a/libg++/iostream/stdio/stdio.h
+++ b/libg++/iostream/stdio/stdio.h
@@ -101,10 +101,10 @@ extern struct _fake_filebuf __std_filebuf_0, __std_filebuf_1, __std_filebuf_2;
 
 #define getc(fp) \
    ((fp)->_gptr >= (fp)->_egptr \
-      && __underflow((struct streambuf*)(fp)) == EOF ? EOF \
+      && __fix_underflow((struct streambuf*)(fp)) == EOF ? EOF \
    : *(unsigned char*)(fp)->_gptr++)
 #define putc(c, fp) \
-   (((fp)->_pptr >= (fp)->_epptr) ? __overflow((struct streambuf*)(fp), (unsigned char)c) \
+   (((fp)->_pptr >= (fp)->_epptr) ? __fix_overflow((struct streambuf*)(fp), (unsigned char)c) \
    : (unsigned char)(*(fp)->_pptr++ = c))
 #define putchar(c) putc(c, stdout)
 #define getchar() getc(stdin)
diff --git a/libg++/iostream/streambuf.C b/libg++/iostream/streambuf.C
index 0038f39..748b6e2 100644
--- a/libg++/iostream/streambuf.C
+++ b/libg++/iostream/streambuf.C
@@ -129,12 +129,7 @@ int streambuf::switch_to_put_mode()
 }
 #endif
 
-#ifdef _G_FRIEND_BUG
-int __underflow(register streambuf *sb) { return __UNDERFLOW(sb); }
-int __UNDERFLOW(register streambuf *sb)
-#else
-int __underflow(register streambuf *sb)
-#endif
+int __fix_underflow(register streambuf *sb)
 {
     if (sb->put_mode())
         if (sb->switch_to_get_mode() == EOF) return EOF;
@@ -199,12 +194,7 @@ int __underflow(register streambuf *sb)
     return sb->underflow();
 }
 
-#ifdef _G_FRIEND_BUG
-int __overflow(register streambuf *sb, int c) { return __OVERFLOW(sb, c); }
-int __OVERFLOW(register streambuf *sb, int c)
-#else
-int __overflow(streambuf* sb, int c)
-#endif
+int __fix_overflow(streambuf* sb, int c)
 {
     return sb->overflow(c);
 }
@@ -233,7 +223,7 @@ int streambuf::xsputn(register const char* s, int n)
 	    }
 	    more -= count;
 	}
-	if (more == 0 || __overflow(this, (unsigned char)*s++) == EOF)
+	if (more == 0 || __fix_overflow(this, (unsigned char)*s++) == EOF)
 	    break;
 	more--;
     }
@@ -289,7 +279,7 @@ int streambuf::xsgetn(char* s, int n)
 	    }
 	    more -= count;
 	}
-	if (more == 0 || __underflow(this) == EOF)
+	if (more == 0 || __fix_underflow(this) == EOF)
 	    break;
     }
     return n - more;
@@ -306,7 +296,7 @@ int streambuf::ignore(int n)
 	    _gptr += count;
 	    more -= count;
 	}
-	if (more == 0 || __underflow(this) == EOF)
+	if (more == 0 || __fix_underflow(this) == EOF)
 	    break;
     }
     return n - more;
diff --git a/libg++/iostream/streambuf.h b/libg++/iostream/streambuf.h
index b2254be..75b8732 100644
--- a/libg++/iostream/streambuf.h
+++ b/libg++/iostream/streambuf.h
@@ -22,6 +22,7 @@
 
 /* #define _G_IO_THROW */ /* Not implemented:  ios::failure */
 
+#define _G_fpos_t _G_off_t
 #include <_G_config.h>
 #ifdef _G_NEED_STDARG_H
 #include <stdarg.h>
@@ -40,12 +41,8 @@ class ostream; class streambuf; class backupbuf;
 #undef open
 #undef close
 
-#ifdef _G_FRIEND_BUG
-extern int __UNDERFLOW(streambuf*);
-extern int __OVERFLOW(streambuf*, int);
-#endif
-extern "C" int __underflow(streambuf*);
-extern "C" int __overflow(streambuf*, int);
+extern "C" int __fix_underflow(streambuf*);
+extern "C" int __fix_overflow(streambuf*, int);
 
 typedef _G_off_t streamoff;
 typedef _G_off_t streampos; // Should perhaps be _G_fpos_t ?
@@ -236,11 +233,7 @@ typedef ios::seek_dir _seek_dir;
 // You are guaranteed to be able to seek back to it if it is saving().
 class streammarker {
     friend class streambuf;
-#ifdef _G_FRIEND_BUG
-    friend int __UNDERFLOW(streambuf*);
-#else
-    friend int __underflow(streambuf*);
-#endif
+    friend int __fix_underflow(streambuf*);
     struct streammarker *_next;  // Only if saving()
     streambuf *_sbuf; // Only valid if saving().
     streampos _spos; // -2: means that _pos is valid.
@@ -291,11 +284,7 @@ struct streambuf : private __streambuf {
     friend class istream;
     friend class ostream;
     friend class streammarker;
-#ifdef _G_FRIEND_BUG
-    friend int __UNDERFLOW(streambuf*);
-#else
-    friend int __underflow(streambuf*);
-#endif
+    friend int __fix_underflow(streambuf*);
   protected:
     static streambuf* _list_all; /* List of open streambufs. */
     streambuf*& xchain() { return _chain; }
@@ -384,16 +373,16 @@ struct streambuf : private __streambuf {
     virtual int set_column(int);
     long sgetline(char* buf, _G_size_t n, char delim, int putback_delim);
     int sbumpc() {
-	if (_gptr >= _egptr && __underflow(this) == EOF) return EOF;
+	if (_gptr >= _egptr && __fix_underflow(this) == EOF) return EOF;
 	else return *(unsigned char*)_gptr++; }
     int sgetc() {
-	if (_gptr >= _egptr && __underflow(this) == EOF) return EOF;
+	if (_gptr >= _egptr && __fix_underflow(this) == EOF) return EOF;
 	else return *(unsigned char*)_gptr; }
     int snextc() {
-	if (_gptr >= _egptr && __underflow(this) == EOF) return EOF;
+	if (_gptr >= _egptr && __fix_underflow(this) == EOF) return EOF;
 	return _gptr++, sgetc(); }
     int sputc(int c) {
-	if (_pptr >= _epptr) return __overflow(this, (unsigned char)c);
+	if (_pptr >= _epptr) return __fix_overflow(this, (unsigned char)c);
 	else return *_pptr++ = c, (unsigned char)c; }
     void stossc() { if (_gptr < _egptr) _gptr++; }
     int vscan(char const *fmt0, _G_va_list ap, ios* stream = NULL);
diff --git a/libg++/src/Integer.cc b/libg++/src/Integer.cc
index 242c67b..b094c1c 100644
--- a/libg++/src/Integer.cc
+++ b/libg++/src/Integer.cc
@@ -39,14 +39,6 @@ Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 #include <new.h>
 #include <builtin.h>
 
-#ifndef HUGE_VAL
-#ifdef HUGE
-#define HUGE_VAL HUGE
-#else
-#define HUGE_VAL DBL_MAX
-#endif
-#endif
-
 /*
  Sizes of shifts for multiple-precision arithmetic.
  These should not be changed unless Integer representation
@@ -425,7 +417,7 @@ double Itodouble(const IntRep* rep)
     while (a != 0)
     {
       if (d >= bound)
-        return (rep->sgn == I_NEGATIVE) ? -HUGE_VAL : HUGE_VAL;
+        return (rep->sgn == I_NEGATIVE) ? -DBL_MAX : DBL_MAX;
       d *= 2.0;
       if (rep->s[i] & a)
         d += 1.0;
diff --git a/libg++/src/Makefile.in b/libg++/src/Makefile.in
index b6add92..b553e28 100644
--- a/libg++/src/Makefile.in
+++ b/libg++/src/Makefile.in
@@ -86,8 +86,8 @@ OBJS =  AllocRing.o Obstack.o builtin.o \
  Normal.o NegExp.o Weibull.o Erlang.o DiscUnif.o \
  Uniform.o Poisson.o HypGeom.o Geom.o Binomial.o \
  RNG.o ACG.o MLCG.o  RndInt.o  \
- Fix.o Fix16.o Fix24.o CursesW.o GetOpt.o $(EH_FILES) \
- new.o chr.o dtoa.o error.o gcd.o hash.o \
+ Fix.o Fix16.o Fix24.o GetOpt.o $(EH_FILES) \
+ new.o chr.o error.o gcd.o hash.o \
  lg.o fmtq.o ioob.o pow.o sqrt.o str.o timer.o \
  math.o compare.o \
  SLList.o DLList.o
diff --git a/libg++/src/SmplHist.cc b/libg++/src/SmplHist.cc
index a1fcd15..3bad7aa 100644
--- a/libg++/src/SmplHist.cc
+++ b/libg++/src/SmplHist.cc
@@ -21,15 +21,9 @@ Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 #include <stream.h>
 #include <SmplHist.h>
 #include <math.h>
-
-#ifndef HUGE_VAL
-#ifdef HUGE
-#define HUGE_VAL HUGE
-#else
 #include <float.h>
+
 #define HUGE_VAL DBL_MAX
-#endif
-#endif
 
 const int SampleHistogramMinimum = -2;
 const int SampleHistogramMaximum = -1;
diff --git a/libg++/src/SmplStat.cc b/libg++/src/SmplStat.cc
index 461bea4..9f47dd0 100644
--- a/libg++/src/SmplStat.cc
+++ b/libg++/src/SmplStat.cc
@@ -21,15 +21,9 @@ Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 #include <stream.h>
 #include <SmplStat.h>
 #include <math.h>
-
-#ifndef HUGE_VAL
-#ifdef HUGE
-#define HUGE_VAL HUGE
-#else
 #include <float.h>
+
 #define HUGE_VAL DBL_MAX
-#endif
-#endif
 
 // error handling
 
diff --git a/libiberty/Makefile.in b/libiberty/Makefile.in
index 3657dc3..1a233d2 100644
--- a/libiberty/Makefile.in
+++ b/libiberty/Makefile.in
@@ -124,21 +124,21 @@ INSTALL_DEST = libdir
 install: install_to_$(INSTALL_DEST)
 
 install_to_libdir: all
-	-parent=`echo $(libdir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(libdir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(libdir) ] ; then true ; else mkdir $(libdir) ; fi
-	$(INSTALL_DATA) $(TARGETLIB) $(libdir)/$(TARGETLIB).n
-	$(RANLIB) $(libdir)/$(TARGETLIB).n
-	mv -f $(libdir)/$(TARGETLIB).n $(libdir)/$(TARGETLIB)
+	-if [ -d $(DESTDIR)$(libdir) ] ; then true ; else mkdir $(DESTDIR)$(libdir) ; fi
+	$(INSTALL_DATA) $(TARGETLIB) $(DESTDIR)$(libdir)/$(TARGETLIB).n
+	$(RANLIB) $(DESTDIR)$(libdir)/$(TARGETLIB).n
+	mv -f $(DESTDIR)$(libdir)/$(TARGETLIB).n $(DESTDIR)$(libdir)/$(TARGETLIB)
 
 install_to_tooldir: all
-	-parent=`echo $(tooldir)|sed -e 's@/[^/]*$$@@'`; \
+	-parent=`echo $(DESTDIR)$(tooldir)|sed -e 's@/[^/]*$$@@'`; \
 	if [ -d $$parent ] ; then true ; else mkdir $$parent ; fi
-	-if [ -d $(tooldir) ] ; then true ; else mkdir $(tooldir) ; fi
-	-if [ -d $(tooldir)/lib ] ; then true ; else mkdir $(tooldir)/lib ; fi
-	$(INSTALL_DATA) $(TARGETLIB) $(tooldir)/lib/$(TARGETLIB).n
-	$(RANLIB) $(tooldir)/lib/$(TARGETLIB).n
-	mv -f $(tooldir)/lib/$(TARGETLIB).n $(tooldir)/lib/$(TARGETLIB)
+	-if [ -d $(DESTDIR)$(tooldir) ] ; then true ; else mkdir $(DESTDIR)$(tooldir) ; fi
+	-if [ -d $(DESTDIR)$(tooldir)/lib ] ; then true ; else mkdir $(DESTDIR)$(tooldir)/lib ; fi
+	$(INSTALL_DATA) $(TARGETLIB) $(DESTDIR)$(tooldir)/lib/$(TARGETLIB).n
+	$(RANLIB) $(DESTDIR)$(tooldir)/lib/$(TARGETLIB).n
+	mv -f $(DESTDIR)$(tooldir)/lib/$(TARGETLIB).n $(DESTDIR)$(tooldir)/lib/$(TARGETLIB)
 
 # The default configuration adds to libiberty all those functions that are
 # missing in libc.  More precisely, it includes whatever $(CC) fails to find.
diff --git a/libiberty/getopt.c b/libiberty/getopt.c
index 8f226cb..4cba73a 100644
--- a/libiberty/getopt.c
+++ b/libiberty/getopt.c
@@ -203,7 +203,7 @@ exchange (argv)
      char **argv;
 {
   int nonopts_size = (last_nonopt - first_nonopt) * sizeof (char *);
-  char **temp = (char **) __alloca (nonopts_size);
+  char **temp = (char **) alloca (nonopts_size);
 
   /* Interchange the two blocks of data in ARGV.  */
 
