diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 074e01e..5916dc3 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -252,7 +252,7 @@ NM_FOR_TARGET = ` \
   fi`
 
 # Dir to search for system headers.  Overridden by cross-make.
-SYSTEM_HEADER_DIR = /usr/include
+SYSTEM_HEADER_DIR = /usr/${CHOST}/include
 
 # Where to find some libiberty headers.
 HASHTAB_H   = $(srcdir)/../include/hashtab.h
@@ -698,7 +698,7 @@ HOST_ERRORS = $(HOST_PREFIX)errors.o
 # -I$(@D) and -I$(srcdir)/$(@D) cause the subdirectory of the file
 # currently being compiled, in both source trees, to be examined as well.
 INCLUDES = -I. -I$(@D) -I$(srcdir) -I$(srcdir)/$(@D) \
-	   -I$(srcdir)/config -I$(srcdir)/../include
+	   -I$(srcdir)/config -I$(srcdir)/../include -isystem ${SYSTEM_HEADER_DIR}
 
 # Always use -I$(srcdir)/config when compiling.
 .c.o:
