diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 89cd993..519845c 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -252,7 +252,7 @@ NM_FOR_TARGET = ` \
   fi`
 
 # Dir to search for system headers.  Overridden by cross-make.
-SYSTEM_HEADER_DIR = /usr/include
+SYSTEM_HEADER_DIR = $(build_tooldir)/include
 
 # Where to find some libiberty headers.
 HASHTAB_H   = $(srcdir)/../include/hashtab.h
@@ -1988,7 +1988,7 @@ PREPROCESSOR_DEFINES = \
   -DGPLUSPLUS_BACKWARD_INCLUDE_DIR=\"$(gcc_gxx_include_dir)/backward\" \
   -DLOCAL_INCLUDE_DIR=\"$(local_includedir)\" \
   -DCROSS_INCLUDE_DIR=\"$(gcc_tooldir)/sys-include\" \
-  -DTOOL_INCLUDE_DIR=\"$(gcc_tooldir)/include\"
+  -DTOOL_INCLUDE_DIR=\"$(build_tooldir)/include\"
 
 LIBCPP_OBJS =	cpplib.o cpplex.o cppmacro.o cppexp.o cppfiles.o \
 		cpphash.o cpperror.o cppinit.o cppdefault.o \
diff --git a/gcc/configure b/gcc/configure
index 19d1984..a03b49c 100755
--- a/gcc/configure
+++ b/gcc/configure
@@ -4366,14 +4366,7 @@ fi
 gcc_version_full=`grep version_string ${gcc_version_trigger} | sed -e 's/.*"\([^"]*\)".*/\1/'`
 gcc_version=`echo ${gcc_version_full} | sed -e 's/\([^ ]*\) .*/\1/'`
 
-# Compile in configure arguments.
-if test -f configargs.h ; then
-	# Being re-configured.
-	gcc_config_arguments=`grep configuration_arguments configargs.h | sed -e 's/.*"\([^"]*\)".*/\1/'`
-	gcc_config_arguments="$gcc_config_arguments : (reconfigured) $TOPLEVEL_CONFIGURE_ARGUMENTS"
-else
-	gcc_config_arguments="$TOPLEVEL_CONFIGURE_ARGUMENTS"
-fi
+gcc_config_arguments="$TOPLEVEL_CONFIGURE_ARGUMENTS"
 cat > configargs.h <<EOF
 /* Generated automatically. */
 static const char configuration_arguments[] = "$gcc_config_arguments";
diff --git a/gcc/configure.in b/gcc/configure.in
index 2315e5a..0b744a9 100644
--- a/gcc/configure.in
+++ b/gcc/configure.in
@@ -821,14 +821,7 @@ changequote(,)dnl
 gcc_version_full=`grep version_string ${gcc_version_trigger} | sed -e 's/.*"\([^"]*\)".*/\1/'`
 gcc_version=`echo ${gcc_version_full} | sed -e 's/\([^ ]*\) .*/\1/'`
 
-# Compile in configure arguments.
-if test -f configargs.h ; then
-	# Being re-configured.
-	gcc_config_arguments=`grep configuration_arguments configargs.h | sed -e 's/.*"\([^"]*\)".*/\1/'`
-	gcc_config_arguments="$gcc_config_arguments : (reconfigured) $TOPLEVEL_CONFIGURE_ARGUMENTS"
-else
-	gcc_config_arguments="$TOPLEVEL_CONFIGURE_ARGUMENTS"
-fi
+gcc_config_arguments="$TOPLEVEL_CONFIGURE_ARGUMENTS"
 cat > configargs.h <<EOF
 /* Generated automatically. */
 static const char configuration_arguments[] = "$gcc_config_arguments";
