diff --git a/gcc/config/alpha/linux.h b/gcc/config/alpha/linux.h
index 84d2657..4d97723 100644
--- a/gcc/config/alpha/linux.h
+++ b/gcc/config/alpha/linux.h
@@ -25,7 +25,7 @@ Boston, MA 02111-1307, USA.  */
 
 #undef CPP_PREDEFINES
 #define CPP_PREDEFINES \
-"-Dlinux -Dunix -Asystem=linux -D_LONGLONG -D__alpha__ " \
+"-Dlinux -Dunix -Asystem=linux -D_LONGLONG -D__alpha__ -D__LP64__ -D_LP64 " \
 SUB_CPP_PREDEFINES
 
 /* The GNU C++ standard library requires that these macros be defined.  */
diff --git a/gcc/config/mips/linux.h b/gcc/config/mips/linux.h
index 39adfe3..e276a8d 100644
--- a/gcc/config/mips/linux.h
+++ b/gcc/config/mips/linux.h
@@ -100,6 +100,7 @@ Boston, MA 02111-1307, USA.  */
 %{mabi=n32: -D_MIPS_SZPTR=32} \
 %{mabi=64: -D_MIPS_SZPTR=64} \
 %{!mabi*: -D_MIPS_SZPTR=32} \
+%{mabi=64: -D__LP64__ -D_LP64} \
 %{!mips*: -U__mips -D__mips} \
 %{mabi=32: -U__mips64} \
 %{mabi=n32: -D__mips64} \
diff --git a/gcc/config/s390/linux64.h b/gcc/config/s390/linux64.h
index fad88cc..da99e81 100644
--- a/gcc/config/s390/linux64.h
+++ b/gcc/config/s390/linux64.h
@@ -34,7 +34,7 @@ Boston, MA 02111-1307, USA.  */
 /* Names to predefine in the preprocessor for this target machine.  */
 
 #undef CPP_PREDEFINES
-#define CPP_PREDEFINES "-Dlinux -Asystem(linux) -Acpu(s390) -Amachine(s390) -D__s390__ -D__s390x__ -Asystem(unix) -Dunix -D__ELF__"
+#define CPP_PREDEFINES "-Dlinux -Asystem(linux) -Acpu(s390) -Amachine(s390) -D__s390__ -D__s390x__ -Asystem(unix) -Dunix -D__ELF__ -D__LP64__ -D_LP64"
 
 #undef STARTFILE_SPEC
 #define STARTFILE_SPEC \
