diff --git a/gcc/c-family/c-opts.c b/gcc/c-family/c-opts.c
index 7c22f3f87..22916fcb8 100644
--- a/gcc/c-family/c-opts.c
+++ b/gcc/c-family/c-opts.c
@@ -659,11 +659,13 @@ c_common_handle_option (size_t scode, const char *arg, int value,
       break;
 
     case OPT_std_c11:
+    case OPT_std_c17: /* Fake c17 for cmdline compatibility. */
       if (!preprocessing_asm_p)
 	set_std_c11 (true /* ISO */);
       break;
 
     case OPT_std_gnu11:
+    case OPT_std_gnu17: /* Fake gnu17 for cmdline compatibility. */
       if (!preprocessing_asm_p)
 	set_std_c11 (false /* ISO */);
       break;
diff --git a/gcc/c-family/c.opt b/gcc/c-family/c.opt
index 27dd7f9c6..e4945eff3 100644
--- a/gcc/c-family/c.opt
+++ b/gcc/c-family/c.opt
@@ -1702,6 +1702,14 @@ std=c1x
 C ObjC Alias(std=c11)
 Deprecated in favor of -std=c11.
 
+std=c17
+C ObjC
+Fake c17 for cmdline compatibility.
+
+std=c18
+C ObjC Alias(std=c17)
+Fake c17 for cmdline compatibility.
+
 std=c89
 C ObjC Alias(std=c90)
 Conform to the ISO 1990 C standard.
@@ -1759,6 +1767,14 @@ std=gnu1x
 C ObjC Alias(std=gnu11)
 Deprecated in favor of -std=gnu11.
 
+std=gnu17
+C ObjC
+Fake gnu17 for cmdline compatibility.
+
+std=gnu18
+C ObjC Alias(std=gnu17)
+Fake gnu17 for cmdline compatibility.
+
 std=gnu89
 C ObjC Alias(std=gnu90)
 Conform to the ISO 1990 C standard with GNU extensions.
