diff --git a/Makefile.in b/Makefile.in
index 5c123fc..a1a270a 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -136,7 +136,7 @@ local_prefix = /usr/local
 # Directory in which to put host dependent programs and libraries
 exec_prefix = $(prefix)
 # Directory in which to put the executable for the command `gcc'
-bindir = $(exec_prefix)/bin
+bindir = $(exec_prefix)/$(target)/gcc-bin/$(version)
 # Directory in which to put the directories used by the compiler.
 libdir = $(exec_prefix)/lib
 # Directory in which the compiler finds executables, libraries, etc.
@@ -144,7 +144,7 @@ libsubdir = $(libdir)/gcc-lib/$(target)/$(version)
 # Extension (if any) to put in installed man-page filename.
 manext = .1
 # Directory in which to put man pages.
-mandir = $(prefix)/man/man1
+mandir = $(prefix)/share/gcc-data/$(target)/$(version)/man/man1
 # Directory in which to find other cross-compilation tools and headers.
 # Used in install-cross.
 tooldir = $(prefix)/$(target)
@@ -1121,7 +1121,7 @@ cccp.o: cccp.c $(CONFIG_H) pcp.h version.c
 # is for compatibility with the current version of libg++.
 	$(CC) $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
 	  -DGCC_INCLUDE_DIR=\"$(libsubdir)/include\" \
-	  -DGPLUSPLUS_INCLUDE_DIR=\"$(libdir)/g++-include\" \
+	  -DGPLUSPLUS_INCLUDE_DIR=\"$(libsubdir)/include/g++\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(local_prefix)/include\" \
 	  -DCROSS_INCLUDE_DIR=\"$(libsubdir)/sys-include\" \
 	  -DTOOL_INCLUDE_DIR=\"$(tooldir)/include\" \
@@ -1144,7 +1144,7 @@ stamp-proto:	$(srcdir)/protoize.c getopt.h $(CONFIG_H)
 	touch stamp-proto
 	$(CC) -c $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
           -DGCC_INCLUDE_DIR=\"$(libsubdir)/include\" \
-          -DGPLUSPLUS_INCLUDE_DIR=\"$(libdir)/g++-include\" \
+          -DGPLUSPLUS_INCLUDE_DIR=\"$(libsubdir)/include/g++\" \
           -DCROSS_INCLUDE_DIR=\"$(libsubdir)/sys-include\" \
 	  -DTOOL_INCLUDE_DIR=\"$(tooldir)/include\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(local_prefix)/include\" \
@@ -1153,7 +1153,7 @@ stamp-proto:	$(srcdir)/protoize.c getopt.h $(CONFIG_H)
 	mv protoize.o unprotoize.o
 	$(CC) -c $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
           -DGCC_INCLUDE_DIR=\"$(libsubdir)/include\" \
-          -DGPLUSPLUS_INCLUDE_DIR=\"$(libdir)/g++-include\" \
+          -DGPLUSPLUS_INCLUDE_DIR=\"$(libsubdir)/include/g++\" \
           -DCROSS_INCLUDE_DIR=\"$(libsubdir)/sys-include\" \
 	  -DTOOL_INCLUDE_DIR=\"$(tooldir)/include\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(local_prefix)/include\" \
