diff --git a/Makefile.in b/Makefile.in
index 1be7362..8c591e5 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -136,7 +136,7 @@ local_prefix = /usr/local
 # Directory in which to put host dependent programs and libraries
 exec_prefix = $(prefix)
 # Directory in which to put the executable for the command `gcc'
-bindir = $(exec_prefix)/bin
+bindir = $(exec_prefix)/$(target)/gcc-bin/$(version)
 # Directory in which to put the directories used by the compiler.
 libdir = $(exec_prefix)/lib
 # Directory in which the compiler finds executables, libraries, etc.
@@ -144,7 +144,7 @@ libsubdir = $(libdir)/gcc-lib/$(target)/$(version)
 # Extension (if any) to put in installed man-page filename.
 manext = .1
 # Directory in which to put man pages.
-mandir = $(prefix)/man/man1
+mandir = $(prefix)/share/gcc-data/$(target)/$(version)/man/man1
 # Directory in which to find other cross-compilation tools and headers.
 # Used in install-cross.
 tooldir = $(prefix)/$(target)
