diff --git a/cccp.c b/cccp.c
index fad9756..85237c5 100644
--- a/cccp.c
+++ b/cccp.c
@@ -1854,8 +1854,8 @@ main (argc, argv)
     fatal ("I/O error on output");
 
   if (errors)
-    exit (FAILURE_EXIT_CODE);
-  exit (SUCCESS_EXIT_CODE);
+    _exit (FAILURE_EXIT_CODE);
+  _exit (SUCCESS_EXIT_CODE);
 
  perror:
   pfatal_with_name (in_fname);
@@ -5784,7 +5784,7 @@ do_error (buf, limit, op, keyword)
   copy[length] = 0;
   SKIP_WHITE_SPACE (copy);
   error ("#error %s", copy);
-  exit (FAILURE_EXIT_CODE);
+  _exit (FAILURE_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
@@ -8460,7 +8460,7 @@ fatal (str, arg)
   fprintf (stderr, "%s: ", progname);
   fprintf (stderr, str, arg);
   fprintf (stderr, "\n");
-  exit (FAILURE_EXIT_CODE);
+  _exit (FAILURE_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -8487,9 +8487,9 @@ pfatal_with_name (name)
 {
   perror_with_name (name);
 #ifdef VMS
-  exit (vaxc$errno);
+  _exit (vaxc$errno);
 #else
-  exit (FAILURE_EXIT_CODE);
+  _exit (FAILURE_EXIT_CODE);
 #endif
 }
 
diff --git a/collect2.c b/collect2.c
index bc118ed..efe8b17 100644
--- a/collect2.c
+++ b/collect2.c
@@ -233,7 +233,7 @@ my_exit (status)
   if (o_file != 0 && o_file[0])
     maybe_unlink (o_file);
 
-  exit (status);
+  _exit (status);
 }
 
 
diff --git a/dbxout.c b/dbxout.c
index 09f6ece..ac234e4 100644
--- a/dbxout.c
+++ b/dbxout.c
@@ -346,7 +346,7 @@ abspath (rel_filename)
 
               	  fprintf (stderr, "%s: invalid file name: %s\n",
 			   pname, rel_filename);
-              	  exit (1);
+              	  _exit (1);
               	}
                     *++outp = '\0';
                     continue;
diff --git a/enquire.c b/enquire.c
index fab663f..49d39a6 100644
--- a/enquire.c
+++ b/enquire.c
@@ -585,13 +585,13 @@ int setmode(s) char *s; {
 #endif
 
 Procedure farewell(bugs) int bugs; {
-	if (bugs == 0) exit(0);
+	if (bugs == 0) _exit(0);
 	printf("\n%sFor hints on dealing with the ", co);
 	if (bugs == 1) printf("problem");
 	else printf("%d problems", bugs);
 	printf(" above\n   see the section 'TROUBLESHOOTING' in the file ");
 	printf("%s%s\n", FILENAME, oc);
-	exit(bugs);
+	_exit(bugs);
 }
 
 /* The program has received a signal where it wasn't expecting one */
diff --git a/gcc.c b/gcc.c
index efb0523..5831efb 100644
--- a/gcc.c
+++ b/gcc.c
@@ -1526,7 +1526,7 @@ pexecute (search_flag, program, argv, not_last)
       /* Exec the program.  */
       (*func) (program, argv);
       perror_exec (program);
-      exit (-1);
+      _exit (-1);
       /* NOTREACHED */
       return 0;
 
@@ -1891,12 +1891,12 @@ process_command (argc, argv)
 	  printf ("*predefines:\n%s\n\n", cpp_predefines);
 	  printf ("*cross_compile:\n%d\n\n", cross_compile);
 
-	  exit (0);
+	  _exit (0);
 	}
       else if (! strcmp (argv[i], "-dumpversion"))
 	{
 	  printf ("%s\n", version_string);
-	  exit (0);
+	  _exit (0);
 	}
       else if (! strcmp (argv[i], "-print-libgcc-file-name"))
 	{
@@ -3182,7 +3182,7 @@ main (argc, argv)
   if (print_libgcc_file_name)
     {
       printf ("%s\n", find_file ("libgcc.a"));
-      exit (0);
+      _exit (0);
     }
 
   /* Obey some of the options.  */
@@ -3339,7 +3339,7 @@ main (argc, argv)
     delete_failure_queue ();
   delete_temp_files ();
 
-  exit (error_count);
+  _exit (error_count);
   /* NOTREACHED */
   return 0;
 }
@@ -3527,7 +3527,7 @@ fatal VPROTO((char *format, ...))
   va_end (ap);
   fprintf (stderr, "\n");
   delete_temp_files ();
-  exit (1);
+  _exit (1);
 }
 
 static void
@@ -3559,7 +3559,7 @@ fatal (msg, arg1, arg2)
 {
   error (msg, arg1, arg2);
   delete_temp_files ();
-  exit (1);
+  _exit (1);
 }
 
 static void
diff --git a/genattr.c b/genattr.c
index c3acfb2..6975b0c 100644
--- a/genattr.c
+++ b/genattr.c
@@ -146,7 +146,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genattr: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -181,7 +181,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -244,7 +244,7 @@ from the machine description file `md'.  */\n\n");
     }
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genattrtab.c b/genattrtab.c
index cfef21d..b46e64c 100644
--- a/genattrtab.c
+++ b/genattrtab.c
@@ -4450,7 +4450,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genattrtab: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -4487,7 +4487,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -4630,7 +4630,7 @@ from the machine description file `md'.  */\n\n");
     write_function_unit_info ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/gencodes.c b/gencodes.c
index 5d84c02..5f2cc50 100644
--- a/gencodes.c
+++ b/gencodes.c
@@ -81,7 +81,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "gencodes: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -111,7 +111,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -153,7 +153,7 @@ from the machine description file `md'.  */\n\n");
   printf ("#endif /* MAX_INSN_CODE */\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genconfig.c b/genconfig.c
index 64fccb9..d8a0ac1 100644
--- a/genconfig.c
+++ b/genconfig.c
@@ -239,7 +239,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genconfig: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -269,7 +269,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -320,7 +320,7 @@ from the machine description file `md'.  */\n\n");
     printf ("#define HAVE_lo_sum\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genemit.c b/genemit.c
index 05f60f3..519f448 100644
--- a/genemit.c
+++ b/genemit.c
@@ -692,7 +692,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genemit: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -722,7 +722,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -789,7 +789,7 @@ from the machine description file `md'.  */\n\n");
   output_init_mov_optab ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genextract.c b/genextract.c
index 81a92b9..3515925 100644
--- a/genextract.c
+++ b/genextract.c
@@ -348,7 +348,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genextract: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -404,7 +404,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -532,7 +532,7 @@ from the machine description file `md'.  */\n\n");
   printf ("    }\n}\n");
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genflags.c b/genflags.c
index edb5443..84df230 100644
--- a/genflags.c
+++ b/genflags.c
@@ -101,7 +101,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genflags: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -131,7 +131,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -154,7 +154,7 @@ from the machine description file `md'.  */\n\n");
     }
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genoutput.c b/genoutput.c
index 76d9693..bc9326c 100644
--- a/genoutput.c
+++ b/genoutput.c
@@ -898,7 +898,7 @@ fatal (s, a1, a2, a3, a4)
   fprintf (stderr, "genoutput: ");
   fprintf (stderr, s, a1, a2, a3, a4);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -937,7 +937,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -971,7 +971,7 @@ main (argc, argv)
   output_epilogue ();
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genpeep.c b/genpeep.c
index eb14a69..aef0808 100644
--- a/genpeep.c
+++ b/genpeep.c
@@ -394,7 +394,7 @@ fatal (s, a1, a2)
   fprintf (stderr, "genpeep: ");
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -426,7 +426,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -483,7 +483,7 @@ from the machine description file `md'.  */\n\n");
   printf ("rtx peep_operand[%d];\n", max_opno + 1);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/genrecog.c b/genrecog.c
index 83a6d26..89d1f70 100644
--- a/genrecog.c
+++ b/genrecog.c
@@ -1630,7 +1630,7 @@ fatal (s, a1, a2)
   fprintf (stderr, s, a1, a2);
   fprintf (stderr, "\n");
   fprintf (stderr, "after %d definitions\n", next_index);
-  exit (FATAL_EXIT_CODE);
+  _exit (FATAL_EXIT_CODE);
 }
 
 /* More 'friendly' abort that prints the line and file.
@@ -1663,7 +1663,7 @@ main (argc, argv)
   if (infile == 0)
     {
       perror (argv[1]);
-      exit (FATAL_EXIT_CODE);
+      _exit (FATAL_EXIT_CODE);
     }
 
   init_rtl ();
@@ -1746,7 +1746,7 @@ from the machine description file `md'.  */\n\n");
   write_subroutine (split_tree.first, SPLIT);
 
   fflush (stdout);
-  exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
+  _exit (ferror (stdout) != 0 ? FATAL_EXIT_CODE : SUCCESS_EXIT_CODE);
   /* NOTREACHED */
   return 0;
 }
diff --git a/getopt.c b/getopt.c
index 49b99ce..3b0b504 100644
--- a/getopt.c
+++ b/getopt.c
@@ -667,7 +667,7 @@ main (argc, argv)
       printf ("\n");
     }
 
-  exit (0);
+  _exit (0);
 }
 
 #endif /* TEST */
diff --git a/getopt1.c b/getopt1.c
index ae7930d..5e208de 100644
--- a/getopt1.c
+++ b/getopt1.c
@@ -152,7 +152,7 @@ main (argc, argv)
       printf ("\n");
     }
 
-  exit (0);
+  _exit (0);
 }
 
 #endif /* TEST */
diff --git a/objc-actions.c b/objc-actions.c
index 24647e8..0252220 100644
--- a/objc-actions.c
+++ b/objc-actions.c
@@ -1331,7 +1331,7 @@ build_ivar_chain (interface)
 	  /* fatal did not work with 2 args...should fix */
 	  error ("Cannot find interface declaration for `%s', superclass of `%s'",
 		 IDENTIFIER_POINTER (super_name), IDENTIFIER_POINTER (my_name));
-	  exit (34);
+	  _exit (34);
         }
       if (super_interface == interface)
         {
@@ -3485,7 +3485,7 @@ start_class (code, class_name, super_name)
 	{
 	  error ("Cannot find interface declaration for `%s'",
 		 IDENTIFIER_POINTER (class_name));
-	  exit (1);
+	  _exit (1);
 	}
       else
         add_category (class_category_is_assoc_with, class);
@@ -3513,7 +3513,7 @@ start_class (code, class_name, super_name)
         {
 	  error ("Cannot find interface declaration for `%s'",
 		 IDENTIFIER_POINTER (class_name));
-	  exit (1);
+	  _exit (1);
         }
     }
   return class;
diff --git a/protoize.c b/protoize.c
index 36be093..bd2d11d 100644
--- a/protoize.c
+++ b/protoize.c
@@ -140,7 +140,7 @@ extern int fprintf (FILE *, const char *, ...);
 extern int printf (const char *, ...);
 extern int open (const char *, int, ...);
 #endif /* 0 */
-extern void exit ();
+extern void _exit ();
 extern pointer_type malloc ();
 extern pointer_type realloc ();
 extern void free ();
@@ -620,7 +620,7 @@ xmalloc (byte_count)
     {
       fprintf (stderr, "\n%s: fatal error: can't allocate %u more bytes of memory\n",
 	       pname, byte_count);
-      exit (1);
+      _exit (1);
       return 0;		/* avoid warnings */
     }
   else
@@ -641,7 +641,7 @@ xrealloc (old_space, byte_count)
     {
       fprintf (stderr, "\n%s: fatal error: can't allocate %u more bytes of memory\n",
 	       pname, byte_count);
-      exit (1);
+      _exit (1);
       return 0;		/* avoid warnings */
     }
   else
@@ -694,7 +694,7 @@ void
 fancy_abort ()
 {
   fprintf (stderr, "%s: internal abort\n", pname);
-  exit (1);
+  _exit (1);
 }
 
 /* Make a duplicate of a given string in a newly allocated area.  */
@@ -785,7 +785,7 @@ usage ()
   fprintf (stderr, "%s: usage '%s [ -VqfnkNlgC ] [ -B <diname> ] [ filename ... ]'\n",
 	   pname, pname);
 #endif /* !defined (UNPROTOIZE) */
-  exit (1);
+  _exit (1);
 }
 
 /* Return true if the given filename (assumed to be an absolute filename)
@@ -1279,7 +1279,7 @@ abspath (cwd, rel_filename)
 
               	  fprintf (stderr, "%s: invalid file name: %s\n",
 			   pname, rel_filename);
-              	  exit (1);
+              	  _exit (1);
               	}
                     *++outp = '\0';
                     continue;
@@ -1449,7 +1449,7 @@ aux_info_corrupted ()
 {
   fprintf (stderr, "\n%s: fatal error: aux info file corrupted at line %d\n",
 	   pname, current_aux_info_lineno);
-  exit (1);
+  _exit (1);
 }
 
 /* ??? This comment is vague.  Say what the condition is for.  */
@@ -1769,7 +1769,7 @@ save_def_or_dec (l, is_syscalls)
 			 def_dec_p->hash_entry->symbol,
 			 def_dec_p->file->hash_entry->symbol,
 			 def_dec_p->line);
-                exit (1);
+                _exit (1);
               }
             free_def_dec (def_dec_p);
             return;
@@ -2080,7 +2080,7 @@ gen_aux_info_file (base_filename)
 	  write (f, ": ", 2);
 	  write (f, sys_errlist[e], strlen (sys_errlist[e]));
 	  write (f, "\n", 1);
-          exit (1);
+          _exit (1);
         }
       return 1;		/* Never executed.  */
     }
@@ -4458,7 +4458,7 @@ main (argc, argv)
     {
       fprintf (stderr, "%s: cannot get working directory: %s\n",
 	       pname, sys_errlist[errno]);
-      exit (1);
+      _exit (1);
     }
 
   /* By default, convert the files in the current directory.  */
@@ -4583,8 +4583,8 @@ main (argc, argv)
       do_processing ();
     }
   if (errors)
-    exit (1);
+    _exit (1);
   else
-    exit (0);
+    _exit (0);
   return 1;
 }
diff --git a/toplev.c b/toplev.c
index a814984..f0ba4bb 100644
--- a/toplev.c
+++ b/toplev.c
@@ -694,7 +694,7 @@ pfatal_with_name (name)
 {
   fprintf (stderr, "%s: ", progname);
   perror (name);
-  exit (35);
+  _exit (35);
 }
 
 void
@@ -702,7 +702,7 @@ fatal_io_error (name)
      char *name;
 {
   fprintf (stderr, "%s: %s: I/O error\n", progname, name);
-  exit (35);
+  _exit (35);
 }
 
 void
@@ -711,7 +711,7 @@ fatal (s, v)
      int v;
 {
   error (s, v);
-  exit (34);
+  _exit (34);
 }
 
 /* Called to give a better error message when we don't have an insn to match
@@ -3074,10 +3074,10 @@ You Lose!  You must define PREFERRED_DEBUGGING_TYPE!
 #endif /* not OS2 */
 
   if (errorcount)
-    exit (FATAL_EXIT_CODE);
+    _exit (FATAL_EXIT_CODE);
   if (sorrycount)
-    exit (FATAL_EXIT_CODE);
-  exit (SUCCESS_EXIT_CODE);
+    _exit (FATAL_EXIT_CODE);
+  _exit (SUCCESS_EXIT_CODE);
   return 34;
 }
 
