diff --git a/Makefile.in b/Makefile.in
index 99c3173..ced98cc 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -651,9 +651,9 @@ libmingwthrd_old.a: lib%.a: %.def
 	$(AR) $(ARFLAGS) $@ $(call extra_objects_filter,$^)
 
 all-mingwrt-dll install-mingwrt-dll-files: mingwm10.dll
-mingwm10.dll: mingwthrd_old.def
+mingwm10.dll: mingwthrd_old.def libmingwex.a
 	$(CC) -shared -o $@ -Wl,--image-base,0x6FBC0000 \
-	  -Wl,--entry,_DllMainCRTStartup@12 $^
+	  -Wl,--entry,_DllMainCRTStartup@12 -L. -lmingwex $^
 
 # Code profiling support, using the GNU profiling library; (note
 # that applications using this library MUST be licensed to comply
