diff --git a/include/stdio.h b/include/stdio.h
index 570a2d2..595b850 100644
--- a/include/stdio.h
+++ b/include/stdio.h
@@ -467,32 +467,6 @@ _CRTIMP __cdecl __MINGW_NOTHROW  int vsprintf (char *, const char *, __VALIST);
 
 #endif	/* !__USE_MINGW_ANSI_STDIO */
 
-#if __GNUC__ && defined _ISOC99_SOURCE
-/* Although MinGW implementations of the ISO-C99 snprintf() and
- * vsnprintf() functions do not conflict with any implementation
- * in MSVCRT.DLL, (because MSVCRT.DLL does not implement either),
- * there are -Wformat attribute conflicts with the GCC built-in
- * prototypes associated with each; by providing the following
- * in-line function implementations, which will override GCC's
- * built-in prototypes, we may avoid these conflicts.
- */
-__mingw_stdio_redirect__
-int snprintf (char *__buf, size_t __len, const char *__format, ...)
-{
-  register int __retval;
-  __builtin_va_list __local_argv; __builtin_va_start( __local_argv, __format );
-  __retval = __mingw_vsnprintf( __buf, __len, __format, __local_argv );
-  __builtin_va_end( __local_argv );
-  return __retval;
-}
-
-__mingw_stdio_redirect__
-int vsnprintf (char *__buf, size_t __len, const char *__format, __VALIST __local_argv)
-{
-  return __mingw_vsnprintf( __buf, __len, __format, __local_argv );
-}
-#endif	/* __GNUC__ && defined _ISOC99_SOURCE */
-
 /* Regardless of user preference, always offer these alternative
  * entry points, for direct access to the MSVCRT implementations,
  * with ms_printf -Wformat checking in each case.
