diff --git a/include/stdio.h b/include/stdio.h
index 570a2d2..3a8b7b0 100644
--- a/include/stdio.h
+++ b/include/stdio.h
@@ -467,7 +467,7 @@ _CRTIMP __cdecl __MINGW_NOTHROW  int vsprintf (char *, const char *, __VALIST);
 
 #endif	/* !__USE_MINGW_ANSI_STDIO */
 
-#if __GNUC__ && defined _ISOC99_SOURCE
+#if (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 2)) && defined _ISOC99_SOURCE
 /* Although MinGW implementations of the ISO-C99 snprintf() and
  * vsnprintf() functions do not conflict with any implementation
  * in MSVCRT.DLL, (because MSVCRT.DLL does not implement either),
