diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 5492e756d..6cf7e10bb 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -194,6 +194,20 @@ GCC_WARN_CXXFLAGS = $(LOOSE_WARN) $($(@D)-warn) $(NOCOMMON_FLAG) $($@-warn)
 build/gengtype-lex.o-warn = -Wno-error
 gengtype-lex.o-warn = -Wno-error
 expmed.o-warn = -Wno-error
+build/genhooks.o-warn = -Wno-error=unused-result
+diagnostic.o-warn = -Wno-error=unused-result
+final.o-warn = -Wno-error=unused-result
+gcc.o-warn = -Wno-error=unused-result
+lto-wrapper.o-warn = -Wno-error=unused-result
+lto/lto.o-warn = -Wno-error=unused-result
+targhooks.o-warn = -Wno-error=unused-result
+toplev.o-warn = -Wno-error=unused-result
+tree-ssa-structalias.o-warn = -Wno-error=unused-result
+fortran/trans.o-warn = -Wno-error=unused-result
+fortran/trans-array.o-warn = -Wno-error=unused-result
+fortran/trans-decl.o-warn = -Wno-error=unused-result
+fortran/trans-expr.o-warn = -Wno-error=unused-result
+fortran/trans-io.o-warn = -Wno-error=unused-result
 
 # All warnings have to be shut off in stage1 if the compiler used then
 # isn't gcc; configure determines that.  WARN_CFLAGS will be either
diff --git a/libcpp/Makefile.in b/libcpp/Makefile.in
index 2353c9c04..28ac0c0b4 100644
--- a/libcpp/Makefile.in
+++ b/libcpp/Makefile.in
@@ -36,7 +36,7 @@ CFLAGS = @CFLAGS@
 WARN_CFLAGS = @warn@ @c_warn@ @WARN_PEDANTIC@ @WERROR@
 CXX = @CXX@
 CXXFLAGS = @CXXFLAGS@
-WARN_CXXFLAGS = @warn@ @WARN_PEDANTIC@ @WERROR@
+WARN_CXXFLAGS = @warn@ @WARN_PEDANTIC@ @WERROR@ -Wno-error=unused-result
 CPP = @CPP@
 CPPFLAGS = @CPPFLAGS@
 EXEEXT = @EXEEXT@
