diff --git a/ld/ldmisc.c b/ld/ldmisc.c
index e204811a..4aa417c3 100644
--- a/ld/ldmisc.c
+++ b/ld/ldmisc.c
@@ -372,9 +372,6 @@ vfinfo (FILE *fp, const char *fmt, va_list ap, bool is_warning)
 				&& filename_cmp (last_file, filename) != 0)
 			    || strcmp (last_function, functionname) != 0)
 			  {
-			    lfinfo (fp, _("%pB: in function `%pT':\n"),
-				    abfd, functionname);
-
 			    last_bfd = abfd;
 			    free (last_file);
 			    last_file = NULL;
