diff --git a/bfd/elflink.c b/bfd/elflink.c
index f8521426..f08af04a 100644
--- a/bfd/elflink.c
+++ b/bfd/elflink.c
@@ -13336,9 +13336,6 @@ bfd_elf_final_link (bfd *abfd, struct bfd_link_info *info)
 		  else if (bfd_link_pde (info))
 		    info->callbacks->einfo
 		      (_("%P: warning: creating DT_TEXTREL in a PDE\n"));
-		  else
-		    info->callbacks->einfo
-		      (_("%P: warning: creating DT_TEXTREL in a PIE\n"));
 		  break;
 		}
 	    }
@@ -15304,12 +15301,6 @@ _bfd_elf_maybe_set_textrel (struct elf_link_hash_entry *h, void *inf)
 				"in read-only section `%pA'\n"),
 			      sec->owner, h->root.root.string, sec);
 
-      if (bfd_link_textrel_check (info))
-	/* xgettext:c-format */
-	info->callbacks->einfo (_("%P: %pB: warning: relocation against `%s' "
-				  "in read-only section `%pA'\n"),
-				sec->owner, h->root.root.string, sec);
-
       /* Not an error, just cut short the traversal.  */
       return false;
     }
diff --git a/gas/config/tc-mips.c b/gas/config/tc-mips.c
index f632e94e..7e1403b4 100644
--- a/gas/config/tc-mips.c
+++ b/gas/config/tc-mips.c
@@ -4003,8 +4003,6 @@ check_fpabi (int fpabi)
       break;
 
     case Val_GNU_MIPS_ABI_FP_OLD_64:
-      as_warn (_(".gnu_attribute %d,%d is no longer supported"),
-	       Tag_GNU_MIPS_ABI_FP, fpabi);
       break;
 
     case Val_GNU_MIPS_ABI_FP_NAN2008:
