diff --git a/gas/config/tc-mips.c b/gas/config/tc-mips.c
index f632e94e..7e1403b4 100644
--- a/gas/config/tc-mips.c
+++ b/gas/config/tc-mips.c
@@ -4003,8 +4003,6 @@ check_fpabi (int fpabi)
       break;
 
     case Val_GNU_MIPS_ABI_FP_OLD_64:
-      as_warn (_(".gnu_attribute %d,%d is no longer supported"),
-	       Tag_GNU_MIPS_ABI_FP, fpabi);
       break;
 
     case Val_GNU_MIPS_ABI_FP_NAN2008:
