diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 59cb73a4..12463aae 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -7078,7 +7078,8 @@ grokdeclarator (const cp_declarator *declarator,
 	 common.  With no options, it is allowed.  With -Wreturn-type,
 	 it is a warning.  It is only an error with -pedantic-errors.  */
       is_main = (funcdef_flag
-		 && dname && MAIN_NAME_P (dname)
+		 && dname && TREE_CODE (dname) == IDENTIFIER_NODE
+		 && MAIN_NAME_P (dname)
 		 && ctype == NULL_TREE
 		 && in_namespace == NULL_TREE
 		 && current_namespace == global_namespace);
diff --git a/gcc/testsuite/g++.dg/template/crash102.C b/gcc/testsuite/g++.dg/template/crash102.C
new file mode 100644
index 00000000..fd3c36ee
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/crash102.C
@@ -0,0 +1,5 @@
+// PR c++/45043
+
+template < typename > class A;
+template < typename T > A < T >::B::~B () // { dg-error "type" }
+{}
