diff --git a/gcc/builtins.c b/gcc/builtins.c
index 88ecc602..0126e583 100644
--- a/gcc/builtins.c
+++ b/gcc/builtins.c
@@ -6691,7 +6691,9 @@ fold_builtin_expect (tree arglist)
 	}
       while (TREE_CODE (inner) == COMPONENT_REF
 	     || TREE_CODE (inner) == ARRAY_REF);
-      if (DECL_P (inner) && DECL_WEAK (inner))
+      if ((TREE_CODE (inner) == VAR_DECL
+           || TREE_CODE (inner) == FUNCTION_DECL)
+	  && DECL_WEAK (inner))
 	return 0;
     }
 
diff --git a/gcc/testsuite/gcc.dg/pr38700.c b/gcc/testsuite/gcc.dg/pr38700.c
new file mode 100644
index 00000000..bfa75b0c
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/pr38700.c
@@ -0,0 +1,11 @@
+/* PR c/38700 */
+/* { dg-do compile } */
+/* { dg-options "-O0" } */
+
+int
+foo ()
+{
+  __SIZE_TYPE__ s = __builtin_expect ((__SIZE_TYPE__)&&L, 0);
+L:
+  return 0;
+}
