diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index d5ca963b..5f1d261c 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -6534,7 +6534,8 @@ tsubst_decl (tree t, tree args, tsubst_flags_t complain)
 	       specialize R.  */
 	    gen_tmpl = most_general_template (DECL_TI_TEMPLATE (t));
 	    argvec = tsubst_template_args (DECL_TI_ARGS
-					   (DECL_TEMPLATE_RESULT (gen_tmpl)),
+                                          (DECL_TEMPLATE_RESULT
+                                                 (DECL_TI_TEMPLATE (t))),
 					   args, complain, in_decl);
 
 	    /* Check to see if we already have this specialization.  */
diff --git a/gcc/testsuite/g++.dg/template/crash94.C b/gcc/testsuite/g++.dg/template/crash94.C
new file mode 100644
index 00000000..810aed0a
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/crash94.C
@@ -0,0 +1,28 @@
+// Origin: PR c++/42697
+// { dg-do compile }
+
+template<class Value_t>
+class fparser
+{
+    template<bool Option>
+    void eval2(Value_t r[2]);
+public:
+    void evaltest();
+};
+
+template<>
+template<bool Option>
+void fparser<int>::eval2(int r[2])
+{
+    struct ObjType {};
+}
+
+
+template<class Value_t>
+void fparser<Value_t>::evaltest
+    ()
+{
+    eval2<false>(0);
+}
+
+template class fparser<int>;
