diff --git a/gcc/dwarf2out.c b/gcc/dwarf2out.c
index 72421402..50b8f9cc 100644
--- a/gcc/dwarf2out.c
+++ b/gcc/dwarf2out.c
@@ -12926,11 +12926,8 @@ force_type_die (tree type)
       else
 	context_die = comp_unit_die;
 
-      type_die = lookup_type_die (type);
-      if (type_die)
-	return type_die;
-      gen_type_die (type, context_die);
-      type_die = lookup_type_die (type);
+      type_die = modified_type_die (type, TYPE_READONLY (type),
+				    TYPE_VOLATILE (type), context_die);
       gcc_assert (type_die);
     }
   return type_die;
diff --git a/gcc/testsuite/g++.dg/debug/pr34895.C b/gcc/testsuite/g++.dg/debug/pr34895.C
new file mode 100644
index 00000000..ca06336c
--- /dev/null
+++ b/gcc/testsuite/g++.dg/debug/pr34895.C
@@ -0,0 +1,19 @@
+// { dg-do compile }
+// { dg-options "-O2 -g" }
+//
+// Copyright (C) 2008 Free Software Foundation, Inc.
+// Contributed by Theodore.Papadopoulo 20 Jan 2008 <Theodore.Papadopoulo@sophia.inria.fr>
+
+struct A {
+    A() { }
+    unsigned operator()() { return 1; }
+};
+struct B: public A {
+    typedef const A base;
+    using base::operator();
+    B() { }
+};
+int
+main() {
+    B b;
+}
