diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index 85c63f06..2cc1635a 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -12916,6 +12916,9 @@ cp_parser_class_name (cp_parser *parser,
 		}
 	      return error_mark_node;
 	    }
+	  else if (decl != error_mark_node
+		   && !parser->scope)
+	    maybe_note_name_used_in_class (identifier, decl);
 	}
     }
   else
diff --git a/gcc/testsuite/g++.dg/lookup/name-clash7.C b/gcc/testsuite/g++.dg/lookup/name-clash7.C
new file mode 100644
index 00000000..5c0690aa
--- /dev/null
+++ b/gcc/testsuite/g++.dg/lookup/name-clash7.C
@@ -0,0 +1,11 @@
+// PR c++/28513
+
+class foo {			// { dg-error "changes meaning" }
+public:
+  typedef int bar;
+};
+
+class baz {
+public:
+  foo::bar foo;			// { dg-error "declaration" }
+};
