diff --git a/gcc/c-format.c b/gcc/c-format.c
index 82f2bc77..e787b2d8 100644
--- a/gcc/c-format.c
+++ b/gcc/c-format.c
@@ -2433,7 +2433,8 @@ init_dynamic_gfc_info (void)
 	  locus = identifier_global_value (locus);
 	  if (locus)
 	    {
-	      if (TREE_CODE (locus) != TYPE_DECL)
+	      if (TREE_CODE (locus) != TYPE_DECL
+		  || TREE_TYPE (locus) == error_mark_node)
 		{
 		  error ("%<locus%> is not defined as a type");
 		  locus = 0;
diff --git a/gcc/testsuite/gcc.dg/format/gcc_gfc-2.c b/gcc/testsuite/gcc.dg/format/gcc_gfc-2.c
new file mode 100644
index 00000000..f3095fa2
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/format/gcc_gfc-2.c
@@ -0,0 +1,13 @@
+/* PR c/35436 */
+/* { dg-do compile } */
+/* { dg-options "-Wformat" } */
+
+typedef void locus[1]; /* { dg-error "array of void" } */
+
+void foo(const char*, ...)
+  __attribute__((__format__(__gcc_gfc__, 1, 2))); /* { dg-error "locus" } */
+
+void bar()
+{
+  foo("%L", 0); /* { dg-warning "format" }  */
+}
