diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 64d6b4ae..81a0ce84 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -6834,6 +6834,11 @@ tsubst_decl (tree t, tree args, tsubst_flags_t complain)
 	     tsubst_copy (DECL_NAME (t), args, complain, in_decl));
 	  if (!r)
 	    r = error_mark_node;
+	  else
+	    {
+	      TREE_PROTECTED (r) = TREE_PROTECTED (t);
+	      TREE_PRIVATE (r) = TREE_PRIVATE (t);
+	    }
 	}
       else
 	{
diff --git a/gcc/testsuite/g++.dg/template/access20.C b/gcc/testsuite/g++.dg/template/access20.C
new file mode 100644
index 00000000..ebf575e6
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/access20.C
@@ -0,0 +1,18 @@
+// PR c++/29470
+
+template <typename T> struct B
+{
+        protected:
+        T v;			// { dg-error "protected" }
+};
+template <typename T> struct D : B<T>
+{
+        protected:
+        using B<T>::v;
+};
+int main()
+{
+        D<int> d;
+        d.v = 0;		// { dg-error "context" }
+        return 0;
+}
