diff --git a/gcc/testsuite/gcc.dg/debug/pr29558.c b/gcc/testsuite/gcc.dg/debug/pr29558.c
new file mode 100644
index 00000000..99386b33
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/debug/pr29558.c
@@ -0,0 +1,30 @@
+/* { dg-do compile } */
+
+void stpi_unpack_16_1(int length, unsigned char *out, unsigned char bit)
+{
+  unsigned char tempin;
+  unsigned char temp[16];
+  for (bit = 128; length > 0; length--) {
+    if (tempin & 128)
+      temp[0] |= bit;
+    else
+      {
+        *out++ = temp[1];
+        *out++ = temp[2];
+        *out++ = temp[3];
+        *out++ = temp[4];
+        *out++ = temp[5];
+        *out++ = temp[6];
+        *out++ = temp[7];
+        *out++ = temp[9];
+        *out++ = temp[10];
+        *out++ = temp[11];
+        *out++ = temp[12];
+        *out++ = temp[13];
+        *out++ = temp[14];
+        *out++ = temp[15];
+        __builtin_memset (temp, 0, 16);
+     }
+  }
+}
+
diff --git a/gcc/var-tracking.c b/gcc/var-tracking.c
index c94c6670..8bc09880 100644
--- a/gcc/var-tracking.c
+++ b/gcc/var-tracking.c
@@ -1445,7 +1445,8 @@ track_expr_p (tree expr)
   if (MEM_P (decl_rtl))
     {
       /* Do not track structures and arrays.  */
-      if (GET_MODE (decl_rtl) == BLKmode)
+      if (GET_MODE (decl_rtl) == BLKmode
+	  || AGGREGATE_TYPE_P (TREE_TYPE (realdecl)))
 	return 0;
       if (MEM_SIZE (decl_rtl)
 	  && INTVAL (MEM_SIZE (decl_rtl)) > MAX_VAR_PARTS)
