diff --git a/gcc/testsuite/gcc.c-torture/compile/20070925-1.c b/gcc/testsuite/gcc.c-torture/compile/20070925-1.c
new file mode 100644
index 00000000..188193bc
--- /dev/null
+++ b/gcc/testsuite/gcc.c-torture/compile/20070925-1.c
@@ -0,0 +1,22 @@
+/* PR tree-optimization/32694 */
+
+typedef signed long long int WordS64;
+typedef unsigned long long int Word64;
+
+int
+foo (Word64 * p)
+{
+  while (1)
+    {
+      WordS64 c = 0x1llu;
+      WordS64 x = *p;
+      if (c >= 0)
+	{
+	  if (x > (WordS64) 0x7FFFFFFFFFFFFFFFll - c)
+	    return 6;
+	}
+      else if (x < (WordS64) 0x8000000000000000ll - c)
+	return 7;
+      p++;
+    }
+}
diff --git a/gcc/tree-ssa-propagate.c b/gcc/tree-ssa-propagate.c
index 71ee02fb..d876dc9a 100644
--- a/gcc/tree-ssa-propagate.c
+++ b/gcc/tree-ssa-propagate.c
@@ -564,7 +564,8 @@ set_rhs (tree *stmt_p, tree expr)
   ssa_op_iter iter;
 
   /* Verify the constant folded result is valid gimple.  */
-  if (TREE_CODE_CLASS (code) == tcc_binary)
+  if (TREE_CODE_CLASS (code) == tcc_binary
+      || TREE_CODE_CLASS (code) == tcc_comparison)
     {
       if (!is_gimple_val (TREE_OPERAND (expr, 0))
 	  || !is_gimple_val (TREE_OPERAND (expr, 1)))
