diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index c6305164..3d63f89a 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -4827,7 +4827,7 @@ make_rtl_for_nonlocal_decl (tree decl, tree init, const char* asmspec)
       /* An in-class declaration of a static data member should be
 	 external; it is only a declaration, and not a definition.  */
       if (init == NULL_TREE)
-	gcc_assert (DECL_EXTERNAL (decl));
+	gcc_assert (DECL_EXTERNAL (decl) || !TREE_PUBLIC (decl));
     }
 
   /* We don't create any RTL for local variables.  */
