diff --git a/gcc/cp/error.c b/gcc/cp/error.c
index a4fa57d0..c3ffe930 100644
--- a/gcc/cp/error.c
+++ b/gcc/cp/error.c
@@ -1898,6 +1898,10 @@ dump_expr (tree t, int flags)
       pp_multiplicative_expression (cxx_pp, t);
       break;
 
+    case OBJ_TYPE_REF:
+      dump_expr (resolve_virtual_fun_from_obj_type_ref (t), flags);
+      break;
+
       /*  This list is incomplete, but should suffice for now.
 	  It is very important that `sorry' does not call
 	  `report_error_function'.  That could cause an infinite loop.  */
diff --git a/gcc/testsuite/g++.dg/other/error20.C b/gcc/testsuite/g++.dg/other/error20.C
new file mode 100644
index 00000000..3c8e8f22
--- /dev/null
+++ b/gcc/testsuite/g++.dg/other/error20.C
@@ -0,0 +1,12 @@
+// PR c++/34275
+// { dg-do compile }
+
+struct A
+{		// { dg-error "candidates" }
+  virtual A foo ();
+};
+
+void bar (A& a)
+{
+  a.foo () = 0; // { dg-error "A::foo\\(\\) = 0" }
+}   
