diff --git a/gcc/fortran/scanner.c b/gcc/fortran/scanner.c
index 2e3ed748..a4e24c38 100644
--- a/gcc/fortran/scanner.c
+++ b/gcc/fortran/scanner.c
@@ -290,7 +290,7 @@ next_char (void)
   if (gfc_current_locus.nextc == NULL)
     return '\n';
 
-  c = *gfc_current_locus.nextc++;
+  c = (unsigned char) *gfc_current_locus.nextc++;
   if (c == '\0')
     {
       gfc_current_locus.nextc--; /* Remain on this line.  */
diff --git a/gcc/testsuite/gfortran.dg/string_0xfe_0xff_1.f90 b/gcc/testsuite/gfortran.dg/string_0xfe_0xff_1.f90
new file mode 100644
index 00000000..b92a4cd3
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/string_0xfe_0xff_1.f90
@@ -0,0 +1,7 @@
+! { dg-do compile }
+! PR 30452 - this used to cause syntax errors due to the presence,
+!            as characters, of bytes 0xfe and 0xff.
+program main
+  if (char (254) /= "þ") call abort
+  if (char (255) /= "ÿ") call abort
+end program main
