diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 4919c313..8ea9d0b8 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -2360,7 +2360,8 @@ reduce_template_parm_level (tree index, tree type, int levels, tree args,
 {
   if (TEMPLATE_PARM_DESCENDANTS (index) == NULL_TREE
       || (TEMPLATE_PARM_LEVEL (TEMPLATE_PARM_DESCENDANTS (index))
-	  != TEMPLATE_PARM_LEVEL (index) - levels))
+	  != TEMPLATE_PARM_LEVEL (index) - levels)
+      || !same_type_p (type, TREE_TYPE (TEMPLATE_PARM_DESCENDANTS (index))))
     {
       tree orig_decl = TEMPLATE_PARM_DECL (index);
       tree decl, t;
diff --git a/gcc/testsuite/g++.dg/template/nontype19.C b/gcc/testsuite/g++.dg/template/nontype19.C
new file mode 100644
index 00000000..1df78b3b
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/nontype19.C
@@ -0,0 +1,19 @@
+// PR c++/42466
+
+template<class IntT, IntT X>
+struct A
+{
+  A();
+
+  template<IntT X2>
+  A(const A<IntT, X2>& other);
+};
+
+int main(int argc, char** argv)
+{
+    A<int, 42> a;
+    A<int, 100> b = a;
+
+    A<unsigned, 42u> c;
+    A<unsigned, 100u> d = c;
+}
