diff --git a/gcc/testsuite/g++.dg/eh/forced1.C b/gcc/testsuite/g++.dg/eh/forced1.C
index b0200813..13020b71 100644
--- a/gcc/testsuite/g++.dg/eh/forced1.C
+++ b/gcc/testsuite/g++.dg/eh/forced1.C
@@ -1,5 +1,6 @@
 // HP-UX libunwind.so doesn't provide _UA_END_OF_STACK.
 // { dg-do run { xfail "ia64-hp-hpux11.*" } }
+// { dg-options "-static" }
 
 // Test that forced unwinding runs all cleanups.  Also tests that
 // rethrowing doesn't call the exception object destructor.
diff --git a/gcc/testsuite/g++.old-deja/g++.eh/terminate2.C b/gcc/testsuite/g++.old-deja/g++.eh/terminate2.C
index 3a311584..485f2e10 100644
--- a/gcc/testsuite/g++.old-deja/g++.eh/terminate2.C
+++ b/gcc/testsuite/g++.old-deja/g++.eh/terminate2.C
@@ -1,4 +1,5 @@
 // { dg-do run  }
+// { dg-options "-static" }
 // Test that an unhandled exception causes us to call terminate.
 
 #include <exception>
diff --git a/libiberty/testsuite/test-demangle.c b/libiberty/testsuite/test-demangle.c
index dcb27d25..c498d97a 100644
--- a/libiberty/testsuite/test-demangle.c
+++ b/libiberty/testsuite/test-demangle.c
@@ -46,7 +46,7 @@ static unsigned int lineno;
 #define LINELEN 80
 
 static void
-getline(buf)
+get_line(buf)
      struct line *buf;
 {
   char *data = buf->data;
@@ -148,12 +148,12 @@ main(argc, argv)
 
   for (;;)
     {
-      getline (&format);
+      get_line (&format);
       if (feof (stdin))
 	break;
 
-      getline (&input);
-      getline (&expect);
+      get_line (&input);
+      get_line (&expect);
 
       tests++;
 
@@ -268,7 +268,7 @@ main(argc, argv)
 
       if (no_params)
 	{
-	  getline (&expect);
+	  get_line (&expect);
 	  result = cplus_demangle (input.data, DMGL_ANSI|DMGL_TYPES);
 
 	  if (result
diff --git a/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-5.cc b/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-5.cc
index 244bae02..50dd0925 100644
--- a/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-5.cc
+++ b/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-5.cc
@@ -1,4 +1,5 @@
 // { dg-require-cxa-atexit "" }
+// { dg-options "-static" }
 
 // Copyright (C) 2004, 2005 Free Software Foundation, Inc.
 //
diff --git a/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-7.cc b/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-7.cc
index 55ff20aa..240de852 100644
--- a/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-7.cc
+++ b/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-7.cc
@@ -1,4 +1,5 @@
 // { dg-require-cxa-atexit "" }
+// { dg-options "-static" }
 
 // Copyright (C) 2004, 2005 Free Software Foundation, Inc.
 //
diff --git a/libstdc++-v3/testsuite/tr1/2_general_utilities/memory/shared_ptr/cons/weak_ptr_expired.cc b/libstdc++-v3/testsuite/tr1/2_general_utilities/memory/shared_ptr/cons/weak_ptr_expired.cc
index d40227ad..9a7fd4b3 100644
--- a/libstdc++-v3/testsuite/tr1/2_general_utilities/memory/shared_ptr/cons/weak_ptr_expired.cc
+++ b/libstdc++-v3/testsuite/tr1/2_general_utilities/memory/shared_ptr/cons/weak_ptr_expired.cc
@@ -1,4 +1,5 @@
 // { dg-do run { xfail *-*-* } }
+// { dg-options "-static" }
 // Copyright (C) 2005 Free Software Foundation
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
