diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 1f058ba..bef8955 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -2899,7 +2899,9 @@ push_template_decl_real (decl, is_friend)
 	    if (current == decl)
 	      current = ctx;
 	    else
-	      current = TYPE_CONTEXT (current);
+	      current = (TYPE_P (current)
+			 ? TYPE_CONTEXT (current)
+			 : DECL_CONTEXT (current));
 	  }
     }
 
diff --git a/gcc/testsuite/g++.dg/template/crash68.C b/gcc/testsuite/g++.dg/template/crash68.C
new file mode 100644
index 0000000..9171f8c
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/crash68.C
@@ -0,0 +1,16 @@
+// PR c++/33035
+
+template<class A> 
+struct a {
+        template<class B> 
+        struct b {
+                template<class C>
+                void f()
+                {
+                        struct g
+                        {
+                                ~g() {}
+                        };
+                }
+        };
+};
