diff --git a/gcc/cp/method.c b/gcc/cp/method.c
index 1c2d2ea..a382ecb 100644
--- a/gcc/cp/method.c
+++ b/gcc/cp/method.c
@@ -849,7 +849,7 @@ synthesize_exception_spec (type, extractor, client)
         continue;
       while (TREE_CODE (type) == ARRAY_TYPE)
   	type = TREE_TYPE (type);
-      if (TREE_CODE (type) != RECORD_TYPE)
+      if (!CLASS_TYPE_P (type))
         continue;
       
       fn = (*extractor) (type, client);
@@ -944,7 +944,7 @@ locate_copy (type, client_)
       int excess;
       int quals;
       
-      parms = TREE_CHAIN (parms);
+      parms = skip_artificial_parms_for (fn, parms);
       if (!parms)
         continue;
       src_type = TREE_VALUE (parms);
