diff --git a/gcc/config/i386/i386.c b/gcc/config/i386/i386.c
index b70b39d..eed396a 100644
--- a/gcc/config/i386/i386.c
+++ b/gcc/config/i386/i386.c
@@ -1786,8 +1786,8 @@ classify_argument (mode, type, classes, bit_offset)
 		     misaligned integers.  */
 		  if (DECL_BIT_FIELD (field))
 		    {
-		      for (i = int_bit_position (field) / 8 / 8;
-			   i < (int_bit_position (field)
+		      for (i = (int_bit_position (field) + (bit_offset % 64)) / 8 / 8;
+			   i < ((int_bit_position (field) + (bit_offset % 64))
 			        + tree_low_cst (DECL_SIZE (field), 0)
 			       	+ 63) / 8 / 8; i++)
 			classes[i] =
