diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 63a78eb..1f058ba 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -2050,7 +2050,8 @@ reduce_template_parm_level (index, type, levels, args, complain)
 {
   if (TEMPLATE_PARM_DESCENDANTS (index) == NULL_TREE
       || (TEMPLATE_PARM_LEVEL (TEMPLATE_PARM_DESCENDANTS (index))
-	  != TEMPLATE_PARM_LEVEL (index) - levels))
+	  != TEMPLATE_PARM_LEVEL (index) - levels)
+      || !same_type_p (type, TREE_TYPE (TEMPLATE_PARM_DESCENDANTS (index))))
     {
       tree decl 
 	= build_decl (TREE_CODE (TEMPLATE_PARM_DECL (index)),
diff --git a/gcc/testsuite/g++.dg/template/nontype19.C b/gcc/testsuite/g++.dg/template/nontype19.C
new file mode 100644
index 0000000..1df78b3
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/nontype19.C
@@ -0,0 +1,19 @@
+// PR c++/42466
+
+template<class IntT, IntT X>
+struct A
+{
+  A();
+
+  template<IntT X2>
+  A(const A<IntT, X2>& other);
+};
+
+int main(int argc, char** argv)
+{
+    A<int, 42> a;
+    A<int, 100> b = a;
+
+    A<unsigned, 42u> c;
+    A<unsigned, 100u> d = c;
+}
