diff --git a/gcc/testsuite/gcc.dg/20030711-1.c b/gcc/testsuite/gcc.dg/20030711-1.c
index 25f515c..391ca05 100644
--- a/gcc/testsuite/gcc.dg/20030711-1.c
+++ b/gcc/testsuite/gcc.dg/20030711-1.c
@@ -5,6 +5,10 @@
 #include <sys/mman.h>
 #include <stdlib.h>
 
+#ifndef MAP_FAILED
+#define MAP_FAILED ((void *) -1)
+#endif
+
 void __attribute__((noinline)) test (const char *p)
 {
   if (__builtin_strncmp (p, "abcdefghijklmnopq", 17) == 0)
