diff --git a/gcc/cp/g++spec.c b/gcc/cp/g++spec.c
index 542ca06..733b662 100644
--- a/gcc/cp/g++spec.c
+++ b/gcc/cp/g++spec.c
@@ -54,7 +54,7 @@ lang_specific_driver (fn, in_argc, in_argv, in_added_libraries)
 
   /* This will be 0 if we encounter a situation where we should not
      link in libstdc++.  */
-  int library = 1;
+  int library = 0;
 
   /* The number of arguments being added to what's in argv, other than
      libraries.  We use this to track the number of times we've inserted
