diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index a06b91a5d..78879a566 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -211,6 +211,12 @@ libgcov-driver-tool.o-warn = -Wno-error
 libgcov-merge-tool.o-warn = -Wno-error
 gimple-match.o-warn = -Wno-unused
 generic-match.o-warn = -Wno-unused
+gcc.o-warn = -Wno-error=unused-result
+build/genhooks.o-warn = -Wno-error=unused-result
+lto-wrapper.o-warn = -Wno-error=unused-result
+lto/lto.o-warn = -Wno-error=unused-result
+toplev.o-warn = -Wno-error=nonnull -Wno-error=format-extra-args
+fortran/trans-openmp.o-warn = -Wno-error=format -Wno-error=format-extra-args
 
 # All warnings have to be shut off in stage1 if the compiler used then
 # isn't gcc; configure determines that.  WARN_CFLAGS will be either
diff --git a/gcc/fortran/frontend-passes.c b/gcc/fortran/frontend-passes.c
index c051cc09d..639e31798 100644
--- a/gcc/fortran/frontend-passes.c
+++ b/gcc/fortran/frontend-passes.c
@@ -138,7 +138,7 @@ gfc_run_passes (gfc_namespace *ns)
  */
 
 static int
-realloc_string_callback (gfc_code **c, int *walk_subtrees,
+realloc_string_callback (gfc_code **c, int *walk_subtrees ATTRIBUTE_UNUSED,
 			 void *data ATTRIBUTE_UNUSED)
 {
   gfc_expr *expr1, *expr2;
diff --git a/gcc/ipa-pure-const.c b/gcc/ipa-pure-const.c
index 89143496b..b5f350ad6 100644
--- a/gcc/ipa-pure-const.c
+++ b/gcc/ipa-pure-const.c
@@ -1161,13 +1161,6 @@ pure_const_read_summary (void)
     }
 }
 
-
-static bool
-ignore_edge (struct cgraph_edge *e)
-{
-  return (!e->can_throw_external);
-}
-
 /* Return true if NODE is self recursive function.
    Indirectly recursive functions appears as non-trivial strongly
    connected components, so we need to care about self recursion
