diff --git a/gcc/c-family/c-opts.c b/gcc/c-family/c-opts.c
index ac3976c..46dcb00 100644
--- a/gcc/c-family/c-opts.c
+++ b/gcc/c-family/c-opts.c
@@ -666,11 +666,13 @@ c_common_handle_option (size_t scode, const char *arg, int value,
       break;
 
     case OPT_std_c11:
+    case OPT_std_c17: /* Fake c17 for cmdline compatibility. */
       if (!preprocessing_asm_p)
 	set_std_c11 (true /* ISO */);
       break;
 
     case OPT_std_gnu11:
+    case OPT_std_gnu17: /* Fake gnu17 for cmdline compatibility. */
       if (!preprocessing_asm_p)
 	set_std_c11 (false /* ISO */);
       break;
diff --git a/gcc/c-family/c.opt b/gcc/c-family/c.opt
index baffb22..157a812 100644
--- a/gcc/c-family/c.opt
+++ b/gcc/c-family/c.opt
@@ -1598,6 +1598,14 @@ std=c1x
 C ObjC Alias(std=c11)
 Deprecated in favor of -std=c11
 
+std=c17
+C ObjC
+Fake c17 for cmdline compatibility.
+
+std=c18
+C ObjC Alias(std=c17)
+Fake c17 for cmdline compatibility.
+
 std=c89
 C ObjC Alias(std=c90)
 Conform to the ISO 1990 C standard
@@ -1654,6 +1662,14 @@ std=gnu1x
 C ObjC Alias(std=gnu11)
 Deprecated in favor of -std=gnu11
 
+std=gnu17
+C ObjC
+Fake gnu17 for cmdline compatibility.
+
+std=gnu18
+C ObjC Alias(std=gnu17)
+Fake gnu17 for cmdline compatibility.
+
 std=gnu89
 C ObjC Alias(std=gnu90)
 Conform to the ISO 1990 C standard with GNU extensions
