diff --git a/Makefile.in b/Makefile.in
index b7a8cb2..357f8dd 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -221,7 +221,7 @@ EXTRA_HEADERS =
 
 # Target to use whe installing assert.h.  Some systems may
 # want to set this empty.
-INSTALL_ASSERT_H = install-assert-h
+INSTALL_ASSERT_H = 
 
 # Set this to `ld' to enable use of collect2.
 # USE_COLLECT2 =
diff --git a/g++ b/g++
index 9d80072..b5def52 100755
--- a/g++
+++ b/g++
@@ -4,7 +4,7 @@
 
 # The compiler name might be different when doing cross-compilation
 # (this should be configured)
-gcc_name=i586-legacy-linux-gnulibc1-gcc-2.4.5
+gcc_name=i486-legacy-linuxaout-gcc-2.4.5
 speclang=-xnone
 
 # replace the command name by the name of the new command
