diff --git a/libio/stdio.h b/libio/stdio.h
index bdc79ea..f85becc 100644
--- a/libio/stdio.h
+++ b/libio/stdio.h
@@ -846,12 +846,6 @@ __BEGIN_NAMESPACE_STD
 extern void perror (const char *__s);
 __END_NAMESPACE_STD
 
-/* Provide the declarations for `sys_errlist' and `sys_nerr' if they
-   are available on this system.  Even if available, these variables
-   should not be used directly.  The `strerror' function provides
-   all the necessary functionality.  */
-#include <bits/sys_errlist.h>
-
 
 #ifdef	__USE_POSIX
 /* Return the system file descriptor for STREAM.  */
diff --git a/posix/sys/types.h b/posix/sys/types.h
index 0d51ae8..2636db6 100644
--- a/posix/sys/types.h
+++ b/posix/sys/types.h
@@ -217,9 +217,6 @@ typedef int register_t __attribute__ ((__mode__ (__word__)));
 
 /* It also defines `fd_set' and the FD_* macros for `select'.  */
 # include <sys/select.h>
-
-/* BSD defines these symbols, so we follow.  */
-# include <sys/sysmacros.h>
 #endif /* Use BSD.  */
 
 
