diff --git a/gcc/toplev.cc b/gcc/toplev.cc
index 8bdb4d9e5..85c7754af 100644
--- a/gcc/toplev.cc
+++ b/gcc/toplev.cc
@@ -2315,10 +2315,8 @@ toplev::main (int argc, char **argv)
      emit some diagnostics here.  */
   invoke_plugin_callbacks (PLUGIN_FINISH, NULL);
 
-  if (flag_diagnostics_generate_patch)
+  if (global_dc->edit_context_ptr)
     {
-      gcc_assert (global_dc->edit_context_ptr);
-
       pretty_printer pp;
       pp_show_color (&pp) = pp_show_color (global_dc->printer);
       global_dc->edit_context_ptr->print_diff (&pp, true);
