diff --git a/gcc/gcc.cc b/gcc/gcc.cc
index 15a24fc54..449071cc0 100644
--- a/gcc/gcc.cc
+++ b/gcc/gcc.cc
@@ -881,7 +881,7 @@ proper position among the other output files.  */
 #endif
 
 #ifdef HAVE_AS_DEBUG_PREFIX_MAP
-#define ASM_MAP " %{fdebug-prefix-map=*:--debug-prefix-map %*}"
+#define ASM_MAP " %{ffile-prefix-map=*:--debug-prefix-map %*} %{fdebug-prefix-map=*:--debug-prefix-map %*}"
 #else
 #define ASM_MAP ""
 #endif
