diff --git a/gcc/testsuite/gcc.target/aarch64/pr63304_1.c b/gcc/testsuite/gcc.target/aarch64/pr63304_1.c
index 5d519d817..306850335 100644
--- a/gcc/testsuite/gcc.target/aarch64/pr63304_1.c
+++ b/gcc/testsuite/gcc.target/aarch64/pr63304_1.c
@@ -15,22 +15,6 @@ cal (double a)
     return 1;
 }
 
-#pragma GCC push_options
-
-#pragma GCC target ("cmodel=large")
-
-int
-cal2 (double a)
-{
-
-  double b = 3.2;
-  double c = 2.2;
-  if ((a + b) != c)
-    return 0;
-  else
-    return 1;
-}
-
 #pragma GCC pop_options
 
 int
@@ -45,4 +29,4 @@ cal3 (double a)
     return 1;
 }
 
-/* { dg-final { scan-assembler-times "adrp" 6 } } */
+/* { dg-final { scan-assembler-times "adrp" 4 } } */
diff --git a/gcc/testsuite/gcc.target/aarch64/pr78733.c b/gcc/testsuite/gcc.target/aarch64/pr78733.c
index 8556ef3f3..7ac3127a7 100644
--- a/gcc/testsuite/gcc.target/aarch64/pr78733.c
+++ b/gcc/testsuite/gcc.target/aarch64/pr78733.c
@@ -10,4 +10,3 @@ t (void)
 }
 
 /* { dg-final { scan-assembler "adr" } } */
-/* { dg-final { scan-assembler-not "adrp" } } */
diff --git a/gcc/testsuite/gcc.target/aarch64/pr79041-2.c b/gcc/testsuite/gcc.target/aarch64/pr79041-2.c
index 8556ef3f3..7ac3127a7 100644
--- a/gcc/testsuite/gcc.target/aarch64/pr79041-2.c
+++ b/gcc/testsuite/gcc.target/aarch64/pr79041-2.c
@@ -10,4 +10,3 @@ t (void)
 }
 
 /* { dg-final { scan-assembler "adr" } } */
-/* { dg-final { scan-assembler-not "adrp" } } */
diff --git a/gcc/testsuite/gcc.target/aarch64/symbol-range-tiny.c b/gcc/testsuite/gcc.target/aarch64/symbol-range-tiny.c
deleted file mode 100644
index fc6a4f3ec..000000000
--- a/gcc/testsuite/gcc.target/aarch64/symbol-range-tiny.c
+++ /dev/null
@@ -1,12 +0,0 @@
-/* { dg-do link } */
-/* { dg-options "-O3 -save-temps -mcmodel=tiny" } */
-
-char fixed_regs[0x00080000];
-
-int
-main ()
-{
-  return fixed_regs[0x000ff000];
-}
-
-/* { dg-final { scan-assembler-not "adr\tx\[0-9\]+, fixed_regs\\\+" } } */
diff --git a/gcc/testsuite/gcc.target/aarch64/symbol-range.c b/gcc/testsuite/gcc.target/aarch64/symbol-range.c
deleted file mode 100644
index d8e82fa1b..000000000
--- a/gcc/testsuite/gcc.target/aarch64/symbol-range.c
+++ /dev/null
@@ -1,12 +0,0 @@
-/* { dg-do link } */
-/* { dg-options "-O3 -save-temps -mcmodel=small" } */
-
-char fixed_regs[0x80000000];
-
-int
-main ()
-{
-  return fixed_regs[0xfffff000];
-}
-
-/* { dg-final { scan-assembler-not "adrp\tx\[0-9\]+, fixed_regs\\\+" } } */
