diff --git a/gcc/config.gcc b/gcc/config.gcc
index a9971722..e276f0c3 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -1553,7 +1553,7 @@ mips*-*-netbsd*)			# NetBSD/mips, either endian.
 mips64*-*-linux*)
 	tm_file="dbxelf.h elfos.h svr4.h linux.h ${tm_file} mips/linux.h mips/linux64.h"
 	tmake_file="${tmake_file} mips/t-linux64"
-	tm_defines="${tm_defines} MIPS_ABI_DEFAULT=ABI_N32"
+	tm_defines="${tm_defines} MIPS_ABI_DEFAULT=ABI_64"
 	gnu_ld=yes
 	gas=yes
 	;;
diff --git a/gcc/config/mips/linux64.h b/gcc/config/mips/linux64.h
index dc893f47..88194849 100644
--- a/gcc/config/mips/linux64.h
+++ b/gcc/config/mips/linux64.h
@@ -22,7 +22,7 @@ along with GCC; see the file COPYING3.  If not see
    in order to make the other specs easier to write.  */
 #define DRIVER_SELF_SPECS \
 "%{!EB:%{!EL:%(endian_spec)}}", \
-"%{!mabi=*: -mabi=n32}"
+"%{!mabi=*: -mabi=64}"
 
 #undef SUBTARGET_ASM_SPEC
 #define SUBTARGET_ASM_SPEC "\
