diff --git a/crossdev b/crossdev
index 192c390..a315622 100755
--- a/crossdev
+++ b/crossdev
@@ -280,8 +280,10 @@ parse_target() {
 		      ;;
 
 		mingw*|*-mingw*)
-		      # bug #584858
-		      die "'${CTARGET}' target is not supported anymore, use i686-w64-mingw32"
+			  KCAT="dev-util"; KPKG="w32api";
+			  LCAT="dev-util"; LPKG="mingw-runtime"
+			  WITH_DEF_HEADERS="yes" # gcc can't boot without headers: bug #693770
+			  GUSE+=" -hardened" # gcc ICEs as libssp/ssp.c:186:1: internal compiler error: in seh_emit_stackalloc
 		      ;;
 
 		spu*) TARCH=ppc64;
