diff --git a/bfd/elflink.c b/bfd/elflink.c
index c4f57cf2..4c4afe82 100644
--- a/bfd/elflink.c
+++ b/bfd/elflink.c
@@ -13756,9 +13756,6 @@ bfd_elf_final_link (bfd *abfd, struct bfd_link_info *info)
 		  else if (bfd_link_pde (info))
 		    info->callbacks->einfo
 		      (_("%P: warning: creating DT_TEXTREL in a PDE\n"));
-		  else
-		    info->callbacks->einfo
-		      (_("%P: warning: creating DT_TEXTREL in a PIE\n"));
 		  break;
 		}
 	    }
diff --git a/gas/config/loongarch-parse.c b/gas/config/loongarch-parse.c
index c6f66944..e3c43971 100644
--- a/gas/config/loongarch-parse.c
+++ b/gas/config/loongarch-parse.c
@@ -289,9 +289,6 @@ emit_bin (int op)
 	  opr1 = (valueT) opr1 << opr2;
 	  break;
 	case RIGHT_OP:
-	  if (opr1 < 0)
-	    as_warn (_("Right shift of negative numbers may be changed "
-		       "from arithmetic right shift to logical right shift!"));
 	  /* Arithmetic right shift.  */
 	  opr1 = opr1 >> opr2;
 	  break;
diff --git a/gas/config/loongarch-parse.y b/gas/config/loongarch-parse.y
index d575748d..cd7a6a92 100644
--- a/gas/config/loongarch-parse.y
+++ b/gas/config/loongarch-parse.y
@@ -237,9 +237,6 @@ emit_bin (int op)
 	  opr1 = (valueT) opr1 << opr2;
 	  break;
 	case RIGHT_OP:
-	  if (opr1 < 0)
-	    as_warn (_("Right shift of negative numbers may be changed "
-		       "from arithmetic right shift to logical right shift!"));
 	  /* Arithmetic right shift.  */
 	  opr1 = opr1 >> opr2;
 	  break;
