diff --git a/gcc/fortran/resolve.c b/gcc/fortran/resolve.c
index b203f363..86ce4b95 100644
--- a/gcc/fortran/resolve.c
+++ b/gcc/fortran/resolve.c
@@ -5505,7 +5505,7 @@ resolve_select (gfc_code *code)
 	/* Strip all other unreachable cases.  */
 	if (body->ext.case_list)
 	  {
-	    for (cp = body->ext.case_list; cp->next; cp = cp->next)
+	    for (cp = body->ext.case_list; cp && cp->next; cp = cp->next)
 	      {
 		if (cp->next->unreachable)
 		  {
diff --git a/gcc/testsuite/gfortran.dg/pr62135.f90 b/gcc/testsuite/gfortran.dg/pr62135.f90
new file mode 100644
index 00000000..d5e674b2
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/pr62135.f90
@@ -0,0 +1,17 @@
+! { dg-do compile }
+! { dg-options -Wsurprising }
+
+   PROGRAM PR62135
+      IMPLICIT NONE
+      CHARACTER*1 :: choice
+      choice = 'x'
+      SELECT CASE (choice)
+         ! This triggered an ICE: an unreachable case clause
+         ! as the last of a list.
+         CASE ('2':'7','9':'0') ! { dg-warning "can never be matched" }
+            WRITE(*,*) "barf"
+         CASE DEFAULT
+            CONTINUE
+      END SELECT
+   END PROGRAM PR62135
+
