diff --git a/gcc/cp/semantics.c b/gcc/cp/semantics.c
index f672592b..41d7c938 100644
--- a/gcc/cp/semantics.c
+++ b/gcc/cp/semantics.c
@@ -4183,6 +4183,7 @@ finish_decltype_type (tree expr, bool id_expression_or_member_access_p)
         case CONST_DECL:
         case PARM_DECL:
         case RESULT_DECL:
+        case TEMPLATE_PARM_INDEX:
           type = TREE_TYPE (expr);
           break;
 
diff --git a/gcc/testsuite/g++.dg/cpp0x/decltype15.C b/gcc/testsuite/g++.dg/cpp0x/decltype15.C
new file mode 100644
index 00000000..5c2d445d
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/decltype15.C
@@ -0,0 +1,13 @@
+// PR c++/38640
+// { dg-do compile }
+// { dg-options "-std=c++0x" }
+
+template<int N> void foo (decltype (N));
+template<long int N> void foo (decltype (N));
+
+void
+bar (void)
+{
+  foo<5> (6);
+  foo<5L> (6L);
+}
