diff --git a/gcc/cp/method.c b/gcc/cp/method.c
index c91fc476..8b6dcb65 100644
--- a/gcc/cp/method.c
+++ b/gcc/cp/method.c
@@ -621,6 +621,7 @@ do_build_copy_constructor (tree fndecl)
 
 	      if (DECL_MUTABLE_P (field))
 		quals &= ~TYPE_QUAL_CONST;
+	      quals |= TYPE_QUALS (expr_type);
 	      expr_type = cp_build_qualified_type (expr_type, quals);
 	    }
 
diff --git a/gcc/testsuite/g++.dg/init/synth2.C b/gcc/testsuite/g++.dg/init/synth2.C
new file mode 100644
index 00000000..82d80af9
--- /dev/null
+++ b/gcc/testsuite/g++.dg/init/synth2.C
@@ -0,0 +1,17 @@
+// PR c++/34180
+
+struct G {
+  G();				// { dg-error "note" }
+  G(G&);			// { dg-error "note" }
+};
+
+class A
+{				// { dg-error "no match" }
+  const G g;
+};
+
+int main()
+{
+  A a;
+  A b = a;			// { dg-error "note" }
+}
