diff --git a/gcc/cp/tree.c b/gcc/cp/tree.c
index 8d802a1e..adc74168 100644
--- a/gcc/cp/tree.c
+++ b/gcc/cp/tree.c
@@ -2401,6 +2401,10 @@ cp_walk_subtrees (tree *tp, int *walk_subtrees_p, walk_tree_fn func,
       break;
 
     case CAST_EXPR:
+    case REINTERPRET_CAST_EXPR:
+    case STATIC_CAST_EXPR:
+    case CONST_CAST_EXPR:
+    case DYNAMIC_CAST_EXPR:
       if (TREE_TYPE (*tp))
 	WALK_SUBTREE (TREE_TYPE (*tp));
 
diff --git a/gcc/testsuite/g++.dg/cpp0x/pr38795.C b/gcc/testsuite/g++.dg/cpp0x/pr38795.C
new file mode 100644
index 00000000..54fb361d
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/pr38795.C
@@ -0,0 +1,13 @@
+// PR c++/38795
+// { dg-do compile }
+// { dg-options "-std=gnu++0x" }
+
+template<typename... T> int foo(int i)
+{
+  return *reinterpret_cast<T*>(i);	// { dg-error "not expanded with|T" }
+}
+
+void bar(int i)
+{
+  foo<int>(i);
+}
