diff --git a/gcc/fortran/expr.c b/gcc/fortran/expr.c
index 671931c9..805fcb72 100644
--- a/gcc/fortran/expr.c
+++ b/gcc/fortran/expr.c
@@ -2272,6 +2272,9 @@ check_init_expr (gfc_expr *e)
 	      m = MATCH_ERROR;
 	    }
 
+	  if (m == MATCH_ERROR)
+	    return FAILURE;
+
 	  /* Try to scalarize an elemental intrinsic function that has an
 	     array argument.  */
 	  isym = gfc_find_function (e->symtree->n.sym->name);
diff --git a/gcc/testsuite/gfortran.dg/initialization_28.f90 b/gcc/testsuite/gfortran.dg/initialization_28.f90
new file mode 100644
index 00000000..f5330534
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/initialization_28.f90
@@ -0,0 +1,9 @@
+! { dg-do compile }
+!
+! PR fortran/50163
+!
+! Contributed by Philip Mason
+!
+character(len=2) :: xx ='aa'
+integer :: iloc=index(xx,'bb') ! { dg-error "has not been declared or is a variable" }
+end
