diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 6579f573..ebe24ade 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -9175,6 +9175,14 @@ grokdeclarator (const cp_declarator *declarator,
 		error ("virtual non-class function %qs", name);
 		virtualp = 0;
 	      }
+	    else if (sfk == sfk_constructor
+		     || sfk == sfk_destructor)
+	      {
+		error (funcdef_flag
+		       ? "%qs defined in a non-class scope"
+		       : "%qs declared in a non-class scope", name);
+		sfk = sfk_none;
+	      }
 	  }
 	else if (TREE_CODE (type) == FUNCTION_TYPE && staticp < 2
 		 && !NEW_DELETE_OPNAME_P (original_name))
diff --git a/gcc/testsuite/g++.dg/parse/ctor9.C b/gcc/testsuite/g++.dg/parse/ctor9.C
new file mode 100644
index 00000000..b80f7289
--- /dev/null
+++ b/gcc/testsuite/g++.dg/parse/ctor9.C
@@ -0,0 +1,7 @@
+/* PR c++/37647 */
+/* { dg-do "compile" } */
+
+struct A
+{
+  A() { void A(); } /* { dg-error "return type specification for constructor invalid|non-class scope" } */
+};
