diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 938b32d0..2aca46bb 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -15450,6 +15450,8 @@ tsubst_initializer_list (tree t, tree argvec)
 		    = tsubst_pack_expansion (expr, argvec,
 					     tf_warning_or_error,
 					     NULL_TREE);
+		  if (expanded_exprs == error_mark_node)
+		    continue;
 
 		  /* Prepend each of the expanded expressions to the
 		     corresponding TREE_LIST in EXPANDED_ARGUMENTS.  */
diff --git a/gcc/testsuite/g++.dg/cpp0x/vt-35243.C b/gcc/testsuite/g++.dg/cpp0x/vt-35243.C
new file mode 100644
index 00000000..4b555744
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/vt-35243.C
@@ -0,0 +1,9 @@
+// { dg-options "-std=c++0x" }
+struct A {};
+
+template<typename... T> struct B : T...
+{
+  B() : T(x)... {} // { dg-error "not declared" }
+};
+
+B<A> b;
