diff --git a/gcc/c-common.c b/gcc/c-common.c
index e809332c..54d0c9e1 100644
--- a/gcc/c-common.c
+++ b/gcc/c-common.c
@@ -6818,6 +6818,7 @@ fold_offsetof_1 (tree expr, tree stop_ref)
       return error_mark_node;
 
     case CALL_EXPR:
+    case TARGET_EXPR:
       error ("cannot apply %<offsetof%> when %<operator[]%> is overloaded");
       return error_mark_node;
 
diff --git a/gcc/testsuite/g++.dg/parse/offsetof8.C b/gcc/testsuite/g++.dg/parse/offsetof8.C
new file mode 100644
index 00000000..9add91e3
--- /dev/null
+++ b/gcc/testsuite/g++.dg/parse/offsetof8.C
@@ -0,0 +1,16 @@
+// PR c++/35321
+
+struct A
+{
+  A operator[] (int);
+};
+
+struct B
+{
+  A a;
+};
+
+void foo()
+{
+  __builtin_offsetof(B, a[0]); /* { dg-error "error: cannot apply 'offsetof' when 'operator\\\[\\\]' is overloaded" } */
+}
