diff --git a/gcc/cp/tree.c b/gcc/cp/tree.c
index 70cf67d7..ea1b81e0 100644
--- a/gcc/cp/tree.c
+++ b/gcc/cp/tree.c
@@ -273,6 +273,9 @@ build_target_expr (tree decl, tree value)
 #ifdef ENABLE_CHECKING
   gcc_assert (VOID_TYPE_P (TREE_TYPE (value))
 	      || TREE_TYPE (decl) == TREE_TYPE (value)
+	      /* On ARM ctors return 'this'.  */
+	      || (TREE_CODE (TREE_TYPE (value)) == POINTER_TYPE
+		  && TREE_CODE (value) == CALL_EXPR)
 	      || useless_type_conversion_p (TREE_TYPE (decl),
 					    TREE_TYPE (value)));
 #endif
