diff --git a/gcc/config.gcc b/gcc/config.gcc
index 21f5359..76e05ee 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -1897,7 +1897,7 @@ mips64*-*-linux*)
 	# that is installed.  What GCC thinks of as the default must
 	# remain as ABI_32 such that the SONAMEs of the libgcc shared
 	# libraries remain compatible with those of mips-linux-gnu.
-	tm_defines="MIPS_ISA_DEFAULT=32 MIPS_ABI_DEFAULT=ABI_32"
+	tm_defines="MIPS_ISA_DEFAULT=64 MIPS_ABI_DEFAULT=ABI_N32"
 	target_cpu_default="MASK_ABICALLS|MASK_FLOAT64|MASK_64BIT"
 	extra_parts="crtbegin.o crtbeginS.o crtbeginT.o crtend.o crtendS.o"
 	gnu_ld=yes
diff --git a/gcc/config/mips/linux64.h b/gcc/config/mips/linux64.h
index 83b3d2c..2564366 100644
--- a/gcc/config/mips/linux64.h
+++ b/gcc/config/mips/linux64.h
@@ -22,7 +22,7 @@ Boston, MA 02111-1307, USA.  */
 #define DRIVER_SELF_SPECS \
 ENDIAN_SPEC, \
 "%{mabi-fake-default:%{!mabi=*:-mabi=32}}", \
-"%{!mabi=*:-mabi=32}", \
+"%{!mabi=*:-mabi=n32}", \
 "%{!mips*:%{!march=*:%{mabi=32:-mips1}%{mabi=n32|mabi=64:-mips3}}}"
 
 #undef SUBTARGET_ASM_SPEC
