diff --git a/gcc/config/rs6000/linux64.h b/gcc/config/rs6000/linux64.h
index 599c692..548dafb 100644
--- a/gcc/config/rs6000/linux64.h
+++ b/gcc/config/rs6000/linux64.h
@@ -160,9 +160,7 @@ Boston, MA 02111-1307, USA.  */
   %{rdynamic:-export-dynamic} \
   %{!dynamic-linker:-dynamic-linker /lib64/ld64.so.1}}}"
 
-#ifdef NATIVE_CROSS
 #define STARTFILE_PREFIX_SPEC "/usr/local/lib64/ /lib64/ /usr/lib64/"
-#endif
 
 #undef  STARTFILE_LINUX_SPEC
 #define STARTFILE_LINUX_SPEC "\
