diff --git a/configure.in b/configure.in
index 3b0f2b5..d347670 100644
--- a/configure.in
+++ b/configure.in
@@ -52,8 +52,7 @@ libstdcxx_version="target-libstdc++-v3"
 host_tools="texinfo byacc flex bison binutils ld gas gcc sid sim gdb make patch prms send-pr gprof etc expect dejagnu ash bash bzip2 m4 autoconf automake libtool grep diff rcs fileutils shellutils time textutils wdiff find uudecode hello tar gzip indent recode release sed utils guile perl gawk findutils snavigator libtool gettext zip fastjar"
 
 # libgcj represents the runtime libraries only used by gcj.
-libgcj="target-libffi \
-	target-boehm-gc \
+libgcj="target-boehm-gc \
 	target-zlib \
 	target-qthreads \
 	target-libjava"
@@ -66,7 +65,8 @@ target_libs="target-libiberty \
 		target-newlib \
 		${libstdcxx_version} \
 		target-libf2c \
-		${libgcj}
+		${libgcj} \
+		target-libffi \
 		target-libobjc"
 
 # these tools are built using the target libs, and are intended to run only
