diff --git a/gcc/c-common.c b/gcc/c-common.c
index 6ec6764..3018c21 100644
--- a/gcc/c-common.c
+++ b/gcc/c-common.c
@@ -4987,6 +4987,15 @@ cb_register_builtins (pfile)
   /* Misc.  */
   builtin_define_with_value ("__VERSION__", version_string, 1);
 
+  /* Definitions for LP64 model.  */
+  if (TYPE_PRECISION (long_integer_type_node) == 64
+      && POINTER_SIZE == 64
+      && TYPE_PRECISION (integer_type_node) == 32)
+    {
+      cpp_define (pfile, "_LP64");
+      cpp_define (pfile, "__LP64__");
+    }
+
   /* Other target-independent built-ins determined by command-line
      options.  */
   if (optimize_size)
diff --git a/gcc/config/alpha/freebsd.h b/gcc/config/alpha/freebsd.h
index 149ce06..f809c62 100644
--- a/gcc/config/alpha/freebsd.h
+++ b/gcc/config/alpha/freebsd.h
@@ -29,7 +29,6 @@ Boston, MA 02111-1307, USA.  */
 #define FBSD_TARGET_CPU_CPP_BUILTINS()		\
   do						\
     {						\
-      builtin_define ("__LP64__");		\
       if (flag_pic)				\
 	{					\
 	  builtin_define ("__PIC__");		\
diff --git a/gcc/config/i386/freebsd64.h b/gcc/config/i386/freebsd64.h
index 4a66034..12ca062 100644
--- a/gcc/config/i386/freebsd64.h
+++ b/gcc/config/i386/freebsd64.h
@@ -23,17 +23,6 @@ Boston, MA 02111-1307, USA.  */
 #undef  TARGET_VERSION
 #define TARGET_VERSION fprintf (stderr, " (FreeBSD/x86-64 ELF)");
 
-#undef  FBSD_TARGET_CPU_CPP_BUILTINS
-#define FBSD_TARGET_CPU_CPP_BUILTINS()			\
-  do							\
-    {							\
-      if (TARGET_64BIT)					\
-      {							\
-	builtin_define ("__LP64__");			\
-      }							\
-    }							\
-  while (0)
-
 /* Provide a LINK_SPEC appropriate for the FreeBSD/x86-64 ELF target.
    This is a copy of LINK_SPEC from <i386/freebsd.h> tweaked for
    the x86-64 target.  */
diff --git a/gcc/config/i386/linux64.h b/gcc/config/i386/linux64.h
index 9d2fe58..1cd382b 100644
--- a/gcc/config/i386/linux64.h
+++ b/gcc/config/i386/linux64.h
@@ -36,11 +36,6 @@ Boston, MA 02111-1307, USA.  */
 	    builtin_define ("__PIC__");				\
 	    builtin_define ("__pic__");				\
 	  }							\
-	if (TARGET_64BIT)					\
-	  {							\
-	    builtin_define ("__LP64__");			\
-	    builtin_define ("_LP64");				\
-	  }							\
     }								\
   while (0)
 
diff --git a/gcc/config/ia64/ia64.h b/gcc/config/ia64/ia64.h
index 2a3e409..a0fc134 100644
--- a/gcc/config/ia64/ia64.h
+++ b/gcc/config/ia64/ia64.h
@@ -40,11 +40,6 @@ do {						\
 	builtin_define("__ia64__");		\
 	builtin_define("__itanium__");		\
 	builtin_define("__ELF__");		\
-	if (!TARGET_ILP32)			\
-	  {					\
-	    builtin_define("_LP64");		\
-	    builtin_define("__LP64__");		\
-	  }					\
 	if (TARGET_BIG_ENDIAN)			\
 	  builtin_define("__BIG_ENDIAN__");	\
 } while (0)
diff --git a/gcc/config/pa/pa.h b/gcc/config/pa/pa.h
index 9686b9e..8647eac 100644
--- a/gcc/config/pa/pa.h
+++ b/gcc/config/pa/pa.h
@@ -365,11 +365,6 @@ do {								\
      builtin_assert("machine=hppa");				\
      builtin_define("__hppa");					\
      builtin_define("__hppa__");				\
-     if (TARGET_64BIT)						\
-       {							\
-	 builtin_define("_LP64");				\
-	 builtin_define("__LP64__");				\
-       }							\
      if (TARGET_PA_20)						\
        builtin_define("_PA_RISC2_0");				\
      else if (TARGET_PA_11)					\
diff --git a/gcc/config/sparc/freebsd.h b/gcc/config/sparc/freebsd.h
index 21f3da2..b2809dd 100644
--- a/gcc/config/sparc/freebsd.h
+++ b/gcc/config/sparc/freebsd.h
@@ -25,14 +25,6 @@ the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */
 #define CPP_CPU64_DEFAULT_SPEC \
   "-D__sparc64__ -D__sparc_v9__ -D__sparcv9 -D__sparc__ -D__arch64__"
 
-#undef  FBSD_TARGET_CPU_CPP_BUILTINS
-#define FBSD_TARGET_CPU_CPP_BUILTINS()		\
-  do						\
-    {						\
-      builtin_define ("__LP64__");		\
-    }						\
-  while (0)
-
 /* Because we include sparc/sysv4.h.  */
 #undef  CPP_PREDEFINES
 /* Do not define it here, we now use TARGET_OS_CPP_BUILTINS.  */
