diff --git a/Makefile.in b/Makefile.in
index 9d0396e..58312a3 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -1389,7 +1389,7 @@ install-float-h: float.h
 # Install limits.h, optionally inheriting from the standard system version.
 install-limits-h: limits.h limitx.h limity.h
 	-rm -f $(DESTDIR)$(libsubdir)/include/limits.h tmp-limits.h
-	if [ -f /usr/include/limits.h ] ; then \
+	if [ -f $(tooldir)/include/limits.h ] ; then \
 	  cat $(srcdir)/limitx.h $(srcdir)/limits.h $(srcdir)/limity.h > tmp-limits.h; \
 	else \
 	  cat $(srcdir)/limits.h > tmp-limits.h; \
diff --git a/gbl-ctors.h b/gbl-ctors.h
index d99f622..e755eca 100644
--- a/gbl-ctors.h
+++ b/gbl-ctors.h
@@ -32,11 +32,11 @@ the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */
 */
 
 #ifdef sun
-extern void on_exit (void*, void*);
+extern int on_exit (void*, void*);
 #define ON_EXIT(FUNC,ARG) on_exit ((FUNC), (ARG))
 #else
 #ifdef HAVE_ATEXIT
-extern void atexit (void (*) (void));
+extern int atexit (void (*) (void));
 #define ON_EXIT(FUNC,ARG) atexit ((FUNC))
 #endif
 #endif
