diff --git a/cp/call.c b/cp/call.c
index d0e800c..799e8f6 100644
--- a/cp/call.c
+++ b/cp/call.c
@@ -31,10 +31,6 @@ Boston, MA 02111-1307, USA.  */
 #include "output.h"
 #include "flags.h"
 
-#ifdef HAVE_STDLIB_H
-#include <stdlib.h>
-#endif
-
 #include "obstack.h"
 #define obstack_chunk_alloc xmalloc
 #define obstack_chunk_free free
diff --git a/cp/cvt.c b/cp/cvt.c
index 2996492..3366734 100644
--- a/cp/cvt.c
+++ b/cp/cvt.c
@@ -33,10 +33,6 @@ Boston, MA 02111-1307, USA.  */
 #include "class.h"
 #include "convert.h"
 
-#ifdef HAVE_STDLIB_H
-#include <stdlib.h>
-#endif
-
 extern tree static_aggregates;
 
 static tree build_thunk PROTO((tree, tree));
diff --git a/cp/decl.c b/cp/decl.c
index f336f06..1f94744 100644
--- a/cp/decl.c
+++ b/cp/decl.c
@@ -42,10 +42,6 @@ Boston, MA 02111-1307, USA.  */
 #include "output.h"
 #include "except.h"
 
-#ifdef HAVE_STDLIB_H
-#include <stdlib.h>
-#endif
-
 #ifdef NEED_DECLARATION_FREE
 extern void free	PROTO((void *));
 #endif
diff --git a/cp/decl2.c b/cp/decl2.c
index dd2bb43..76eb0b3 100644
--- a/cp/decl2.c
+++ b/cp/decl2.c
@@ -39,10 +39,6 @@ Boston, MA 02111-1307, USA.  */
 #include "except.h"
 #include "expr.h"
 
-#ifdef HAVE_STDLIB_H
-#include <stdlib.h>
-#endif
-
 #ifdef HAVE_STRING_H
 #include <string.h>
 #endif
diff --git a/cp/g++spec.c b/cp/g++spec.c
index 66b2756..8dc049f 100644
--- a/cp/g++spec.c
+++ b/cp/g++spec.c
@@ -25,10 +25,6 @@ Boston, MA 02111-1307, USA.  */
 
 #include "gansidecl.h"
 
-#ifdef HAVE_STDLIB_H
-#include <stdlib.h>
-#endif
-
 #ifdef HAVE_STRING_H
 #include <string.h>
 #endif
@@ -59,7 +55,7 @@ lang_specific_driver (fn, in_argc, in_argv)
 
   /* This will be 0 if we encounter a situation where we should not
      link in libstdc++.  */
-  int library = 1;
+  int library = 0;
 
   /* The number of arguments being added to what's in argv, other than
      libraries.  We use this to track the number of times we've inserted
diff --git a/cp/lex.c b/cp/lex.c
index f634b21..8e79e08 100644
--- a/cp/lex.c
+++ b/cp/lex.c
@@ -51,13 +51,7 @@ Boston, MA 02111-1307, USA.  */
 #include <locale.h>
 #endif
 
-#ifdef HAVE_STDLIB_H
-#ifndef MULTIBYTE_CHARS
-#include <stdlib.h>
-#endif
-#else
 extern double atof ();
-#endif
 
 #ifdef HAVE_STRING_H
 #include <string.h>
diff --git a/cp/pt.c b/cp/pt.c
index 9f42d00..6be482e 100644
--- a/cp/pt.c
+++ b/cp/pt.c
@@ -39,10 +39,6 @@ Boston, MA 02111-1307, USA.  */
 #include "defaults.h"
 #include "except.h"
 
-#ifdef HAVE_STDLIB_H
-#include <stdlib.h>
-#endif
-
 extern struct obstack permanent_obstack;
 
 extern int lineno;
diff --git a/cp/repo.c b/cp/repo.c
index c1d756c..6f8ca01 100644
--- a/cp/repo.c
+++ b/cp/repo.c
@@ -35,11 +35,7 @@ Boston, MA 02111-1307, USA.  */
 #ifdef HAVE_STRING_H
 #include <string.h>
 #endif
-#ifdef HAVE_STDLIB_H
-#include <stdlib.h>
-#else
 extern char * getenv ();
-#endif
 
 #ifdef NEED_DECLARATION_RINDEX
 extern char *rindex ();
diff --git a/cp/tree.c b/cp/tree.c
index 98b1a6d..61ae93c 100644
--- a/cp/tree.c
+++ b/cp/tree.c
@@ -32,10 +32,6 @@ Boston, MA 02111-1307, USA.  */
 #include <varargs.h>
 #endif
 
-#ifdef HAVE_STDLIB_H
-#include <stdlib.h>
-#endif
-
 #ifdef NEED_DECLARATION_FREE
 extern void free	PROTO((void *));
 #endif
diff --git a/cp/xref.c b/cp/xref.c
index 40cdf42..8596126 100644
--- a/cp/xref.c
+++ b/cp/xref.c
@@ -28,10 +28,6 @@ Boston, MA 02111-1307, USA.  */
 
 #include <ctype.h>
 
-#ifdef HAVE_STDLIB_H
-#include <stdlib.h>
-#endif
-
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
 #endif
