diff --git a/gcc/config.gcc b/gcc/config.gcc
index 9503b96..cf2f7a6 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -1774,7 +1774,7 @@ mips*-*-netbsd*)			# NetBSD/mips, either endian.
 mips64*-*-linux* | mipsisa64*-*-linux*)
 	tm_file="dbxelf.h elfos.h gnu-user.h linux.h glibc-stdint.h ${tm_file} mips/gnu-user.h mips/gnu-user64.h mips/linux64.h"
 	tmake_file="${tmake_file} mips/t-linux64"
-	tm_defines="${tm_defines} MIPS_ABI_DEFAULT=ABI_N32"
+	tm_defines="${tm_defines} MIPS_ABI_DEFAULT=ABI_64"
 	case ${target} in
 		mips64el-st-linux-gnu)
 			tm_file="${tm_file} mips/st.h"
