diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 686e801ad..92eccc71d 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -198,6 +198,48 @@ gengtype-lex.o-warn = -Wno-error
 # mips-tfile.c contains -Wcast-qual warnings.
 mips-tfile.o-warn = -Wno-error
 expmed.o-warn = -Wno-error
+build/genhooks.o-warn = -Wno-error=unused-result
+diagnostic.o-warn = -Wno-error=unused-result
+final.o-warn = -Wno-error=unused-result
+toplev.o-warn = -Wno-error=unused-result
+targhooks.o-warn = -Wno-error=unused-result
+tree-ssa-structalias.o-warn = -Wno-error=unused-result
+fortran/trans.o-warn = -Wno-error=unused-result
+fortran/trans-array.o-warn = -Wno-error=unused-result
+fortran/trans-expr.o-warn = -Wno-error=unused-result
+fortran/trans-io.o-warn = -Wno-error=unused-result
+lto/lto.o-warn = -Wno-error=unused-result
+m68k-common.o-warn = -Wno-error=unused-result
+arm.o-warn = -Wno-error=format-security
+c-convert.o-warn = -Wno-error=format-security
+c-decl.o-warn = -Wno-error=format-security
+c-family/c-common.o-warn = -Wno-error=format-security
+c-typeck.o-warn = -Wno-error=format-security
+collect2.o-warn = -Wno-error=format-security
+cp/cvt.o-warn = -Wno-error=format-security
+cp/decl.o-warn = -Wno-error=format-security
+cp/parser.o-warn = -Wno-error=format-security
+cp/pt.o-warn = -Wno-error=format-security
+cp/typeck.o-warn = -Wno-error=format-security
+dwarf2asm.o-warn = -Wno-error=format-security
+dwarf2out.o-warn = -Wno-error=format-security
+fold-const.o-warn = -Wno-error=format-security
+fortran/decl.o-warn = -Wno-error=format-security
+fortran/frontend-passes.o-warn = -Wno-error=format-security
+fortran/intrinsic.o-warn = -Wno-error=format-security
+fortran/match.o-warn = -Wno-error=format-security
+fortran/matchexp.o-warn = -Wno-error=format-security
+fortran/module.o-warn = -Wno-error=format-security
+fortran/openmp.o-warn = -Wno-error=format-security
+fortran/primary.o-warn = -Wno-error=format-security
+fortran/symbol.o-warn = -Wno-error=format-security
+m68k.o-warn = -Wno-error=format-security
+opts.o-warn = -Wno-error=format-security
+tree-sra.o-warn = -Wno-error=format-security
+tree-ssa-uninit.o-warn = -Wno-error=format-security
+gcc.o-warn = -Wno-error=unused-result -Wno-error=format-security
+lto-wrapper.o-warn = -Wno-error=unused-result -Wno-error=format-security
+fortran/trans-decl.o-warn = -Wno-error=unused-result -Wno-error=format-security
 
 # All warnings have to be shut off in stage1 if the compiler used then
 # isn't gcc; configure determines that.  WARN_CFLAGS will be either
diff --git a/libcpp/Makefile.in b/libcpp/Makefile.in
index d6df46d62..21fd6bfcf 100644
--- a/libcpp/Makefile.in
+++ b/libcpp/Makefile.in
@@ -36,7 +36,7 @@ CFLAGS = @CFLAGS@
 WARN_CFLAGS = @warn@ @c_warn@ @WARN_PEDANTIC@ @WERROR@
 CXX = @CXX@
 CXXFLAGS = @CXXFLAGS@
-WARN_CXXFLAGS = @warn@ @WARN_PEDANTIC@ @WERROR@
+WARN_CXXFLAGS = @warn@ @WARN_PEDANTIC@ @WERROR@ -Wno-error=format-security -Wno-error=unused-result
 CPP = @CPP@
 CPPFLAGS = @CPPFLAGS@
 EXEEXT = @EXEEXT@
