diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 686e801ad..0a119612f 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -187,8 +187,8 @@ VALGRIND_DRIVER_DEFINES = @valgrind_path_defines@
 # This is how we control whether or not the additional warnings are applied.
 .-warn = $(STRICT_WARN)
 build-warn = $(STRICT_WARN)
-GCC_WARN_CFLAGS = $(LOOSE_WARN) $(C_LOOSE_WARN) $($(@D)-warn) $(if $(filter-out $(STRICT_WARN),$($(@D)-warn)),,$(C_STRICT_WARN)) $(NOCOMMON_FLAG) $($@-warn)
-GCC_WARN_CXXFLAGS = $(LOOSE_WARN) $($(@D)-warn) $(NOCOMMON_FLAG) $($@-warn)
+GCC_WARN_CFLAGS = $(LOOSE_WARN) $(C_LOOSE_WARN) $($(@D)-warn) $(if $(filter-out $(STRICT_WARN),$($(@D)-warn)),,$(C_STRICT_WARN)) $(NOCOMMON_FLAG) $($@-warn) -Wno-error=unused-result
+GCC_WARN_CXXFLAGS = $(LOOSE_WARN) $($(@D)-warn) $(NOCOMMON_FLAG) $($@-warn) -Wno-error=format-security -Wno-error=unused-result
 
 # These files are to have specific diagnostics suppressed, or are not to
 # be subject to -Werror:
diff --git a/libcpp/Makefile.in b/libcpp/Makefile.in
index d6df46d62..21fd6bfcf 100644
--- a/libcpp/Makefile.in
+++ b/libcpp/Makefile.in
@@ -36,7 +36,7 @@ CFLAGS = @CFLAGS@
 WARN_CFLAGS = @warn@ @c_warn@ @WARN_PEDANTIC@ @WERROR@
 CXX = @CXX@
 CXXFLAGS = @CXXFLAGS@
-WARN_CXXFLAGS = @warn@ @WARN_PEDANTIC@ @WERROR@
+WARN_CXXFLAGS = @warn@ @WARN_PEDANTIC@ @WERROR@ -Wno-error=format-security -Wno-error=unused-result
 CPP = @CPP@
 CPPFLAGS = @CPPFLAGS@
 EXEEXT = @EXEEXT@
