diff --git a/gcc/cp/ChangeLog b/gcc/cp/ChangeLog
index 8cdcfb471..f5619af34 100644
--- a/gcc/cp/ChangeLog
+++ b/gcc/cp/ChangeLog
@@ -2,12 +2,6 @@
 
 	* GCC 4.7.4 released.
 
-2014-05-30  Jason Merrill  <jason@redhat.com>
-
-	PR c++/56947
-	* pt.c (instantiate_decl): Don't defer instantiation of a nested
-	function.
-
 2014-05-07  Jakub Jelinek  <jakub@redhat.com>
 
 	Backported from mainline
diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index e1cc59d5e..f2b2f9bcc 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -18191,7 +18191,6 @@ instantiate_decl (tree d, int defer_ok,
      if the variable has a constant value the referring expression can
      take advantage of that fact.  */
   if (TREE_CODE (d) == VAR_DECL
-      || decl_function_context (d)
       || DECL_DECLARED_CONSTEXPR_P (d))
     defer_ok = 0;
 
