diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index 51adf9189..5fdb7d740 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -4059,8 +4059,6 @@ cp_parser_skip_to_pragma_eol (cp_parser* parser, cp_token *pragma_tok)
 
   if (pragma_tok)
     {
-      /* Ensure that the pragma is not parsed again.  */
-      cp_lexer_purge_tokens_after (parser->lexer, pragma_tok);
       parser->lexer->in_pragma = false;
     }
 }
diff --git a/gcc/testsuite/g++.dg/gomp/pr104623.C b/gcc/testsuite/g++.dg/gomp/pr104623.C
new file mode 100644
index 000000000..725e19e8c
--- /dev/null
+++ b/gcc/testsuite/g++.dg/gomp/pr104623.C
@@ -0,0 +1,9 @@
+// PR c++/104623
+// { dg-do compile }
+
+void
+foo ()
+{
+  struct __attribute__() a	// { dg-error "expected primary-expression before" }
+  #pragma omp task
+}
