diff --git a/gcc/toplev.c b/gcc/toplev.c
index ea0a2a1b0..856a0064a 100644
--- a/gcc/toplev.c
+++ b/gcc/toplev.c
@@ -2354,10 +2354,8 @@ toplev::main (int argc, char **argv)
      emit some diagnostics here.  */
   invoke_plugin_callbacks (PLUGIN_FINISH, NULL);
 
-  if (flag_diagnostics_generate_patch)
+  if (global_dc->edit_context_ptr)
     {
-      gcc_assert (global_dc->edit_context_ptr);
-
       pretty_printer pp;
       pp_show_color (&pp) = pp_show_color (global_dc->printer);
       global_dc->edit_context_ptr->print_diff (&pp, true);
