diff --git a/gcc/gcc.c b/gcc/gcc.c
index f386bb98e..a78e14072 100644
--- a/gcc/gcc.c
+++ b/gcc/gcc.c
@@ -871,7 +871,7 @@ proper position among the other output files.  */
 #endif
 
 #ifdef HAVE_AS_DEBUG_PREFIX_MAP
-#define ASM_MAP " %{fdebug-prefix-map=*:--debug-prefix-map %*}"
+#define ASM_MAP " %{ffile-prefix-map=*:--debug-prefix-map %*} %{fdebug-prefix-map=*:--debug-prefix-map %*}"
 #else
 #define ASM_MAP ""
 #endif
