diff --git a/libstdc++-v3/include/bits/basic_string.h b/libstdc++-v3/include/bits/basic_string.h
index a0763843f..d011bee47 100644
--- a/libstdc++-v3/include/bits/basic_string.h
+++ b/libstdc++-v3/include/bits/basic_string.h
@@ -2915,7 +2915,6 @@ _GLIBCXX_BEGIN_NAMESPACE_CXX11
       template<typename _Tp>
 	_If_sv<_Tp, int>
 	compare(size_type __pos, size_type __n, const _Tp& __svt) const
-	noexcept(is_same<_Tp, __sv_type>::value)
 	{
 	  __sv_type __sv = __svt;
 	  return __sv_type(*this).substr(__pos, __n).compare(__sv);
@@ -2935,7 +2934,6 @@ _GLIBCXX_BEGIN_NAMESPACE_CXX11
 	_If_sv<_Tp, int>
 	compare(size_type __pos1, size_type __n1, const _Tp& __svt,
 		size_type __pos2, size_type __n2 = npos) const
-	noexcept(is_same<_Tp, __sv_type>::value)
 	{
 	  __sv_type __sv = __svt;
 	  return __sv_type(*this)
@@ -5854,7 +5852,6 @@ _GLIBCXX_END_NAMESPACE_CXX11
       template<typename _Tp>
 	_If_sv<_Tp, int>
 	compare(size_type __pos, size_type __n, const _Tp& __svt) const
-	noexcept(is_same<_Tp, __sv_type>::value)
 	{
 	  __sv_type __sv = __svt;
 	  return __sv_type(*this).substr(__pos, __n).compare(__sv);
@@ -5874,7 +5871,6 @@ _GLIBCXX_END_NAMESPACE_CXX11
 	_If_sv<_Tp, int>
 	compare(size_type __pos1, size_type __n1, const _Tp& __svt,
 		size_type __pos2, size_type __n2 = npos) const
-	noexcept(is_same<_Tp, __sv_type>::value)
 	{
 	  __sv_type __sv = __svt;
 	  return __sv_type(*this)
diff --git a/libstdc++-v3/testsuite/21_strings/basic_string/operations/compare/char/123991.cc b/libstdc++-v3/testsuite/21_strings/basic_string/operations/compare/char/123991.cc
new file mode 100644
index 000000000..53ef2d460
--- /dev/null
+++ b/libstdc++-v3/testsuite/21_strings/basic_string/operations/compare/char/123991.cc
@@ -0,0 +1,56 @@
+// { dg-do run { target c++17 } }
+
+// Bug 123991 - std::string::compare crashes instead of throwing
+
+#include <string>
+#include <string_view>
+#include <stdexcept>
+#include <testsuite_hooks.h>
+
+void
+test_compare_3arg()
+{
+  std::string_view sv;
+  std::string s;
+  static_assert( ! noexcept(s.compare(0, 0, sv)) );
+#ifdef __cpp_exceptions
+  try
+  {
+    (void) s.compare(1, 0, sv);
+    VERIFY(false);
+  }
+  catch (const std::out_of_range&)
+  { }
+#endif
+}
+
+void
+test_compare_5arg()
+{
+  std::string_view sv;
+  std::string s;
+  static_assert( ! noexcept(s.compare(0, 0, sv, 0, 0)) );
+#ifdef __cpp_exceptions
+  try
+  {
+    (void) s.compare(1, 0, sv, 0, 0);
+    VERIFY(false);
+  }
+  catch (const std::out_of_range&)
+  { }
+
+  try
+  {
+    (void) s.compare(0, 0, sv, 1, 0);
+    VERIFY(false);
+  }
+  catch (const std::out_of_range&)
+  { }
+#endif
+}
+
+int main()
+{
+  test_compare_3arg();
+  test_compare_5arg();
+}
diff --git a/libstdc++-v3/testsuite/21_strings/basic_string/operations/compare/wchar_t/123991.cc b/libstdc++-v3/testsuite/21_strings/basic_string/operations/compare/wchar_t/123991.cc
new file mode 100644
index 000000000..c4188bbff
--- /dev/null
+++ b/libstdc++-v3/testsuite/21_strings/basic_string/operations/compare/wchar_t/123991.cc
@@ -0,0 +1,56 @@
+// { dg-do run { target c++17 } }
+
+// Bug 123991 - std::string::compare crashes instead of throwing
+
+#include <string>
+#include <string_view>
+#include <stdexcept>
+#include <testsuite_hooks.h>
+
+void
+test_compare_3arg()
+{
+  std::wstring_view sv;
+  std::wstring s;
+  static_assert( ! noexcept(s.compare(0, 0, sv)) );
+#ifdef __cpp_exceptions
+  try
+  {
+    (void) s.compare(1, 0, sv);
+    VERIFY(false);
+  }
+  catch (const std::out_of_range&)
+  { }
+#endif
+}
+
+void
+test_compare_5arg()
+{
+  std::wstring_view sv;
+  std::wstring s;
+  static_assert( ! noexcept(s.compare(0, 0, sv, 0, 0)) );
+#ifdef __cpp_exceptions
+  try
+  {
+    (void) s.compare(1, 0, sv, 0, 0);
+    VERIFY(false);
+  }
+  catch (const std::out_of_range&)
+  { }
+
+  try
+  {
+    (void) s.compare(0, 0, sv, 1, 0);
+    VERIFY(false);
+  }
+  catch (const std::out_of_range&)
+  { }
+#endif
+}
+
+int main()
+{
+  test_compare_3arg();
+  test_compare_5arg();
+}
