diff --git a/gcc/fortran/trans-expr.c b/gcc/fortran/trans-expr.c
index e47fd2ce3..9b2d2d8fd 100644
--- a/gcc/fortran/trans-expr.c
+++ b/gcc/fortran/trans-expr.c
@@ -6094,7 +6094,8 @@ gfc_trans_subcomponent_assign (tree dest, gfc_component * cm, gfc_expr * expr)
     {
       gfc_init_se (&se, NULL);
       /* Pointer component.  */
-      if (cm->attr.dimension && !cm->attr.proc_pointer)
+      if ((cm->attr.dimension || cm->attr.codimension)
+	  && !cm->attr.proc_pointer)
 	{
 	  /* Array pointer.  */
 	  if (expr->expr_type == EXPR_NULL)
@@ -6131,7 +6132,8 @@ gfc_trans_subcomponent_assign (tree dest, gfc_component * cm, gfc_expr * expr)
 					gfc_class_initializer (&cm->ts, expr));
       gfc_add_expr_to_block (&block, tmp);
     }
-  else if (cm->attr.dimension && !cm->attr.proc_pointer)
+  else if ((cm->attr.dimension || cm->attr.codimension)
+	   && !cm->attr.proc_pointer)
     {
       if (cm->attr.allocatable && expr->expr_type == EXPR_NULL)
  	gfc_conv_descriptor_data_set (&block, dest, null_pointer_node);
diff --git a/gcc/testsuite/gfortran.dg/coarray/alloc_comp_3.f90 b/gcc/testsuite/gfortran.dg/coarray/alloc_comp_3.f90
new file mode 100644
index 000000000..cf2d54203
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/coarray/alloc_comp_3.f90
@@ -0,0 +1,23 @@
+! { dg-do compile }
+!
+! PR fortran/60881
+!
+! Contributed by Damian Rouson
+!
+! Was ICEing before
+!
+program main
+  implicit none
+  type co_object
+    logical :: defined=.false.
+    real, allocatable :: dummy_to_facilitate_extension[:]
+  end type
+  type, extends(co_object) :: global_field
+  end type
+  type(global_field) T
+  call assign_local_field(T)
+contains
+  subroutine assign_local_field(lhs)
+    type(global_field) lhs
+  end subroutine
+end program
