diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index 132382443..78168439c 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -31831,10 +31831,7 @@ c_parse_file (void)
   static bool already_called = false;
 
   if (already_called)
-    {
-      sorry ("inter-module optimizations not implemented for C++");
-      return;
-    }
+    fatal_error ("inter-module optimizations not implemented for C++");
   already_called = true;
 
   the_parser = cp_parser_new ();
