diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index 1eec11c15..76a43091b 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -22069,9 +22069,6 @@ cp_parser_std_attribute_spec (cp_parser *parser)
 
       if (!cp_parser_parse_definitely (parser))
 	{
-	  gcc_assert (alignas_expr == error_mark_node
-		      || alignas_expr == NULL_TREE);
-
 	  alignas_expr =
 	    cp_parser_assignment_expression (parser, /*cast_p=*/false,
 					     /**cp_id_kind=*/NULL);
diff --git a/gcc/testsuite/g++.dg/cpp0x/pr78341.C b/gcc/testsuite/g++.dg/cpp0x/pr78341.C
new file mode 100644
index 000000000..af9063840
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/pr78341.C
@@ -0,0 +1,4 @@
+// PR c++/78341
+// { dg-do compile { target c++11 } }
+
+alignas (alignas double // { dg-error "" }
