diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index 35f8d6f88..82e7844c6 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -23680,8 +23680,9 @@ cp_parser_late_parsing_for_member (cp_parser* parser, tree member_function)
   maybe_begin_member_template_processing (member_function);
 
   /* If the body of the function has not yet been parsed, parse it
-     now.  */
-  if (DECL_PENDING_INLINE_P (member_function))
+     now.  Except if the tokens have been purged (PR c++/39751).  */
+  if (DECL_PENDING_INLINE_P (member_function)
+      && !DECL_PENDING_INLINE_INFO (member_function)->first->purged_p)
     {
       tree function_scope;
       cp_token_cache *tokens;
diff --git a/gcc/testsuite/g++.dg/parse/pr39751.C b/gcc/testsuite/g++.dg/parse/pr39751.C
new file mode 100644
index 000000000..cfe803c6b
--- /dev/null
+++ b/gcc/testsuite/g++.dg/parse/pr39751.C
@@ -0,0 +1,10 @@
+/* { dg-do compile } */
+
+template < typename >
+struct A
+{
+A < struct
+{
+f () :
+
+// { dg-excess-errors "" }
