diff --git a/gcc/fortran/frontend-passes.c b/gcc/fortran/frontend-passes.c
index 668ab2255..84d8fcf09 100644
--- a/gcc/fortran/frontend-passes.c
+++ b/gcc/fortran/frontend-passes.c
@@ -887,6 +887,9 @@ optimize_binop_array_assignment (gfc_code *c, gfc_expr **rhs, bool seen_op)
 {
   gfc_expr *e;
 
+  if (!*rhs)
+    return false;
+
   e = *rhs;
   if (e->expr_type == EXPR_OP)
     {
diff --git a/gcc/testsuite/gfortran.dg/pr77694.f90 b/gcc/testsuite/gfortran.dg/pr77694.f90
new file mode 100644
index 000000000..d9dc01b90
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/pr77694.f90
@@ -0,0 +1,7 @@
+! { dg-do compile }
+! { dg-options -O }
+program p
+   logical x(2), y(2)
+   x = .true.
+   y = .nt. x   ! { dg-error "Unknown operator" }
+end
