diff --git a/gcc/cp/semantics.c b/gcc/cp/semantics.c
index aed275f85..8b6195659 100644
--- a/gcc/cp/semantics.c
+++ b/gcc/cp/semantics.c
@@ -5114,6 +5114,8 @@ finish_omp_reduction_clause (tree c, bool *need_default_ctor, bool *need_dtor)
       id = OVL_CURRENT (id);
       mark_used (id);
       tree body = DECL_SAVED_TREE (id);
+      if (!body)
+	return true;
       if (TREE_CODE (body) == STATEMENT_LIST)
 	{
 	  tree_stmt_iterator tsi;
diff --git a/gcc/testsuite/g++.dg/gomp/pr59628.C b/gcc/testsuite/g++.dg/gomp/pr59628.C
new file mode 100644
index 000000000..518200b9c
--- /dev/null
+++ b/gcc/testsuite/g++.dg/gomp/pr59628.C
@@ -0,0 +1,13 @@
+// PR c++/59628
+// { dg-do compile }
+// { dg-options "-fopenmp" }
+
+struct A { int i; };
+
+void foo()
+{
+  A a;
+  #pragma omp declare reduction (+: A: omp_out.i +: omp_in.i)  // { dg-error "expected" }
+  #pragma omp parallel reduction (+: a)
+  ;
+}
