diff --git a/gcc/emit-rtl.c b/gcc/emit-rtl.c
index 3041b9e7a..e82382247 100644
--- a/gcc/emit-rtl.c
+++ b/gcc/emit-rtl.c
@@ -3524,6 +3524,11 @@ try_split (rtx pat, rtx trial, int last)
 		   next = NEXT_INSN (next))
 		if (NOTE_KIND (next) == NOTE_INSN_CALL_ARG_LOCATION)
 		  {
+		    /* Advance after to the next instruction if it is about to
+		       be removed */
+		    if (after == next)
+		      after = NEXT_INSN(after);
+
 		    remove_insn (next);
 		    add_insn_after (next, insn, NULL);
 		    break;
