diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 1f5bbb3a8..2771ecf6e 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -214,14 +214,26 @@ GCC_WARN_CXXFLAGS = $(LOOSE_WARN) $($(@D)-warn) $(NOCOMMON_FLAG) $($@-warn)
 # These files are to have specific diagnostics suppressed, or are not to
 # be subject to -Werror:
 # flex output may yield harmless "no previous prototype" warnings
+build/gengtype.o-warn = -Wno-format-overflow
 build/gengtype-lex.o-warn = -Wno-error
+gengtype.o-warn = -Wno-format-overflow
 gengtype-lex.o-warn = -Wno-error
 libgcov-util.o-warn = -Wno-error
 libgcov-driver-tool.o-warn = -Wno-error
 libgcov-merge-tool.o-warn = -Wno-error
+aarch64.o-warn = -Wno-error
 gimple-match.o-warn = -Wno-unused
 generic-match.o-warn = -Wno-unused
 dfp.o-warn = -Wno-strict-aliasing
+arm.o-warn = -Wno-error=format-diag
+build/genhooks.o-warn = -Wno-error=unused-result
+gcc.o-warn = -Wno-error=unused-result
+ggc-common.o-warn = -Wno-error=deprecated-declarations
+lto/lto-common.o-warn = -Wno-error=unused-result
+lto-wrapper.o-warn = -Wno-error=unused-result
+m68k-common.o-warn = -Wno-error=unused-result
+objc/objc-next-runtime-abi-02.o-warn = -Wno-error=unused-result
+objcp/objc-next-runtime-abi-02.o-warn = -Wno-error=unused-result
 
 # All warnings have to be shut off in stage1 if the compiler used then
 # isn't gcc; configure determines that.  WARN_CFLAGS will be either
diff --git a/gcc/config/i386/i386-expand.cc b/gcc/config/i386/i386-expand.cc
index 5c8d9c556..d326c73ae 100644
--- a/gcc/config/i386/i386-expand.cc
+++ b/gcc/config/i386/i386-expand.cc
@@ -19712,6 +19712,8 @@ static bool expand_vec_perm_interleave3 (struct expand_vec_perm_d *d);
    a two vector permutation into a single vector permutation by using
    an interleave operation to merge the vectors.  */
 
+#pragma GCC push_options
+#pragma GCC optimize ("-O2")
 static bool
 expand_vec_perm_interleave2 (struct expand_vec_perm_d *d)
 {
@@ -20025,6 +20027,7 @@ expand_vec_perm_interleave2 (struct expand_vec_perm_d *d)
   emit_insn (seq);
   return true;
 }
+#pragma GCC pop_options
 
 /* A subroutine of ix86_expand_vec_perm_const_1.  Try to simplify
    a single vector cross-lane permutation into vpermq followed
diff --git a/gcc/fortran/symbol.cc b/gcc/fortran/symbol.cc
index e7de5a8f5..0e11963d6 100644
--- a/gcc/fortran/symbol.cc
+++ b/gcc/fortran/symbol.cc
@@ -3481,8 +3481,8 @@ gfc_get_ha_sym_tree (const char *name, gfc_symtree **result)
 int
 gfc_get_ha_symbol (const char *name, gfc_symbol **result)
 {
-  int i;
-  gfc_symtree *st;
+  int i = 0;
+  gfc_symtree *st = NULL;
 
   i = gfc_get_ha_sym_tree (name, &st);
 
diff --git a/libcpp/charset.cc b/libcpp/charset.cc
index ca8b7cf7a..3d70a71a8 100644
--- a/libcpp/charset.cc
+++ b/libcpp/charset.cc
@@ -224,6 +224,8 @@ one_utf8_to_cppchar (const uchar **inbufp, size_t *inbytesleftp,
   return 0;
 }
 
+#pragma GCC push_options
+#pragma GCC optimize ("-O2")
 static inline int
 one_cppchar_to_utf8 (cppchar_t c, uchar **outbufp, size_t *outbytesleftp)
 {
@@ -257,6 +259,7 @@ one_cppchar_to_utf8 (cppchar_t c, uchar **outbufp, size_t *outbytesleftp)
   *outbufp = outbuf;
   return 0;
 }
+#pragma GCC pop_options
 
 /* The following four functions transform one character between the two
    encodings named in the function name.  All have the signature
diff --git a/libdecnumber/decNumber.c b/libdecnumber/decNumber.c
index 22ac359a6..48f0b07b4 100644
--- a/libdecnumber/decNumber.c
+++ b/libdecnumber/decNumber.c
@@ -927,9 +927,9 @@ decNumber * decNumberCompareTotalMag(decNumber *res, const decNumber *lhs,
 				     const decNumber *rhs, decContext *set) {
   uInt status=0;		   /* accumulator */
   uInt needbytes;		   /* for space calculations */
-  decNumber bufa[D2N(DECBUFFER+1)];/* +1 in case DECBUFFER=0 */
+  decNumber bufa[D2N(DECBUFFER+1)] = {0};/* +1 in case DECBUFFER=0 */
   decNumber *allocbufa=NULL;	   /* -> allocated bufa, iff allocated */
-  decNumber bufb[D2N(DECBUFFER+1)];
+  decNumber bufb[D2N(DECBUFFER+1)] = {0};
   decNumber *allocbufb=NULL;	   /* -> allocated bufb, iff allocated */
   decNumber *a, *b;		   /* temporary pointers */
 
@@ -947,6 +947,7 @@ decNumber * decNumberCompareTotalMag(decNumber *res, const decNumber *lhs,
 	if (allocbufa==NULL) {		/* hopeless -- abandon */
 	  status|=DEC_Insufficient_storage;
 	  break;}
+	memset(allocbufa, 0x00, needbytes);
 	a=allocbufa;			/* use the allocated space */
 	}
       decNumberCopy(a, lhs);		/* copy content */
@@ -961,6 +962,7 @@ decNumber * decNumberCompareTotalMag(decNumber *res, const decNumber *lhs,
 	if (allocbufb==NULL) {		/* hopeless -- abandon */
 	  status|=DEC_Insufficient_storage;
 	  break;}
+	memset(allocbufb, 0x00, needbytes);
 	b=allocbufb;			/* use the allocated space */
 	}
       decNumberCopy(b, rhs);		/* copy content */
@@ -3343,6 +3345,8 @@ const char *decNumberClassToString(enum decClass eclass) {
 /* All fields are updated as required.	This is a utility operation,  */
 /* so special values are unchanged and no error is possible.	      */
 /* ------------------------------------------------------------------ */
+#pragma GCC push_options
+#pragma GCC optimize ("-O2")
 decNumber * decNumberCopy(decNumber *dest, const decNumber *src) {
 
   #if DECCHECK
@@ -3370,6 +3374,7 @@ decNumber * decNumberCopy(decNumber *dest, const decNumber *src) {
     }
   return dest;
   } /* decNumberCopy */
+#pragma GCC pop_options
 
 /* ------------------------------------------------------------------ */
 /* decNumberCopyAbs -- quiet absolute value operator		      */
diff --git a/libstdc++-v3/config/locale/gnu/ctype_members.cc b/libstdc++-v3/config/locale/gnu/ctype_members.cc
index bd190712e..9382d7629 100644
--- a/libstdc++-v3/config/locale/gnu/ctype_members.cc
+++ b/libstdc++-v3/config/locale/gnu/ctype_members.cc
@@ -221,7 +221,7 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
   ctype<wchar_t>::
   do_narrow(wchar_t __wc, char __dfault) const
   {
-    if (__wc >= 0 && __wc < 128 && _M_narrow_ok)
+    if (__wc < 128 && _M_narrow_ok)
       return _M_narrow[__wc];
 #if __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ > 2)
     __c_locale __old = __uselocale(_M_c_locale_ctype);
@@ -244,7 +244,7 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
     if (_M_narrow_ok)
       while (__lo < __hi)
 	{
-	  if (*__lo >= 0 && *__lo < 128)
+	  if (*__lo < 128)
 	    *__dest = _M_narrow[*__lo];
 	  else
 	    {
diff --git a/libstdc++-v3/libsupc++/eh_call.cc b/libstdc++-v3/libsupc++/eh_call.cc
index e880b4bf9..ce9c762cc 100644
--- a/libstdc++-v3/libsupc++/eh_call.cc
+++ b/libstdc++-v3/libsupc++/eh_call.cc
@@ -60,6 +60,8 @@ __cxa_call_terminate(_Unwind_Exception* ue_header) throw ()
 
 
 #ifdef __ARM_EABI_UNWINDER__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 // The ARM EABI __cxa_call_unexpected has the same semantics as the generic
 // routine, but the exception specification has a different format.
 extern "C" void
@@ -158,4 +160,5 @@ __cxa_call_unexpected(void* exc_obj_in)
       __terminate(terminateHandler);
     }
 }
+#pragma GCC diagnostic pop
 #endif // __ARM_EABI_UNWINDER__
diff --git a/libstdc++-v3/libsupc++/pure.cc b/libstdc++-v3/libsupc++/pure.cc
index 30b4e57de..bab21fe21 100644
--- a/libstdc++-v3/libsupc++/pure.cc
+++ b/libstdc++-v3/libsupc++/pure.cc
@@ -46,13 +46,19 @@
 extern "C" void
 __cxxabiv1::__cxa_pure_virtual (void)
 {
-  writestr ("pure virtual method called\n");
+#if _GLIBCXX_HOSTED && _GLIBCXX_VERBOSE
+  auto ret = writestr ("pure virtual method called\n");
+  (void)ret;
+#endif
   std::terminate ();
 }
 
 extern "C" void
 __cxxabiv1::__cxa_deleted_virtual (void)
 {
-  writestr ("deleted virtual method called\n");
+#if _GLIBCXX_HOSTED && _GLIBCXX_VERBOSE
+  auto ret = writestr ("deleted virtual method called\n");
+  (void)ret;
+#endif
   std::terminate ();
 }
diff --git a/libstdc++-v3/src/c++17/floating_to_chars.cc b/libstdc++-v3/src/c++17/floating_to_chars.cc
index 4599d68a3..5dbc88e81 100644
--- a/libstdc++-v3/src/c++17/floating_to_chars.cc
+++ b/libstdc++-v3/src/c++17/floating_to_chars.cc
@@ -85,6 +85,10 @@ using F128_type = __float128;
 using F128_type = void;
 #endif
 
+#if !defined __SIZEOF_INT128__
+#pragma GCC diagnostic ignored "-Wunused-function"
+#endif
+
 namespace
 {
 #if defined __SIZEOF_INT128__
