diff --git a/stl/stl_config.h b/stl/stl_config.h
index ebf6464..446b101 100644
--- a/stl/stl_config.h
+++ b/stl/stl_config.h
@@ -263,6 +263,14 @@
 #  define __stl_debug_do(expr)
 # endif
 
+#if defined(__GNUC__) && __GNUC__ == 2 && __GNUC_MINOR__ < 8
+#ifndef __STL_HACK_DUMMY
+const int __dummy = 0;
+#define __STL_HACK_DUMMY
+#endif
+#define __PUT_STATIC_DATA_MEMBERS_HERE
+#endif
+
 // tuning of static template data members workaround
 # if ( __STL_STATIC_TEMPLATE_DATA < 1 )
 // ignore __PUT directive in this case
