diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index e844ae248..890f1c731 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -25643,6 +25643,12 @@ cp_parser_std_attribute_spec (cp_parser *parser)
       if (alignas_expr == error_mark_node)
 	return error_mark_node;
 
+      /* Missing ')' means the code cannot possibly be valid; go ahead
+	 and commit to make sure we issue a hard error.  */
+      if (cp_parser_uncommitted_to_tentative_parse_p (parser)
+	  && cp_lexer_next_token_is_not (parser->lexer, CPP_CLOSE_PAREN))
+	cp_parser_commit_to_tentative_parse (parser);
+
       if (!parens.require_close (parser))
 	return error_mark_node;
 
diff --git a/gcc/testsuite/g++.dg/parse/alignas1.C b/gcc/testsuite/g++.dg/parse/alignas1.C
new file mode 100644
index 000000000..8f361018f
--- /dev/null
+++ b/gcc/testsuite/g++.dg/parse/alignas1.C
@@ -0,0 +1,10 @@
+// PR c++/64235
+// { dg-do compile { target c++11 } }
+
+struct S {
+  double d;
+};
+
+struct alignas(sizeof(S) S1 { }; // { dg-error "expected '\\)' before 'S1'" }
+struct alignas(16 S2 { }; // { dg-error "expected '\\)' before 'S2'" }
+struct alignas(int S3 { }; // { dg-error "expected '\\)' before 'S3'" }
