diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 110d0e26e..470d5c4a6 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -4492,7 +4492,9 @@ process_template_parm (tree list, location_t parm_loc, tree parm,
      process_template_parm could fail. */
   tree reqs = finish_shorthand_constraint (parm, constr);
 
-  pushdecl (decl);
+  decl = pushdecl (decl);
+  if (!is_non_type)
+    parm = decl;
 
   if (defval && TREE_CODE (defval) == OVERLOAD)
     lookup_keep (defval, true);
diff --git a/gcc/testsuite/g++.dg/template/pr90138.C b/gcc/testsuite/g++.dg/template/pr90138.C
new file mode 100644
index 000000000..473b73db1
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/pr90138.C
@@ -0,0 +1,5 @@
+// PR c++/90138
+
+template <, typename T, typename typename, typename T>	// { dg-error "expected" }
+struct S;	// { dg-error "no default" }
+// { dg-error "two or more" "" { target *-*-* } .-2 }
