diff --git a/include/asm-i386/posix_types.h b/include/asm-i386/posix_types.h
index 712ef70..cdbf1d3 100644
--- a/include/asm-i386/posix_types.h
+++ b/include/asm-i386/posix_types.h
@@ -58,6 +58,6 @@ typedef struct {
 		__asm__ __volatile__("cld ; rep ; stosl" \
 			:"=m" (*(__kernel_fd_set *) (fdsetp)) \
 			:"a" (0), "c" (__FDSET_LONGS), \
-			"D" ((__kernel_fd_set *) (fdsetp)) :"cx","di")
+			"D" ((__kernel_fd_set *) (fdsetp)))
 
 #endif
