diff --git a/gcc/ipa-reference.c b/gcc/ipa-reference.c
index cc790fdbc..2629426cb 100644
--- a/gcc/ipa-reference.c
+++ b/gcc/ipa-reference.c
@@ -665,9 +665,6 @@ analyze_function (struct cgraph_node *fn)
   tree decl = fn->decl;
   struct function *this_cfun = DECL_STRUCT_FUNCTION (decl);
   basic_block this_block;
-#ifdef ENABLE_CHECKING
-  tree step;
-#endif
   ipa_reference_local_vars_info_t local;
 
   if (dump_file)
@@ -717,20 +714,6 @@ analyze_function (struct cgraph_node *fn)
   if (local->calls_read_all)
     BITMAP_FREE (local->statics_read);
 
-
-#ifdef ENABLE_CHECKING
-  /* Verify that all local initializers was expanded by gimplifier.  */
-  for (step = DECL_STRUCT_FUNCTION (decl)->local_decls;
-       step;
-       step = TREE_CHAIN (step))
-    {
-      tree var = TREE_VALUE (step);
-      if (TREE_CODE (var) == VAR_DECL
-	  && DECL_INITIAL (var)
-	  && !TREE_STATIC (var))
-	gcc_unreachable ();
-    }
-#endif
   pop_cfun ();
   current_function_decl = NULL;
 }
