diff --git a/gcc/c-common.c b/gcc/c-common.c
index 0eac72727..924539564 100644
--- a/gcc/c-common.c
+++ b/gcc/c-common.c
@@ -6566,7 +6566,7 @@ handle_mode_attribute (tree *node, tree name, tree args,
 	  return NULL_TREE;
 	}
 
-      *node = typefm;
+      *node = build_qualified_type (typefm, TYPE_QUALS (type));
     }
 
   return NULL_TREE;
diff --git a/gcc/testsuite/c-c++-common/pr79641.c b/gcc/testsuite/c-c++-common/pr79641.c
new file mode 100644
index 000000000..5d2c9b4d3
--- /dev/null
+++ b/gcc/testsuite/c-c++-common/pr79641.c
@@ -0,0 +1,4 @@
+/* PR c++/79641 */
+/* { dg-do compile } */
+
+const int __attribute__((__mode__ (__QI__))) i = 0;
