diff --git a/gcc/fortran/iresolve.c b/gcc/fortran/iresolve.c
index a2ed88ca7..693c59fd5 100644
--- a/gcc/fortran/iresolve.c
+++ b/gcc/fortran/iresolve.c
@@ -142,6 +142,9 @@ gfc_resolve_adjustl (gfc_expr *f, gfc_expr *string)
 {
   f->ts.type = BT_CHARACTER;
   f->ts.kind = string->ts.kind;
+  if (string->ts.u.cl)
+    f->ts.u.cl = gfc_new_charlen (gfc_current_ns, string->ts.u.cl);
+
   f->value.function.name = gfc_get_string ("__adjustl_s%d", f->ts.kind);
 }
 
@@ -151,6 +154,9 @@ gfc_resolve_adjustr (gfc_expr *f, gfc_expr *string)
 {
   f->ts.type = BT_CHARACTER;
   f->ts.kind = string->ts.kind;
+  if (string->ts.u.cl)
+    f->ts.u.cl = gfc_new_charlen (gfc_current_ns, string->ts.u.cl);
+
   f->value.function.name = gfc_get_string ("__adjustr_s%d", f->ts.kind);
 }
 
diff --git a/gcc/testsuite/gfortran.dg/adjustl_1.f90 b/gcc/testsuite/gfortran.dg/adjustl_1.f90
new file mode 100644
index 000000000..1238da8c2
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/adjustl_1.f90
@@ -0,0 +1,8 @@
+! { dg-do  run }
+! PR 52749 - this used to ICE.
+! Original test case by Stefan Mauerberger.
+PROGRAM test
+  character(len=10) :: u
+  WRITE(unit=u,fmt='(3A)') PACK(ADJUSTL([" a", " b"]), [.TRUE., .FALSE.])
+  if (u .ne. 'a    ') call abort
+END PROGRAM test
