diff --git a/gcc/ira-color.c b/gcc/ira-color.c
index b9114e6a9..440c9eec4 100644
--- a/gcc/ira-color.c
+++ b/gcc/ira-color.c
@@ -286,7 +286,8 @@ update_copy_costs (ira_allocno_t allocno, bool decr_p)
 	    gcc_unreachable ();
 
 	  cover_class = ALLOCNO_COVER_CLASS (another_allocno);
-	  if (! ira_reg_classes_intersect_p[rclass][cover_class]
+	  if (! TEST_HARD_REG_BIT (reg_class_contents[cover_class],
+				   hard_regno)
 	      || ALLOCNO_ASSIGNED_P (another_allocno))
 	    continue;
 
diff --git a/gcc/testsuite/gcc.c-torture/compile/pr46002.c b/gcc/testsuite/gcc.c-torture/compile/pr46002.c
new file mode 100644
index 000000000..27a3a3a28
--- /dev/null
+++ b/gcc/testsuite/gcc.c-torture/compile/pr46002.c
@@ -0,0 +1,10 @@
+/* { dg-options "-fira-algorithm=priority" } */
+char **
+foo (char **p, char *cmp, unsigned i)
+{
+  for (; *p; p++)
+    if (__builtin_strncmp (*p, cmp, i))
+      if (i == __builtin_strlen (*p))
+	break;
+  return p;
+}
