diff --git a/gcc/predict.c b/gcc/predict.c
index 617fd5cc0..2a6a25bd0 100644
--- a/gcc/predict.c
+++ b/gcc/predict.c
@@ -1327,9 +1327,17 @@ strip_predict_hints (void)
 		  && gimple_call_num_args (stmt) == 2)
 		{
 		  var = gimple_call_lhs (stmt);
-		  ass_stmt = gimple_build_assign (var, gimple_call_arg (stmt, 0));
-
-		  gsi_replace (&bi, ass_stmt, true);
+		  if (var)
+		    {
+		      ass_stmt
+			= gimple_build_assign (var, gimple_call_arg (stmt, 0));
+		      gsi_replace (&bi, ass_stmt, true);
+		    }
+		  else
+		    {
+		      gsi_remove (&bi, true);
+		      continue;
+		    }
 		}
 	    }
 	  gsi_next (&bi);
diff --git a/gcc/testsuite/gcc.dg/pr46288.c b/gcc/testsuite/gcc.dg/pr46288.c
new file mode 100644
index 000000000..24c8874fe
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/pr46288.c
@@ -0,0 +1,8 @@
+/* { dg-do compile } */
+/* { dg-options "-O -fno-tree-dce" } */
+
+void foo (int i)
+{
+  __builtin_expect (i, i);
+}
+
