diff --git a/gcc/cp/method.c b/gcc/cp/method.c
index 088c9e7c1..71b231033 100644
--- a/gcc/cp/method.c
+++ b/gcc/cp/method.c
@@ -1035,6 +1035,12 @@ defaultable_fn_check (tree fn)
 {
   special_function_kind kind = sfk_none;
 
+  if (template_parm_scope_p ())
+    {
+      error ("a template cannot be defaulted");
+      return false;
+    }
+
   if (DECL_CONSTRUCTOR_P (fn))
     {
       if (FUNCTION_FIRST_USER_PARMTYPE (fn) == void_list_node)
diff --git a/gcc/testsuite/g++.dg/cpp0x/defaulted24.C b/gcc/testsuite/g++.dg/cpp0x/defaulted24.C
new file mode 100644
index 000000000..307bf94ab
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/defaulted24.C
@@ -0,0 +1,6 @@
+// PR c++/48280
+// { dg-options -std=c++0x }
+
+struct S {
+  template < typename > S (const S &) = default; // { dg-error "" }
+};
