diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index 772b5eae9..68cff6f87 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -3917,7 +3917,8 @@ cp_parser_unqualified_id (cp_parser* parser,
 	    && (cp_lexer_peek_nth_token (parser->lexer, 2)->type
 		!= CPP_LESS)
 	    && (token->u.value == TYPE_IDENTIFIER (scope)
-		|| constructor_name_p (token->u.value, scope)))
+		|| (CLASS_TYPE_P (scope)
+		    && constructor_name_p (token->u.value, scope))))
 	  {
 	    cp_lexer_consume_token (parser->lexer);
 	    return build_nt (BIT_NOT_EXPR, scope);
diff --git a/gcc/testsuite/g++.dg/cpp0x/enum18.C b/gcc/testsuite/g++.dg/cpp0x/enum18.C
new file mode 100644
index 000000000..5575ca6ce
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/enum18.C
@@ -0,0 +1,8 @@
+// PR c++/47277
+// { dg-options -std=c++0x }
+
+int main(void) {
+           enum e {};
+           e ev;
+           ev.e::~e_u();	// { dg-error "" }
+}
