diff --git a/gcc/cp/call.c b/gcc/cp/call.c
index c902c7338..47155e4e9 100644
--- a/gcc/cp/call.c
+++ b/gcc/cp/call.c
@@ -6225,7 +6225,7 @@ build_new_method_call (tree instance, tree fns, VEC(tree,gc) **args,
     *fn_p = NULL_TREE;
 
   if (error_operand_p (instance)
-      || error_operand_p (fns))
+      || !fns || error_operand_p (fns))
     return error_mark_node;
 
   if (!BASELINK_P (fns))
diff --git a/gcc/testsuite/g++.dg/template/crash95.C b/gcc/testsuite/g++.dg/template/crash95.C
new file mode 100644
index 000000000..959a9d066
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/crash95.C
@@ -0,0 +1,11 @@
+// PR c++/43705
+
+template < typename > struct S
+{
+  template < > struct S < int > // { dg-error "explicit|specialization|template|parameter" }
+  {
+    S(int);
+  };
+};
+
+S < int > s(0);
