diff --git a/gcc/config/arm/arm.c b/gcc/config/arm/arm.c
index 8c102515e..cf95ff045 100644
--- a/gcc/config/arm/arm.c
+++ b/gcc/config/arm/arm.c
@@ -21510,10 +21510,13 @@ arm_order_regs_for_local_alloc (void)
 void
 arm_optimization_options (int level, int size ATTRIBUTE_UNUSED)
 {
-  /* Enable section anchors by default at -O1 or higher.
-     Use 2 to distinguish from an explicit -fsection-anchors
-     given on the command line.  */
-  if (level > 0)
+  /* Enable section anchors by default.
+     Skip section anchors for Objective C and Objective C++
+     until front-ends fixed.
+     Do not enable section anchors without toplevel reorder.  */
+  if (level > 0
+      && lang_hooks.name[4] != 'O'
+      && flag_toplevel_reorder != 0)
     flag_section_anchors = 2;
 }
 
