diff --git a/gcc/testsuite/obj-c++.dg/lookup-2.mm b/gcc/testsuite/obj-c++.dg/lookup-2.mm
index 4d56cf25c..5eaf38996 100644
--- a/gcc/testsuite/obj-c++.dg/lookup-2.mm
+++ b/gcc/testsuite/obj-c++.dg/lookup-2.mm
@@ -1,5 +1,4 @@
 /* { dg-do run } */
-/* { dg-xfail-if "PR23614" { "*-*-*" } { "*" } { "-fnext-runtime" } } */
 /* { dg-xfail-run-if "Needs OBJC2 ABI" { *-*-darwin* && { lp64 && { ! objc2 } } } { "-fnext-runtime" } { "" } } */
 #include "../objc-obj-c++-shared/Object1.h"
 #include <stdlib.h>
@@ -38,7 +37,7 @@ MyWidget gWidget;
 @implementation Container
 + (MyWidget *)elementForView:(Foo *)view
 {
-    MyWidget *widget = nil;
+    MyWidget *widget = 0;
     if ([view conformsTo:@protocol(MyProto)]) {
         widget = [(Foo <MyProto> *)view widget];
     }
