diff --git a/bfd/elflink.c b/bfd/elflink.c
index 2b1450fa..51f60cf4 100644
--- a/bfd/elflink.c
+++ b/bfd/elflink.c
@@ -7169,28 +7169,6 @@ warning: enabling an executable stack because of -z execstack command line optio
 
       if (notesec || info->stacksize > 0)
 	{
-	  if (exec)
-	    {
-	      if (info->warn_execstack != 0)
-		{
-		  /* PR 29072: Because an executable stack is a serious
-		     security risk, make sure that the user knows that it is
-		     being enabled despite the fact that it was not requested
-		     on the command line.  */
-		  if (noteobj)
-		    _bfd_error_handler (_("\
-warning: %s: requires executable stack (because the .note.GNU-stack section is executable)"),
-		       bfd_get_filename (noteobj));
-		  else if (emptyobj)
-		    {
-		      _bfd_error_handler (_("\
-warning: %s: missing .note.GNU-stack section implies executable stack"),
-					  bfd_get_filename (emptyobj));
-		      _bfd_error_handler (_("\
-NOTE: This behaviour is deprecated and will be removed in a future version of the linker"));
-		    }
-		}
-	    }
 	  elf_stack_flags (output_bfd) = PF_R | PF_W | exec;
 	}
 
@@ -13380,9 +13358,6 @@ bfd_elf_final_link (bfd *abfd, struct bfd_link_info *info)
 		  else if (bfd_link_pde (info))
 		    info->callbacks->einfo
 		      (_("%P: warning: creating DT_TEXTREL in a PDE\n"));
-		  else
-		    info->callbacks->einfo
-		      (_("%P: warning: creating DT_TEXTREL in a PIE\n"));
 		  break;
 		}
 	    }
