diff --git a/gcc/config.gcc b/gcc/config.gcc
index d22f34bc..b248355a 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -1557,7 +1557,7 @@ mips64*-*-linux*)
 	# that is installed.  What GCC thinks of as the default must
 	# remain as ABI_32 such that the SONAMEs of the libgcc shared
 	# libraries remain compatible with those of mips-linux-gnu.
-	tm_defines="MIPS_ISA_DEFAULT=1 MIPS_ABI_DEFAULT=ABI_32"
+	tm_defines="MIPS_ISA_DEFAULT=64 MIPS_ABI_DEFAULT=ABI_64"
 	target_cpu_default="MASK_ABICALLS|MASK_FLOAT64|MASK_64BIT"
 	extra_parts="crtbegin.o crtbeginS.o crtbeginT.o crtend.o crtendS.o"
 	gnu_ld=yes
diff --git a/gcc/config/mips/linux64.h b/gcc/config/mips/linux64.h
index 58121eb3..31d7cb76 100644
--- a/gcc/config/mips/linux64.h
+++ b/gcc/config/mips/linux64.h
@@ -24,7 +24,7 @@ Boston, MA 02111-1307, USA.  */
    meant to override that.  However, for correct behavior at build
    time, we also need t-linux64 to get the build-time specs in line
    with the setting in config.gcc.  */
-#define DRIVER_DEFAULT_ABI_SELF_SPEC "%{!mabi=*:-mabi=n32}"
+#define DRIVER_DEFAULT_ABI_SELF_SPEC "%{!mabi=*:-mabi=64}"
 #undef SUBTARGET_EXTRA_SPECS
 #define SUBTARGET_EXTRA_SPECS \
   { "driver_default_abi_self_spec", DRIVER_DEFAULT_ABI_SELF_SPEC },
