diff --git a/gcc/config.gcc b/gcc/config.gcc
index d22f34bc..b248355a 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -1557,7 +1557,7 @@ mips64*-*-linux*)
 	# that is installed.  What GCC thinks of as the default must
 	# remain as ABI_32 such that the SONAMEs of the libgcc shared
 	# libraries remain compatible with those of mips-linux-gnu.
-	tm_defines="MIPS_ISA_DEFAULT=1 MIPS_ABI_DEFAULT=ABI_32"
+	tm_defines="MIPS_ISA_DEFAULT=64 MIPS_ABI_DEFAULT=ABI_64"
 	target_cpu_default="MASK_ABICALLS|MASK_FLOAT64|MASK_64BIT"
 	extra_parts="crtbegin.o crtbeginS.o crtbeginT.o crtend.o crtendS.o"
 	gnu_ld=yes
diff --git a/gcc/config/mips/linux64.h b/gcc/config/mips/linux64.h
index 58121eb3..31d7cb76 100644
--- a/gcc/config/mips/linux64.h
+++ b/gcc/config/mips/linux64.h
@@ -24,7 +24,7 @@ Boston, MA 02111-1307, USA.  */
    meant to override that.  However, for correct behavior at build
    time, we also need t-linux64 to get the build-time specs in line
    with the setting in config.gcc.  */
-#define DRIVER_DEFAULT_ABI_SELF_SPEC "%{!mabi=*:-mabi=n32}"
+#define DRIVER_DEFAULT_ABI_SELF_SPEC "%{!mabi=*:-mabi=64}"
 #undef SUBTARGET_EXTRA_SPECS
 #define SUBTARGET_EXTRA_SPECS \
   { "driver_default_abi_self_spec", DRIVER_DEFAULT_ABI_SELF_SPEC },
diff --git a/gcc/config/mips/t-linux64 b/gcc/config/mips/t-linux64
index 402838b1..1896f491 100644
--- a/gcc/config/mips/t-linux64
+++ b/gcc/config/mips/t-linux64
@@ -1,6 +1,6 @@
-MULTILIB_OPTIONS = mabi=32/mabi=n32/mabi=64
-MULTILIB_DIRNAMES = o32 32 64
-MULTILIB_OSDIRNAMES = ../lib ../lib32 ../lib64
+MULTILIB_OPTIONS = mabi=n32/mabi=32/mabi=64
+MULTILIB_DIRNAMES = n32 32 64
+MULTILIB_OSDIRNAMES = ../lib32 ../lib ../lib64
 
 EXTRA_MULTILIB_PARTS=crtbegin.o crtend.o crtbeginS.o crtendS.o crtbeginT.o
 
@@ -17,27 +17,3 @@ tp-bit.c: $(srcdir)/config/fp-bit.c
 	echo '# define TFLOAT' >> tp-bit.c
 	cat $(srcdir)/config/fp-bit.c >> tp-bit.c
 	echo '#endif' >> tp-bit.c
-
-# This arranges for us to extract the built-in specs of the compiler
-# into specs.install.  It's this file that we're going to install as
-# specs.  Howver, at toolchain build time, we must use specs that
-# enforce the defaults that the compiler expects, namely, the o32 ABI,
-# instead of n32, that we want to use as the default post install.
-SPECS = specs.install
-specs: specs.install
-	# Changing the build-time default ABI from n32 to o32
-	# The post-install default ABI will remain n32
-	@if test "x`grep '!mabi=\*:-mabi=n32' specs.install`" = \
-		"x%{!mabi=*:-mabi=n32}"; then \
-	  sed 's,\(!mabi=\*:-mabi\)=n32,\1=32,' < specs.install > $@ ; \
-	else \
-	  echo failed, gcc/config/mips/t-linux64 needs updating >&2; \
-	  false; \
-	fi
-
-# This arranges for anyone who uses gcc -print-multi-lib post
-# installation to get lists of options for each multilib to get
-# options that actually work.
-MULTILIB_EXTRA_OPTS = mabi-fake-default
-CRTSTUFF_T_CFLAGS = -mabi-fake-default
-CRTSTUFF_T_CFLAGS_S = -mabi-fake-default
