diff --git a/gcc/config/rs6000/rs6000-protos.h b/gcc/config/rs6000/rs6000-protos.h
index 3303ecd4..68def585 100644
--- a/gcc/config/rs6000/rs6000-protos.h
+++ b/gcc/config/rs6000/rs6000-protos.h
@@ -105,7 +105,7 @@ extern int registers_ok_for_quad_peep (rtx, rtx);
 extern int addrs_ok_for_quad_peep (rtx, rtx);
 extern bool gpr_or_gpr_p (rtx, rtx);
 extern enum reg_class secondary_reload_class (enum reg_class,
-					      enum machine_mode, rtx);
+					      rtx);
 extern int ccr_bit (rtx, int);
 extern int extract_MB (rtx);
 extern int extract_ME (rtx);
diff --git a/gcc/config/rs6000/rs6000.c b/gcc/config/rs6000/rs6000.c
index 6909ceff..4c62c0f3 100644
--- a/gcc/config/rs6000/rs6000.c
+++ b/gcc/config/rs6000/rs6000.c
@@ -8723,7 +8723,6 @@ addrs_ok_for_quad_peep (rtx addr1, rtx addr2)
 
 enum reg_class
 secondary_reload_class (enum reg_class class,
-			enum machine_mode mode,
 			rtx in)
 {
   int regno;
diff --git a/gcc/config/rs6000/rs6000.h b/gcc/config/rs6000/rs6000.h
index 01a97da4..f616a520 100644
--- a/gcc/config/rs6000/rs6000.h
+++ b/gcc/config/rs6000/rs6000.h
@@ -1493,7 +1493,7 @@ enum reg_class
    NO_REGS is returned.  */
 
 #define SECONDARY_RELOAD_CLASS(CLASS, MODE, IN) \
-  secondary_reload_class (CLASS, MODE, IN)
+  secondary_reload_class (CLASS, IN)
 
 /* If we are copying between FP or AltiVec registers and anything
    else, we need a memory location.  */
diff --git a/gcc/bb-reorder.c b/gcc/bb-reorder.c
index a8b3280d..1633b760 100644
--- a/gcc/bb-reorder.c
+++ b/gcc/bb-reorder.c
@@ -183,7 +183,7 @@ find_traces (int *n_traces, struct trace *traces)
   /* Find the traces.  */
   for (i = 0; i < N_ROUNDS; i++)
     {
-      gcov_type count_threshold;
+      gcov_type count_threshold = 0;
 
       if (rtl_dump_file)
 	fprintf (rtl_dump_file, "STC - round %d\n", i + 1);
diff --git a/gcc/bb-reorder.c b/gcc/bb-reorder.c
diff --git a/gcc/f/data.c b/gcc/f/data.c
index 2040f0ab..7de8210b 100644
--- a/gcc/f/data.c
+++ b/gcc/f/data.c
@@ -1393,9 +1393,9 @@ ffedata_gather_ (ffestorag mst, ffestorag st)
 
       for (;;)
 	{
-	  ffetargetOffset unexp;
-	  ffetargetOffset siz;
-	  ffebitCount length;
+	  ffetargetOffset unexp = 0;
+	  ffetargetOffset siz = 0;
+	  ffebitCount length = 0;
 	  bool value;
 
 	  ffebit_test (bits, source_offset, &value, &length);
