diff --git a/configure b/configure
index ed4abe084..f9a878780 100755
--- a/configure
+++ b/configure
@@ -2765,7 +2765,6 @@ libgcj="target-libffi \
 # the host libraries and the host tools (which may be a cross compiler)
 # Note that libiberty is not a target library.
 target_libraries="target-libgcc \
-		target-libbacktrace \
 		target-libgloss \
 		target-newlib \
 		target-libgomp \
diff --git a/configure.ac b/configure.ac
index 15f2b101f..4f3ae0025 100644
--- a/configure.ac
+++ b/configure.ac
@@ -152,7 +152,6 @@ libgcj="target-libffi \
 # the host libraries and the host tools (which may be a cross compiler)
 # Note that libiberty is not a target library.
 target_libraries="target-libgcc \
-		target-libbacktrace \
 		target-libgloss \
 		target-newlib \
 		target-libgomp \
