diff --git a/gcc/cgraphunit.c b/gcc/cgraphunit.c
index 0f933bdc7..f512f2d26 100644
--- a/gcc/cgraphunit.c
+++ b/gcc/cgraphunit.c
@@ -2123,7 +2123,7 @@ cgraph_node::assemble_thunks_and_aliases (void)
 	cgraph_node *thunk = e->caller;
 
 	e = e->next_caller;
-	thunk->expand_thunk (true, false);
+	thunk->expand_thunk (!rtl_dump_and_exit, false);
 	thunk->assemble_thunks_and_aliases ();
       }
     else
diff --git a/gcc/testsuite/g++.dg/debug/pr106261.C b/gcc/testsuite/g++.dg/debug/pr106261.C
new file mode 100644
index 000000000..6dee7e68f
--- /dev/null
+++ b/gcc/testsuite/g++.dg/debug/pr106261.C
@@ -0,0 +1,36 @@
+// PR debug/106261
+// { dg-do compile }
+// { dg-options "-dx -fno-dwarf2-cfi-asm" }
+
+struct A
+{
+  virtual void foo ();
+  int a;
+};
+class C : virtual public A
+{
+};
+struct B
+{
+  A *b;
+
+  B (A *x) : b (x) { b->foo (); }
+};
+struct E
+{
+  virtual ~E ();
+};
+class D : public C, E
+{
+};
+struct F : D
+{
+  F (int);
+
+  static void bar ()
+  {
+    F a (0);
+    B b (&a);
+  }
+};
+void baz () { F::bar (); }
