diff --git a/gcc/function.c b/gcc/function.c
index 81f052fa8..18f130064 100644
--- a/gcc/function.c
+++ b/gcc/function.c
@@ -3660,6 +3660,8 @@ assign_parms (tree fndecl)
      now that all parameters have been copied out of hard registers.  */
   emit_insn (all.first_conversion_insn);
 
+  do_pending_stack_adjust ();
+
   /* Estimate reload stack alignment from scalar return mode.  */
   if (SUPPORTS_STACK_ALIGNMENT)
     {
diff --git a/gcc/testsuite/gcc.target/i386/pr117296.c b/gcc/testsuite/gcc.target/i386/pr117296.c
new file mode 100644
index 000000000..1b4f6e806
--- /dev/null
+++ b/gcc/testsuite/gcc.target/i386/pr117296.c
@@ -0,0 +1,12 @@
+/* PR target/117296 */
+/* { dg-do compile } */
+/* { dg-options "-O1" } */
+/* { dg-additional-options "-mtune=k6 -mstringop-strategy=libcall -ffloat-store" { target ia32 } } */
+
+int x;
+
+void
+foo (_Complex double c)
+{
+ lab:;
+}
