diff --git a/gcc/testsuite/g++.dg/torture/pr123603.C b/gcc/testsuite/g++.dg/torture/pr123603.C
new file mode 100644
index 000000000..f971a8217
--- /dev/null
+++ b/gcc/testsuite/g++.dg/torture/pr123603.C
@@ -0,0 +1,20 @@
+// { dg-do compile }
+
+void _setjmp();
+void a() __attribute__((__noreturn__));
+struct b {
+  b(int c) {
+    if (c)
+      a();
+  }
+  ~b();
+};
+int d;
+void l(long);
+void e() {
+  b a(d);
+  for (int f = 0; f < 10 ; ++f) {
+    l(f - 1);
+    _setjmp();
+  }
+}
diff --git a/gcc/tree-ssa-pre.c b/gcc/tree-ssa-pre.c
index 5583db6db..3b93a7d43 100644
--- a/gcc/tree-ssa-pre.c
+++ b/gcc/tree-ssa-pre.c
@@ -2698,7 +2698,12 @@ find_or_generate_expression (basic_block block, tree op, gimple_seq *stmts)
   if (leader)
     {
       if (leader->kind == NAME)
-	return PRE_EXPR_NAME (leader);
+	{
+	  tree name = PRE_EXPR_NAME (leader);
+	  if (SSA_NAME_OCCURS_IN_ABNORMAL_PHI (name))
+	    return NULL_TREE;
+	  return name;
+	}
       else if (leader->kind == CONSTANT)
 	return PRE_EXPR_CONSTANT (leader);
 
