diff --git a/gcc/testsuite/gcc.target/arm/neon-vceq_p64.c b/gcc/testsuite/gcc.target/arm/neon-vceq_p64.c
index 21a6a78a2..e1ecd94e5 100644
--- a/gcc/testsuite/gcc.target/arm/neon-vceq_p64.c
+++ b/gcc/testsuite/gcc.target/arm/neon-vceq_p64.c
@@ -1,4 +1,5 @@
 /* { dg-do run } */
+/* { dg-options "-Wno-format" } */
 /* { dg-require-effective-target arm_crypto_ok } */
 /* { dg-require-effective-target arm_neon_hw } */
 /* { dg-add-options arm_crypto } */
diff --git a/gcc/testsuite/gcc.target/arm/neon-vtst_p64.c b/gcc/testsuite/gcc.target/arm/neon-vtst_p64.c
index 3a0b117c2..391aa54a6 100644
--- a/gcc/testsuite/gcc.target/arm/neon-vtst_p64.c
+++ b/gcc/testsuite/gcc.target/arm/neon-vtst_p64.c
@@ -1,4 +1,5 @@
 /* { dg-do run } */
+/* { dg-options "-Wno-format" } */
 /* { dg-require-effective-target arm_crypto_ok } */
 /* { dg-require-effective-target arm_neon_hw } */
 /* { dg-add-options arm_crypto } */
diff --git a/gcc/testsuite/gcc.target/arm/pr51534.c b/gcc/testsuite/gcc.target/arm/pr51534.c
index f675a444a..3711b45ee 100644
--- a/gcc/testsuite/gcc.target/arm/pr51534.c
+++ b/gcc/testsuite/gcc.target/arm/pr51534.c
@@ -3,7 +3,7 @@
 
 /* { dg-do assemble } */
 /* { dg-require-effective-target arm_neon_ok } */
-/* { dg-options "-save-temps -mfloat-abi=hard -O3" } */
+/* { dg-options "-save-temps -O3" } */
 /* { dg-add-options arm_neon } */
 
 #include <arm_neon.h>
diff --git a/gcc/testsuite/gcc.target/arm/pr51968.c b/gcc/testsuite/gcc.target/arm/pr51968.c
index 781470223..122fdfd29 100644
--- a/gcc/testsuite/gcc.target/arm/pr51968.c
+++ b/gcc/testsuite/gcc.target/arm/pr51968.c
@@ -1,6 +1,6 @@
 /* PR target/51968 */
-/* { dg-do compile } */
-/* { dg-options "-O2 -march=armv7-a -mfloat-abi=softfp -mfpu=neon" } */
+/* { dg-do compile { target { arm*-*-*eabihf* } } } */
+/* { dg-options "-O2 -march=armv7-a -mfpu=neon" } */
 /* { dg-require-effective-target arm_neon_ok } */
 #include <arm_neon.h>
 
diff --git a/gcc/testsuite/gfortran.dg/ieee/ieee_9.f90 b/gcc/testsuite/gfortran.dg/ieee/ieee_9.f90
deleted file mode 100644
index 5e0ac36f0..000000000
--- a/gcc/testsuite/gfortran.dg/ieee/ieee_9.f90
+++ /dev/null
@@ -1,71 +0,0 @@
-! { dg-do run { xfail arm*-*-gnueabi arm*-*-gnueabihf } }
-! { dg-skip-if "PR89639" { hppa*-*-linux* } }
-program foo
-   use ieee_arithmetic
-   use iso_fortran_env
-   integer i, p
-   real x
-   x = 4
-   i = 4
-
-   if (int8 > 0) then
-      if (real32 > 0) then
-         p = int(ieee_scalb(real(x, real32), int(i, int8)))
-         if (p /= 64) stop 1
-      endif
-      if (real64 > 0) then
-         p = int(ieee_scalb(real(x, real64), int(i, int8)))
-         if (p /= 64) stop 2
-      endif
-      if (real128 > 0) then
-         p = int(ieee_scalb(real(x, real128), int(i, int8)))
-         if (p /= 64) stop 3
-      end if
-   end if
-
-   if (int16 > 0) then
-      if (real32 > 0) then
-         p = int(ieee_scalb(real(x, real32), int(i, int16)))
-         if (p /= 64) stop 4
-      endif
-      if (real64 > 0) then
-         p = int(ieee_scalb(real(x, real64), int(i, int16)))
-         if (p /= 64) stop 5
-      endif
-      if (real128 > 0) then
-         p = int(ieee_scalb(real(x, real128), int(i, int16)))
-         if (p /= 64) stop 6
-      end if
-   end if
-
-   if (int32 > 0) then
-      if (real32 > 0) then
-         p = int(ieee_scalb(real(x, real32), int(i, int32)))
-         if (p /= 64) stop 7
-      endif
-      if (real64 > 0) then
-         p = int(ieee_scalb(real(x, real64), int(i, int32)))
-         if (p /= 64) stop 8
-      endif
-      if (real128 > 0) then
-         p = int(ieee_scalb(real(x, real128), int(i, int32)))
-         if (p /= 64) stop 9
-      end if
-   end if
-
-   if (int64 > 0) then
-      if (real32 > 0) then
-         p = int(ieee_scalb(real(x, real32), int(i, int64)))
-         if (p /= 64) stop 10
-      endif
-      if (real64 > 0) then
-         p = int(ieee_scalb(real(x, real64), int(i, int64)))
-         if (p /= 64) stop 11
-      endif
-      if (real128 > 0) then
-         p = int(ieee_scalb(real(x, real128), int(i, int64)))
-         if (p /= 64) stop 12
-      end if
-   end if
-
-end program foo
