diff --git a/gcc/testsuite/gcc.target/loongarch/cmov_fi.c b/gcc/testsuite/gcc.target/loongarch/cmov_fi.c
index 634c40367..ac9fadbdb 100644
--- a/gcc/testsuite/gcc.target/loongarch/cmov_fi.c
+++ b/gcc/testsuite/gcc.target/loongarch/cmov_fi.c
@@ -1,6 +1,6 @@
 /* Test asm const. */
 /* { dg-do compile } */
-/* { dg-options "-O2 -S" } */
+/* { dg-options "-O2 -S -Wno-format" } */
 /* { dg-final { scan-assembler-times "main:.*movgr2fr.*movgr2fr.*fsel.*movfr2gr.*" 1 } } */
 #include <stdio.h>
 
diff --git a/gcc/testsuite/gfortran.dg/elemental_subroutine_3.f90 b/gcc/testsuite/gfortran.dg/elemental_subroutine_3.f90
index e8c1c5664..47fad7ed6 100644
--- a/gcc/testsuite/gfortran.dg/elemental_subroutine_3.f90
+++ b/gcc/testsuite/gfortran.dg/elemental_subroutine_3.f90
@@ -1,4 +1,5 @@
 ! { dg-do run }
+! { dg-options "-Wno-overflow" }
 ! Test the fix for PR25746, in which dependency checking was not being
 ! done for elemental subroutines and therefore for interface assignments.
 !
