diff --git a/gcc/config/alpha/linux.h b/gcc/config/alpha/linux.h
index 7e87e66..403e8c8 100644
--- a/gcc/config/alpha/linux.h
+++ b/gcc/config/alpha/linux.h
@@ -25,7 +25,7 @@ Boston, MA 02111-1307, USA.  */
 
 #undef CPP_PREDEFINES
 #define CPP_PREDEFINES \
-"-D__gnu_linux__ -Dlinux -Dunix -Asystem=linux -D_LONGLONG -D__alpha__ " \
+"-D__gnu_linux__ -Dlinux -Dunix -Asystem=linux -D_LONGLONG -D__alpha__ -D__LP64__ -D_LP64" \
 SUB_CPP_PREDEFINES
 
 /* The GNU C++ standard library requires that these macros be defined.  */
diff --git a/gcc/config/i386/linux64.h b/gcc/config/i386/linux64.h
index 6070750..52b81f4 100644
--- a/gcc/config/i386/linux64.h
+++ b/gcc/config/i386/linux64.h
@@ -28,7 +28,7 @@ Boston, MA 02111-1307, USA.  */
 #define CPP_PREDEFINES "-D__ELF__ -Dunix -D__gnu_linux__ -Dlinux -Asystem(posix)"
 
 #undef CPP_SPEC
-#define CPP_SPEC "%(cpp_cpu) %{fPIC:-D__PIC__ -D__pic__} %{fpic:-D__PIC__ -D__pic__} %{posix:-D_POSIX_SOURCE} %{pthread:-D_REENTRANT} %{!m32:-D__LONG_MAX__=9223372036854775807L}"
+#define CPP_SPEC "%(cpp_cpu) %{fPIC:-D__PIC__ -D__pic__} %{fpic:-D__PIC__ -D__pic__} %{posix:-D_POSIX_SOURCE} %{pthread:-D_REENTRANT} %{!m32:-D__LONG_MAX__=9223372036854775807L -D__LP64__ -D_LP64}"
 
 /* Provide a LINK_SPEC.  Here we provide support for the special GCC
    options -static and -shared, which allow us to link things in one
diff --git a/gcc/config/mips/linux.h b/gcc/config/mips/linux.h
index f1ff69e..550f9d3 100644
--- a/gcc/config/mips/linux.h
+++ b/gcc/config/mips/linux.h
@@ -180,6 +180,7 @@ void FN ()							\
 %{mabi=n32: -D_MIPS_SZPTR=32} \
 %{mabi=64: -D_MIPS_SZPTR=64} \
 %{!mabi*: -D_MIPS_SZPTR=32} \
+%{mabi=64: -D__LP64__ -D_LP64} \
 %{!mips*: -U__mips -D__mips} \
 %{mabi=32: -U__mips64} \
 %{mabi=n32: -D__mips64} \
diff --git a/gcc/config/rs6000/linux64.h b/gcc/config/rs6000/linux64.h
index 37d73fa..ed291b0 100644
--- a/gcc/config/rs6000/linux64.h
+++ b/gcc/config/rs6000/linux64.h
@@ -107,7 +107,7 @@ Boston, MA 02111-1307, USA.  */
 #define CPP_PREDEFINES \
  "-D_PPC_ -D__PPC__ -D_PPC64_ -D__PPC64__ -D__powerpc__ -D__powerpc64__ \
   -D_PIC_ -D__PIC__ -D_BIG_ENDIAN -D__BIG_ENDIAN__ -D__ELF__ \
-  -D__LONG_MAX__=9223372036854775807L \
+  -D__LONG_MAX__=9223372036854775807L -D__LP64__ -D_LP64 \
   -Acpu=powerpc64 -Amachine=powerpc64"
 
 #undef  CPP_OS_DEFAULT_SPEC
diff --git a/gcc/config/s390/linux.h b/gcc/config/s390/linux.h
index d46dbdd..f7ed672 100644
--- a/gcc/config/s390/linux.h
+++ b/gcc/config/s390/linux.h
@@ -61,7 +61,7 @@ Boston, MA 02111-1307, USA.  */
   "-D__SIZE_TYPE__=long\\ unsigned\\ int -D__PTRDIFF_TYPE__=int"
 #define CPP_ARCH64_SPEC \
   "-D__SIZE_TYPE__=long\\ unsigned\\ int -D__PTRDIFF_TYPE__=long\\ int \
-   -D__s390x__ -D__LONG_MAX__=9223372036854775807L"
+   -D__s390x__ -D__LONG_MAX__=9223372036854775807L -D__LP64__ -D_LP64"
 
 #ifdef DEFAULT_TARGET_64BIT
 #undef  CPP_SPEC
diff --git a/gcc/config/sparc/linux64.h b/gcc/config/sparc/linux64.h
index e1a1286..5d72e7b 100644
--- a/gcc/config/sparc/linux64.h
+++ b/gcc/config/sparc/linux64.h
@@ -173,6 +173,7 @@ ENDFILE_SPEC_COMMON
 #define CPP_SUBTARGET_SPEC "\
 %{fPIC:-D__PIC__ -D__pic__} \
 %{fpic:-D__PIC__ -D__pic__} \
+%{!m32:-D__LP64__ -D_LP64} \
 %{posix:-D_POSIX_SOURCE} \
 %{pthread:-D_REENTRANT} \
 "
