diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 1a2a35fa4..4be03127e 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -8878,7 +8878,6 @@ grokdeclarator (const cp_declarator *declarator,
 	  && TYPE_ANONYMOUS_P (type)
 	  && cp_type_quals (type) == TYPE_UNQUALIFIED)
 	{
-	  tree oldname = TYPE_NAME (type);
 	  tree t;
 
 	  /* Replace the anonymous name with the real name everywhere.  */
diff --git a/gcc/tree-ssa-remove-local-statics.c b/gcc/tree-ssa-remove-local-statics.c
index d9755fee1..5e88c9e11 100644
--- a/gcc/tree-ssa-remove-local-statics.c
+++ b/gcc/tree-ssa-remove-local-statics.c
@@ -835,7 +835,7 @@ execute_rls (void)
 
       if (dump_file)
         fprintf (dump_file, "found %d static variables to consider\n",
-                 stats.n_statics);
+                 (int)(stats.n_statics));
 
       initialize_block_and_statement_dataflow (stats.n_statics);
 
@@ -845,7 +845,7 @@ execute_rls (void)
 
       if (dump_file)
         fprintf (dump_file, "removed %d static variables\n",
-                 stats.n_optimizable);
+                 (int)(stats.n_optimizable));
     }
 
   rls_done ();
