diff --git a/gcc/config/sparc/sparc.c b/gcc/config/sparc/sparc.c
index 6e640f6a2..8bcd15b89 100644
--- a/gcc/config/sparc/sparc.c
+++ b/gcc/config/sparc/sparc.c
@@ -7848,7 +7848,7 @@ sparc_profile_hook (int labelno)
 #if defined (OBJECT_FORMAT_ELF)
 static void
 sparc_elf_asm_named_section (const char *name, unsigned int flags,
-			     tree decl)
+			     tree decl ATTRIBUTE_UNUSED)
 {
   fprintf (asm_out_file, "\t.section\t\"%s\"", name);
 
