diff --git a/gcc/config/arm/arm.c b/gcc/config/arm/arm.c
index 6808ee17d..b10378e73 100644
--- a/gcc/config/arm/arm.c
+++ b/gcc/config/arm/arm.c
@@ -19626,13 +19626,6 @@ arm_mangle_type (const_tree type)
   if (TARGET_AAPCS_BASED 
       && lang_hooks.types_compatible_p (CONST_CAST_TREE (type), va_list_type))
     {
-      static bool warned;
-      if (!warned && warn_psabi)
-	{
-	  warned = true;
-	  inform (input_location,
-		  "the mangling of %<va_list%> has changed in GCC 4.4");
-	}
       return "St9__va_list";
     }
 
