diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index bdef8483..b9e3b90a 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -12606,7 +12606,7 @@ cp_parser_class_specifier (cp_parser* parser)
   unsigned saved_num_template_parameter_lists;
   tree old_scope = NULL_TREE;
   tree scope = NULL_TREE;
-  tree bases;
+  tree bases = NULL_TREE;
 
   push_deferring_access_checks (dk_no_deferred);
 
diff --git a/gcc/gcc.c b/gcc/gcc.c
index 23b53a09..16c667aa 100644
--- a/gcc/gcc.c
+++ b/gcc/gcc.c
@@ -5934,7 +5934,7 @@ retry_ice (const char *prog, const char **argv)
 
   for (attempt = 0; attempt < RETRY_ICE_ATTEMPTS + 1; ++attempt)
     {
-      int fd;
+      int fd = -1;
       int status;
 
       temp_filenames[attempt * 2] = make_temp_file (".out");
