From 5932804f2534c0c8a44394a502e1e6809abff375 Mon Sep 17 00:00:00 2001
From: Jonathan Yong <10walls@gmail.com>
Date: Tue, 8 Sep 2020 15:21:56 +0200
Subject: [PATCH 08/12] Cygwin: g++ time

---
 libstdc++-v3/acinclude.m4 | 3 +++
 libstdc++-v3/configure    | 3 +++
 2 files changed, 6 insertions(+)

diff --git a/libstdc++-v3/acinclude.m4 b/libstdc++-v3/acinclude.m4
index e2a12607d5d..894c6f3230c 100644
--- a/libstdc++-v3/acinclude.m4
+++ b/libstdc++-v3/acinclude.m4
@@ -1418,7 +1418,10 @@ AC_DEFUN([GLIBCXX_ENABLE_LIBSTDCXX_TIME], [
 
     case "${target_os}" in
       cygwin*)
+        ac_has_clock_monotonic=yes
+        ac_has_clock_realtime=yes
         ac_has_nanosleep=yes
+        ac_has_sched_yield=yes
         ;;
       darwin*)
         ac_has_nanosleep=yes
diff --git a/libstdc++-v3/configure b/libstdc++-v3/configure
index 64f2552afe5..5218810d855 100755
--- a/libstdc++-v3/configure
+++ b/libstdc++-v3/configure
@@ -20584,7 +20584,10 @@ ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
 
     case "${target_os}" in
       cygwin*)
+        ac_has_clock_monotonic=yes
+        ac_has_clock_realtime=yes
         ac_has_nanosleep=yes
+        ac_has_sched_yield=yes
         ;;
       darwin*)
         ac_has_nanosleep=yes
-- 
2.45.1

