diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig01.C b/gcc/testsuite/g++.old-deja/g++.gb/sig01.C
index a3e9277..9daf2aa 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig01.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig01.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr assignment
 // Test assignment to local signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig02.C b/gcc/testsuite/g++.old-deja/g++.gb/sig02.C
index f739a28..e361336 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig02.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig02.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr assignment
 // Test assigning objects of different classes to local signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig03.C b/gcc/testsuite/g++.old-deja/g++.gb/sig03.C
index d791c7e..65b2e99 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig03.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig03.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr overloading
 // Test overloading of signature member functions.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig04.C b/gcc/testsuite/g++.old-deja/g++.gb/sig04.C
index af54959..727dd56 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig04.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig04.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr this
 // Test passing of 'this' pointer through signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig05.C b/gcc/testsuite/g++.old-deja/g++.gb/sig05.C
index 3023e34..c282f01 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig05.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig05.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr initialization
 // Test initialization of a local signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig06.C b/gcc/testsuite/g++.old-deja/g++.gb/sig06.C
index b6476d5..e0d9788 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig06.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig06.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr initialization
 // Test initialization of a local signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig07.C b/gcc/testsuite/g++.old-deja/g++.gb/sig07.C
index 18a47be..15777b6 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig07.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig07.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr default-argument
 // Test calling a signature member function with default argument.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig08.C b/gcc/testsuite/g++.old-deja/g++.gb/sig08.C
index 2d5d14d..2d077a6 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig08.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig08.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr argument return-value
 // Test passing a signature pointer to a function and returning it from one.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig09.C b/gcc/testsuite/g++.old-deja/g++.gb/sig09.C
index a2e3b33..b3fc86d 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig09.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig09.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr assignment initialization new
 // Test assignment to/initialization of signature pointer with run-time value.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig10.C b/gcc/testsuite/g++.old-deja/g++.gb/sig10.C
index e16f2ee..65ac189 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig10.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig10.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr argument return-value new
 // Test passing a signature pointer to a function and returning it from one.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig11.C b/gcc/testsuite/g++.old-deja/g++.gb/sig11.C
index 892695f..4da427f 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig11.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig11.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr array
 // Test a signature pointer array.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig12.C b/gcc/testsuite/g++.old-deja/g++.gb/sig12.C
index 2747a80..c7215f3 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig12.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig12.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr structure
 // Test a signature pointer structure.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig13.C b/gcc/testsuite/g++.old-deja/g++.gb/sig13.C
index f5da9f8..3ff2350 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig13.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig13.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr assignment initialization
 // Test assignment of/initialization with same type signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig14.C b/gcc/testsuite/g++.old-deja/g++.gb/sig14.C
index 6ce1d02..e13f118 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig14.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig14.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr assignment initialization
 // Test assignment of/initialization with different type signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig15.C b/gcc/testsuite/g++.old-deja/g++.gb/sig15.C
index f0ae52e..c166507 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig15.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig15.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr cast
 // Test casting a signature pointer to a class pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig16.C b/gcc/testsuite/g++.old-deja/g++.gb/sig16.C
index 35475e9..5d860b1 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig16.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig16.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr cast
 // Test casting a class pointer to an anonymous signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig17.C b/gcc/testsuite/g++.old-deja/g++.gb/sig17.C
index 6d1d1d9..acc4278 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig17.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig17.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr cast
 // Test casting a class pointer through a chain of signature pointers.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig18.C b/gcc/testsuite/g++.old-deja/g++.gb/sig18.C
index ba0ec86..7466c12 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig18.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig18.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr assignment virtual
 // Test assignment of object from virtual class to signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig19.C b/gcc/testsuite/g++.old-deja/g++.gb/sig19.C
index 160a4a3..848c268 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig19.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig19.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr initialization virtual
 // Test initialization of signature pointer with object from abstract class.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig20.C b/gcc/testsuite/g++.old-deja/g++.gb/sig20.C
index a8228de..f053533 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig20.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig20.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr initialization virtual
 // Test initialization of signature pointer with object from virtual class.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig21.C b/gcc/testsuite/g++.old-deja/g++.gb/sig21.C
index faad45b..2af1db8 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig21.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig21.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr assignment virtual
 // Test assignment of objects of different types to same signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig22.C b/gcc/testsuite/g++.old-deja/g++.gb/sig22.C
index 25b0a1a..1df2c32 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig22.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig22.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr assignment initialization
 // Test assignment to/init of sigptr that requires copying sigtable slots.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig23.C b/gcc/testsuite/g++.old-deja/g++.gb/sig23.C
index 767cfa3..6a5e626 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig23.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig23.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr assignment local-signature
 // Test assignment to signature pointer of local signature.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig24.C b/gcc/testsuite/g++.old-deja/g++.gb/sig24.C
index 2dca6f9..ae408ed 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig24.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig24.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr opaque-type
 // Test calling member functions taking/returning opaque type through sigptr.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig25.C b/gcc/testsuite/g++.old-deja/g++.gb/sig25.C
index d00ef57..876dfa3 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig25.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig25.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr default-implementation
 // Test calling default implementation through signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig26.C b/gcc/testsuite/g++.old-deja/g++.gb/sig26.C
index 399737b..ae56443 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig26.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig26.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr default-implementation
 // Test calling default implementation through signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig27.C b/gcc/testsuite/g++.old-deja/g++.gb/sig27.C
index 7b9e7d2..0b36c7b 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig27.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig27.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr multiple-inheritance
 // Test class defined by multiple inheritance as implementation of signature.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig28.C b/gcc/testsuite/g++.old-deja/g++.gb/sig28.C
index 4edff68..3918a68 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig28.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig28.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr multiple-inheritance
 // Test correct adjustment of `this' pointer in case of multiple inheritance.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig29.C b/gcc/testsuite/g++.old-deja/g++.gb/sig29.C
index e5c5e39..7293f3f 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig29.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig29.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr multiple-inheritance
 // Test calling virtual function from MI class through signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig30.C b/gcc/testsuite/g++.old-deja/g++.gb/sig30.C
index 3e93e11..a64766f 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig30.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig30.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr multiple-inheritance
 // Test calling overwritten virtual functions through signature pointer.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.gb/sig31.C b/gcc/testsuite/g++.old-deja/g++.gb/sig31.C
index 90fa9e9..78dac22 100644
--- a/gcc/testsuite/g++.old-deja/g++.gb/sig31.C
+++ b/gcc/testsuite/g++.old-deja/g++.gb/sig31.C
@@ -1,4 +1,4 @@
-// Special g++ Options: -fhandle-signatures
+// Special g++ Options: -fhandle-signatures -Wno-deprecated
 // GROUPS passed gb sigptr assignment sigsigtable
 // Test use of multiple signature-signature tables for same signature pair.
 
diff --git a/gcc/testsuite/g++.old-deja/g++.robertl/eb131.C b/gcc/testsuite/g++.old-deja/g++.robertl/eb131.C
index d966fd9..1768dd0 100644
--- a/gcc/testsuite/g++.old-deja/g++.robertl/eb131.C
+++ b/gcc/testsuite/g++.old-deja/g++.robertl/eb131.C
@@ -10,7 +10,7 @@ struct a {
 	void bar( double );
 	void bar( float );
 
-  void foo( void (a::*member)(float) ); // ERROR - candidate
+  void foo( void (a::*member)(float) );
 };
 
 a::a()
diff --git a/gcc/testsuite/gcc.misc-tests/mg.exp b/gcc/testsuite/gcc.misc-tests/mg.exp
index 845e3e6..dacdf1c 100644
--- a/gcc/testsuite/gcc.misc-tests/mg.exp
+++ b/gcc/testsuite/gcc.misc-tests/mg.exp
@@ -24,4 +24,4 @@ load_lib mike-gcc.exp
 prebase
 set actions none-of-the-above
 set compiler_output "mg.o ?: .*mg.c \[ \\\n\]*nonexist.h"
-postbase mg.c "" "" "-MM -MG 2>/dev/null"
+postbase mg.c "" "" "-MM -MG"
