diff --git a/Makefile.in b/Makefile.in
index 7af1784..8b3352d 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -233,7 +233,7 @@ CC_FOR_TARGET = ` \
 # variable is passed down to the gcc Makefile, where it is used to
 # build libgcc2.a.  We define it here so that it can itself be
 # overridden on the command line.
-GCC_FOR_TARGET = $$r/gcc/xgcc -B$$r/gcc/ -B$(build_tooldir)/bin/ -I$(build_tooldir)/include
+GCC_FOR_TARGET = $$r/gcc/xgcc -B$$r/gcc/ -B$(build_tooldir)/bin/ -I$(build_tooldir)/include -isystem /usr/${CHOST}/include
 
 CHILL_FOR_TARGET = ` \
   if [ -f $$r/gcc/xgcc ] ; then \
diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 85bf9f8..0f5ae25 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -218,7 +218,7 @@ RANLIB_TEST_FOR_TARGET = \
        && [ -f /usr/bin/ranlib -o -f /bin/ranlib ] )
 
 # Dir to search for system headers.  Overridden by cross-make.
-SYSTEM_HEADER_DIR = /usr/include
+SYSTEM_HEADER_DIR = /usr/${CHOST}/include
 
 # Control whether to run fixproto.
 STMP_FIXPROTO = stmp-fixproto
@@ -593,7 +593,7 @@ HOST_PRINT = $(HOST_PREFIX)print-rtl.o
 # Both . and srcdir are used, in that order,
 # so that tm.h and config.h will be found in the compilation
 # subdirectory rather than in the source directory.
-INCLUDES = -I. -I$(srcdir) -I$(srcdir)/config -I$(srcdir)/../include
+INCLUDES = -I. -I$(srcdir) -I$(srcdir)/config -I$(srcdir)/../include -isystem ${SYSTEM_HEADER_DIR}
 
 # Always use -I$(srcdir)/config when compiling.
 .c.o:
diff --git a/libio/Makefile.in b/libio/Makefile.in
index 1d4123b..1a3b675 100644
--- a/libio/Makefile.in
+++ b/libio/Makefile.in
@@ -48,7 +48,7 @@ STDIO_WRAP_OBJECTS =  iofdopen.o iofflush.o iofgets.o iofopen.o iofprintf.o iofp
 
 IOSTREAM_OBJECTS = builtinbuf.o filebuf.o fstream.o \
   indstream.o ioassign.o ioextend.o iomanip.o iostream.o \
-  isgetline.o isgetsb.o isscan.o \
+  isgetline.o isgetsb.o isscan.o iovfscanf.o \
   osform.o procbuf.o sbform.o sbgetline.o sbscan.o \
   stdiostream.o stdstrbufs.o stdstreams.o stream.o streambuf.o strstream.o \
   PlotFile.o SFile.o parsestream.o pfstream.o editbuf.o
