diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 7793a88..0d263fd 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -106,9 +106,7 @@ LN_S=@LN_S@
 # These permit overriding just for certain files.
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_DATA = @INSTALL_DATA@
-MAKEINFO = `if [ -f $(objdir)/../texinfo/makeinfo/Makefile ] ; \
-	then echo $(objdir)/../texinfo/makeinfo/makeinfo ; \
-	else echo makeinfo ; fi`
+MAKEINFO = makeinfo
 MAKEINFOFLAGS =
 TEXI2DVI = texi2dvi
 # For GNUmake: let us decide what gets passed to recursive makes.
@@ -218,7 +218,7 @@ RANLIB_TEST_FOR_TARGET = \
        && [ -f /usr/bin/ranlib -o -f /bin/ranlib ] )
 
 # Dir to search for system headers.  Overridden by cross-make.
-SYSTEM_HEADER_DIR = /usr/include
+SYSTEM_HEADER_DIR = $(build_tooldir)/include
 
 # Control whether to run fixproto.
 STMP_FIXPROTO = stmp-fixproto
@@ -1975,7 +1975,7 @@ cccp.o: cccp.c $(CONFIG_H) intl.h pcp.h version.c config.status system.h \
 	  -DGPLUSPLUS_INCLUDE_DIR=\"$(gcc_gxx_include_dir)\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(includedir)\" \
 	  -DCROSS_INCLUDE_DIR=\"$(gcc_tooldir)/sys-include\" \
-	  -DTOOL_INCLUDE_DIR=\"$(gcc_tooldir)/include\" \
+	  -DTOOL_INCLUDE_DIR=\"$(build_tooldir)/include\" \
 	  -c `echo $(srcdir)/cccp.c | sed 's,^\./,,'`
 
 LIBCPP_OBJS =	cpplib.o cpphash.o cppalloc.o cpperror.o cppexp.o cppfiles.o \
@@ -2009,7 +2009,7 @@ cppinit.o:  cppinit.c $(CONFIG_H) cpplib.h intl.h system.h \
 	  -DGPLUSPLUS_INCLUDE_DIR=\"$(gcc_gxx_include_dir)\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(includedir)\" \
 	  -DCROSS_INCLUDE_DIR=\"$(gcc_tooldir)/sys-include\" \
-	  -DTOOL_INCLUDE_DIR=\"$(gcc_tooldir)/include\" \
+	  -DTOOL_INCLUDE_DIR=\"$(build_tooldir)/include\" \
 	  -c `echo $(srcdir)/cppinit.c | sed 's,^\./,,'`
 
 # Note for the stamp targets, we run the program `true' instead of
@@ -2031,7 +2031,7 @@ protoize.o: protoize.c $(srcdir)/../include/getopt.h $(CONFIG_H) system.h \
           -DGCC_INCLUDE_DIR=\"$(libsubdir)/include\" \
 	  -DGPLUSPLUS_INCLUDE_DIR=\"$(gcc_gxx_include_dir)\" \
 	  -DCROSS_INCLUDE_DIR=\"$(gcc_tooldir)/sys-include\" \
-	  -DTOOL_INCLUDE_DIR=\"$(gcc_tooldir)/include\" \
+	  -DTOOL_INCLUDE_DIR=\"$(build_tooldir)/include\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(includedir)\" \
 	  -DSTD_PROTO_DIR=\"$(libsubdir)\" \
 	  $(srcdir)/protoize.c
@@ -2042,7 +2042,7 @@ unprotoize.o: unprotoize.c protoize.c $(srcdir)/../include/getopt.h \
           -DGCC_INCLUDE_DIR=\"$(libsubdir)/include\" \
 	  -DGPLUSPLUS_INCLUDE_DIR=\"$(gcc_gxx_include_dir)\" \
 	  -DCROSS_INCLUDE_DIR=\"$(gcc_tooldir)/sys-include\" \
-	  -DTOOL_INCLUDE_DIR=\"$(gcc_tooldir)/include\" \
+	  -DTOOL_INCLUDE_DIR=\"$(build_tooldir)/include\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(includedir)\" \
 	  -DSTD_PROTO_DIR=\"$(libsubdir)\" \
 	  $(srcdir)/unprotoize.c
diff --git a/libio/Makefile.in b/libio/Makefile.in
index 4342beb..125b57b 100644
--- a/libio/Makefile.in
+++ b/libio/Makefile.in
@@ -48,7 +48,7 @@ STDIO_WRAP_OBJECTS =  iofdopen.o iofflush.o iofgets.o iofopen.o iofprintf.o iofp
 
 IOSTREAM_OBJECTS = builtinbuf.o filebuf.o fstream.o \
   indstream.o ioassign.o ioextend.o iomanip.o iostream.o \
-  isgetline.o isgetsb.o isscan.o \
+  isgetline.o isgetsb.o isscan.o iovfscanf.o \
   osform.o procbuf.o sbform.o sbgetline.o sbscan.o \
   stdiostream.o stdstrbufs.o stdstreams.o stream.o streambuf.o strstream.o \
   PlotFile.o SFile.o parsestream.o pfstream.o editbuf.o
diff --git a/libstdc++/config/linux.ml b/libstdc++/config/linux.ml
index 14dcb45..f9ed8d9 100644
--- a/libstdc++/config/linux.ml
+++ b/libstdc++/config/linux.ml
@@ -1,10 +1,5 @@
 # Elf with shared libm, so we can link it into the shared libstdc++.
 
-ARLIB   = libstdc++-$(LIBSTDCXX_INTERFACE)$(LIBC_INTERFACE)$(CXX_INTERFACE)-$(VERSION).a
-MARLINK = libstdc++$(LIBC_INTERFACE)$(CXX_INTERFACE).a.$(LIBSTDCXX_INTERFACE)
-SHLIB   = libstdc++-$(LIBSTDCXX_INTERFACE)$(LIBC_INTERFACE)$(CXX_INTERFACE)-$(VERSION).so
-MSHLINK = libstdc++$(LIBC_INTERFACE)$(CXX_INTERFACE).so.$(LIBSTDCXX_INTERFACE)
-
 LIBS    = $(ARLIB) marlink $(ARLINK) $(SHLIB) mshlink $(SHLINK)
 SHFLAGS = -Wl,-soname,$(MSHLINK)
 SHDEPS  = -lm
