diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 85bf9f8..183f2b6 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -218,7 +218,7 @@ RANLIB_TEST_FOR_TARGET = \
        && [ -f /usr/bin/ranlib -o -f /bin/ranlib ] )
 
 # Dir to search for system headers.  Overridden by cross-make.
-SYSTEM_HEADER_DIR = /usr/include
+SYSTEM_HEADER_DIR = /usr/${CHOST}/include
 
 # Control whether to run fixproto.
 STMP_FIXPROTO = stmp-fixproto
diff --git a/libio/Makefile.in b/libio/Makefile.in
index 1d4123b..1a3b675 100644
--- a/libio/Makefile.in
+++ b/libio/Makefile.in
@@ -48,7 +48,7 @@ STDIO_WRAP_OBJECTS =  iofdopen.o iofflush.o iofgets.o iofopen.o iofprintf.o iofp
 
 IOSTREAM_OBJECTS = builtinbuf.o filebuf.o fstream.o \
   indstream.o ioassign.o ioextend.o iomanip.o iostream.o \
-  isgetline.o isgetsb.o isscan.o \
+  isgetline.o isgetsb.o isscan.o iovfscanf.o \
   osform.o procbuf.o sbform.o sbgetline.o sbscan.o \
   stdiostream.o stdstrbufs.o stdstreams.o stream.o streambuf.o strstream.o \
   PlotFile.o SFile.o parsestream.o pfstream.o editbuf.o
