diff --git a/Makefile.in b/Makefile.in
index 5248a26..e20f57b 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -233,7 +233,7 @@ CC_FOR_TARGET = ` \
 # variable is passed down to the gcc Makefile, where it is used to
 # build libgcc2.a.  We define it here so that it can itself be
 # overridden on the command line.
-GCC_FOR_TARGET = $$r/gcc/xgcc -B$$r/gcc/ -B$(build_tooldir)/bin/ -I$(build_tooldir)/include
+GCC_FOR_TARGET = $$r/gcc/xgcc -B$$r/gcc/ -B$(build_tooldir)/bin/ -I$(build_tooldir)/include -isystem /usr/${CHOST}/include
 
 CHILL_FOR_TARGET = ` \
   if [ -f $$r/gcc/xgcc ] ; then \
diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 7793a88..0d263fd 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -218,7 +218,7 @@ RANLIB_TEST_FOR_TARGET = \
        && [ -f /usr/bin/ranlib -o -f /bin/ranlib ] )
 
 # Dir to search for system headers.  Overridden by cross-make.
-SYSTEM_HEADER_DIR = /usr/include
+SYSTEM_HEADER_DIR = /usr/${CHOST}/include
 
 # Control whether to run fixproto.
 STMP_FIXPROTO = stmp-fixproto
@@ -593,7 +593,7 @@ HOST_PRINT = $(HOST_PREFIX)print-rtl.o
 # Both . and srcdir are used, in that order,
 # so that tm.h and config.h will be found in the compilation
 # subdirectory rather than in the source directory.
-INCLUDES = -I. -I$(srcdir) -I$(srcdir)/config -I$(srcdir)/../include
+INCLUDES = -I. -I$(srcdir) -I$(srcdir)/config -I$(srcdir)/../include -isystem ${SYSTEM_HEADER_DIR}
 
 # Always use -I$(srcdir)/config when compiling.
 .c.o:
diff --git a/libf2c/libF77/Makefile.in b/libf2c/libF77/Makefile.in
index 74c69a0..679d4c0 100644
--- a/libf2c/libF77/Makefile.in
+++ b/libf2c/libF77/Makefile.in
@@ -46,7 +46,7 @@ SHELL = /bin/sh
 
 #### End of system configuration section. ####
 
-ALL_CFLAGS = -I. -I$(srcdir) -I$(G2C_H_DIR) -I$(F2C_H_DIR) $(CPPFLAGS) $(DEFS) $(CFLAGS)
+ALL_CFLAGS = -I. -I$(srcdir) -I$(G2C_H_DIR) -I$(F2C_H_DIR) $(CPPFLAGS) $(DEFS) $(CFLAGS) -isystem /usr/${CHOST}/include
 
 .SUFFIXES:
 .SUFFIXES: .c .o
diff --git a/libf2c/libI77/Makefile.in b/libf2c/libI77/Makefile.in
index 1ff6b60..d529e84 100644
--- a/libf2c/libI77/Makefile.in
+++ b/libf2c/libI77/Makefile.in
@@ -47,7 +47,7 @@ SHELL = /bin/sh
 #### End of system configuration section. ####
 
 ALL_CFLAGS = -I. -I$(srcdir) -I$(G2C_H_DIR) -I$(F2C_H_DIR) $(CPPFLAGS) \
-             $(DEFS) $(CFLAGS)
+             $(DEFS) $(CFLAGS) -isystem /usr/${CHOST}/include
 
 .SUFFIXES:
 .SUFFIXES: .c .o
diff --git a/libf2c/libU77/Makefile.in b/libf2c/libU77/Makefile.in
index 3468a82..65345ca 100644
--- a/libf2c/libU77/Makefile.in
+++ b/libf2c/libU77/Makefile.in
@@ -47,7 +47,7 @@ SHELL = /bin/sh
 
 # fio.h is in libI77.  config.h is in `.'.
 ALL_CFLAGS = -I. -I$(srcdir) -I$(F2C_H_DIR)/libI77 -I$(G2C_H_DIR) \
-             -I$(F2C_H_DIR) $(CPPFLAGS) $(DEFS) $(CFLAGS)
+             -I$(F2C_H_DIR) $(CPPFLAGS) $(DEFS) $(CFLAGS) -isystem /usr/${CHOST}/include
 
 # This could probably be done more elegantly, but it's currently
 # just for running the u77-test test.
diff --git a/libio/Makefile.in b/libio/Makefile.in
index 4342beb..125b57b 100644
--- a/libio/Makefile.in
+++ b/libio/Makefile.in
@@ -48,7 +48,7 @@ STDIO_WRAP_OBJECTS =  iofdopen.o iofflush.o iofgets.o iofopen.o iofprintf.o iofp
 
 IOSTREAM_OBJECTS = builtinbuf.o filebuf.o fstream.o \
   indstream.o ioassign.o ioextend.o iomanip.o iostream.o \
-  isgetline.o isgetsb.o isscan.o \
+  isgetline.o isgetsb.o isscan.o iovfscanf.o \
   osform.o procbuf.o sbform.o sbgetline.o sbscan.o \
   stdiostream.o stdstrbufs.o stdstreams.o stream.o streambuf.o strstream.o \
   PlotFile.o SFile.o parsestream.o pfstream.o editbuf.o
diff --git a/libio/config/mtsafe.mt b/libio/config/mtsafe.mt
index 5848851..1a62e6e 100644
--- a/libio/config/mtsafe.mt
+++ b/libio/config/mtsafe.mt
@@ -1,2 +1,2 @@
 # tell we want the mt-safe version
-MT_CFLAGS = -D_IO_MTSAFE_IO
+MT_CFLAGS =
diff --git a/libio/configure.in b/libio/configure.in
index ac49429..15b65d7 100644
--- a/libio/configure.in
+++ b/libio/configure.in
@@ -142,8 +142,8 @@ case "${target}" in
     XCXXINCLUDES='-I. -I$(srcdir) -I$(srcdir)/../winsup/include'
     ;;
   *)
-    XCINCLUDES='-I. -I$(srcdir)'
-    XCXXINCLUDES='-I. -I$(srcdir)'
+    XCINCLUDES='-I. -I$(srcdir) -isystem /usr/${CHOST}/include'
+    XCXXINCLUDES='-I. -I$(srcdir) -isystem /usr/${CHOST}/include'
     ;;
 esac
 MOSTLYCLEAN='*.o pic stamp-picdir core iostream.list'
diff --git a/libobjc/Makefile.in b/libobjc/Makefile.in
index cf8a04e..791224c 100644
--- a/libobjc/Makefile.in
+++ b/libobjc/Makefile.in
@@ -72,7 +72,7 @@ CC1OBJ = `$(CC) -print-prog-name=cc1obj`
 
 INCLUDES = -I$(srcdir)/objc  -I$(srcdir)/$(MULTISRCTOP)../gcc \
   -I$(srcdir)/$(MULTISRCTOP)../gcc/config -I$(MULTIBUILDTOP)../../gcc \
-  -I$(srcdir)/$(MULTISRCTOP)../include
+  -I$(srcdir)/$(MULTISRCTOP)../include -isystem /usr/${CHOST}/include
 
 OBJC_GCFLAGS=-DOBJC_WITH_GC=1
 OBJC_THREAD_FILE=thr-@OBJC_THREAD_FILE@
diff --git a/libstdc++/config/linux.mt b/libstdc++/config/linux.mt
index 5848851..1a62e6e 100644
--- a/libstdc++/config/linux.mt
+++ b/libstdc++/config/linux.mt
@@ -1,2 +1,2 @@
 # tell we want the mt-safe version
-MT_CFLAGS = -D_IO_MTSAFE_IO
+MT_CFLAGS =
diff --git a/libstdc++/configure.in b/libstdc++/configure.in
index bd2c9c7..e6ca8c2 100644
--- a/libstdc++/configure.in
+++ b/libstdc++/configure.in
@@ -128,7 +128,7 @@ ${moveifchange} temp.mt target-mkfrag
 LIBDIR=yes
 TO_TOPDIR=../
 ALL='libs'
-XCXXINCLUDES="-I${srcdir} -I${srcdir}/stl -I${TO_TOPDIR}libio -I${srcdir}/${TO_TOPDIR}libio"
+XCXXINCLUDES="-I${srcdir} -I${srcdir}/stl -I${TO_TOPDIR}libio -I${srcdir}/${TO_TOPDIR}libio -isystem /usr/${CHOST}/include"
 MOSTLYCLEAN='*.o pic stamp-picdir core so_locations $(MOSTLYCLEAN_JUNK)'
 CLEAN='$(CLEAN_JUNK)'
 EXTRA_DISTCLEAN='target-mkfrag'
