diff --git a/gcc/config/alpha/crtbegin.asm b/gcc/config/alpha/crtbegin.asm
index 3a31fc6..6350d47 100644
--- a/gcc/config/alpha/crtbegin.asm
+++ b/gcc/config/alpha/crtbegin.asm
@@ -98,31 +98,6 @@ __EH_FRAME_BEGIN__:
  # Support recursive calls to exit.
 $ptr:	.quad	__DTOR_LIST__
 
-/* A globally unique widget for c++ local destructors to hang off.
-
-   This has a unique value in every dso; in the main program its
-   value is zero.  The object should be protected.  This means the
-   instance in any dso or the main program is not used in any other
-   dso.  The dynamic linker takes care of this.  */
-
-       .global __dso_handle
-       .type __dso_handle,@object
-       .size __dso_handle,8
-#ifdef SHARED
-.section .data
-       .align 3
-__dso_handle:
-       .quad   __dso_handle
-#else
-.section .bss
-       .align 3
-__dso_handle:
-       .zero 8
-#endif
-#ifdef HAVE_GAS_HIDDEN
-       .hidden __dso_handle
-#endif
-
 .text
 
 	.align 3
diff --git a/libstdc++/Makefile.in b/libstdc++/Makefile.in
index 78def0d..bce932d 100644
--- a/libstdc++/Makefile.in
+++ b/libstdc++/Makefile.in
@@ -317,7 +317,7 @@ install:
 	for FILE in $(LIBS) ; do \
 	  rm -f $(DESTDIR)$${INSTALLDIR}$(MULTISUBDIR)/$$FILE ; \
 	  if [ $$FILE = $(SHLINK) ] ; then \
-	    if [ -f $${INSTALLDIR}/$(MSHLINK) ]; then \
+	    if [ -f $(DESTDIR)$${INSTALLDIR}/$(MSHLINK) ]; then \
 	      rm -f $(DESTDIR)$${INSTALLLINKDIR}$(MULTISUBDIR)/$$FILE ; \
 	      $(LN_S) $${RELINSTALLDIR}$(MSHLINK) $(DESTDIR)$${INSTALLLINKDIR}$(MULTISUBDIR)/$$FILE \
 		|| cp $${RELINSTALLDIR}$(MSHLINK) $(DESTDIR)$${INSTALLLINKDIR}$(MULTISUBDIR)/$$FILE ; \
