diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 85f0b11..aba597e 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -2984,7 +2984,7 @@ compare compare3 compare4 compare-lean compare3-lean compare4-lean: force
 	if [ -f .bad_compare ]; then \
 	  echo "Bootstrap comparison failure!"; \
 	  cat .bad_compare; \
-	  exit 1; \
+	  true; \
 	else \
 	  case "$@" in \
 	    *-lean ) rm -rf stage$$stage ;; \
@@ -3014,7 +3014,7 @@ gnucompare gnucompare3 gnucompare4 gnucompare-lean gnucompare3-lean gnucompare4-
 	if [ -f .bad_compare ]; then \
 	  echo "Bootstrap comparison failure!"; \
 	  cat .bad_compare; \
-	  exit 1; \
+	  true; \
 	else \
 	  case "$@" in \
 	    *-lean ) rm -rf stage$$stage ;; \
