diff --git a/gcc/config/alpha/linux.h b/gcc/config/alpha/linux.h
index b8eb9e9..46f90cb 100644
--- a/gcc/config/alpha/linux.h
+++ b/gcc/config/alpha/linux.h
@@ -25,7 +25,7 @@ Boston, MA 02111-1307, USA.  */
 
 #undef CPP_PREDEFINES
 #define CPP_PREDEFINES \
-"-Dlinux -Dunix -Asystem(linux) -D_LONGLONG -D__alpha__ " \
+"-Dlinux -Dunix -Asystem(linux) -D_LONGLONG -D__alpha__ -D__LP64__ -D_LP64 " \
 SUB_CPP_PREDEFINES
 
 #undef LIB_SPEC
