diff --git a/math/math.h b/math/math.h
index a1631a8c..6a7978ca 100644
--- a/math/math.h
+++ b/math/math.h
@@ -902,7 +902,7 @@ extern int signgam;
 	       long double: FUNC ## l ARGS,	\
 	       __MATH_TG_F64X (FUNC, ARGS)	\
 	       _Float128: FUNC ## f128 ARGS)
-# else
+# elif __GNUC_PREREQ (4,4)
 #  if __HAVE_FLOATN_NOT_TYPEDEF
 #   error "Non-typedef _FloatN but no _Generic."
 #  endif
@@ -917,6 +917,13 @@ extern int signgam;
        (__builtin_types_compatible_p (__typeof (TG_ARG), long double),	\
 	FUNC ## l ARGS,							\
 	FUNC ## f128 ARGS)))
+# else
+#  define __MATH_TG(TG_ARG, FUNC, ARGS)		\
+   (sizeof (TG_ARG) == sizeof (float)		\
+    ? FUNC ## f ARGS				\
+    : sizeof (TG_ARG) == sizeof (double)		\
+    ? FUNC ARGS					\
+    : FUNC ## l ARGS)
 # endif
 #else
 # define __MATH_TG(TG_ARG, FUNC, ARGS)		\
diff --git a/math/tgmath.h b/math/tgmath.h
index d3654e8c..97cdcde6 100644
--- a/math/tgmath.h
+++ b/math/tgmath.h
@@ -59,7 +59,9 @@
 # endif
 # if ((__HAVE_FLOAT64X && !__HAVE_FLOAT128)		\
       || (__HAVE_FLOAT128 && !__HAVE_FLOAT64X))
-#  error "Unsupported combination of types for <tgmath.h>."
+#  if __GNUC_PREREQ (4, 4)
+#   error "Unsupported combination of types for <tgmath.h>."
+#  endif
 # endif
 # define __TGMATH_1_NARROW_D(F, X)		\
   (F ## l (X))
