diff --git a/sysdeps/aarch64/fpu/bits/math-vector.h b/sysdeps/aarch64/fpu/bits/math-vector.h
index 04837bdc..8e1fb761 100644
--- a/sysdeps/aarch64/fpu/bits/math-vector.h
+++ b/sysdeps/aarch64/fpu/bits/math-vector.h
@@ -91,7 +91,7 @@
 # define __DECL_SIMD_tanf __DECL_SIMD_aarch64
 #endif
 
-#if __GNUC_PREREQ(9, 0)
+#if __GNUC_PREREQ(9, 0) && (!defined(__cplusplus) || __cplusplus < 201703L )
 #  define __ADVSIMD_VEC_MATH_SUPPORTED
 typedef __Float32x4_t __f32x4_t;
 typedef __Float64x2_t __f64x2_t;
