diff --git a/sysdeps/loongarch/Makefile b/sysdeps/loongarch/Makefile
index b00c090f..7726327f 100644
--- a/sysdeps/loongarch/Makefile
+++ b/sysdeps/loongarch/Makefile
@@ -47,11 +47,3 @@ endif
 ASFLAGS-.os += \
   $(pic-ccflag) \
   # ASFLAGS-.os
-
-# All the objects in lib*_nonshared.a need to be compiled with medium code
-# model or large applications may fail to link.
-ifeq (yes,$(have-cmodel-medium))
-CFLAGS-.oS += \
-  -mcmodel=medium \
-  # CFLAGS-.oS
-endif
