diff --git a/libg++/src/Intdouble.cc b/libg++/src/Intdouble.cc
index 8db9061..6e7e5ac 100644
--- a/libg++/src/Intdouble.cc
+++ b/libg++/src/Intdouble.cc
@@ -25,13 +25,7 @@ Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 #include <math.h>
 #include <limits.h>
 
-#ifndef HUGE_VAL
-#ifdef HUGE
-#define HUGE_VAL HUGE
-#else
 #define HUGE_VAL DBL_MAX
-#endif
-#endif
 
 // convert to a double 
 
diff --git a/libg++/src/Makefile.in b/libg++/src/Makefile.in
index 0f3f946..6fa3bd2 100644
--- a/libg++/src/Makefile.in
+++ b/libg++/src/Makefile.in
@@ -43,7 +43,7 @@ OBJS =  AllocRing.o Obstack.o builtin.o \
  Normal.o NegExp.o Weibull.o Erlang.o DiscUnif.o \
  Uniform.o Poisson.o HypGeom.o Geom.o Binomial.o \
  RNG.o ACG.o MLCG.o  RndInt.o  \
- Fix.o Fix16.o Fix24.o CursesW.o GetOpt.o $(EH_FILES) \
+ Fix.o Fix16.o Fix24.o GetOpt.o $(EH_FILES) \
  new.o chr.o error.o gcd.o hash.o \
  lg.o fmtq.o ioob.o pow.o sqrt.o str.o timer.o \
  math.o compare.o $(BIT_OBJS) \
diff --git a/libg++/src/SmplHist.cc b/libg++/src/SmplHist.cc
index a1fcd15..3bad7aa 100644
--- a/libg++/src/SmplHist.cc
+++ b/libg++/src/SmplHist.cc
@@ -21,15 +21,9 @@ Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 #include <stream.h>
 #include <SmplHist.h>
 #include <math.h>
-
-#ifndef HUGE_VAL
-#ifdef HUGE
-#define HUGE_VAL HUGE
-#else
 #include <float.h>
+
 #define HUGE_VAL DBL_MAX
-#endif
-#endif
 
 const int SampleHistogramMinimum = -2;
 const int SampleHistogramMaximum = -1;
diff --git a/libg++/src/SmplStat.cc b/libg++/src/SmplStat.cc
index 461bea4..9f47dd0 100644
--- a/libg++/src/SmplStat.cc
+++ b/libg++/src/SmplStat.cc
@@ -21,15 +21,9 @@ Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 #include <stream.h>
 #include <SmplStat.h>
 #include <math.h>
-
-#ifndef HUGE_VAL
-#ifdef HUGE
-#define HUGE_VAL HUGE
-#else
 #include <float.h>
+
 #define HUGE_VAL DBL_MAX
-#endif
-#endif
 
 // error handling
 
diff --git a/libio/floatconv.c b/libio/floatconv.c
index 912c283..23a904d 100644
--- a/libio/floatconv.c
+++ b/libio/floatconv.c
@@ -1285,14 +1285,7 @@ _IO_strtod
                         if (e1 > DBL_MAX_10_EXP) {
  ovfl:
                                 errno = ERANGE;
-#if defined(sun) && !defined(__svr4__)
-/* SunOS defines HUGE_VAL as __infinity(), which is in libm. */
-#undef HUGE_VAL
-#endif
-#ifndef HUGE_VAL
-#define HUGE_VAL        1.7976931348623157E+308
-#endif
-                                rv = HUGE_VAL;
+                                rv = DBL_MAX;
                                 goto ret;
                                 }
                         if (e1 >>= 4) {
diff --git a/libio/gen-params b/libio/gen-params
index d01d718..59ea512 100755
--- a/libio/gen-params
+++ b/libio/gen-params
@@ -285,8 +285,6 @@ done
 
 cat <<!EOF!
 typedef ${clock_t-int /* default */} ${macro_prefix}clock_t;
-typedef ${dev_t-int /* default */} ${macro_prefix}dev_t;
-typedef ${fpos_t-long /* default */} ${macro_prefix}fpos_t;
 typedef ${gid_t-int /* default */} ${macro_prefix}gid_t;
 typedef ${ino_t-int /* default */} ${macro_prefix}ino_t;
 typedef ${mode_t-int /* default */} ${macro_prefix}mode_t;
@@ -297,7 +295,6 @@ typedef ${pid_t-int /* default */} ${macro_prefix}pid_t;
 #define __PTRDIFF_TYPE__ ${ptrdiff_t-long int /* default */}
 #endif
 typedef __PTRDIFF_TYPE__ ${macro_prefix}ptrdiff_t;
-typedef ${sigset_t-int /* default */} ${macro_prefix}sigset_t;
 #ifndef __SIZE_TYPE__
 #define __SIZE_TYPE__ ${size_t-unsigned long /* default */}
 #endif
diff --git a/libio/libio.h b/libio/libio.h
index f108151..5206eba 100644
--- a/libio/libio.h
+++ b/libio/libio.h
@@ -30,8 +30,9 @@ the executable file might be covered by the GNU General Public License. */
 
 #if 1
 #include <_G_config.h>
-#define _IO_pos_t _G_fpos_t /* obsolete */
-#define _IO_fpos_t _G_fpos_t
+#define _G_fpos_t _G_off_t
+#define _IO_pos_t _G_off_t
+#define _IO_fpos_t _G_off_t
 #define _IO_size_t _G_size_t
 #define _IO_ssize_t _G_ssize_t
 #define _IO_off_t _G_off_t
