diff --git a/libio/libio.h b/libio/libio.h
index 5206eba..f2ab38e 100644
--- a/libio/libio.h
+++ b/libio/libio.h
@@ -224,9 +224,11 @@ extern int __overflow __P((_IO_FILE*, int));
 
 extern unsigned __adjust_column __P((unsigned start, const char *line, int count));
 
-#define _IO_getc(_fp) \
+#define _IO_getc_unlocked(_fp) \
        ((_fp)->_IO_read_ptr >= (_fp)->_IO_read_end ? __uflow(_fp) \
 	: *(unsigned char*)(_fp)->_IO_read_ptr++)
+extern int _IO_getc (_IO_FILE *__fp);
+
 #define _IO_peekc(_fp) \
        ((_fp)->_IO_read_ptr >= (_fp)->_IO_read_end \
 	  && __underflow(_fp) == EOF ? EOF \
