diff --git a/gcc/gcc.c b/gcc/gcc.c
index b3e7909..1ba3e4c 100644
--- a/gcc/gcc.c
+++ b/gcc/gcc.c
@@ -641,7 +641,7 @@ static const char *asm_options =
 "%a %Y %{c:%W{o*}%{!o*:-o %w%b%O}}%{!c:-o %d%w%u%O}";
 
 static const char *invoke_as =
-"%{!S:-o %{|!pipe:%g.s} |\n as %(asm_options) %{!pipe:%g.s} %A }";
+"%{!S:-o %{|!pipe:%g.s} |\n i586-pc-msdosdjgpp-as %(asm_options) %{!pipe:%g.s} %A }";
 
 /* Some compilers have limits on line lengths, and the multilib_select
    and/or multilib_matches strings can be very long, so we build them at
