diff --git a/configure b/configure
index bd83b9b2..d0abe7d5 100755
--- a/configure
+++ b/configure
@@ -2770,7 +2770,6 @@ libgcj="target-libffi \
 # the host libraries and the host tools (which may be a cross compiler)
 # Note that libiberty is not a target library.
 target_libraries="target-libgcc \
-		target-libbacktrace \
 		target-libgloss \
 		target-newlib \
 		target-libgomp \
diff --git a/configure.ac b/configure.ac
index 4f70caca..08ae66c9 100644
--- a/configure.ac
+++ b/configure.ac
@@ -152,7 +152,6 @@ libgcj="target-libffi \
 # the host libraries and the host tools (which may be a cross compiler)
 # Note that libiberty is not a target library.
 target_libraries="target-libgcc \
-		target-libbacktrace \
 		target-libgloss \
 		target-newlib \
 		target-libgomp \
