diff --git a/gcc/cp/name-lookup.c b/gcc/cp/name-lookup.c
index f3a12ed64..5022d10f5 100644
--- a/gcc/cp/name-lookup.c
+++ b/gcc/cp/name-lookup.c
@@ -1095,8 +1095,10 @@ pushdecl_maybe_friend_1 (tree x, bool is_friend)
                                || TREE_CODE (x) == TYPE_DECL)))
                    /* Don't check for internally generated vars unless
                       it's an implicit typedef (see create_implicit_typedef
-                      in decl.c).  */
-		   && (!DECL_ARTIFICIAL (x) || DECL_IMPLICIT_TYPEDEF_P (x)))
+                      in decl.c) or anonymous union variable.  */
+		   && (!DECL_ARTIFICIAL (x)
+		       || DECL_IMPLICIT_TYPEDEF_P (x)
+		       || (VAR_P (x) && DECL_ANON_UNION_VAR_P (x))))
 	    {
 	      bool nowarn = false;
 
diff --git a/gcc/testsuite/g++.dg/warn/Wshadow-12.C b/gcc/testsuite/g++.dg/warn/Wshadow-12.C
new file mode 100644
index 000000000..95dc5beb9
--- /dev/null
+++ b/gcc/testsuite/g++.dg/warn/Wshadow-12.C
@@ -0,0 +1,9 @@
+// PR c++/72707
+// { dg-do compile }
+
+void
+foo (double x)
+{
+  union { int x; };	// { dg-error "shadows a parameter" }
+  x = 0;
+}
