diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index ffd06d1b8..bcf57abcf 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -17935,6 +17935,9 @@ cp_parser_initializer (cp_parser* parser, bool* is_direct_init,
       init = error_mark_node;
     }
 
+  if (check_for_bare_parameter_packs (init))
+    init = error_mark_node;
+
   return init;
 }
 
diff --git a/gcc/testsuite/g++.dg/cpp0x/variadic168.C b/gcc/testsuite/g++.dg/cpp0x/variadic168.C
new file mode 100644
index 000000000..202897e4b
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/variadic168.C
@@ -0,0 +1,18 @@
+// PR c++/69095
+// { dg-do compile { target c++11 } }
+
+struct B1 {
+  template <typename Ret, typename... Args, unsigned = sizeof(Args)> // { dg-error "parameter packs not expanded" }
+  void insert(Ret);
+};
+
+struct B2 {
+  template <typename Ret, typename... Args>
+  void insert(Ret, unsigned = sizeof(Args)); // { dg-error "parameter packs not expanded" }
+};
+
+template <typename Ret, typename... Args, unsigned = sizeof(Args)> // { dg-error "parameter packs not expanded" }
+void insert1(Ret);
+
+template <typename Ret, typename... Args>
+void insert2(Ret, unsigned = sizeof(Args)); // { dg-error "parameter packs not expanded" }
