diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index e703d779b..e48deb6d5 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -14936,6 +14936,9 @@ instantiate_template_1 (tree tmpl, tree orig_args, tsubst_flags_t complain)
       return NULL_TREE;
     }
 
+  if (targ_ptr == error_mark_node)
+    return error_mark_node;
+
   /* Check to see if we already have this specialization.  */
   gen_tmpl = most_general_template (tmpl);
   if (tmpl != gen_tmpl)
diff --git a/gcc/testsuite/g++.dg/cpp0x/pr58674.C b/gcc/testsuite/g++.dg/cpp0x/pr58674.C
new file mode 100644
index 000000000..fe97c6de9
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/pr58674.C
@@ -0,0 +1,18 @@
+// PR c++/58674
+// { dg-do compile { target c++11 } }
+
+template<int> struct A {};
+
+template<int N> using B = A<N>;
+
+template<typename T> struct C
+{
+  B<T::i> b;  // { dg-error "not usable" }
+};
+
+struct X
+{
+  static const int i;
+};
+
+C<X> c;
