diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 90c504937..3bb975e37 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -10573,7 +10573,9 @@ grokdeclarator (const cp_declarator *declarator,
 	      {
 		/* C++ allows static class members.  All other work
 		   for this is done by grokfield.  */
-		decl = build_lang_decl_loc (declarator->id_loc,
+		decl = build_lang_decl_loc (declarator
+					    ? declarator->id_loc
+					    : input_location,
 					    VAR_DECL, unqualified_id, type);
 		set_linkage_for_static_data_member (decl);
 		/* Even if there is an in-class initialization, DECL
diff --git a/gcc/testsuite/g++.dg/parse/bitfield4.C b/gcc/testsuite/g++.dg/parse/bitfield4.C
new file mode 100644
index 000000000..e10fa5f33
--- /dev/null
+++ b/gcc/testsuite/g++.dg/parse/bitfield4.C
@@ -0,0 +1,6 @@
+// PR c++/58700
+
+struct A
+{
+  static int : 4;  // { dg-error "bit-field" }
+};
