diff --git a/gcc/lto/lto-lang.c b/gcc/lto/lto-lang.c
index 0e5c121f5..089ea290e 100644
--- a/gcc/lto/lto-lang.c
+++ b/gcc/lto/lto-lang.c
@@ -1191,8 +1191,8 @@ lto_init (void)
     }
   else
     {
-      lto_define_builtins (va_list_type_node,
-			   build_reference_type (va_list_type_node));
+      lto_define_builtins (build_reference_type (va_list_type_node),
+			   va_list_type_node);
     }
 
   targetm.init_builtins ();
