diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index e4347e1b7..fb72eadb5 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -11890,6 +11890,7 @@ lookup_and_check_tag (enum tag_types tag_code, tree name,
       /* First try ordinary name lookup, ignoring hidden class name
 	 injected via friend declaration.  */
       decl = lookup_name_prefer_type (name, 2);
+      decl = strip_using_decl (decl);
       /* If that fails, the name will be placed in the smallest
 	 non-class, non-function-prototype scope according to 3.3.1/5.
 	 We may already have a hidden name declared as friend in this
diff --git a/gcc/testsuite/g++.dg/lookup/using54.C b/gcc/testsuite/g++.dg/lookup/using54.C
new file mode 100644
index 000000000..680bdadaa
--- /dev/null
+++ b/gcc/testsuite/g++.dg/lookup/using54.C
@@ -0,0 +1,16 @@
+// PR c++/60894
+
+struct B
+{
+  struct S {};
+};
+
+struct D : B
+{
+  using B::S;
+  void doIt(struct S&);
+};
+
+void D::doIt(struct S&)
+{
+}
