diff --git a/Makefile.in b/Makefile.in
index abbf991..c74a8a6 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -1251,7 +1251,7 @@ c-common.o : c-common.c $(CONFIG_H) $(TREE_H) c-tree.h c-lex.h flags.h
 # Language-independent files.
 
 DRIVER_DEFINES = \
-  -DSTANDARD_STARTFILE_PREFIX=\"$(libdir)/\" \
+  -DSTANDARD_STARTFILE_PREFIX=\"$(build_tooldir)/lib/\" \
   -DSTANDARD_EXEC_PREFIX=\"$(libdir)/gcc-lib/\" \
   -DDEFAULT_TARGET_VERSION=\"$(version)\" \
   -DDEFAULT_TARGET_MACHINE=\"$(target_alias)\" \
diff --git a/cccp.c b/cccp.c
index 28ff4ca..62d1a0b 100644
--- a/cccp.c
+++ b/cccp.c
@@ -87,10 +87,6 @@ extern char *getenv ();
 # define __attribute__(x)
 #endif
 
-#ifndef STANDARD_INCLUDE_DIR
-# define STANDARD_INCLUDE_DIR "/usr/include"
-#endif
-
 /* By default, colon separates directories in a path.  */
 #ifndef PATH_SEPARATOR
 # define PATH_SEPARATOR ':'
@@ -530,7 +526,6 @@ static struct default_include {
 #ifndef STANDARD_INCLUDE_COMPONENT
 #define STANDARD_INCLUDE_COMPONENT 0
 #endif
-    { STANDARD_INCLUDE_DIR, STANDARD_INCLUDE_COMPONENT, 0, 0 },
 #endif /* not CROSS_COMPILE */
     { 0, 0, 0, 0 }
     };
diff --git a/cpplib.c b/cpplib.c
index 4875601..f121858 100644
--- a/cpplib.c
+++ b/cpplib.c
@@ -136,9 +136,6 @@ extern char *update_path ();
 #define PATH_SEPARATOR ':'
 #endif
 
-#ifndef STANDARD_INCLUDE_DIR
-#define STANDARD_INCLUDE_DIR "/usr/include"
-#endif
 #ifndef INCLUDE_LEN_FUDGE
 #define INCLUDE_LEN_FUDGE 0
 #endif
@@ -398,7 +395,6 @@ static struct default_include {
 #ifndef STANDARD_INCLUDE_COMPONENT
 #define STANDARD_INCLUDE_COMPONENT 0
 #endif
-    { STANDARD_INCLUDE_DIR, STANDARD_INCLUDE_COMPONENT, 0, 0 },
 #endif /* not CROSS_COMPILE */
     { 0, 0, 0, 0 }
     };
diff --git a/protoize.c b/protoize.c
index 34f8fe5..57a6854 100644
--- a/protoize.c
+++ b/protoize.c
@@ -298,13 +298,6 @@ static const unexpansion unexpansions[] = {
 
 static const int hash_mask = (HASH_TABLE_SIZE - 1);
 
-/* Make a table of default system include directories
-   just as it is done in cccp.c.  */
-
-#ifndef STANDARD_INCLUDE_DIR
-#define STANDARD_INCLUDE_DIR "/usr/include"
-#endif
-
 #ifndef LOCAL_INCLUDE_DIR
 #define LOCAL_INCLUDE_DIR "/usr/local/include"
 #endif
@@ -341,7 +334,6 @@ struct default_include { const char *fname;
 #ifdef SYSTEM_INCLUDE_DIR
     { SYSTEM_INCLUDE_DIR, 0, 0, 0 },
 #endif
-    { STANDARD_INCLUDE_DIR, 0, 0, 0},
 #endif /* not CROSS_COMPILE */
     { 0, 0, 0, 0}
     };
