diff --git a/bfd/elf32-ppc.c b/bfd/elf32-ppc.c
index ba9e59ab..cdd9bd97 100644
--- a/bfd/elf32-ppc.c
+++ b/bfd/elf32-ppc.c
@@ -10361,7 +10361,7 @@ ppc_elf_finish_dynamic_sections (bfd *output_bfd,
 #define ELF_ARCH		bfd_arch_powerpc
 #define ELF_TARGET_ID		PPC32_ELF_DATA
 #define ELF_MACHINE_CODE	EM_PPC
-#define ELF_MAXPAGESIZE		0x10000
+#define ELF_MAXPAGESIZE		0x1000
 #define ELF_COMMONPAGESIZE	0x1000
 #define elf_info_to_howto	ppc_elf_info_to_howto
 
diff --git a/bfd/elf32-sh.c b/bfd/elf32-sh.c
index 28589600..2dd9278b 100644
--- a/bfd/elf32-sh.c
+++ b/bfd/elf32-sh.c
@@ -6642,7 +6642,7 @@ sh_elf_encode_eh_address (bfd *abfd,
 #undef	TARGET_LITTLE_NAME
 #define	TARGET_LITTLE_NAME		"elf32-shl-nbsd"
 #undef	ELF_MAXPAGESIZE
-#define	ELF_MAXPAGESIZE			0x10000
+#define	ELF_MAXPAGESIZE			0x1000
 #undef	ELF_COMMONPAGESIZE
 #undef	elf_symbol_leading_char
 #define	elf_symbol_leading_char		0
diff --git a/bfd/elf32-sparc.c b/bfd/elf32-sparc.c
index b61b4bb9..9c0f2d73 100644
--- a/bfd/elf32-sparc.c
+++ b/bfd/elf32-sparc.c
@@ -222,7 +222,7 @@ elf32_sparc_reloc_type_class (const struct bfd_link_info *info,
 #define ELF_MACHINE_CODE EM_SPARC
 #define ELF_MACHINE_ALT1 EM_SPARC32PLUS
 #define ELF_OSABI	ELFOSABI_GNU
-#define ELF_MAXPAGESIZE 0x10000
+#define ELF_MAXPAGESIZE 0x2000
 #define ELF_COMMONPAGESIZE 0x2000
 
 #define bfd_elf32_bfd_merge_private_bfd_data \
diff --git a/bfd/elf64-alpha.c b/bfd/elf64-alpha.c
index 199505d1..1c580a57 100644
--- a/bfd/elf64-alpha.c
+++ b/bfd/elf64-alpha.c
@@ -5402,7 +5402,7 @@ static const struct elf_size_info alpha_elf_size_info =
 #define ELF_ARCH		bfd_arch_alpha
 #define ELF_TARGET_ID		ALPHA_ELF_DATA
 #define ELF_MACHINE_CODE	EM_ALPHA
-#define ELF_MAXPAGESIZE	0x10000
+#define ELF_MAXPAGESIZE	0x2000
 #define ELF_COMMONPAGESIZE	0x2000
 
 #define bfd_elf64_bfd_link_hash_table_create \
diff --git a/bfd/elf64-ppc.c b/bfd/elf64-ppc.c
index 7e41085e..aa2e3f42 100644
--- a/bfd/elf64-ppc.c
+++ b/bfd/elf64-ppc.c
@@ -73,7 +73,7 @@ static bfd_vma opd_entry_value
 #define ELF_ARCH		bfd_arch_powerpc
 #define ELF_TARGET_ID		PPC64_ELF_DATA
 #define ELF_MACHINE_CODE	EM_PPC64
-#define ELF_MAXPAGESIZE		0x10000
+#define ELF_MAXPAGESIZE		0x1000
 #define ELF_COMMONPAGESIZE	0x1000
 #define elf_info_to_howto	ppc64_elf_info_to_howto
 
diff --git a/bfd/elf64-sparc.c b/bfd/elf64-sparc.c
index c73f2934..ed358c5b 100644
--- a/bfd/elf64-sparc.c
+++ b/bfd/elf64-sparc.c
@@ -890,7 +890,7 @@ static const struct elf_size_info elf64_sparc_size_info =
 #define ELF_ARCH	bfd_arch_sparc
 #define ELF_TARGET_ID	SPARC_ELF_DATA
 #define ELF_OSABI	ELFOSABI_GNU
-#define ELF_MAXPAGESIZE 0x100000
+#define ELF_MAXPAGESIZE 0x2000
 #define ELF_COMMONPAGESIZE 0x2000
 
 /* This is the official ABI value.  */
diff --git a/bfd/elfnn-loongarch.c b/bfd/elfnn-loongarch.c
index 2c67cccd..427947dd 100644
--- a/bfd/elfnn-loongarch.c
+++ b/bfd/elfnn-loongarch.c
@@ -7064,7 +7064,7 @@ elf_loongarch64_hash_symbol (struct elf_link_hash_entry *h)
 #define ELF_TARGET_ID LARCH_ELF_DATA
 #define ELF_MACHINE_CODE EM_LOONGARCH
 #define ELF_MINPAGESIZE 0x1000
-#define ELF_MAXPAGESIZE 0x10000
+#define ELF_MAXPAGESIZE 0x4000
 #define ELF_COMMONPAGESIZE 0x4000
 #define bfd_elfNN_bfd_reloc_type_lookup loongarch_reloc_type_lookup
 #define bfd_elfNN_bfd_link_hash_table_create				  \
