diff --git a/gcc/ada/adaint.c b/gcc/ada/adaint.c
index dc94d63d5..e1d52a15f 100644
--- a/gcc/ada/adaint.c
+++ b/gcc/ada/adaint.c
@@ -73,6 +73,18 @@
 #define HOST_OBJECT_SUFFIX ".obj"
 #endif
 
+#if defined (__MINGW32__)
+
+#if defined (RTX)
+#include <windows.h>
+#include <Rtapi.h>
+#else
+#include "mingw32.h"
+
+/* Current code page to use, set in initialize.c.  */
+UINT CurrentCodePage;
+#endif
+
 #ifdef IN_RTS
 #include "tconfig.h"
 #include "tsystem.h"
@@ -107,18 +119,6 @@
 extern "C" {
 #endif
 
-#if defined (__MINGW32__)
-
-#if defined (RTX)
-#include <windows.h>
-#include <Rtapi.h>
-#else
-#include "mingw32.h"
-
-/* Current code page to use, set in initialize.c.  */
-UINT CurrentCodePage;
-#endif
-
 #include <sys/utime.h>
 
 /* For isalpha-like tests in the compiler, we're expected to resort to
diff --git a/gcc/ada/initialize.c b/gcc/ada/initialize.c
index 00c4d04e1..e38662390 100644
--- a/gcc/ada/initialize.c
+++ b/gcc/ada/initialize.c
@@ -40,6 +40,11 @@
 #include "vxWorks.h"
 #endif
 
+#if defined (__MINGW32__)
+#include "mingw32.h"
+#include <windows.h>
+#endif
+
 #ifdef IN_RTS
 #include "tconfig.h"
 #include "tsystem.h"
@@ -62,8 +67,6 @@ extern "C" {
 /******************************************/
 
 #if defined (__MINGW32__)
-#include "mingw32.h"
-#include <windows.h>
 
 extern void __gnat_init_float (void);
 extern void __gnat_install_SEH_handler (void *);
diff --git a/gcc/ada/mingw32.h b/gcc/ada/mingw32.h
index 67bfd2ccc..5ca32f87f 100644
--- a/gcc/ada/mingw32.h
+++ b/gcc/ada/mingw32.h
@@ -36,8 +36,6 @@
 #ifndef _MINGW32_H
 #define _MINGW32_H
 
-#include <_mingw.h>
-
 #ifndef RTX
 #define GNAT_UNICODE_SUPPORT
 #define _UNICODE /* For C runtime */
diff --git a/libgfortran/io/unix.c b/libgfortran/io/unix.c
index 185d0dca1..f418e4d77 100644
--- a/libgfortran/io/unix.c
+++ b/libgfortran/io/unix.c
@@ -98,6 +98,12 @@ id_from_fd (const int fd)
 }
 
 #endif /* HAVE_WORKING_STAT */
+
+
+/* On mingw, we don't use umask in tempfile_open(), because it
+   doesn't support the user/group/other-based permissions.  */
+#undef HAVE_UMASK
+
 #endif /* __MINGW32__ */
 
 
