diff --git a/Makefile.in b/Makefile.in
index b4d4c5b..16ae427 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -116,7 +116,7 @@ USER_H = $(srcdir)/ginclude/stdarg.h $(srcdir)/ginclude/stddef.h \
 
 # Target to use whe installing assert.h.  Some systems may
 # want to set this empty.
-INSTALL_ASSERT_H = install-assert-h
+INSTALL_ASSERT_H =
 
 # The GCC to use for compiling libgcc2.a, enquire, and libgcc1-test.
 # Usually the one we just built.
diff --git a/cp/g++.c b/cp/g++.c
index f7ebd5b..aa23d94 100644
--- a/cp/g++.c
+++ b/cp/g++.c
@@ -51,7 +51,7 @@ Boston, MA 02111-1307, USA.  */
    Makefile should compile this file with the proper name
    (e.g., "i386-aout-gcc").  */
 #ifndef GCC_NAME
-#define GCC_NAME "i586-legacy-linux-gnulibc1-gcc-2.7.2"
+#define GCC_NAME "i486-legacy-linuxaout-gcc-2.7.2"
 #endif
 
 /* This bit is set if we saw a `-xfoo' language specification.  */
@@ -389,7 +389,7 @@ main (argc, argv)
 
   /* This will be 0 if we encounter a situation where we should not
      link in libstdc++, or 2 if we should link in libg++ as well.  */
-  int library = 1;
+  int library = 0;
 
   /* Used to track options that take arguments, so we don't go wrapping
      those with -xc++/-xnone.  */
