diff --git a/Makefile.in b/Makefile.in
index d5c20b0..b930d42 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -141,19 +141,19 @@ local_prefix = /usr/local
 # Directory in which to put host dependent programs and libraries
 exec_prefix = $(prefix)
 # Directory in which to put the executable for the command `gcc'
-bindir = $(exec_prefix)/bin
+bindir = $(exec_prefix)/$(target)/gcc-bin/$(version)
 # Directory in which to put the directories used by the compiler.
 libdir = $(exec_prefix)/lib
 # Directory in which the compiler finds executables, libraries, etc.
 libsubdir = $(libdir)/gcc-lib/$(target)/$(version)
 # Directory for header file assert.h.
-includedir = $(prefix)/include
+includedir = $(libsubdir)/include
 # assertdir is overridden in cross-make.
 assertdir = $(includedir)
 # Extension (if any) to put in installed man-page filename.
 manext = .1
 # Directory in which to put man pages.
-mandir = $(prefix)/man/man1
+mandir = $(prefix)/share/gcc-data/$(target)/$(version)/man/man1
 # Directory in which to find other cross-compilation tools and headers.
 # Used in install-cross.
 tooldir = $(prefix)/$(target)
