diff --git a/Makefile.in b/Makefile.in
index e645671..869f4ee 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -141,19 +141,21 @@ local_prefix = /usr/local
 # Directory in which to put host dependent programs and libraries
 exec_prefix = $(prefix)
 # Directory in which to put the executable for the command `gcc'
-bindir = $(exec_prefix)/bin
+bindir = $(exec_prefix)/$(target)/gcc-bin/$(version)
 # Directory in which to put the directories used by the compiler.
 libdir = $(exec_prefix)/lib
 # Directory in which the compiler finds executables, libraries, etc.
 libsubdir = $(libdir)/gcc-lib/$(target)/$(version)
 # Directory for header file assert.h.
-includedir = $(prefix)/include
+includedir = $(libsubdir)/include
+# Directory in which the compiler finds g++ includes.
+gxx_include_dir= $(includedir)/g++
 # assertdir is overridden in cross-make.
 assertdir = $(includedir)
 # Extension (if any) to put in installed man-page filename.
 manext = .1
 # Directory in which to put man pages.
-mandir = $(prefix)/man/man1
+mandir = $(prefix)/share/gcc-data/$(target)/$(version)/man/man1
 # Directory in which to find other cross-compilation tools and headers.
 # Used in install-cross.
 tooldir = $(prefix)/$(target)
@@ -1188,7 +1190,7 @@ cccp.o: cccp.c $(CONFIG_H) pcp.h version.c
 # is for compatibility with the current version of libg++.
 	$(CC) $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
 	  -DGCC_INCLUDE_DIR=\"$(libsubdir)/include\" \
-	  -DGPLUSPLUS_INCLUDE_DIR=\"$(libdir)/g++-include\" \
+	  -DGPLUSPLUS_INCLUDE_DIR=\"$(gxx_include_dir)\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(local_prefix)/include\" \
 	  -DCROSS_INCLUDE_DIR=\"$(libsubdir)/sys-include\" \
 	  -DTOOL_INCLUDE_DIR=\"$(tooldir)/include\" \
@@ -1210,7 +1212,7 @@ unprotoize.o:	stamp-proto ;
 stamp-proto:	$(srcdir)/protoize.c getopt.h $(CONFIG_H)
 	$(CC) -c $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
           -DGCC_INCLUDE_DIR=\"$(libsubdir)/include\" \
-          -DGPLUSPLUS_INCLUDE_DIR=\"$(libdir)/g++-include\" \
+          -DGPLUSPLUS_INCLUDE_DIR=\"$(gxx_include_dir)\" \
           -DCROSS_INCLUDE_DIR=\"$(libsubdir)/sys-include\" \
 	  -DTOOL_INCLUDE_DIR=\"$(tooldir)/include\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(local_prefix)/include\" \
@@ -1219,7 +1221,7 @@ stamp-proto:	$(srcdir)/protoize.c getopt.h $(CONFIG_H)
 	mv protoize.o unprotoize.o
 	$(CC) -c $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
           -DGCC_INCLUDE_DIR=\"$(libsubdir)/include\" \
-          -DGPLUSPLUS_INCLUDE_DIR=\"$(libdir)/g++-include\" \
+          -DGPLUSPLUS_INCLUDE_DIR=\"$(gxx_include_dir)\" \
           -DCROSS_INCLUDE_DIR=\"$(libsubdir)/sys-include\" \
 	  -DTOOL_INCLUDE_DIR=\"$(tooldir)/include\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(local_prefix)/include\" \
