diff --git a/Makefile.in b/Makefile.in
index 9b327e2..74fea79 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -114,7 +114,7 @@ RANLIB_TEST_FOR_TARGET = [ -f $(TARGET_TOOLPREFIX)ranlib ]
 CROSS_TOOLS =
 
 # Dir to search for system headers.  Overridden by cross-make.
-SYSTEM_HEADER_DIR = /usr/include
+SYSTEM_HEADER_DIR = $(tooldir)/include
 
 # There may be a premade insn-attrtab.c for this machine.
 # (You could rebuild it with genattrtab as usual, but it takes a long time.)
diff --git a/gbl-ctors.h b/gbl-ctors.h
index 13048ce..acbf7fb 100644
--- a/gbl-ctors.h
+++ b/gbl-ctors.h
@@ -32,11 +32,11 @@ the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */
 */
 
 #ifdef HAVE_ATEXIT
-extern void atexit (void (*) (void));
+extern int atexit (void (*) (void));
 #define ON_EXIT(FUNC,ARG) atexit ((FUNC))
 #else
 #ifdef sun
-extern void on_exit (void*, void*);
+extern int on_exit (void*, void*);
 #define ON_EXIT(FUNC,ARG) on_exit ((FUNC), (ARG))
 #endif
 #endif
