diff --git a/Makefile.in b/Makefile.in
index 6a271f2..33ad1a6 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -53,16 +53,16 @@ INSTALL_XFORM = $(INSTALL) -t='$(program_transform_name)'
 AS = as
 AR = ar
 AR_FLAGS = rc
-CC = cc
+CC = @CC@
 
 # We don't specify -g -O because many compilers don't support -g -O,
 # and/or -O is broken in and of itself.
-CFLAGS = -g
+CFLAGS = -g -O2
 
-CXX = gcc
+CXX = @CXX@
 
 # Use -O to stress test the compiler.
-CXXFLAGS = -g -O
+CXXFLAGS = -g -O2
 
 RANLIB = ranlib
 NM = nm
@@ -207,6 +207,7 @@ BASE_FLAGS_TO_PASS = \
 	"CC_FOR_BUILD=$(CC_FOR_BUILD)" \
 	"CC_FOR_TARGET=$(CC_FOR_TARGET)" \
 	"CFLAGS=$(CFLAGS)" \
+	"DESTDIR=$(DESTDIR)" \
 	"CXX_FOR_BUILD=$(CXX_FOR_BUILD)" \
 	"CXXFLAGS=$(CXXFLAGS)" \
 	"CXX_FOR_TARGET=$(CXX_FOR_TARGET)" \
@@ -854,9 +855,9 @@ all-xiberty: all-gcc all-newlib
 ### other supporting targets
 
 MAKEDIRS= \
-	$(prefix) \
-	$(exec_prefix) \
-	$(tooldir)
+	$(DESTDIR)$(prefix) \
+	$(DESTDIR)$(exec_prefix) \
+	$(DESTDIR)$(tooldir)
 
 .PHONY: install-dirs
 install-dirs:
diff --git a/etc/Makefile.in b/etc/Makefile.in
index 3b57840..beff8d7 100644
--- a/etc/Makefile.in
+++ b/etc/Makefile.in
@@ -42,13 +42,13 @@ DVIFILES = configure.dvi standards.dvi cfg-paper.dvi
 all:
 
 install:  $(srcdir)/configure.man
-	$(INSTALL_DATA) $(srcdir)/configure.man $(man1dir)/configure.1
+	$(INSTALL_DATA) $(srcdir)/configure.man $(DESTDIR)$(man1dir)/configure.1
 
 info: $(INFOFILES)
 
 install-info:
 	for i in *.info* ; do \
-	  $(INSTALL_DATA) $$i $(infodir) ; \
+	  $(INSTALL_DATA) $$i $(DESTDIR)$(infodir) ; \
 	done
 
 dvi : $(DVIFILES)
diff --git a/libg++/Makefile.in b/libg++/Makefile.in
index 5d1fe0e..5e16050 100644
--- a/libg++/Makefile.in
+++ b/libg++/Makefile.in
@@ -154,14 +154,14 @@ installcheck: check
 
 .PHONY: install
 install:
-	$(INSTALL_DATA) $(TARGETLIB) $(INSTALLDIR)/$(TARGETLIB)
-	$(RANLIB) $(INSTALLDIR)/$(TARGETLIB)
-	chmod a-x $(INSTALLDIR)/$(TARGETLIB)
+	$(INSTALL_DATA) $(TARGETLIB) $(DESTDIR)$(INSTALLDIR)/$(TARGETLIB)
+	$(RANLIB) $(DESTDIR)$(INSTALLDIR)/$(TARGETLIB)
+	chmod a-x $(DESTDIR)$(INSTALLDIR)/$(TARGETLIB)
 	cd $(srcdir)/src; \
 	for FILE in *.h gen/*.ccP gen/*.hP; do \
-		rm -f $(gxx_includedir)/$$FILE ; \
-		$(INSTALL_DATA) $$FILE $(gxx_includedir)/$$FILE || exit 1; \
-		chmod a-x,a+r $(gxx_includedir)/$$FILE ; \
+		rm -f $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+		$(INSTALL_DATA) $$FILE $(DESTDIR)$(gxx_includedir)/$$FILE || exit 1; \
+		chmod a-x,a+r $(DESTDIR)$(gxx_includedir)/$$FILE ; \
 	done
 	if [ "x$(GPERF)" != "x" ]; then \
 	  cd gperf ; $(MAKE) $(FLAGS_TO_PASS) install; \
diff --git a/libg++/genclass/Makefile.in b/libg++/genclass/Makefile.in
index c6a04f5..d73fe87 100644
--- a/libg++/genclass/Makefile.in
+++ b/libg++/genclass/Makefile.in
@@ -20,4 +20,4 @@ gentest: $(srcdir)/gentest.sh $(srcdir)/expected.out
 
 .PHONY: install
 install:
-	$(INSTALL_PROGRAM) genclass $(bindir)/genclass
+	$(INSTALL_PROGRAM) genclass $(DESTDIR)$(bindir)/genclass
diff --git a/libg++/gperf/Makefile.in b/libg++/gperf/Makefile.in
index 36453a1..5b2ba8d 100644
--- a/libg++/gperf/Makefile.in
+++ b/libg++/gperf/Makefile.in
@@ -32,7 +32,7 @@ src_all:
 install:
 	rootme=`pwd` ; export rootme ; cd src ; \
 		$(MAKE) $(FLAGS_TO_PASS) install
-	$(INSTALL_DATA) $(srcdir)/gperf.1 $(man1dir)/gperf.1
+	$(INSTALL_DATA) $(srcdir)/gperf.1 $(DESTDIR)$(man1dir)/gperf.1
 
 distrib: force
 	cd ..; rm -f gperf.tar.Z; tar cvf gperf.tar gperf; compress gperf.tar;\
diff --git a/libg++/gperf/src/Makefile.in b/libg++/gperf/src/Makefile.in
index 396ae65..d15ff5c 100644
--- a/libg++/gperf/src/Makefile.in
+++ b/libg++/gperf/src/Makefile.in
@@ -34,4 +34,4 @@ $(TARGETPROG): $(OBJECTS)
 
 .PHONY: install
 install: $(TARGETPROG)
-	$(INSTALL_PROGRAM) $(TARGETPROG) $(bindir)/$(TARGETPROG)
+	$(INSTALL_PROGRAM) $(TARGETPROG) $(DESTDIR)$(bindir)/$(TARGETPROG)
diff --git a/libiberty/Makefile.in b/libiberty/Makefile.in
index 6923867..51b7f9b 100644
--- a/libiberty/Makefile.in
+++ b/libiberty/Makefile.in
@@ -130,14 +130,14 @@ INSTALL_DEST = libdir
 install: install_to_$(INSTALL_DEST)
 
 install_to_libdir: all
-	$(INSTALL_DATA) $(TARGETLIB) $(libdir)/$(TARGETLIB).n
-	( cd $(libdir) ; $(RANLIB) $(libdir)/$(TARGETLIB).n )
-	mv -f $(libdir)/$(TARGETLIB).n $(libdir)/$(TARGETLIB)
+	$(INSTALL_DATA) $(TARGETLIB) $(DESTDIR)$(libdir)/$(TARGETLIB).n
+	( cd $(DESTDIR)$(libdir) ; $(RANLIB) $(DESTDIR)$(libdir)/$(TARGETLIB).n )
+	mv -f $(DESTDIR)$(libdir)/$(TARGETLIB).n $(DESTDIR)$(libdir)/$(TARGETLIB)
 
 install_to_tooldir: all
-	$(INSTALL_DATA) $(TARGETLIB) $(tooldir)/lib/$(TARGETLIB).n
-	( cd $(tooldir) ; $(RANLIB) $(tooldir)/lib/$(TARGETLIB).n )
-	mv -f $(tooldir)/lib/$(TARGETLIB).n $(tooldir)/lib/$(TARGETLIB)
+	$(INSTALL_DATA) $(TARGETLIB) $(DESTDIR)$(tooldir)/lib/$(TARGETLIB).n
+	( cd $(DESTDIR)$(tooldir) ; $(RANLIB) $(DESTDIR)$(tooldir)/lib/$(TARGETLIB).n )
+	mv -f $(DESTDIR)$(tooldir)/lib/$(TARGETLIB).n $(DESTDIR)$(tooldir)/lib/$(TARGETLIB)
 
 # The default configuration adds to libiberty all those functions that are
 # missing in libc.  More precisely, it includes whatever $(CC) fails to find.
diff --git a/libio/Makefile.in b/libio/Makefile.in
index 036e089..0031bcf 100644
--- a/libio/Makefile.in
+++ b/libio/Makefile.in
@@ -81,14 +81,14 @@ _G_config.h: ${srcdir}/gen-params
 
 USER_INCLUDES = *.h
 install:
-	rm -f $(tooldir)/include/_G_config.h ; \
-	  $(INSTALL_DATA) _G_config.h $(tooldir)/include/_G_config.h || exit 1;
+	rm -f $(DESTDIR)$(tooldir)/include/_G_config.h ; \
+	  $(INSTALL_DATA) _G_config.h $(DESTDIR)$(tooldir)/include/_G_config.h || exit 1;
 	cd $(srcdir); \
 	for FILE in $(USER_INCLUDES) ; do \
-		rm -f $(gxx_includedir)/$$FILE ; \
-		$(INSTALL_DATA) $$FILE $(gxx_includedir)/$$FILE ; \
-		chmod a-x $(gxx_includedir)/$$FILE ; \
+		rm -f $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+		$(INSTALL_DATA) $$FILE $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+		chmod a-x $(DESTDIR)$(gxx_includedir)/$$FILE ; \
 	done
-	$(INSTALL_DATA) libiostream.a $(INSTALLDIR)/libiostream.a
-	$(RANLIB) $(INSTALLDIR)/libiostream.a
-	chmod a-x $(INSTALLDIR)/libiostream.a
+	$(INSTALL_DATA) libiostream.a $(DESTDIR)$(INSTALLDIR)/libiostream.a
+	$(RANLIB) $(DESTDIR)$(INSTALLDIR)/libiostream.a
+	chmod a-x $(DESTDIR)$(INSTALLDIR)/libiostream.a
