diff --git a/libio/libio.h b/libio/libio.h
index 869c121..0938023 100644
--- a/libio/libio.h
+++ b/libio/libio.h
@@ -231,10 +231,12 @@ extern int __overflow _PARAMS((_IO_FILE*, int));
 
 extern unsigned __adjust_column _PARAMS((unsigned start, const char *line, int count));
 
-#define _IO_getc(_fp) \
+#define _IO_getc_unlocked(_fp) \
        ((_fp)->_IO_read_ptr >= (_fp)->_IO_read_end \
 	  && __underflow(_fp) == EOF ? EOF \
 	: *(unsigned char*)(_fp)->_IO_read_ptr++)
+extern int _IO_getc (_IO_FILE *__fp);
+
 #define _IO_peekc(_fp) \
        ((_fp)->_IO_read_ptr >= (_fp)->_IO_read_end \
 	  && __underflow(_fp) == EOF ? EOF \
