diff --git a/gcc/config.gcc b/gcc/config.gcc
index 6dc2427bc..ad0cbc9ad 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -1881,7 +1881,7 @@ mips*-*-netbsd*)			# NetBSD/mips, either endian.
 mips64*-*-linux* | mipsisa64*-*-linux*)
 	tm_file="dbxelf.h elfos.h gnu-user.h linux.h glibc-stdint.h ${tm_file} mips/linux.h mips/linux64.h"
 	tmake_file="${tmake_file} mips/t-linux64 mips/t-libgcc-mips16"
-	tm_defines="${tm_defines} MIPS_ABI_DEFAULT=ABI_N32"
+	tm_defines="${tm_defines} MIPS_ABI_DEFAULT=ABI_64"
 	case ${target} in
 		mips64el-st-linux-gnu)
 			tm_file="${tm_file} mips/st.h"
diff --git a/gcc/config/mips/linux64.h b/gcc/config/mips/linux64.h
index 987a9915c..9271e7bce 100644
--- a/gcc/config/mips/linux64.h
+++ b/gcc/config/mips/linux64.h
@@ -26,7 +26,7 @@ along with GCC; see the file COPYING3.  If not see
   BASE_DRIVER_SELF_SPECS, \
   LINUX_DRIVER_SELF_SPECS \
   " %{!EB:%{!EL:%(endian_spec)}}" \
-  " %{!mabi=*: -mabi=n32}"
+  " %{!mabi=*: -mabi=64}"
 
 #undef LIB_SPEC
 #define LIB_SPEC "\
