diff --git a/gcc/gengtype-state.c b/gcc/gengtype-state.c
index 07c880258..0510b194b 100644
--- a/gcc/gengtype-state.c
+++ b/gcc/gengtype-state.c
@@ -303,7 +303,7 @@ read_a_state_token (void)
       obstack_1grow (&id_obstack, (char) 0);
       ids = XOBFINISH (&id_obstack, char *);
       sid = state_ident_by_name (ids, INSERT);
-      obstack_free (&id_obstack, ids);
+      obstack_free (&id_obstack, NULL);
       ids = NULL;
       tk = XCNEW (struct state_token_st);
       tk->stok_kind = STOK_NAME;
@@ -408,7 +408,7 @@ read_a_state_token (void)
       tk->stok_file = state_path;
       tk->stok_next = NULL;
       strcpy (tk->stok_un.stok_string, cstr);
-      obstack_free (&bstring_obstack, cstr);
+      obstack_free (&bstring_obstack, NULL);
 
       return tk;
     }
diff --git a/gcc/gengtype.c b/gcc/gengtype.c
index 6c0ca4a0b..fb2033341 100644
--- a/gcc/gengtype.c
+++ b/gcc/gengtype.c
@@ -1925,7 +1925,7 @@ matching_file_name_substitute (const char *filnam, regmatch_t pmatch[10],
   obstack_1grow (&str_obstack, '\0');
   rawstr = XOBFINISH (&str_obstack, char *);
   str = xstrdup (rawstr);
-  obstack_free (&str_obstack, rawstr);
+  obstack_free (&str_obstack, NULL);
   DBGPRINTF ("matched replacement %s", str);
   rawstr = NULL;
   return str;
