diff --git a/bfd/elflink.c b/bfd/elflink.c
index 6346d7e2..fd9ec192 100644
--- a/bfd/elflink.c
+++ b/bfd/elflink.c
@@ -13684,9 +13684,6 @@ bfd_elf_final_link (bfd *abfd, struct bfd_link_info *info)
 		  else if (bfd_link_pde (info))
 		    info->callbacks->einfo
 		      (_("%P: warning: creating DT_TEXTREL in a PDE\n"));
-		  else
-		    info->callbacks->einfo
-		      (_("%P: warning: creating DT_TEXTREL in a PIE\n"));
 		  break;
 		}
 	    }
