diff --git a/Makefile.in b/Makefile.in
index 9b30686..227ece9 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -68,7 +68,7 @@ HOST_PREFIX =
 HOST_PREFIX_1 = loser-
 
 # These flag values are normally overridden by the configure script.
-CFLAGS = -g
+CFLAGS = -g -O2
 CXXFLAGS = -g -O2
 
 LIBCFLAGS = $(CFLAGS)
@@ -323,6 +323,7 @@ BASE_FLAGS_TO_PASS = \
 	"CC_FOR_BUILD=$(CC_FOR_BUILD)" \
 	"CC_FOR_TARGET=$(CC_FOR_TARGET)" \
 	"CFLAGS=$(CFLAGS)" \
+	"DESTDIR=$(DESTDIR)" \
 	"CFLAGS_FOR_TARGET=$(CFLAGS_FOR_TARGET)" \
 	"CXX_FOR_BUILD=$(CXX_FOR_BUILD)" \
 	"CXXFLAGS=$(CXXFLAGS)" \
@@ -1466,8 +1467,8 @@ install-target: $(INSTALL_TARGET_MODULES)
 ### other supporting targets
 
 MAKEDIRS= \
-	$(prefix) \
-	$(exec_prefix)
+	$(DESTDIR)$(prefix) \
+	$(DESTDIR)$(exec_prefix)
 .PHONY: installdirs
 installdirs: mkinstalldirs
 	$(SHELL) $(srcdir)/mkinstalldirs $(MAKEDIRS)
diff --git a/libiberty/Makefile.in b/libiberty/Makefile.in
index 29eeab0..a6a02f9 100644
--- a/libiberty/Makefile.in
+++ b/libiberty/Makefile.in
@@ -103,6 +103,7 @@ FLAGS_TO_PASS = \
 	"AR_FLAGS=$(AR_FLAGS)" \
 	"CC=$(CC)" \
 	"CFLAGS=$(CFLAGS)" \
+	"DESTDIR=$(DESTDIR)" \
 	"LIBCFLAGS=$(LIBCFLAGS)" \
 	"EXTRA_OFILES=$(EXTRA_OFILES)" \
 	"HDEFINES=$(HDEFINES)" \
@@ -171,15 +172,15 @@ INSTALL_DEST = libdir
 install: install_to_$(INSTALL_DEST)
 
 install_to_libdir: all
-	$(INSTALL_DATA) $(TARGETLIB) $(libdir)$(MULTISUBDIR)/$(TARGETLIB).n
-	( cd $(libdir)$(MULTISUBDIR) ; $(RANLIB) $(TARGETLIB).n )
-	mv -f $(libdir)$(MULTISUBDIR)/$(TARGETLIB).n $(libdir)$(MULTISUBDIR)/$(TARGETLIB)
+	$(INSTALL_DATA) $(TARGETLIB) $(DESTDIR)$(libdir)$(MULTISUBDIR)/$(TARGETLIB).n
+	( cd $(DESTDIR)$(libdir)$(MULTISUBDIR) ; $(RANLIB) $(TARGETLIB).n )
+	mv -f $(DESTDIR)$(libdir)$(MULTISUBDIR)/$(TARGETLIB).n $(DESTDIR)$(libdir)$(MULTISUBDIR)/$(TARGETLIB)
 	@$(MULTIDO) $(FLAGS_TO_PASS) multi-do DO=install
 
 install_to_tooldir: all
-	$(INSTALL_DATA) $(TARGETLIB) $(tooldir)/lib$(MULTISUBDIR)/$(TARGETLIB).n
-	( cd $(tooldir)/lib$(MULTISUBDIR) ; $(RANLIB) $(TARGETLIB).n )
-	mv -f $(tooldir)/lib$(MULTISUBDIR)/$(TARGETLIB).n $(tooldir)/lib$(MULTISUBDIR)/$(TARGETLIB)
+	$(INSTALL_DATA) $(TARGETLIB) $(DESTDIR)$(tooldir)/lib$(MULTISUBDIR)/$(TARGETLIB).n
+	( cd $(DESTDIR)$(tooldir)/lib$(MULTISUBDIR) ; $(RANLIB) $(TARGETLIB).n )
+	mv -f $(DESTDIR)$(tooldir)/lib$(MULTISUBDIR)/$(TARGETLIB).n $(DESTDIR)$(tooldir)/lib$(MULTISUBDIR)/$(TARGETLIB)
 	@$(MULTIDO) $(FLAGS_TO_PASS) multi-do DO=install
 
 # The default configuration adds to libiberty all those functions that are
diff --git a/libio/Makefile.in b/libio/Makefile.in
index 2e3fb06..dd49f22 100644
--- a/libio/Makefile.in
+++ b/libio/Makefile.in
@@ -91,6 +91,10 @@ tpipe: tpipe.o libio.a
 iostream.list: stamp-picdir $(_G_CONFIG_H) $(LIBIOSTREAM_DEP)
 	@echo "$(LIBIOSTREAM_USE)"> iostream.list
 
+# The "pic" subdir must be created before we can put any object
+# files into it.
+$(LIBIOSTREAM_DEP) $(LIBIO_OBJECTS): stamp-picdir
+
 stmp-stdio:
 	@rootme=`pwd`/ ; export rootme; cd stdio ; \
 	  $(MAKE) $(FLAGS_TO_PASS) stdio.list
@@ -107,15 +111,15 @@ install:
 	rootme=`pwd`/ ; export rootme ; \
 	if [ -z "$(MULTISUBDIR)" ]; then \
 	  if [ "$(_G_CONFIG_H)" != "" ]; then \
-	    rm -f $(tooldir)/include/_G_config.h ; \
-	    $(INSTALL_DATA) _G_config.h $(tooldir)/include/_G_config.h || exit 1; \
+	    rm -f $(DESTDIR)$(tooldir)/include/_G_config.h ; \
+	    $(INSTALL_DATA) _G_config.h $(DESTDIR)$(tooldir)/include/_G_config.h || exit 1; \
 	  else true; \
 	  fi ; \
 	  cd $(srcdir); \
 	  for FILE in $(USER_INCLUDES); do if [ $$FILE != _G_config.h ]; then \
-	    rm -f $(gxx_include_dir)/$$FILE ; \
-	    $(INSTALL_DATA) $$FILE $(gxx_include_dir)/$$FILE ; \
-	    chmod a-x $(gxx_include_dir)/$$FILE ; \
+	    rm -f $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
+	    $(INSTALL_DATA) $$FILE $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
+	    chmod a-x $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
 	  else true; fi; done ; \
 	else true; \
 	fi
diff --git a/libio/config/linux.mt b/libio/config/linux.mt
index 0585d90..da20dc2 100644
--- a/libio/config/linux.mt
+++ b/libio/config/linux.mt
@@ -1,7 +1,7 @@
 # Use the libio which comes with the local libc.
 
 # That is where we keep the g++ header files.
-gxx_include_dir =$(prefix)/include/g++
+gxx_include_dir =$(libdir)/include/g++-v2
 
 # Comment this out to avoid including the stdio functions in libiostream.a:
 # LIBIOSTREAM_OBJECTS = $(IO_OBJECTS) $(IOSTREAM_OBJECTS) $(STDIO_WRAP_OBJECTS) $(OSPRIM_OBJECTS)
diff --git a/libstdc++/Makefile.in b/libstdc++/Makefile.in
index bd35db9..3f32935 100644
--- a/libstdc++/Makefile.in
+++ b/libstdc++/Makefile.in
@@ -250,46 +250,46 @@ install:
 	if [ -z "$(MULTISUBDIR)" ]; then \
 	  cd $(srcdir); \
 	  for FILE in $(HEADERS); do \
-	    rm -f $(gxx_include_dir)/$$FILE ; \
+	    rm -f $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
 	    if [ -f stl/$$FILE ]; then \
-	      $(INSTALL_DATA) stl/$$FILE $(gxx_include_dir)/$$FILE ; \
+	      $(INSTALL_DATA) stl/$$FILE $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
 	    else \
-	      $(INSTALL_DATA) $$FILE $(gxx_include_dir)/$$FILE ; \
+	      $(INSTALL_DATA) $$FILE $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
 	    fi ; \
-	    chmod a-x $(gxx_include_dir)/$$FILE ; \
+	    chmod a-x $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
 	  done ; \
 	  for FILE in *.h std/*.*; do \
-	    rm -f $(gxx_include_dir)/$$FILE ; \
-	    $(INSTALL_DATA) $$FILE $(gxx_include_dir)/$$FILE ; \
-	    chmod a-x $(gxx_include_dir)/$$FILE ; \
+	    rm -f $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
+	    $(INSTALL_DATA) $$FILE $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
+	    chmod a-x $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
 	  done ; \
 	  cd stl; \
 	  for FILE in *.h; do \
-	    rm -f $(gxx_include_dir)/$$FILE ; \
-	    $(INSTALL_DATA) $$FILE $(gxx_include_dir)/$$FILE ; \
-	    chmod a-x $(gxx_include_dir)/$$FILE ; \
+	    rm -f $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
+	    $(INSTALL_DATA) $$FILE $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
+	    chmod a-x $(DESTDIR)$(gxx_include_dir)/$$FILE ; \
 	  done ; \
 	else true ; \
 	fi
 	rootme=`pwd`/ ; export rootme ; \
-	rm -f $(INSTALLDIR)$(MULTISUBDIR)/$(SHLINK) ; \
+	rm -f $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$(SHLINK) ; \
 	for FILE in $(LIBS) ; do \
-	  rm -f $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	  rm -f $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  if [ $$FILE = $(SHLINK) ] ; then \
-	    ln -f -s $(SHLIB) $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    ln -f -s $(SHLIB) $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  elif [ $$FILE = mshlink ]; then \
 	    for FILE in $(MSHLINK) ; do \
-	      rm -f $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
-	      ln -f -s $(SHLIB) $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	      rm -f $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	      ln -f -s $(SHLIB) $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	    done; \
 	  elif [ $$FILE = $(SHLIB) ]; then \
-	    $(INSTALL_PROGRAM) $$FILE $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    $(INSTALL_PROGRAM) $$FILE $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	    : On the HP, shared libraries must be mode 555. ;\
-	    chmod 555 $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    chmod 555 $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  else \
-	    $(INSTALL_DATA) $$FILE $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
-	    $(RANLIB) $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
-	    chmod a-x $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    $(INSTALL_DATA) $$FILE $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    $(RANLIB) $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    chmod a-x $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  fi ; \
 	done
 	@rootme=`pwd`/ ; export rootme ; \
