diff --git a/gcc/objc/objc-act.c b/gcc/objc/objc-act.c
index fb4335da7..cc9ac8c94 100644
--- a/gcc/objc/objc-act.c
+++ b/gcc/objc/objc-act.c
@@ -2342,10 +2342,17 @@ build_module_initializer_routine (void)
 #endif
 
   objc_push_parm (build_decl (PARM_DECL, NULL_TREE, void_type_node));
+#ifdef OBJCPLUS
+  objc_start_function (get_identifier (TAG_GNUINIT),
+		       build_function_type (void_type_node,
+					    OBJC_VOID_AT_END),
+		       NULL_TREE, NULL_TREE);
+#else
   objc_start_function (get_identifier (TAG_GNUINIT),
 		       build_function_type (void_type_node,
 					    OBJC_VOID_AT_END),
 		       NULL_TREE, objc_get_parm_info (0));
+#endif
 
   body = c_begin_compound_stmt (true);
   add_stmt (build_function_call
diff --git a/gcc/testsuite/obj-c++.dg/comp-types-10.mm b/gcc/testsuite/obj-c++.dg/comp-types-10.mm
index 0a8f1c9b2..c7f0cb635 100644
--- a/gcc/testsuite/obj-c++.dg/comp-types-10.mm
+++ b/gcc/testsuite/obj-c++.dg/comp-types-10.mm
@@ -1,6 +1,5 @@
 /* Yet another mysterious gimplifier crasher.  */
 /* { dg-do compile } */
-/* { dg-xfail-if "PR23716" { *-*-* } { "*" } { "" } } */
 /* { dg-prune-output ".*internal compiler error.*" } */
 /* { dg-options "-O3" } */
 
