diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 4747fa8d7..c17d88a5b 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -712,6 +712,12 @@ check_specialization_namespace (tree tmpl)
      function, member class or static data member of a class template
      shall be declared in the namespace of which the class template is
      a member.  */
+  if (current_scope() != DECL_CONTEXT (tmpl)
+      && !at_namespace_scope_p ())
+    {
+      error ("specialization of %qD must appear at namespace scope", tmpl);
+      return false;
+    }
   if (is_associated_namespace (current_namespace, tpl_ns))
     /* Same or super-using namespace.  */
     return true;
diff --git a/gcc/testsuite/g++.dg/template/spec37.C b/gcc/testsuite/g++.dg/template/spec37.C
new file mode 100644
index 000000000..2c01eb028
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/spec37.C
@@ -0,0 +1,6 @@
+// PR c++/28300
+
+template<typename> struct A
+{
+    template<typename T> struct A<T*>; // { dg-error "namespace scope" }
+};
