diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 72d1b76f3..697d85f3b 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -8510,6 +8510,8 @@ grokdeclarator (const cp_declarator *declarator,
 	      type = build_memfn_type (type,
 				       declarator->u.pointer.class_type,
 				       memfn_quals);
+	      if (type == error_mark_node)
+		return error_mark_node;
 	      memfn_quals = TYPE_UNQUALIFIED;
 	    }
 
diff --git a/gcc/testsuite/g++.dg/template/crash103.C b/gcc/testsuite/g++.dg/template/crash103.C
new file mode 100644
index 000000000..9f3e224e7
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/crash103.C
@@ -0,0 +1,4 @@
+// PR c++/45665
+
+template < typename > struct S;
+void (S <0>::*ptr) (); // { dg-error "type" }
