diff --git a/gcc/fortran/expr.c b/gcc/fortran/expr.c
index fedee3637..cb0feb3c1 100644
--- a/gcc/fortran/expr.c
+++ b/gcc/fortran/expr.c
@@ -1215,7 +1215,12 @@ find_array_section (gfc_expr *expr, gfc_ref *ref)
 	    }
 
 	  gcc_assert (begin->rank == 1);
-	  gcc_assert (begin->shape);
+	  /* Zero-sized arrays have no shape and no elements, stop early.  */
+	  if (!begin->shape) 
+	    {
+	      mpz_init_set_ui (nelts, 0);
+	      break;
+	    }
 
 	  vecsub[d] = begin->value.constructor;
 	  mpz_set (ctr[d], vecsub[d]->expr->value.integer);
diff --git a/gcc/testsuite/gfortran.dg/zero_sized_6.f90 b/gcc/testsuite/gfortran.dg/zero_sized_6.f90
new file mode 100644
index 000000000..30ed8dfaa
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/zero_sized_6.f90
@@ -0,0 +1,6 @@
+! { dg-do "compile" }
+! PR38709 - ICE-on-invalid on zero-sized array in init-expr.
+
+  INTEGER, PARAMETER :: a(1) = (/ 1 /)
+  INTEGER, PARAMETER :: i = a(shape(1))   ! { dg-error "Incompatible ranks" }
+END
