diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 76e039968..70f421f00 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -13533,6 +13533,10 @@ unify (tree tparms, tree targs, tree parm, tree arg, int strict)
       FOR_EACH_CONSTRUCTOR_VALUE (CONSTRUCTOR_ELTS (arg), i, elt)
 	{
 	  int elt_strict = strict;
+
+	  if (elt == error_mark_node)
+	    return 1;
+
 	  if (!BRACE_ENCLOSED_INITIALIZER_P (elt))
 	    {
 	      tree type = TREE_TYPE (elt);
diff --git a/gcc/testsuite/g++.dg/cpp0x/initlist31.C b/gcc/testsuite/g++.dg/cpp0x/initlist31.C
new file mode 100644
index 000000000..ffc985567
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/initlist31.C
@@ -0,0 +1,13 @@
+// PR c++/43028
+// { dg-options "-std=c++0x" }
+
+#include <initializer_list>
+
+struct string { string(std::initializer_list<char>) { } };
+
+void f() {
+  auto y =
+  {
+    string(Equation()) // { dg-error "not declared" }
+  }; // { dg-error "unable to deduce" }
+}
