diff --git a/libstdc++-v3/libsupc++/exception_ptr.h b/libstdc++-v3/libsupc++/exception_ptr.h
index b24ec21d7..ad5325f8f 100644
--- a/libstdc++-v3/libsupc++/exception_ptr.h
+++ b/libstdc++-v3/libsupc++/exception_ptr.h
@@ -44,6 +44,8 @@ extern "C++" {
 
 namespace std 
 {
+  class type_info;
+
   /**
    * @addtogroup exceptions
    * @{
@@ -140,7 +142,7 @@ namespace std
       friend bool 
       operator==(const exception_ptr&, const exception_ptr&) throw();
 
-      const type_info*
+      const class std::type_info*
       __cxa_exception_type() const throw();
     };
 
