diff --git a/gcc/testsuite/gcc.target/arm/neon-vset_lanes8.c b/gcc/testsuite/gcc.target/arm/neon-vset_lanes8.c
index e87102edb..a7bbb0711 100644
--- a/gcc/testsuite/gcc.target/arm/neon-vset_lanes8.c
+++ b/gcc/testsuite/gcc.target/arm/neon-vset_lanes8.c
@@ -2,8 +2,7 @@
 
 /* { dg-do run } */
 /* { dg-require-effective-target arm_neon_hw } */
-/* { dg-options "-O0" } */
-/* { dg-add-options arm_neon } */
+/* { dg-options "-O0 -mfpu=neon -mfloat-abi=softfp" } */
 
 #include "arm_neon.h"
 #include <stdlib.h>
