diff --git a/libio/Makefile.in b/libio/Makefile.in
index a4be7bb..0ce0cd9 100644
--- a/libio/Makefile.in
+++ b/libio/Makefile.in
@@ -55,6 +55,9 @@ OSPRIM_OBJECTS = ioprims.o iostrerror.o cleanup.o
 LIBIOSTREAM_OBJECTS = $(IO_OBJECTS) $(IOSTREAM_OBJECTS) $(OSPRIM_OBJECTS)
 LIBIO_OBJECTS = $(IO_OBJECTS) $(STDIO_WRAP_OBJECTS) $(OSPRIM_OBJECTS)
 
+LIBIOSTREAM_DEP = $(LIBIOSTREAM_OBJECTS) stdio.list
+LIBIOSTREAM_USE = $(LIBIOSTREAM_OBJECTS) `cat stdio.list`
+
 USER_INCLUDES = *.h
 
 #### package, host, target, and site dependent Makefile fragments come in here.
@@ -65,9 +68,9 @@ libio.a: $(_G_CONFIG_H) $(LIBIO_OBJECTS)
 	$(AR) $(AR_FLAGS) libio.a $(LIBIO_OBJECTS)
 	$(RANLIB) libio.a
 
-libiostream.a: $(_G_CONFIG_H) $(LIBIOSTREAM_OBJECTS)
+libiostream.a: $(_G_CONFIG_H) $(LIBIOSTREAM_DEP)
 	-rm -rf libiostream.a
-	$(AR) $(AR_FLAGS) libiostream.a $(LIBIOSTREAM_OBJECTS)
+	$(AR) $(AR_FLAGS) libiostream.a $(LIBIOSTREAM_USE)
 	$(RANLIB) libiostream.a
 
 test: test.o libio.a
@@ -75,8 +78,17 @@ test: test.o libio.a
 tpipe: tpipe.o libio.a
 	$(CC) -o tpipe tpipe.o libio.a
 
-iostream.list: $(_G_CONFIG_H) $(LIBIOSTREAM_OBJECTS)
-	@echo "$(LIBIOSTREAM_OBJECTS)"> iostream.list
+iostream.list: $(_G_CONFIG_H) $(LIBIOSTREAM_DEP)
+	@echo "$(LIBIOSTREAM_USE)"> iostream.list
+
+stdio/stdio.list:
+	@rootme=`pwd`/ ; export rootme; cd stdio ; \
+	  $(MAKE) $(FLAGS_TO_PASS) stdio.list
+
+stdio.list: stdio/stdio.list
+	rm -f tstdio.list
+	sed 's,\([a-z_]*\.o\),stdio/\1,g' stdio/stdio.list > tstdio.list
+	mv tstdio.list stdio.list
 
 # The CC=$(CXX) is a kludge to force use of gcc.
 _G_config.h: ${srcdir}/gen-params
diff --git a/libio/stdio/Makefile.in b/libio/stdio/Makefile.in
index 1c6e7ba..0e4ea2b 100644
--- a/libio/stdio/Makefile.in
+++ b/libio/stdio/Makefile.in
@@ -7,10 +7,6 @@ srcdir = .
 STDIO_RENAMED_OBJECTS = ... filled in by configure ...
 # These are the files that a libc would want.
 STDIO_OBJECTS = $(STDIO_RENAMED_OBJECTS) \
-  clearerr.o fdopen.o feof.o ferror.o fgetc.o fileno.o \
-  fputc.o freopen.o fseek.o getc.o getchar.o getdelim.o getline.o getw.o \
-  popen.o putc.o putchar.o putw.o rewind.o \
-  setbuf.c setfileno.o setlinebuf.o snprintf.o \
   vfprintf.o vfscanf.o vprintf.o vscanf.o vsnprintf.o
 
 CC_FOR_STDIO=$(CC)
diff --git a/libio/stdio/configure.in b/libio/stdio/configure.in
index e0fa15f..1efb9bb 100644
--- a/libio/stdio/configure.in
+++ b/libio/stdio/configure.in
@@ -11,6 +11,7 @@ package_makefile_frag=../Make.pack
 
 # per-target:
 
+LIBDIR=yes
 TO_TOPDIR=../../
 ALL=nothing
 MOSTLYCLEAN='*.o core stdio.list'
@@ -26,9 +27,9 @@ MOSTLYCLEAN='*.o core stdio.list'
 
 # TODO: remove rename tmpfile tmpnam
 
-stdio_renames="fclose fflush fgetpos fgets fopen fprintf fputs fread \
-  fscanf fsetpos ftell fwrite getdelim gets perror printf puts \
-  scanf setbuffer setvbuf sprintf sscanf ungetc vsprintf vsscanf"
+stdio_renames="fprintf \
+  fscanf \
+  vsprintf vsscanf"
 stdio_objects=""
 
 for file in $stdio_renames ; do
