diff --git a/gcc/config.gcc b/gcc/config.gcc
index b4c8259b..fe568b35 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -1480,7 +1480,7 @@ mips*-*-netbsd*)			# NetBSD/mips, either endian.
 mips64*-*-linux*)
 	tm_file="dbxelf.h elfos.h svr4.h linux.h ${tm_file} mips/linux.h mips/linux64.h"
 	tmake_file="${tmake_file} mips/t-linux64"
-	tm_defines="MIPS_ABI_DEFAULT=ABI_N32"
+	tm_defines="MIPS_ABI_DEFAULT=ABI_64"
 	gnu_ld=yes
 	gas=yes
 	;;
diff --git a/gcc/config/mips/linux64.h b/gcc/config/mips/linux64.h
index 25455b43..3459329b 100644
--- a/gcc/config/mips/linux64.h
+++ b/gcc/config/mips/linux64.h
@@ -23,7 +23,7 @@ Boston, MA 02110-1301, USA.  */
    in order to make the other specs easier to write.  */
 #define DRIVER_SELF_SPECS \
 "%{!EB:%{!EL:%(endian_spec)}}", \
-"%{!mabi=*: -mabi=n32}"
+"%{!mabi=*: -mabi=64}"
 
 #undef SUBTARGET_ASM_SPEC
 #define SUBTARGET_ASM_SPEC "\
