diff --git a/Makefile.in b/Makefile.in
index b982ecce..9b970d40 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -35475,7 +35475,7 @@ compare:
 
 
 
-.PHONY: bootstrap
+.PHONY: bootstrap bootstrap-lean
 bootstrap:
 	echo stage3 > stage_final
 	@r=`${PWD_COMMAND}`; export r; \
@@ -35485,7 +35485,16 @@ bootstrap:
 	@r=`${PWD_COMMAND}`; export r; \
 	s=`cd $(srcdir); ${PWD_COMMAND}`; export s; \
 	$(MAKE) $(TARGET_FLAGS_TO_PASS) all-host all-target
-	@: $(MAKE); $(stage)
+
+bootstrap-lean:
+	echo stage3 > stage_final
+	@r=`${PWD_COMMAND}`; export r; \
+	s=`cd $(srcdir); ${PWD_COMMAND}`; export s; \
+	$(MAKE) $(RECURSE_FLAGS_TO_PASS) stage3-bubble
+	@: $(MAKE); $(unstage)
+	@r=`${PWD_COMMAND}`; export r; \
+	s=`cd $(srcdir); ${PWD_COMMAND}`; export s; \
+	$(MAKE) $(TARGET_FLAGS_TO_PASS) all-host all-target
 
 
 # Rules to wipe a stage and all the following ones, also used for cleanstrap
