diff --git a/Makefile.in b/Makefile.in
index c553e6b0..db3d60bf 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -35354,67 +35354,6 @@ stage3-start::
 @endif zlib
 
 stage3-end::
-	@rm -f stage_current
-@if bfd
-	@cd $(HOST_SUBDIR); set bfd stage3-bfd ; \
-	@UNDO_LINK_TO_DIR@  ; \
-	set prev-bfd stage2-bfd ; \
-	@UNDO_LINK_TO_DIR@ 
-@endif bfd
-@if opcodes
-	@cd $(HOST_SUBDIR); set opcodes stage3-opcodes ; \
-	@UNDO_LINK_TO_DIR@  ; \
-	set prev-opcodes stage2-opcodes ; \
-	@UNDO_LINK_TO_DIR@ 
-@endif opcodes
-@if binutils
-	@cd $(HOST_SUBDIR); set binutils stage3-binutils ; \
-	@UNDO_LINK_TO_DIR@  ; \
-	set prev-binutils stage2-binutils ; \
-	@UNDO_LINK_TO_DIR@ 
-@endif binutils
-@if gas
-	@cd $(HOST_SUBDIR); set gas stage3-gas ; \
-	@UNDO_LINK_TO_DIR@  ; \
-	set prev-gas stage2-gas ; \
-	@UNDO_LINK_TO_DIR@ 
-@endif gas
-@if gcc
-	@cd $(HOST_SUBDIR); set gcc stage3-gcc ; \
-	@UNDO_LINK_TO_DIR@  ; \
-	set prev-gcc stage2-gcc ; \
-	@UNDO_LINK_TO_DIR@ 
-@endif gcc
-@if intl
-	@cd $(HOST_SUBDIR); set intl stage3-intl ; \
-	@UNDO_LINK_TO_DIR@  ; \
-	set prev-intl stage2-intl ; \
-	@UNDO_LINK_TO_DIR@ 
-@endif intl
-@if ld
-	@cd $(HOST_SUBDIR); set ld stage3-ld ; \
-	@UNDO_LINK_TO_DIR@  ; \
-	set prev-ld stage2-ld ; \
-	@UNDO_LINK_TO_DIR@ 
-@endif ld
-@if libcpp
-	@cd $(HOST_SUBDIR); set libcpp stage3-libcpp ; \
-	@UNDO_LINK_TO_DIR@  ; \
-	set prev-libcpp stage2-libcpp ; \
-	@UNDO_LINK_TO_DIR@ 
-@endif libcpp
-@if libiberty
-	@cd $(HOST_SUBDIR); set libiberty stage3-libiberty ; \
-	@UNDO_LINK_TO_DIR@  ; \
-	set prev-libiberty stage2-libiberty ; \
-	@UNDO_LINK_TO_DIR@ 
-@endif libiberty
-@if zlib
-	@cd $(HOST_SUBDIR); set zlib stage3-zlib ; \
-	@UNDO_LINK_TO_DIR@  ; \
-	set prev-zlib stage2-zlib ; \
-	@UNDO_LINK_TO_DIR@ 
-@endif zlib
 
 # Bubble a bugfix through all the stages up to stage 3.  They
 # are remade, but not reconfigured.  The next stage (if any) will not
