diff --git a/Makefile.in b/Makefile.in
index 914dbe1..febfeb1 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -127,7 +127,7 @@ prefix = /usr/local
 # NOTE: local_prefix *should not* default from prefix.
 local_prefix = /usr/local
 # Directory in which to put the executable for the command `gcc'
-bindir = $(prefix)/bin
+bindir = $(exec_prefix)/$(target)/gcc-bin/$(version)
 # Directory in which to put the directories used by the compiler.
 libdir = $(prefix)/lib
 # Directory in which the compiler finds executables, libraries, etc.
@@ -135,7 +135,7 @@ libsubdir = $(libdir)/gcc/$(target)/$(version)
 # Extension (if any) to put in installed man-page filename.
 manext = .1
 # Directory in which to put man pages.
-mandir = $(prefix)/man/man1
+mandir = $(prefix)/share/gcc-data/$(target)/$(version)/man/man1
 # Directory in which to find other cross-compilation tools and headers.
 # Used in install-cross.
 tooldir = $(prefix)/$(target)
