diff --git a/Makefile.in b/Makefile.in
index 888819c..9a4d92b 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -129,7 +129,7 @@ local_prefix = /usr/local
 # Directory in which to put host dependent programs and libraries
 exec_prefix = $(prefix)
 # Directory in which to put the executable for the command `gcc'
-bindir = $(prefix)/bin
+bindir = $(exec_prefix)/$(target)/gcc-bin/$(version)
 # Directory in which to put the directories used by the compiler.
 libdir = $(exec_prefix)/lib
 # Directory in which the compiler finds executables, libraries, etc.
@@ -137,7 +137,7 @@ libsubdir = $(libdir)/gcc-lib/$(target)/$(version)
 # Extension (if any) to put in installed man-page filename.
 manext = .1
 # Directory in which to put man pages.
-mandir = $(prefix)/man/man1
+mandir = $(prefix)/share/gcc-data/$(target)/$(version)/man/man1
 # Directory in which to find other cross-compilation tools and headers.
 # Used in install-cross.
 tooldir = $(prefix)/$(target)
@@ -1039,7 +1039,7 @@ cccp.o: cccp.c $(CONFIG_H) pcp.h version.c
 # is for compatibility with the current version of libg++.
 	$(CC) $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
           -DGCC_INCLUDE_DIR=\"$(libsubdir)/include\" \
-          -DGPLUSPLUS_INCLUDE_DIR=\"$(libdir)/g++-include\" \
+          -DGPLUSPLUS_INCLUDE_DIR=\"$(libsubdir)/include/g++\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(local_prefix)/include\" \
           -DCROSS_INCLUDE_DIR=\"$(libsubdir)/sys-include\" \
           -DTOOL_INCLUDE_DIR=\"$(tooldir)/include\" \
@@ -1054,7 +1054,7 @@ protoize: protoize.o getopt.o getopt1.o version.o $(LIBDEPS)
 protoize.o:	$(srcdir)/protoize.c getopt.h $(CONFIG_H)
 	$(CC) -c $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
           -DGCC_INCLUDE_DIR=\"$(libsubdir)/include\" \
-          -DGPLUSPLUS_INCLUDE_DIR=\"$(libdir)/g++-include\" \
+          -DGPLUSPLUS_INCLUDE_DIR=\"$(libsubdir)/include/g++\" \
           -DCROSS_INCLUDE_DIR=\"$(libsubdir)/sys-include\" \
           -DTOOL_INCLUDE_DIR=\"$(tooldir)/include\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(local_prefix)/include\" \
@@ -1068,7 +1068,7 @@ unprotoize.o:	$(srcdir)/protoize.c getopt.h $(CONFIG_H)
 	cp $(srcdir)/protoize.c unprotoize.c
 	$(CC) -c $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
           -DGCC_INCLUDE_DIR=\"$(libsubdir)/include\" \
-          -DGPLUSPLUS_INCLUDE_DIR=\"$(libdir)/g++-include\" \
+          -DGPLUSPLUS_INCLUDE_DIR=\"$(libsubdir)/include/g++\" \
           -DCROSS_INCLUDE_DIR=\"$(libsubdir)/sys-include\" \
 	  -DTOOL_INCLUDE_DIR=\"$(tooldir)/include\" \
 	  -DLOCAL_INCLUDE_DIR=\"$(local_prefix)/include\" \
