diff --git a/libio/config/linux.mt b/libio/config/linux.mt
index 25d29c5..a3bcd5b 100644
--- a/libio/config/linux.mt
+++ b/libio/config/linux.mt
@@ -6,9 +6,9 @@
 # things just won't build with the bottom section uncommented.
 
 # Comment this out to avoid including the stdio functions in libiostream.a:
-LIBIOSTREAM_OBJECTS = $(IO_OBJECTS) $(IOSTREAM_OBJECTS) $(STDIO_WRAP_OBJECTS) $(OSPRIM_OBJECTS)
-LIBIOSTREAM_DEP = $(LIBIOSTREAM_OBJECTS) stdio.list
-LIBIOSTREAM_USE = $(LIBIOSTREAM_OBJECTS) `cat stdio.list`
+# LIBIOSTREAM_OBJECTS = $(IO_OBJECTS) $(IOSTREAM_OBJECTS) $(STDIO_WRAP_OBJECTS) $(OSPRIM_OBJECTS)
+# LIBIOSTREAM_DEP = $(LIBIOSTREAM_OBJECTS) stdio.list
+# LIBIOSTREAM_USE = $(LIBIOSTREAM_OBJECTS) `cat stdio.list`
 
 # Comment the above and uncomment the below to use the code in the Linux libc:
 # We have _G_config.h in /usr/include.
@@ -16,7 +16,7 @@ LIBIOSTREAM_USE = $(LIBIOSTREAM_OBJECTS) `cat stdio.list`
 
 # We have those in libc.a.
 # IO_OBJECTS=
-# STDIO_WRAP_OBJECTS=
+STDIO_WRAP_OBJECTS=
 # OSPRIM_OBJECTS=
 
 # We have the rest in /usr/include.
diff --git a/libio/libio.h b/libio/libio.h
index 84a8217..b2b6367 100644
--- a/libio/libio.h
+++ b/libio/libio.h
@@ -28,8 +28,9 @@ the executable file might be covered by the GNU General Public License. */
 #define _IO_STDIO_H
 
 #include <_G_config.h>
-#define _IO_pos_t _G_fpos_t /* obsolete */
-#define _IO_fpos_t _G_fpos_t
+#define _G_fpos_t _G_off_t
+#define _IO_pos_t _G_off_t
+#define _IO_fpos_t _G_off_t
 #define _IO_size_t _G_size_t
 #define _IO_ssize_t _G_ssize_t
 #define _IO_off_t _G_off_t
