diff --git a/libgfortran/intrinsics/eoshift0.c b/libgfortran/intrinsics/eoshift0.c
index a8a891b2..74f13dfa 100644
--- a/libgfortran/intrinsics/eoshift0.c
+++ b/libgfortran/intrinsics/eoshift0.c
@@ -90,6 +90,8 @@ eoshift0 (gfc_array_char * ret, const gfc_array_char * array,
 
   extent[0] = 1;
   count[0] = 0;
+  sstride[0] = -1;
+  rstride[0] = -1;
   n = 0;
   for (dim = 0; dim < GFC_DESCRIPTOR_RANK (array); dim++)
     {
diff --git a/libgfortran/intrinsics/eoshift2.c b/libgfortran/intrinsics/eoshift2.c
index 334d2741..c41301ec 100644
--- a/libgfortran/intrinsics/eoshift2.c
+++ b/libgfortran/intrinsics/eoshift2.c
@@ -94,6 +94,9 @@ eoshift2 (gfc_array_char *ret, const gfc_array_char *array,
 
   extent[0] = 1;
   count[0] = 0;
+  sstride[0] = -1;
+  rstride[0] = -1;
+  bstride[0] = -1;
   n = 0;
   for (dim = 0; dim < GFC_DESCRIPTOR_RANK (array); dim++)
     {
