diff --git a/gcc/cp/call.c b/gcc/cp/call.c
index 44222b8a..106f79fc 100644
--- a/gcc/cp/call.c
+++ b/gcc/cp/call.c
@@ -4386,7 +4386,7 @@ convert_like_real (conversion *convs, tree expr, tree fn, int argnum,
   switch (convs->kind)
     {
     case ck_rvalue:
-      expr = convert_bitfield_to_declared_type (expr);
+      expr = decay_conversion (expr);
       if (! IS_AGGR_TYPE (totype))
 	return expr;
       /* Else fall through.  */
diff --git a/gcc/testsuite/g++.dg/overload/rvalue1.C b/gcc/testsuite/g++.dg/overload/rvalue1.C
new file mode 100644
index 00000000..41805949
--- /dev/null
+++ b/gcc/testsuite/g++.dg/overload/rvalue1.C
@@ -0,0 +1,10 @@
+// PR c++/42655
+
+void unused(const bool &) { }
+
+int main() {
+  volatile bool x = false;
+  unused(!!x); // type of "!x" is bool
+  unused(!x); // type of "!x" is bool
+}
+
