diff --git a/gcc/objc/objc-act.c b/gcc/objc/objc-act.c
index ea2c709a..419f1e29 100644
--- a/gcc/objc/objc-act.c
+++ b/gcc/objc/objc-act.c
@@ -2336,10 +2336,17 @@ build_module_initializer_routine (void)
 #endif
 
   objc_push_parm (build_decl (PARM_DECL, NULL_TREE, void_type_node));
+#ifdef OBJCPLUS
+  objc_start_function (get_identifier (TAG_GNUINIT),
+		       build_function_type (void_type_node,
+					    OBJC_VOID_AT_END),
+		       NULL_TREE, NULL_TREE);
+#else
   objc_start_function (get_identifier (TAG_GNUINIT),
 		       build_function_type (void_type_node,
 					    OBJC_VOID_AT_END),
 		       NULL_TREE, objc_get_parm_info (0));
+#endif
 
   body = c_begin_compound_stmt (true);
   add_stmt (build_function_call
