diff --git a/gcc/config/i386/winnt-cxx.c b/gcc/config/i386/winnt-cxx.c
index 1b857427..6e509331 100755
--- a/gcc/config/i386/winnt-cxx.c
+++ b/gcc/config/i386/winnt-cxx.c
@@ -88,7 +88,7 @@ i386_pe_type_dllimport_p (tree decl)
         ignore the class attribute.  */
     else if (TREE_CODE (decl) == VAR_DECL
 	     && TREE_STATIC (decl) && TREE_PUBLIC (decl)
-	     && !DECL_EXTERNAL (decl))
+	     && DECL_NOT_REALLY_EXTERN (decl))
       {
 	if (!DECL_VIRTUAL_P (decl))
 	     error ("definition of static data member %q+D of "
diff --git a/gcc/cp/optimize.c b/gcc/cp/optimize.c
index 916e1607..479024b2 100644
--- a/gcc/cp/optimize.c
+++ b/gcc/cp/optimize.c
@@ -115,6 +115,7 @@ maybe_clone_body (tree fn)
       TREE_PUBLIC (clone) = TREE_PUBLIC (fn);
       DECL_VISIBILITY (clone) = DECL_VISIBILITY (fn);
       DECL_VISIBILITY_SPECIFIED (clone) = DECL_VISIBILITY_SPECIFIED (fn);
+      DECL_DLLIMPORT_P (clone) = DECL_DLLIMPORT_P (fn);
 
       /* Adjust the parameter names and locations.  */
       parm = DECL_ARGUMENTS (fn);
