diff --git a/gcc/fortran/trans-decl.c b/gcc/fortran/trans-decl.c
index 7662107c..5caf4047 100644
--- a/gcc/fortran/trans-decl.c
+++ b/gcc/fortran/trans-decl.c
@@ -2895,7 +2895,9 @@ generate_dependency_declarations (gfc_symbol *sym)
   int i;
 
   if (sym->ts.type == BT_CHARACTER
-	&& sym->ts.cl->length->expr_type != EXPR_CONSTANT)
+      && sym->ts.cl
+      && sym->ts.cl->length
+      && sym->ts.cl->length->expr_type != EXPR_CONSTANT)
     generate_expr_decls (sym, sym->ts.cl->length);
 
   if (sym->as && sym->as->rank)
diff --git a/gcc/testsuite/gfortran.dg/result_in_spec_3.f90 b/gcc/testsuite/gfortran.dg/result_in_spec_3.f90
new file mode 100644
index 00000000..ffdf8da4
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/result_in_spec_3.f90
@@ -0,0 +1,15 @@
+! { dg-do compile }
+! { dg-options "-std=gnu" }
+! PR fortran/34248
+!
+! There was an ICE for assumed-length functions
+! if RESULT(...) was used and no value assigned
+! to the result variable.
+!
+character(*) FUNCTION test() RESULT(ctab)
+  ctab = "Hello"
+END function test
+
+FUNCTION test2() RESULT(res)
+  character(*) :: res
+END function test2
