diff --git a/gcc/gimplify.c b/gcc/gimplify.c
index f03f2e64..9d52e6b5 100644
--- a/gcc/gimplify.c
+++ b/gcc/gimplify.c
@@ -2435,7 +2435,11 @@ gimplify_cond_expr (tree *expr_p, tree *pre_p, fallback_t fallback)
     {
       tree result;
 
-      if ((fallback & fb_lvalue) == 0)
+      /* If an rvalue is ok or we do not require an lvalue, avoid creating
+	 an addressable temporary.  */
+      if (((fallback & fb_rvalue)
+	   || !(fallback & fb_lvalue))
+	  && !TREE_ADDRESSABLE (type))
 	{
 	  result = tmp2 = tmp = create_tmp_var (TREE_TYPE (expr), "iftmp");
 	  ret = GS_ALL_DONE;
diff --git a/gcc/testsuite/gcc.c-torture/compile/pr30132.c b/gcc/testsuite/gcc.c-torture/compile/pr30132.c
new file mode 100644
index 00000000..adffae77
--- /dev/null
+++ b/gcc/testsuite/gcc.c-torture/compile/pr30132.c
@@ -0,0 +1,5 @@
+double testit(double _Complex* t)
+{
+  return *t==0.0?0.0:-*t;
+}
+
