diff --git a/gcc/cp/tree.c b/gcc/cp/tree.c
index f387cf21..079f9d33 100644
--- a/gcc/cp/tree.c
+++ b/gcc/cp/tree.c
@@ -2204,6 +2204,10 @@ decl_linkage (tree decl)
   if (!DECL_NAME (decl))
     return lk_none;
 
+  /* Fields have no linkage.  */
+  if (TREE_CODE (decl) == FIELD_DECL)
+    return lk_none;
+
   /* Things that are TREE_PUBLIC have external linkage.  */
   if (TREE_PUBLIC (decl))
     return lk_external;
diff --git a/gcc/testsuite/g++.dg/inherit/pr30297.C b/gcc/testsuite/g++.dg/inherit/pr30297.C
new file mode 100644
index 00000000..222d9ab8
--- /dev/null
+++ b/gcc/testsuite/g++.dg/inherit/pr30297.C
@@ -0,0 +1,11 @@
+// Regression test for ICE from PR c++/30297.
+
+struct A
+{
+  int i;
+};
+
+extern "C" struct B : A
+{
+  A::i;
+};
