diff --git a/gcc/testsuite/g++.dg/ext/vector12.C b/gcc/testsuite/g++.dg/ext/vector12.C
new file mode 100644
index 00000000..e0b6d2ba
--- /dev/null
+++ b/gcc/testsuite/g++.dg/ext/vector12.C
@@ -0,0 +1,4 @@
+// PR c++/34917
+// { dg-do compile }
+
+const int i __attribute ((vector_size (8))) = {};
diff --git a/gcc/tree.c b/gcc/tree.c
index 5550d2d1..aa9dd2e8 100644
--- a/gcc/tree.c
+++ b/gcc/tree.c
@@ -3428,6 +3428,8 @@ build_type_attribute_qual_variant (tree ttype, tree attribute, int quals)
       ntype = type_hash_canon (hashcode, ntype);
       ttype = build_qualified_type (ntype, quals);
     }
+  else if (TYPE_QUALS (ttype) != quals)
+    ttype = build_qualified_type (ttype, quals);
 
   return ttype;
 }
