diff --git a/gcc/testsuite/gcc.c-torture/compile/pr35431.c b/gcc/testsuite/gcc.c-torture/compile/pr35431.c
new file mode 100644
index 00000000..ffe9d068
--- /dev/null
+++ b/gcc/testsuite/gcc.c-torture/compile/pr35431.c
@@ -0,0 +1,15 @@
+void bar();
+
+void foo(int i)
+{
+  __complex__ int k = 0;
+
+  if (i)
+    k = 1;
+
+  for (i = 0; i < 1; ++i)
+    ;
+
+  if (k)
+    bar();
+}
diff --git a/gcc/tree-ssa-phiopt.c b/gcc/tree-ssa-phiopt.c
index df35d78e..7e674181 100644
--- a/gcc/tree-ssa-phiopt.c
+++ b/gcc/tree-ssa-phiopt.c
@@ -382,6 +382,11 @@ conditional_replacement (basic_block cond_bb, basic_block middle_bb,
   tree new_var = NULL;
   tree new_var1;
 
+  /* FIXME: Gimplification of complex type is too hard for now.  */
+  if (TREE_CODE (TREE_TYPE (arg0)) == COMPLEX_TYPE
+      || TREE_CODE (TREE_TYPE (arg1)) == COMPLEX_TYPE)
+    return false;
+
   /* The PHI arguments have the constants 0 and 1, then convert
      it to the conditional.  */
   if ((integer_zerop (arg0) && integer_onep (arg1))
