diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 855add98..7a2761cb 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -3314,6 +3314,9 @@ push_template_decl_real (tree decl, bool is_friend)
 
 	    if (current == decl)
 	      current = ctx;
+	    else if (current == NULL_TREE)
+	      /* Can happen in erroneous input.  */
+	      break;
 	    else
 	      current = TYPE_CONTEXT (current);
 	  }
diff --git a/gcc/testsuite/g++.dg/template/error-recovery1.C b/gcc/testsuite/g++.dg/template/error-recovery1.C
new file mode 100644
index 00000000..ef9dc141
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/error-recovery1.C
@@ -0,0 +1,9 @@
+// PR c++/43076
+
+struct S;
+template < typename > struct T
+{
+  template < typename >
+  template < bool > struct T < S > // { dg-error "" }
+  {
+    void f () {			// { dg-error "" }
