diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 51531c13..bd655a63 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -1304,7 +1304,7 @@ register_specialization (tree spec, tree tmpl, tree args, bool is_friend)
      template it is specializing.  */
   if (DECL_TEMPLATE_SPECIALIZATION (spec)
       && !check_specialization_namespace (tmpl))
-    DECL_CONTEXT (spec) = FROB_CONTEXT (decl_namespace_context (tmpl));
+    DECL_CONTEXT (spec) = DECL_CONTEXT (tmpl);
 
   if (!optimize_specialization_lookup_p (tmpl))
     DECL_TEMPLATE_SPECIALIZATIONS (tmpl)
diff --git a/gcc/testsuite/g++.dg/template/spec36.C b/gcc/testsuite/g++.dg/template/spec36.C
new file mode 100644
index 00000000..71ff3a5c
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/spec36.C
@@ -0,0 +1,16 @@
+/* PR c++/38089 */
+/* { dg-do "compile" } */
+
+struct basic_string
+{
+  basic_string(const int __s);
+};
+namespace MyNS {
+  class MyClass {
+    template <typename T>
+    T test() { } /* { dg-error "from definition" } */
+  };
+}
+template <>
+basic_string MyNS::MyClass::test() /* { dg-error "specialization of" } */
+{ return 1; }
