diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 00f465cc..7135be25 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -11853,6 +11853,11 @@ do_decl_instantiation (tree decl, tree storage)
 	 VAR_DECLs so we do the lookup here.  Probably, grokdeclarator
 	 should handle VAR_DECLs as it currently handles
 	 FUNCTION_DECLs.  */
+      if (!DECL_CLASS_SCOPE_P (decl))
+	{
+	  error ("%qD is not a static data member of a class template", decl);
+	  return;
+	}
       result = lookup_field (DECL_CONTEXT (decl), DECL_NAME (decl), 0, false);
       if (!result || TREE_CODE (result) != VAR_DECL)
 	{
diff --git a/gcc/testsuite/g++.dg/template/pr30659.C b/gcc/testsuite/g++.dg/template/pr30659.C
new file mode 100644
index 00000000..86aa1521
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/pr30659.C
@@ -0,0 +1,3 @@
+// PR c++/30659
+
+extern "C" template A<char> foo(); // { dg-error "forbids|static data|expected" }
