diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index a6094554..16e4043b 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -7627,7 +7627,10 @@ grokdeclarator (const cp_declarator *declarator,
 	    warning (0, "%qs initialized and declared %<extern%>", name);
 	}
       else
-	error ("%qs has both %<extern%> and initializer", name);
+	  {
+	    error ("%qs has both %<extern%> and initializer", name);
+	    return error_mark_node;
+	  }
     }
   else if (storage_class == sc_extern && funcdef_flag
 	   && ! toplevel_bindings_p ())
diff --git a/gcc/testsuite/g++.dg/parse/crash43.C b/gcc/testsuite/g++.dg/parse/crash43.C
new file mode 100644
index 00000000..84fe8c51
--- /dev/null
+++ b/gcc/testsuite/g++.dg/parse/crash43.C
@@ -0,0 +1,9 @@
+// PR c++/34600
+
+namespace N
+{
+  void foo()
+  {
+    extern int i = 0; // { dg-error "error: 'i' has both 'extern' and initializer" }
+  }
+}
