diff --git a/gcc/expr.c b/gcc/expr.c
index f0f654d6..39524b97 100644
--- a/gcc/expr.c
+++ b/gcc/expr.c
@@ -4796,6 +4796,9 @@ count_type_elements (tree type, bool allow_flexarr)
     case REFERENCE_TYPE:
       return 1;
 
+    case ERROR_MARK:
+      return 0;
+
     case VOID_TYPE:
     case METHOD_TYPE:
     case FUNCTION_TYPE:
diff --git a/gcc/testsuite/g++.dg/parse/struct-4.C b/gcc/testsuite/g++.dg/parse/struct-4.C
new file mode 100644
index 00000000..559411a1
--- /dev/null
+++ b/gcc/testsuite/g++.dg/parse/struct-4.C
@@ -0,0 +1,13 @@
+/* PR c/35437 */
+/* { dg-do "compile" } */
+
+struct A
+{
+  int i;
+  struct A a; /* { dg-error "has incomplete type" } */
+};
+
+void foo()
+{
+  struct A b = { 0 };
+}
diff --git a/gcc/testsuite/gcc.dg/struct-parse-2.c b/gcc/testsuite/gcc.dg/struct-parse-2.c
new file mode 100644
index 00000000..559411a1
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/struct-parse-2.c
@@ -0,0 +1,13 @@
+/* PR c/35437 */
+/* { dg-do "compile" } */
+
+struct A
+{
+  int i;
+  struct A a; /* { dg-error "has incomplete type" } */
+};
+
+void foo()
+{
+  struct A b = { 0 };
+}
