diff --git a/gcc/cp/parser.c b/gcc/cp/parser.c
index 36b13586..f69ec434 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -9680,7 +9680,6 @@ cp_parser_explicit_specialization (cp_parser* parser)
   if (!begin_specialization ())
     {
       end_specialization ();
-      cp_parser_skip_to_end_of_block_or_statement (parser);
       return;
     }
 
diff --git a/gcc/testsuite/g++.dg/template/pr39425.C b/gcc/testsuite/g++.dg/template/pr39425.C
new file mode 100644
index 00000000..a063e05c
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/pr39425.C
@@ -0,0 +1,18 @@
+// PR c++/39425
+// { dg-do compile }
+
+class a {
+
+  template<unsigned int s>
+    struct _rec {
+      static const char size = _rec< (s >> 1) >::size;
+    };
+
+  template<>	// { dg-error "explicit" }
+  struct _rec <0> {
+    static const char size = 0;
+  };
+
+  static const unsigned int value = _rec < 1 >::size;
+
+}		// { dg-error "unqualified-id" }
diff --git a/gcc/testsuite/g++.dg/template/spec33.C b/gcc/testsuite/g++.dg/template/spec33.C
index 809d4f01..7b7a7519 100644
--- a/gcc/testsuite/g++.dg/template/spec33.C
+++ b/gcc/testsuite/g++.dg/template/spec33.C
@@ -3,5 +3,5 @@
 struct A
 {
     template<int> static void foo   () {}
-    template<>    static void foo<0>() {}  // { dg-error "explicit" }
+    template<>    static void foo<0>() {}  // { dg-error "explicit|template" }
 }; 
