diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 5f515fe5..c9e2e388 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -7870,7 +7870,8 @@ grokdeclarator (const cp_declarator *declarator,
 	  type_quals = TYPE_UNQUALIFIED;
 
 	  if (declarator->kind == cdk_ptrmem
-	      && (TREE_CODE (type) == FUNCTION_TYPE || memfn_quals))
+	      && (TREE_CODE (type) == FUNCTION_TYPE
+		  || (memfn_quals && TREE_CODE (type) == METHOD_TYPE)))
 	    {
 	      memfn_quals |= cp_type_quals (type);
 	      type = build_memfn_type (type,
diff --git a/gcc/testsuite/g++.dg/other/ptrmem9.C b/gcc/testsuite/g++.dg/other/ptrmem9.C
new file mode 100644
index 00000000..c4d3656a
--- /dev/null
+++ b/gcc/testsuite/g++.dg/other/ptrmem9.C
@@ -0,0 +1,5 @@
+// PR c++/33969
+// { dg-do compile }
+
+struct A;
+void (*A::* fp)() const; // { dg-error "invalid in variable declaration" }
