diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 4feb185b..1f028378 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -2629,9 +2629,8 @@ current_template_args (void)
 		    t = TREE_TYPE (t);
 		  else
 		    t = DECL_INITIAL (t);
+		  TREE_VEC_ELT (a, i) = t;
 		}
-
-	      TREE_VEC_ELT (a, i) = t;
 	    }
 	}
 
diff --git a/gcc/testsuite/g++.dg/template/crash55.C b/gcc/testsuite/g++.dg/template/crash55.C
index 377603dc..7cf9f1ea 100644
--- a/gcc/testsuite/g++.dg/template/crash55.C
+++ b/gcc/testsuite/g++.dg/template/crash55.C
@@ -1,6 +1,6 @@
 //PR c++/27668
 
 template<typename class T, T = T()> // { dg-error "nested-name-specifier|two or more|valid type" }
-struct A {};                        // { dg-error "definition|template" }
+struct A {};
 
 template<int> void foo(A<int>);     // { dg-error "mismatch|constant|template argument" }
diff --git a/gcc/testsuite/g++.dg/template/void10.C b/gcc/testsuite/g++.dg/template/void10.C
index 56e0b6d2..4904a281 100644
--- a/gcc/testsuite/g++.dg/template/void10.C
+++ b/gcc/testsuite/g++.dg/template/void10.C
@@ -7,4 +7,4 @@ template<void> struct A                 // { dg-error "not a valid type" }
 
 template<typename> struct B {};
 
-B<int> b;                              // { dg-error "template argument|invalid type" }
+B<int> b;
diff --git a/gcc/testsuite/g++.dg/template/void11.C b/gcc/testsuite/g++.dg/template/void11.C
new file mode 100644
index 00000000..1aad7a15
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/void11.C
@@ -0,0 +1,12 @@
+// PR c++/31446
+
+template<void> struct A // { dg-error "valid type" }
+
+{
+  template<int> friend void foo();
+};
+
+void bar()
+{
+  foo<0>(); // { dg-error "not declared|primary-expression" }
+}
diff --git a/gcc/testsuite/g++.dg/template/void2.C b/gcc/testsuite/g++.dg/template/void2.C
index 05a8186e..eceb3621 100644
--- a/gcc/testsuite/g++.dg/template/void2.C
+++ b/gcc/testsuite/g++.dg/template/void2.C
@@ -3,7 +3,7 @@
 
 template<int> struct A
 {
-  template<void> friend class X;  // { dg-error "void" }
+  template<void> friend class X;  // { dg-error "void|valid type" }
 };
 
-A<0> a; 
+A<0> a;
