diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 1f028378..1ff51577 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -5686,7 +5686,7 @@ tsubst_friend_function (tree decl, tree args)
 	  tree fn = check_classfn (context,
 				   new_friend, NULL_TREE);
 
-	  if (fn && fn != error_mark_node)
+	  if (fn)
 	    new_friend = fn;
 	}
     }
diff --git a/gcc/testsuite/g++.dg/template/crash69.C b/gcc/testsuite/g++.dg/template/crash69.C
new file mode 100644
index 00000000..957f1e19
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/crash69.C
@@ -0,0 +1,12 @@
+// PR c++/31132
+
+template<typename T> class A
+{
+  static int i; // { dg-error "is private" }
+  friend int T::foo(); // { dg-error "does not match" }
+};
+
+struct B
+{
+  void foo() { A<B>::i; } // { dg-error "within|candidate" }
+};
