diff --git a/gcc/cp/cp-gimplify.c b/gcc/cp/cp-gimplify.c
index fb6e3061..2f10d241 100644
--- a/gcc/cp/cp-gimplify.c
+++ b/gcc/cp/cp-gimplify.c
@@ -740,6 +740,23 @@ cp_genericize (tree fndecl)
       DECL_BY_REFERENCE (t) = 1;
       TREE_ADDRESSABLE (t) = 0;
       relayout_decl (t);
+      if (DECL_NAME (t))
+	{
+	  /* Adjust DECL_VALUE_EXPR of the original var.  */
+	  tree outer = outer_curly_brace_block (current_function_decl);
+	  tree var;
+
+	  if (outer)
+	    for (var = BLOCK_VARS (outer); var; var = TREE_CHAIN (var))
+	      if (DECL_NAME (t) == DECL_NAME (var)
+		  && DECL_HAS_VALUE_EXPR_P (var)
+		  && DECL_VALUE_EXPR (var) == t)
+		{
+		  tree val = convert_from_reference (t);
+		  SET_DECL_VALUE_EXPR (var, val);
+		  break;
+		}
+	}
     }
 
   /* If we're a clone, the body is already GIMPLE.  */
diff --git a/gcc/cp/cp-tree.h b/gcc/cp/cp-tree.h
index 36eb4532..a45b790e 100644
--- a/gcc/cp/cp-tree.h
+++ b/gcc/cp/cp-tree.h
@@ -3938,6 +3938,7 @@ extern void start_preparsed_function		(tree, tree, int);
 extern int start_function			(cp_decl_specifier_seq *, const cp_declarator *, tree);
 extern tree begin_function_body			(void);
 extern void finish_function_body		(tree);
+extern tree outer_curly_brace_block		(tree);
 extern tree finish_function			(int);
 extern tree start_method			(cp_decl_specifier_seq *, const cp_declarator *, tree);
 extern tree finish_method			(tree);
diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 0cb54673..009764ed 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -11251,7 +11251,7 @@ finish_function_body (tree compstmt)
    of curly braces, skipping the artificial block created for constructor
    initializers.  */
 
-static tree
+tree
 outer_curly_brace_block (tree fndecl)
 {
   tree block = BLOCK_SUBBLOCKS (DECL_INITIAL (fndecl));
diff --git a/gcc/testsuite/g++.dg/guality/pr46815.C b/gcc/testsuite/g++.dg/guality/pr46815.C
new file mode 100644
index 00000000..41b92199
--- /dev/null
+++ b/gcc/testsuite/g++.dg/guality/pr46815.C
@@ -0,0 +1,25 @@
+// PR debug/46815
+// { dg-do run }
+// { dg-options "-g" }
+// { dg-skip-if "" { *-*-* }  { "*" } { "-O0" } }
+
+struct S
+{
+  int i;
+  S () { i = 42; }
+  virtual void foo (void) {}
+};
+
+S
+bar ()
+{
+  S s;
+  return s;	// { dg-final { gdb-test 17 "s.i" "42" } }
+}
+
+int
+main ()
+{
+  S s = bar ();
+  return s.i - 42;
+}
