diff --git a/gcc/combine.c b/gcc/combine.c
index 14bac70e..1e91035f 100644
--- a/gcc/combine.c
+++ b/gcc/combine.c
@@ -6117,7 +6117,8 @@ make_extraction (enum machine_mode mode, rtx inner, HOST_WIDE_INT pos,
 	return new;
 
       if (GET_CODE (new) == CONST_INT)
-	return gen_int_mode (INTVAL (new), mode);
+	return simplify_unary_operation (unsignedp ? ZERO_EXTEND : SIGN_EXTEND,
+					 mode, new, tmode);
 
       /* If we know that no extraneous bits are set, and that the high
 	 bit is not set, convert the extraction to the cheaper of
diff --git a/gcc/testsuite/gcc.c-torture/execute/pr43438.c b/gcc/testsuite/gcc.c-torture/execute/pr43438.c
new file mode 100644
index 00000000..aa901ce0
--- /dev/null
+++ b/gcc/testsuite/gcc.c-torture/execute/pr43438.c
@@ -0,0 +1,23 @@
+extern void abort (void);
+
+static unsigned char g_2 = 1;
+static int g_9;
+static int *l_8 = &g_9;
+
+static void func_12(int p_13)
+{
+  int * l_17 = &g_9;
+  *l_17 &= 0 < p_13;
+}
+
+int main(void)
+{
+  unsigned char l_11 = 254;
+  *l_8 |= g_2;
+  l_11 |= *l_8;
+  func_12(l_11);
+  if (g_9 != 1)
+    abort ();
+  return 0;
+} 
+
