diff --git a/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-5.cc b/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-5.cc
index 244bae02..50dd0925 100644
--- a/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-5.cc
+++ b/libstdc++-v3/testsuite/ext/mt_allocator/deallocate_local_thread-5.cc
@@ -1,4 +1,5 @@
 // { dg-require-cxa-atexit "" }
+// { dg-options "-static-libstdc++ -static-libgcc" }
 
 // Copyright (C) 2004, 2005 Free Software Foundation, Inc.
 //
