diff --git a/libstdc++-v3/acinclude.m4 b/libstdc++-v3/acinclude.m4
index b380a26..924ab4c 100644
--- a/libstdc++-v3/acinclude.m4
+++ b/libstdc++-v3/acinclude.m4
@@ -895,11 +895,8 @@ AC_DEFUN([GLIBCXX_ENABLE_C99], [
     AC_LANG_SAVE
     AC_LANG_CPLUSPLUS
 
-    # Use -std=c++98 (instead of -std=gnu++98) because leaving __STRICT_ANSI__
-    # undefined may cause fake C99 facilities, like pre-standard snprintf,
-    # to be spuriously enabled.
     ac_save_CXXFLAGS="$CXXFLAGS"
-    CXXFLAGS="$CXXFLAGS -std=c++98"
+    CXXFLAGS="$CXXFLAGS"
     ac_save_LIBS="$LIBS"
     ac_save_gcc_no_link="$gcc_no_link"
 
@@ -1649,10 +1646,8 @@ AC_DEFUN([GLIBCXX_CHECK_C99_TR1], [
   AC_LANG_SAVE
   AC_LANG_CPLUSPLUS
 
-  # Use -std=c++98 because the default (-std=gnu++98) leaves __STRICT_ANSI__
-  # undefined and fake C99 facilities may be spuriously enabled.
   ac_save_CXXFLAGS="$CXXFLAGS"
-  CXXFLAGS="$CXXFLAGS -std=c++98"
+  CXXFLAGS="$CXXFLAGS"
 
   # Check for the existence of <complex.h> complex math functions used
   # by tr1/complex.
diff --git a/libstdc++-v3/configure b/libstdc++-v3/configure
index 61457e9..bd8ec9d 100755
--- a/libstdc++-v3/configure
+++ b/libstdc++-v3/configure
@@ -16765,12 +16765,8 @@ ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
 ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
 ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
 
-
-    # Use -std=c++98 (instead of -std=gnu++98) because leaving __STRICT_ANSI__
-    # undefined may cause fake C99 facilities, like pre-standard snprintf,
-    # to be spuriously enabled.
     ac_save_CXXFLAGS="$CXXFLAGS"
-    CXXFLAGS="$CXXFLAGS -std=c++98"
+    CXXFLAGS="$CXXFLAGS"
     ac_save_LIBS="$LIBS"
     ac_save_gcc_no_link="$gcc_no_link"
 
@@ -19936,11 +19932,8 @@ ac_compile='$CXX -c $CXXFLAGS $CPPFLAGS conftest.$ac_ext >&5'
 ac_link='$CXX -o conftest$ac_exeext $CXXFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
 ac_compiler_gnu=$ac_cv_cxx_compiler_gnu
 
-
-  # Use -std=c++98 because the default (-std=gnu++98) leaves __STRICT_ANSI__
-  # undefined and fake C99 facilities may be spuriously enabled.
   ac_save_CXXFLAGS="$CXXFLAGS"
-  CXXFLAGS="$CXXFLAGS -std=c++98"
+  CXXFLAGS="$CXXFLAGS"
 
   # Check for the existence of <complex.h> complex math functions used
   # by tr1/complex.
