diff --git a/libstdc++/new b/libstdc++/new
index 5563e3e..753fa20 100644
--- a/libstdc++/new
+++ b/libstdc++/new
@@ -26,7 +26,7 @@ void operator delete (void *);
 void operator delete[] (void *);
 
 // default placement versions of operator new
-inline void *operator new(size_t, void *place) { return place; }
-inline void *operator new[](size_t, void *place) { return place; }
+static void *operator new(size_t, void *place) { return place; }
+static void *operator new[](size_t, void *place) { return place; }
 
 #endif
