diff --git a/gcc/cp/class.c b/gcc/cp/class.c
index 398d341b..b1ab4e18 100644
--- a/gcc/cp/class.c
+++ b/gcc/cp/class.c
@@ -4129,6 +4129,10 @@ type_requires_array_cookie (tree type)
       second_parm = TREE_CHAIN (TYPE_ARG_TYPES (TREE_TYPE (fn)));
       if (second_parm == void_list_node)
 	return false;
+      /* Do not consider this function if its second argument is an
+	 ellipsis.  */
+      if (!second_parm)
+	continue;
       /* Otherwise, if we have a two-argument function and the second
 	 argument is `size_t', it will be the usual deallocation
 	 function -- unless there is one-argument function, too.  */
diff --git a/gcc/testsuite/g++.dg/other/dtor2.C b/gcc/testsuite/g++.dg/other/dtor2.C
new file mode 100644
index 00000000..45933912
--- /dev/null
+++ b/gcc/testsuite/g++.dg/other/dtor2.C
@@ -0,0 +1,7 @@
+/* PR c++/35317 */
+/* { dg-do "compile" } */
+
+struct A
+{
+  void operator delete[] (void*, ...);
+};
