diff --git a/gcc/testsuite/g++.dg/warn/Wunused-12.C b/gcc/testsuite/g++.dg/warn/Wunused-12.C
new file mode 100644
index 00000000..44c6bd85
--- /dev/null
+++ b/gcc/testsuite/g++.dg/warn/Wunused-12.C
@@ -0,0 +1,9 @@
+// APPLE LOCAL file mainline 2005-10-10
+// PR c++/24302
+// { dg-options "-Wunused" }
+
+static union
+{
+  unsigned char FLT4ARR[4]; 
+  float FLT4;
+}; // { dg-warning "used" }
diff --git a/gcc/toplev.c b/gcc/toplev.c
index 4e10fd6c..e5f63f75 100644
--- a/gcc/toplev.c
+++ b/gcc/toplev.c
@@ -1710,7 +1710,7 @@ check_global_declarations (tree *vec, int len)
 	  && ! TREE_USED (decl)
 	  /* The TREE_USED bit for file-scope decls is kept in the identifier,
 	     to handle multiple external decls in different scopes.  */
-	  && ! TREE_USED (DECL_NAME (decl))
+	  && ! (DECL_NAME (decl) && TREE_USED (DECL_NAME (decl)))
 	  && ! DECL_EXTERNAL (decl)
 	  && ! TREE_PUBLIC (decl)
 	  /* A volatile variable might be used in some non-obvious way.  */
