diff --git a/gcc/cp/typeck.c b/gcc/cp/typeck.c
index 5ca4189a..4bca4898 100644
--- a/gcc/cp/typeck.c
+++ b/gcc/cp/typeck.c
@@ -3758,7 +3758,7 @@ build_unary_op (enum tree_code code, tree xarg, int noconvert)
       else if (!(arg = build_expr_type_conversion (WANT_INT | WANT_ENUM,
 						   arg, true)))
 	errstring = "wrong type argument to bit-complement";
-      else if (!noconvert)
+      else if (!noconvert && CP_INTEGRAL_TYPE_P (TREE_TYPE (arg)))
 	arg = perform_integral_promotions (arg);
       break;
 
diff --git a/gcc/testsuite/g++.dg/ext/vector3.C b/gcc/testsuite/g++.dg/ext/vector3.C
new file mode 100644
index 00000000..1a67cf76
--- /dev/null
+++ b/gcc/testsuite/g++.dg/ext/vector3.C
@@ -0,0 +1,12 @@
+/* { dg-do compile } */
+/* { dg-options "" } */
+
+// PR c++/28302
+
+int __attribute__((vector_size(8))) x;
+
+void foo()
+{
+  ~x;
+}
+
