diff --git a/fold-const.c b/fold-const.c
index 3feee8a..18781e7 100644
--- a/fold-const.c
+++ b/fold-const.c
@@ -2856,9 +2856,9 @@ make_range (exp, pin_p, plow, phigh)
 			       convert (type, integer_zero_node), 0, high, 1);
 	  n_high = range_binop (MINUS_EXPR, type,
 				convert (type, integer_zero_node), 0, low, 0);
-	  low = n_low, high = n_high;
-	  exp = arg0;
-	  continue;
+	  if (n_high != 0 && TREE_OVERFLOW (n_high))
+	    break;
+	  goto normalize;
 
 	case BIT_NOT_EXPR:
 	  /* ~ X -> -X - 1  */
@@ -2882,6 +2882,7 @@ make_range (exp, pin_p, plow, phigh)
 	      || (n_high != 0 && TREE_OVERFLOW (n_high)))
 	    break;
 
+	normalize:
 	  /* Check for an unsigned range which has wrapped around the maximum
 	     value thus making n_high < n_low, and normalize it.  */
 	  if (n_low && n_high && tree_int_cst_lt (n_high, n_low))
