diff --git a/gcc/opts.c b/gcc/opts.c
index f01771025..330b09555 100644
--- a/gcc/opts.c
+++ b/gcc/opts.c
@@ -1749,6 +1749,16 @@ common_handle_option (struct gcc_options *opts,
       opts->x_warn_maybe_uninitialized = value;
       break;
 
+    case OPT_fwrapv:
+      if (value)
+	opts->x_flag_trapv = 0;
+      break;
+
+    case OPT_ftrapv:
+      if (value)
+	opts->x_flag_wrapv = 0;
+      break;
+
     default:
       /* If the flag was handled in a standard way, assume the lack of
 	 processing here is intentional.  */
