diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index a651d0e71..220b08f0f 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -16010,6 +16010,8 @@ unify (tree tparms, tree targs, tree parm, tree arg, int strict,
 	  if (!BRACE_ENCLOSED_INITIALIZER_P (elt))
 	    {
 	      tree type = TREE_TYPE (elt);
+	      if (type == error_mark_node)
+		return unify_invalid (explain_p);
 	      /* It should only be possible to get here for a call.  */
 	      gcc_assert (elt_strict & UNIFY_ALLOW_OUTER_LEVEL);
 	      elt_strict |= maybe_adjust_types_for_deduction
diff --git a/gcc/testsuite/g++.dg/cpp0x/auto44.C b/gcc/testsuite/g++.dg/cpp0x/auto44.C
new file mode 100644
index 000000000..687f154a0
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/auto44.C
@@ -0,0 +1,10 @@
+// PR c++/58614
+// { dg-do compile { target c++11 } }
+
+#include <initializer_list>
+
+void foo()
+{
+  i;  // { dg-error "not declared" }
+  auto j = { i };  // { dg-error "unable to deduce" }
+}
