diff --git a/gcc/cp/typeck2.c b/gcc/cp/typeck2.c
index 8bc2e59eb..f169a5a7e 100644
--- a/gcc/cp/typeck2.c
+++ b/gcc/cp/typeck2.c
@@ -1661,7 +1661,7 @@ build_functional_cast (tree exp, tree parms, tsubst_flags_t complain)
   tree type;
   VEC(tree,gc) *parmvec;
 
-  if (exp == error_mark_node || parms == error_mark_node)
+  if (error_operand_p (exp) || parms == error_mark_node)
     return error_mark_node;
 
   if (TREE_CODE (exp) == TYPE_DECL)
diff --git a/gcc/testsuite/g++.dg/cpp0x/auto39.C b/gcc/testsuite/g++.dg/cpp0x/auto39.C
new file mode 100644
index 000000000..dfa1fb4e9
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/auto39.C
@@ -0,0 +1,6 @@
+// PR c++/58560
+// { dg-do compile { target c++11 } }
+
+typedef auto T;     // { dg-error "typedef declared 'auto'" }
+
+void foo() { T(); }
