diff --git a/gcc/cp/friend.c b/gcc/cp/friend.c
index 86c0c10bb..90426af17 100644
--- a/gcc/cp/friend.c
+++ b/gcc/cp/friend.c
@@ -240,6 +240,9 @@ make_friend_class (tree type, tree friend_type, bool complain)
 
   friend_type = cv_unqualified (friend_type);
 
+  if (check_for_bare_parameter_packs (friend_type))
+    return;
+
   if (friend_depth)
     /* If the TYPE is a template then it makes sense for it to be
        friends with itself; this means that each instantiation is
diff --git a/gcc/testsuite/g++.dg/cpp0x/variadic127.C b/gcc/testsuite/g++.dg/cpp0x/variadic127.C
new file mode 100644
index 000000000..2e0d593ac
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/variadic127.C
@@ -0,0 +1,15 @@
+// PR c++/51148
+// { dg-do compile { target c++11 } }
+
+template<typename... Types>
+struct S
+{};
+
+template<typename... Types>
+struct T
+{
+  friend class S<Types>;     // { dg-error "parameter packs not expanded" }
+};
+
+int main()
+{}
