diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index dc5a72954..be58dc98d 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -9892,19 +9892,6 @@ grokdeclarator (const cp_declarator *declarator,
 
       if (decl_context != TYPENAME)
 	{
-	  /* A cv-qualifier-seq shall only be part of the function type
-	     for a non-static member function. [8.3.5/4 dcl.fct] */
-	  if (type_memfn_quals (type) != TYPE_UNQUALIFIED
-	      && (current_class_type == NULL_TREE || staticp) )
-	    {
-	      error (staticp
-                     ? G_("qualified function types cannot be used to "
-                          "declare static member functions")
-                     : G_("qualified function types cannot be used to "
-                          "declare free functions"));
-	      type = TYPE_MAIN_VARIANT (type);
-	    }
-
 	  /* The qualifiers on the function type become the qualifiers on
 	     the non-static member function. */
 	  memfn_quals |= type_memfn_quals (type);
diff --git a/gcc/testsuite/g++.dg/other/cv_func.C b/gcc/testsuite/g++.dg/other/cv_func.C
index 941cb8d53..2eb9f039a 100644
--- a/gcc/testsuite/g++.dg/other/cv_func.C
+++ b/gcc/testsuite/g++.dg/other/cv_func.C
@@ -3,7 +3,7 @@
 typedef int FIC(int) const;
 typedef int FI(int);
 
-FIC f; // { dg-error "qualified" }
+FIC f; // { dg-error "cv-qualifier" }
 struct S {
   FIC f; // OK
 
diff --git a/gcc/testsuite/g++.dg/other/cv_func3.C b/gcc/testsuite/g++.dg/other/cv_func3.C
new file mode 100644
index 000000000..0fd47882f
--- /dev/null
+++ b/gcc/testsuite/g++.dg/other/cv_func3.C
@@ -0,0 +1,10 @@
+// PR c++/58810
+
+typedef int F() const;
+
+F f;           // { dg-error "cv-qualifier" }
+
+struct A
+{
+  friend F f;  // { dg-error "cv-qualifier" }
+};
diff --git a/gcc/testsuite/g++.dg/parse/fn-typedef2.C b/gcc/testsuite/g++.dg/parse/fn-typedef2.C
index c9c7f060d..6bb83024b 100644
--- a/gcc/testsuite/g++.dg/parse/fn-typedef2.C
+++ b/gcc/testsuite/g++.dg/parse/fn-typedef2.C
@@ -4,4 +4,4 @@ typedef void ft() const;
 typedef void V;
 typedef V ft() const;
 
-ft f;				// { dg-error "qualified" }
+ft f;				// { dg-error "cv-qualifier" }
