diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index be58dc98d..0441c659b 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -5591,11 +5591,8 @@ check_initializer (tree decl, tree init, int flags, VEC(tree,gc) **cleanups)
 	      /* Declared constexpr, but no suitable initializer; massage
 		 init appropriately so we can pass it into store_init_value
 		 for the error.  */
-	      if (init && BRACE_ENCLOSED_INITIALIZER_P (init))
-		init = finish_compound_literal (type, init,
-						tf_warning_or_error);
-	      else if (CLASS_TYPE_P (type)
-		       && (!init || TREE_CODE (init) == TREE_LIST))
+	      if (CLASS_TYPE_P (type)
+		  && (!init || TREE_CODE (init) == TREE_LIST))
 		{
 		  init = build_functional_cast (type, init, tf_none);
 		  if (init != error_mark_node)
diff --git a/gcc/testsuite/g++.dg/cpp0x/constexpr-array6.C b/gcc/testsuite/g++.dg/cpp0x/constexpr-array6.C
new file mode 100644
index 000000000..16eacdde4
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/constexpr-array6.C
@@ -0,0 +1,15 @@
+// PR c++/58611
+// { dg-do compile { target c++11 } }
+
+struct A
+{
+  int i;
+  constexpr A() {}		// { dg-error "A::i" }
+};
+
+struct B
+{
+  A a;
+};
+
+constexpr B b[] = { {} };	// { dg-error "A::A" }
