diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index ff1b39056..2d5f1cc1a 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -5483,7 +5483,7 @@ convert_nontype_argument (tree type, tree expr, tsubst_flags_t complain)
 		       TREE_TYPE (TREE_TYPE (addr))));
 
 	  expr = TREE_OPERAND (addr, 0);
-	  expr_type = TREE_TYPE (expr);
+      expr_type = TREE_TYPE (probe_type);
 	}
     }
 
diff --git a/gcc/testsuite/g++.dg/template/nontype26.C b/gcc/testsuite/g++.dg/template/nontype26.C
new file mode 100644
index 000000000..763d987dd
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/nontype26.C
@@ -0,0 +1,20 @@
+// PR c++/31671
+
+template<int& i> void doit() {
+  i = 0;
+}
+
+template<const int& i> class X {
+public:
+    void foo() {
+      doit<i>();  // { dg-error "cv-qualification|no matching" }
+    }
+};
+
+int i = 0;
+
+X<i> x;
+
+int main() {
+  x.foo();
+}
