diff --git a/gcc/cp/name-lookup.c b/gcc/cp/name-lookup.c
index 89b69a690..af1cf34f4 100644
--- a/gcc/cp/name-lookup.c
+++ b/gcc/cp/name-lookup.c
@@ -965,7 +965,8 @@ pushdecl_maybe_friend_1 (tree x, bool is_friend)
 	      && (DECL_EXTERNAL (decl) || TREE_PUBLIC (decl))
 	      /* If different sort of thing, we already gave an error.  */
 	      && TREE_CODE (decl) == TREE_CODE (x)
-	      && !same_type_p (TREE_TYPE (x), TREE_TYPE (decl)))
+	      && !comptypes (TREE_TYPE (x), TREE_TYPE (decl),
+			     COMPARE_REDECLARATION))
 	    {
 	      permerror (input_location, "type mismatch with previous external decl of %q#D", x);
 	      permerror (input_location, "previous external decl of %q+#D", decl);
diff --git a/gcc/testsuite/g++.dg/other/redecl3.C b/gcc/testsuite/g++.dg/other/redecl3.C
new file mode 100644
index 000000000..6d35bf53e
--- /dev/null
+++ b/gcc/testsuite/g++.dg/other/redecl3.C
@@ -0,0 +1,9 @@
+// PR c++/22556
+
+extern int foo[]; // OK
+int foo[] = {1,2,3};
+extern int foo[]; // OK
+
+void bar(){
+  extern int foo[]; // g++: ERROR -- SHOULD BE OK
+}
