diff --git a/include/asm-i386/types.h b/include/asm-i386/types.h
index ec3df86..dbcd872 100644
--- a/include/asm-i386/types.h
+++ b/include/asm-i386/types.h
@@ -98,6 +98,6 @@ typedef unsigned long long u64;
 		__asm__ __volatile__("cld ; rep ; stosl" \
 			:"=m" (*(fd_set *) (fdsetp)) \
 			:"a" (0), "c" (__FDSET_INTS), \
-			"D" ((fd_set *) (fdsetp)) :"cx","di")
+			"D" ((fd_set *) (fdsetp)))
 
 #endif
diff --git a/include/gnu/types.h b/include/gnu/types.h
index b708710..7f62122 100644
--- a/include/gnu/types.h
+++ b/include/gnu/types.h
@@ -104,7 +104,7 @@ typedef struct __fd_set {
 	__asm__ __volatile__("cld ; rep ; stosl" \
         	:"=m" (*(__fd_set *) (fdsetp)) \
         	:"a" (0), "c" (__FDSET_LONGS), \
-		"D" ((__fd_set *) (fdsetp)) :"cx","di")
+		"D" ((__fd_set *) (fdsetp)))
 
 #endif /* __NFDBITS */
 
diff --git a/include/string.h b/include/string.h
index 2ef8e4e..b2c5797 100644
--- a/include/string.h
+++ b/include/string.h
@@ -32,12 +32,9 @@ __BEGIN_DECLS
 #define	__need_NULL
 #include <stddef.h>
 
-/* `memcpy' is a built-in function for gcc 2.x. */
-#if defined(__STDC__) && __GNUC__ < 2
 /* Copy N bytes of SRC to DEST.  */
 extern __ptr_t memcpy __P ((__ptr_t __dest, __const __ptr_t __src,
 			    size_t __n));
-#endif
 
 /* Copy N bytes of SRC to DEST, guaranteeing
    correct behavior for overlapping strings.  */
@@ -61,12 +58,9 @@ extern __ptr_t memccpy __P ((__ptr_t __dest, __const __ptr_t __src,
 /* Set N bytes of S to C.  */
 extern __ptr_t memset __P ((__ptr_t __s, int __c, size_t __n));
 
-/* `memcmp' is a built-in function for gcc 2.x. */
-#if defined(__STDC__) && __GNUC__ < 2
 /* Compare N bytes of S1 and S2.  */
 extern int memcmp __P ((__const __ptr_t __s1, __const __ptr_t __s2,
 			size_t __n));
-#endif
 
 /* Search N bytes of S for C.  */
 extern __ptr_t memchr __P ((__const __ptr_t __s, int __c, size_t __n));
