diff --git a/usr/include/gnu/types.h b/usr/include/gnu/types.h
index b708710..ad8dd44 100644
--- a/usr/include/gnu/types.h
+++ b/usr/include/gnu/types.h
@@ -101,10 +101,14 @@ typedef struct __fd_set {
 	__result; })
 
 #define __FD_ZERO(fdsetp) \
+do { \
+	int __d0, __d1; \
 	__asm__ __volatile__("cld ; rep ; stosl" \
-        	:"=m" (*(__fd_set *) (fdsetp)) \
-        	:"a" (0), "c" (__FDSET_LONGS), \
-		"D" ((__fd_set *) (fdsetp)) :"cx","di")
+			:"=m" (*(__kernel_fd_set *) (fdsetp)), \
+			  "=&c" (__d0), "=&D" (__d1) \
+			:"a" (0), "1" (__FDSET_LONGS), \
+			"2" ((__kernel_fd_set *) (fdsetp)) : "memory"); \
+} while (0)
 
 #endif /* __NFDBITS */
 
diff --git a/usr/include/paths.h b/usr/include/paths.h
index c14e02a..9ed1ad3 100644
--- a/usr/include/paths.h
+++ b/usr/include/paths.h
@@ -36,15 +36,11 @@
 #ifndef _PATHS_H_
 #define	_PATHS_H_
 
-#if 0
-#define	__PATH_ETC_INET	"/usr/etc/inet"
-#else
 #define	__PATH_ETC_INET	"/etc"
-#endif
 
 /* Default search path. */
-#define	_PATH_DEFPATH		"/usr/local/bin:/usr/bin:/bin:."
-#define _PATH_DEFPATH_ROOT	"/sbin:/bin:/usr/sbin:/usr/bin"
+#define	_PATH_DEFPATH	     "/usr/local/bin:/usr/bin:/bin:."
+#define _PATH_DEFPATH_ROOT   "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
 
 #define	_PATH_BSHELL	"/bin/sh"
 #define	_PATH_CONSOLE	"/dev/console"
@@ -62,7 +58,7 @@
 #define	_PATH_SENDMAIL	"/usr/sbin/sendmail"
 #define	_PATH_SHELLS	"/etc/shells"
 #define	_PATH_TTY	"/dev/tty"
-#define	_PATH_UNIX	"/vmlinux"
+#define	_PATH_UNIX	"/vmlinuz"
 #define	_PATH_VI	"/usr/bin/vi"
 
 /* Provide trailing slash, since mostly used for building pathnames. */
@@ -73,21 +69,13 @@
 
 #define _PATH_KLOG	"/proc/kmsg"
 #define _PATH_LOGCONF	__PATH_ETC_INET"/syslog.conf"
-#if 0
-#define _PATH_LOGPID	__PATH_ETC_INET"/syslog.pid"
-#else
 #define _PATH_LOGPID	"/var/run/syslog.pid"
-#endif
 #define _PATH_LOG	"/dev/log"
 #define _PATH_CONSOLE	"/dev/console"
 
-#if 0
-#define _PATH_UTMP	"/etc/utmp"
-#else
-#define _PATH_UTMP	"/var/adm/utmp"
-#endif
-#define _PATH_WTMP	"/var/adm/wtmp"
-#define _PATH_LASTLOG	"/var/adm/lastlog"
+#define _PATH_UTMP	"/var/run/utmp"
+#define _PATH_WTMP	"/var/log/wtmp"
+#define _PATH_LASTLOG	"/var/log/lastlog"
 
 #define _PATH_LOCALE	"/usr/lib/locale"
 
diff --git a/usr/include/stdlib.h b/usr/include/stdlib.h
index 0cd4acf..6900899 100644
--- a/usr/include/stdlib.h
+++ b/usr/include/stdlib.h
@@ -97,7 +97,7 @@ extern unsigned long int strtoul __P((__const char *__nptr,
 					 char **__endptr, int __base));
 
 #ifdef	__OPTIMIZE__
-#if defined(__cplusplus) || __GNUC__ >= 2
+#if defined(__cplusplus) || (__GNUC__ >= 2 && __STDC__)
 extern __inline double atof (__const char *__nptr)
 	{ return strtod(__nptr, (char **) NULL); }
 extern __inline int atoi (__const char *__nptr)
@@ -144,7 +144,7 @@ extern void * initstate __P((unsigned int __seed, void * __statebuf,
 extern void * setstate __P((void * __statebuf));
 
 #ifdef	__OPTIMIZE__
-#if defined(__cplusplus) || __GNUC__ >= 2
+#if defined(__cplusplus) || (__GNUC__ >= 2 && __STDC__)
 extern __inline long int random()
 	{ return __random(); }
 extern __inline void srandom(unsigned int __seed)
@@ -164,7 +164,7 @@ extern __inline __ptr_t setstate (__ptr_t __statebuf)
 #endif	/* Use BSD.  */
 
 #ifdef	__OPTIMIZE__
-#if defined(__cplusplus) || __GNUC__ >= 2
+#if defined(__cplusplus) || (__GNUC__ >= 2 && __STDC__)
 extern __inline int rand ()
 	{ return (int) __random(); }
 extern __inline void srand(unsigned int __seed)
@@ -191,7 +191,7 @@ extern void free __P((void * __ptr));
 extern void cfree __P((void * __ptr));
 
 #ifdef	__OPTIMIZE__
-#if defined(__cplusplus) || __GNUC__ >= 2
+#if defined(__cplusplus) || (__GNUC__ >= 2 && __STDC__)
 extern __inline void cfree (__ptr_t __ptr)
 	{ free (__ptr); }
 #else
@@ -320,7 +320,7 @@ extern int mbtowc __P((wchar_t *__pwc, __const char *__s, size_t __n));
 extern int wctomb __P((char *__s, wchar_t __wchar));
 
 #ifdef	__OPTIMIZE__
-#if defined(__cplusplus) || __GNUC__ >= 2
+#if defined(__cplusplus) || (__GNUC__ >= 2 && __STDC__)
 extern __inline int mblen (__const char *__s, size_t __n)
 	{ return mbtowc((wchar_t *) NULL, __s, __n); }
 #else
diff --git a/usr/include/string.h b/usr/include/string.h
index 2ef8e4e..cddbfdf 100644
--- a/usr/include/string.h
+++ b/usr/include/string.h
@@ -32,8 +32,7 @@ __BEGIN_DECLS
 #define	__need_NULL
 #include <stddef.h>
 
-/* `memcpy' is a built-in function for gcc 2.x. */
-#if defined(__STDC__) && __GNUC__ < 2
+#if defined(__STDC__)
 /* Copy N bytes of SRC to DEST.  */
 extern __ptr_t memcpy __P ((__ptr_t __dest, __const __ptr_t __src,
 			    size_t __n));
@@ -58,11 +57,12 @@ extern __ptr_t memccpy __P ((__ptr_t __dest, __const __ptr_t __src,
 #endif /* SVID.  */
 
 
+#if defined(__STDC__)
 /* Set N bytes of S to C.  */
 extern __ptr_t memset __P ((__ptr_t __s, int __c, size_t __n));
+#endif
 
-/* `memcmp' is a built-in function for gcc 2.x. */
-#if defined(__STDC__) && __GNUC__ < 2
+#if defined(__STDC__)
 /* Compare N bytes of S1 and S2.  */
 extern int memcmp __P ((__const __ptr_t __s1, __const __ptr_t __s2,
 			size_t __n));
diff --git a/usr/include/sys/cdefs.h b/usr/include/sys/cdefs.h
index 728faf3..afffee3 100644
--- a/usr/include/sys/cdefs.h
+++ b/usr/include/sys/cdefs.h
@@ -35,9 +35,9 @@ Cambridge, MA 02139, USA.  */
    parameters and have no side effects, or (2) don't return.  */
 
 /*
-#if __GNUC__ < 2 || (__GNUC__ == 2 && __GNUC_MINOR__ < 5) || defined(__cplusplus)
+#if __GNUC__ < 2 || (__GNUC__ == 2 && __GNUC_MINOR__ < 6) || defined(__cplusplus)
 */
-#if __GNUC__ < 2 || (__GNUC__ == 2 && __GNUC_MINOR__ < 5)
+#if __GNUC__ < 2 || (__GNUC__ == 2 && __GNUC_MINOR__ < 6)
   /* Old GCC way. */
 #ifndef	__CONSTVALUE
 #define	__CONSTVALUE	__const__
@@ -99,10 +99,12 @@ Cambridge, MA 02139, USA.  */
 
 #endif	/* GCC.  */
 
-#if defined (__GNUC__) && __GNUC__ < 2
-/* In GCC version 2, (__extension__ EXPR) will not complain
-   about GCC extensions used in EXPR under -ansi or -pedantic.  */
-#define	__extension__
+/* It is possible to compile containing GCC extensions even if GCC is
+   run in pedantic mode if the uses are carefully marked using the
+   `__extension__' keyword.  But this is not generally available before
+   version 2.8.  */
+#if __GNUC__ < 2 || (__GNUC__ == 2 && __GNUC_MINOR__ < 8)
+# define __extension__		/* Ignore */
 #endif
 
 /* For these things, GCC behaves the ANSI way normally,
diff --git a/usr/include/sys/mman.h b/usr/include/sys/mman.h
index f7499f4..36d661a 100644
--- a/usr/include/sys/mman.h
+++ b/usr/include/sys/mman.h
@@ -8,7 +8,9 @@
 #ifndef MAP_ANON
 #define MAP_ANON	MAP_ANONYMOUS	   /* idem */
 #endif
+#ifndef MAP_FILE
 #define MAP_FILE	0x00	   /* The 'normal' way: mapped from file */
+#endif
 
 __BEGIN_DECLS
 
diff --git a/usr/include/sys/uio.h b/usr/include/sys/uio.h
index 3017125..81dbe43 100644
--- a/usr/include/sys/uio.h
+++ b/usr/include/sys/uio.h
@@ -26,9 +26,11 @@ __BEGIN_DECLS
 #define	__need_size_t
 #include <stddef.h>
 
+#include <linux/uio.h>
 
 /* Structure describing a section of memory.  */
 
+#if 0
 struct iovec
 {
   /* Starting address.  */
@@ -36,6 +38,7 @@ struct iovec
   /* Length in bytes.  */
   size_t iov_len;
 };
+#endif
 
 
 /* Read data from file descriptor FD, and put the result in the
