diff --git a/gcc/cp/typeck.c b/gcc/cp/typeck.c
index c33fdeaa2..3ebb24aab 100644
--- a/gcc/cp/typeck.c
+++ b/gcc/cp/typeck.c
@@ -2058,6 +2058,7 @@ is_bitfield_expr_with_lowered_type (const_tree exp)
 
     case MODIFY_EXPR:
     case SAVE_EXPR:
+    case CLEANUP_POINT_EXPR:
       return is_bitfield_expr_with_lowered_type (TREE_OPERAND (exp, 0));
 
     case COMPONENT_REF:
diff --git a/gcc/testsuite/g++.dg/cpp0x/enum44.C b/gcc/testsuite/g++.dg/cpp0x/enum44.C
new file mode 100644
index 000000000..92408c922
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/enum44.C
@@ -0,0 +1,30 @@
+// PR c++/103825
+// { dg-do compile { target c++11 } }
+
+enum class Type { Pawn };
+struct Piece {
+  Type type : 4;
+};
+
+void
+foo ()
+{
+  switch (Piece().type)
+    case Type::Pawn:;
+
+  auto x = Piece().type;
+  switch (x)
+    case Type::Pawn:;
+}
+
+enum class En {A};
+struct St {En field :1;};
+
+void
+bar ()
+{
+  volatile St s = {En::A};
+  switch(s.field) {
+    case En::A : break;
+  }
+}
