diff --git a/gcc/fortran/trans-decl.c b/gcc/fortran/trans-decl.c
index 897928a32..d084c3e2e 100644
--- a/gcc/fortran/trans-decl.c
+++ b/gcc/fortran/trans-decl.c
@@ -4837,6 +4837,11 @@ gfc_trans_deferred_vars (gfc_symbol * proc_sym, gfc_wrapped_block * block)
 					NULL_TREE);
 		  continue;
 		}
+	      else if (sym->attr.codimension && !sym->attr.dimension)
+		{
+		  /* Scalar coarrays do not need array allocation.  */
+		  continue;
+		}
 	      else
 		{
 		  gfc_save_backend_locus (&loc);
diff --git a/gcc/testsuite/gfortran.dg/pr93715.f90 b/gcc/testsuite/gfortran.dg/pr93715.f90
new file mode 100644
index 000000000..5dba371be
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/pr93715.f90
@@ -0,0 +1,11 @@
+! { dg-do compile }
+! { dg-options "-fcoarray=single" }
+!
+! PR fortran/93715
+! A scalar coarray used in asynchronous I/O caused an ICE in
+! gfc_trans_auto_array_allocation.
+
+program p
+   integer :: a, b[*]
+   read (1, *, asynchronous='yes') a, b
+end
