diff --git a/gcc/testsuite/gcc.target/aarch64/symbol-range-tiny.c b/gcc/testsuite/gcc.target/aarch64/symbol-range-tiny.c
deleted file mode 100644
index fc6a4f3ec..000000000
--- a/gcc/testsuite/gcc.target/aarch64/symbol-range-tiny.c
+++ /dev/null
@@ -1,12 +0,0 @@
-/* { dg-do link } */
-/* { dg-options "-O3 -save-temps -mcmodel=tiny" } */
-
-char fixed_regs[0x00080000];
-
-int
-main ()
-{
-  return fixed_regs[0x000ff000];
-}
-
-/* { dg-final { scan-assembler-not "adr\tx\[0-9\]+, fixed_regs\\\+" } } */
diff --git a/gcc/testsuite/gcc.target/aarch64/symbol-range.c b/gcc/testsuite/gcc.target/aarch64/symbol-range.c
deleted file mode 100644
index d8e82fa1b..000000000
--- a/gcc/testsuite/gcc.target/aarch64/symbol-range.c
+++ /dev/null
@@ -1,12 +0,0 @@
-/* { dg-do link } */
-/* { dg-options "-O3 -save-temps -mcmodel=small" } */
-
-char fixed_regs[0x80000000];
-
-int
-main ()
-{
-  return fixed_regs[0xfffff000];
-}
-
-/* { dg-final { scan-assembler-not "adrp\tx\[0-9\]+, fixed_regs\\\+" } } */
diff --git a/gcc/testsuite/gfortran.dg/pr95690.f90 b/gcc/testsuite/gfortran.dg/pr95690.f90
index 2da08912a..1be277638 100644
--- a/gcc/testsuite/gfortran.dg/pr95690.f90
+++ b/gcc/testsuite/gfortran.dg/pr95690.f90
@@ -2,8 +2,8 @@
 module m
 contains
    subroutine s
-      print *, (erfc) ! { dg-error "not a floating constant" }
-   end
+      print *, (erfc)
+   end ! { dg-error "not a floating constant" }
    function erfc()
    end
 end
