diff --git a/Makefile.in b/Makefile.in
index ccb427d..c9e4b9f 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -166,7 +166,7 @@ RANLIB_FOR_TARGET = $(TARGET_TOOLPREFIX)ranlib
 RANLIB_TEST_FOR_TARGET = [ -f $(TARGET_TOOLPREFIX)ranlib ]
 
 # Dir to search for system headers.  Overridden by cross-make.
-SYSTEM_HEADER_DIR = /usr/include
+SYSTEM_HEADER_DIR = $(tooldir)/include
 
 # Control whether to run fixproto.
 STMP_FIXPROTO = stmp-fixproto
