diff --git a/config/alpha/linux.h b/config/alpha/linux.h
index 32428df..b4075ad 100644
--- a/config/alpha/linux.h
+++ b/config/alpha/linux.h
@@ -28,7 +28,7 @@ the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */
 #undef CPP_PREDEFINES
 #define CPP_PREDEFINES "\
 -D__alpha -D__alpha__ -D__linux__ -D__linux -D_LONGLONG -Dlinux -Dunix \
--Asystem(linux) -Acpu(alpha) -Amachine(alpha)"
+-Asystem(linux) -Acpu(alpha) -Amachine(alpha) -D__LP64__ -D_LP64"
 
 /* We don't actually need any of these; the MD_ vars are ignored
    anyway for cross-compilers, and the other specs won't get picked up
