diff --git a/config/alpha/elf.h b/config/alpha/elf.h
index 343244d..4726791 100644
--- a/config/alpha/elf.h
+++ b/config/alpha/elf.h
@@ -35,7 +35,7 @@ Boston, MA 02111-1307, USA.    */
 
 #undef CPP_PREDEFINES
 #define CPP_PREDEFINES "\
--D__alpha -D__alpha__ -D__linux__ -D__linux -D_LONGLONG -Dlinux -Dunix \
+-D__alpha -D__alpha__ -D__linux__ -D__linux -D_LONGLONG -Dlinux -Dunix -D__LP64__ -D_LP64 \
 -Asystem(linux) -Acpu(alpha) -Amachine(alpha) -D__ELF__"
 
 #undef LINK_SPEC
