diff --git a/bfd/elflink.c b/bfd/elflink.c
index 29b89d4a..26a45a36 100644
--- a/bfd/elflink.c
+++ b/bfd/elflink.c
@@ -13795,9 +13795,6 @@ _bfd_elf_final_link (bfd *abfd, struct bfd_link_info *info)
 		  else if (bfd_link_pde (info))
 		    info->callbacks->einfo
 		      (_("%P: warning: creating DT_TEXTREL in a PDE\n"));
-		  else
-		    info->callbacks->einfo
-		      (_("%P: warning: creating DT_TEXTREL in a PIE\n"));
 		  break;
 		}
 	    }
diff --git a/gas/config/loongarch-parse.c b/gas/config/loongarch-parse.c
index 861b88b2..ce4795eb 100644
--- a/gas/config/loongarch-parse.c
+++ b/gas/config/loongarch-parse.c
@@ -290,9 +290,6 @@ emit_bin (int op)
 	  opr1 = (valueT) opr1 << opr2;
 	  break;
 	case RIGHT_OP:
-	  if (opr1 < 0)
-	    as_warn (_("Right shift of negative numbers may be changed "
-		       "from arithmetic right shift to logical right shift!"));
 	  /* Arithmetic right shift.  */
 	  opr1 = opr1 >> opr2;
 	  break;
diff --git a/gas/config/loongarch-parse.y b/gas/config/loongarch-parse.y
index 41c95ac9..59546ff1 100644
--- a/gas/config/loongarch-parse.y
+++ b/gas/config/loongarch-parse.y
@@ -238,9 +238,6 @@ emit_bin (int op)
 	  opr1 = (valueT) opr1 << opr2;
 	  break;
 	case RIGHT_OP:
-	  if (opr1 < 0)
-	    as_warn (_("Right shift of negative numbers may be changed "
-		       "from arithmetic right shift to logical right shift!"));
 	  /* Arithmetic right shift.  */
 	  opr1 = opr1 >> opr2;
 	  break;
