diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 9f6ed8125..25d139a2c 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -213,7 +213,9 @@ GCC_WARN_CXXFLAGS = $(LOOSE_WARN) $($(@D)-warn) $(NOCOMMON_FLAG) $($@-warn)
 # These files are to have specific diagnostics suppressed, or are not to
 # be subject to -Werror:
 # flex output may yield harmless "no previous prototype" warnings
+build/gengtype.o-warn = -Wno-format-overflow
 build/gengtype-lex.o-warn = -Wno-error
+gengtype.o-warn = -Wno-format-overflow
 gengtype-lex.o-warn = -Wno-error
 libgcov-util.o-warn = -Wno-error
 libgcov-driver-tool.o-warn = -Wno-error
@@ -221,6 +223,15 @@ libgcov-merge-tool.o-warn = -Wno-error
 gimple-match.o-warn = -Wno-unused
 generic-match.o-warn = -Wno-unused
 dfp.o-warn = -Wno-strict-aliasing
+arm.o-warn = -Wno-error=format-diag
+build/genhooks.o-warn = -Wno-error=unused-result
+gcc.o-warn = -Wno-error=unused-result
+ggc-common.o-warn = -Wno-error=deprecated-declarations
+lto/lto-common.o-warn = -Wno-error=unused-result
+lto-wrapper.o-warn = -Wno-error=unused-result
+m68k-common.o-warn = -Wno-error=unused-result
+objc/objc-next-runtime-abi-02.o-warn = -Wno-error=unused-result
+objcp/objc-next-runtime-abi-02.o-warn = -Wno-error=unused-result
 
 # All warnings have to be shut off in stage1 if the compiler used then
 # isn't gcc; configure determines that.  WARN_CFLAGS will be either
diff --git a/gcc/calls.c b/gcc/calls.c
index 45c137cee..e5e2a916f 100644
--- a/gcc/calls.c
+++ b/gcc/calls.c
@@ -1446,7 +1446,7 @@ maybe_warn_alloc_args_overflow (tree fn, tree exp, tree args[2], int idx[2])
 	}
     }
 
-  if (!argrange[0])
+  if (!argrange[0][0])
     return;
 
   /* For a two-argument alloc_size, validate the product of the two
diff --git a/gcc/config/aarch64/aarch64.c b/gcc/config/aarch64/aarch64.c
index e6bf5ae38..bdab04d7e 100644
--- a/gcc/config/aarch64/aarch64.c
+++ b/gcc/config/aarch64/aarch64.c
@@ -23244,7 +23244,7 @@ aarch64_simd_clone_compute_vecsize_and_simdlen (struct cgraph_node *node,
 					struct cgraph_simd_clone *clonei,
 					tree base_type, int num)
 {
-  tree t, ret_type, arg_type;
+  tree t, ret_type;
   unsigned int elt_bits, vec_bits, count;
 
   if (!TARGET_SIMD)
diff --git a/gcc/gensupport.c b/gcc/gensupport.c
index f2ad54f0c..ef468a2f1 100644
--- a/gcc/gensupport.c
+++ b/gcc/gensupport.c
@@ -1499,8 +1499,9 @@ alter_attrs_for_subst_insn (class queue_elem * elem, int n_dup)
       switch (GET_CODE (sub))
 	{
 	case SET_ATTR:
-	  if (strchr (XSTR (sub, 1), ',') != NULL)
+	  if (strchr (XSTR (sub, 1), ',') != NULL) {
 	    XSTR (sub, 1) = duplicate_alternatives (XSTR (sub, 1), n_dup);
+	}
 	    break;
 
 	case SET_ATTR_ALTERNATIVE:
diff --git a/gcc/lra-constraints.c b/gcc/lra-constraints.c
index 30d088afb..2fe131923 100644
--- a/gcc/lra-constraints.c
+++ b/gcc/lra-constraints.c
@@ -4751,8 +4751,9 @@ multi_block_pseudo_p (int regno)
   unsigned int uid;
   bitmap_iterator bi;
 
-  if (regno < FIRST_PSEUDO_REGISTER)
+  if (regno < FIRST_PSEUDO_REGISTER) {
     return false;
+  }
 
     EXECUTE_IF_SET_IN_BITMAP (&lra_reg_info[regno].insn_bitmap, 0, uid, bi)
       if (bb == NULL)
diff --git a/gcc/sel-sched-ir.c b/gcc/sel-sched-ir.c
index 4f1e4afdc..49a6debf7 100644
--- a/gcc/sel-sched-ir.c
+++ b/gcc/sel-sched-ir.c
@@ -701,6 +701,7 @@ merge_fences (fence_t f, insn_t insn,
 	  FENCE_ISSUE_MORE (f) = issue_rate;
         }
       else
+		{
         if (candidate->src == BLOCK_FOR_INSN (last_scheduled_insn))
           {
             state_free (FENCE_STATE (f));
@@ -721,6 +722,7 @@ merge_fences (fence_t f, insn_t insn,
             gcc_assert (BLOCK_FOR_INSN (insn)->prev_bb
                         != BLOCK_FOR_INSN (last_scheduled_insn));
           }
+		}
 
         /* Find edge of first predecessor (last_scheduled_insn_old->insn).  */
         FOR_EACH_SUCC_1 (succ, si, last_scheduled_insn_old,
diff --git a/gcc/tree-vect-stmts.c b/gcc/tree-vect-stmts.c
index 2ef71ec41..3a2084012 100644
--- a/gcc/tree-vect-stmts.c
+++ b/gcc/tree-vect-stmts.c
@@ -7553,10 +7553,10 @@ vectorizable_store (stmt_vec_info stmt_info, gimple_stmt_iterator *gsi,
 
   if (memory_access_type == VMAT_GATHER_SCATTER && gs_info.decl)
     {
-      tree vec_oprnd0 = NULL_TREE, vec_oprnd1 = NULL_TREE, src;
+      tree vec_oprnd0 = NULL_TREE, vec_oprnd1 = NULL_TREE, src = NULL_TREE;
       tree arglist = TYPE_ARG_TYPES (TREE_TYPE (gs_info.decl));
-      tree rettype, srctype, ptrtype, idxtype, masktype, scaletype;
-      tree ptr, var, scale, vec_mask;
+      tree rettype = NULL_TREE, srctype = NULL_TREE, ptrtype = NULL_TREE, idxtype = NULL_TREE, masktype = NULL_TREE, scaletype = NULL_TREE;
+      tree ptr = NULL_TREE, var = NULL_TREE, scale = NULL_TREE, vec_mask = NULL_TREE;
       tree mask_arg = NULL_TREE, mask_op = NULL_TREE, perm_mask = NULL_TREE;
       tree mask_halfvectype = mask_vectype;
       edge pe = loop_preheader_edge (loop);
diff --git a/libdecnumber/decNumber.c b/libdecnumber/decNumber.c
index 6c9a58173..bc9c5a45a 100644
--- a/libdecnumber/decNumber.c
+++ b/libdecnumber/decNumber.c
@@ -927,9 +927,9 @@ decNumber * decNumberCompareTotalMag(decNumber *res, const decNumber *lhs,
 				     const decNumber *rhs, decContext *set) {
   uInt status=0;		   /* accumulator */
   uInt needbytes;		   /* for space calculations */
-  decNumber bufa[D2N(DECBUFFER+1)];/* +1 in case DECBUFFER=0 */
+  decNumber bufa[D2N(DECBUFFER+1)] = {0};/* +1 in case DECBUFFER=0 */
   decNumber *allocbufa=NULL;	   /* -> allocated bufa, iff allocated */
-  decNumber bufb[D2N(DECBUFFER+1)];
+  decNumber bufb[D2N(DECBUFFER+1)] = {0};
   decNumber *allocbufb=NULL;	   /* -> allocated bufb, iff allocated */
   decNumber *a, *b;		   /* temporary pointers */
 
@@ -947,6 +947,7 @@ decNumber * decNumberCompareTotalMag(decNumber *res, const decNumber *lhs,
 	if (allocbufa==NULL) {		/* hopeless -- abandon */
 	  status|=DEC_Insufficient_storage;
 	  break;}
+	memset(allocbufa, 0x00, needbytes);
 	a=allocbufa;			/* use the allocated space */
 	}
       decNumberCopy(a, lhs);		/* copy content */
@@ -961,6 +962,7 @@ decNumber * decNumberCompareTotalMag(decNumber *res, const decNumber *lhs,
 	if (allocbufb==NULL) {		/* hopeless -- abandon */
 	  status|=DEC_Insufficient_storage;
 	  break;}
+	memset(allocbufb, 0x00, needbytes);
 	b=allocbufb;			/* use the allocated space */
 	}
       decNumberCopy(b, rhs);		/* copy content */
@@ -3343,6 +3345,8 @@ const char *decNumberClassToString(enum decClass eclass) {
 /* All fields are updated as required.	This is a utility operation,  */
 /* so special values are unchanged and no error is possible.	      */
 /* ------------------------------------------------------------------ */
+#pragma GCC push_options
+#pragma GCC optimize ("-O2")
 decNumber * decNumberCopy(decNumber *dest, const decNumber *src) {
 
   #if DECCHECK
@@ -3370,6 +3374,7 @@ decNumber * decNumberCopy(decNumber *dest, const decNumber *src) {
     }
   return dest;
   } /* decNumberCopy */
+#pragma GCC pop_options
 
 /* ------------------------------------------------------------------ */
 /* decNumberCopyAbs -- quiet absolute value operator		      */
diff --git a/libstdc++-v3/config/locale/gnu/ctype_members.cc b/libstdc++-v3/config/locale/gnu/ctype_members.cc
index 61ab0a02c..b62302edc 100644
--- a/libstdc++-v3/config/locale/gnu/ctype_members.cc
+++ b/libstdc++-v3/config/locale/gnu/ctype_members.cc
@@ -221,7 +221,7 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
   ctype<wchar_t>::
   do_narrow(wchar_t __wc, char __dfault) const
   {
-    if (__wc >= 0 && __wc < 128 && _M_narrow_ok)
+    if (__wc < 128 && _M_narrow_ok)
       return _M_narrow[__wc];
 #if __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ > 2)
     __c_locale __old = __uselocale(_M_c_locale_ctype);
@@ -244,7 +244,7 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
     if (_M_narrow_ok)
       while (__lo < __hi)
 	{
-	  if (*__lo >= 0 && *__lo < 128)
+	  if (*__lo < 128)
 	    *__dest = _M_narrow[*__lo];
 	  else
 	    {
diff --git a/libstdc++-v3/libsupc++/pure.cc b/libstdc++-v3/libsupc++/pure.cc
index b943214de..c97f66888 100644
--- a/libstdc++-v3/libsupc++/pure.cc
+++ b/libstdc++-v3/libsupc++/pure.cc
@@ -46,13 +46,19 @@
 extern "C" void
 __cxxabiv1::__cxa_pure_virtual (void)
 {
-  writestr ("pure virtual method called\n");
+#if _GLIBCXX_HOSTED && _GLIBCXX_VERBOSE
+  auto ret = writestr ("pure virtual method called\n");
+  (void)ret;
+#endif
   std::terminate ();
 }
 
 extern "C" void
 __cxxabiv1::__cxa_deleted_virtual (void)
 {
-  writestr ("deleted virtual method called\n");
+#if _GLIBCXX_HOSTED && _GLIBCXX_VERBOSE
+  auto ret = writestr ("deleted virtual method called\n");
+  (void)ret;
+#endif
   std::terminate ();
 }
