diff --git a/Makefile b/Makefile
index e750af1..230a20f 100644
--- a/Makefile
+++ b/Makefile
@@ -176,7 +176,7 @@ DIR = ../gcc
 # List of extra executables that should be compiled for this target machine
 # that are used for linking.
 # The rules for compiling them should be in the t-* file for the machine.
-EXTRA_PARTS = crtbegin.o crtend.o
+EXTRA_PARTS =
 
 # Dependency on obstack, alloca, malloc or whatever library facilities
 # are not installed in the system libraries.
diff --git a/gcc.c b/gcc.c
index ae59b29..0c55fca 100644
--- a/gcc.c
+++ b/gcc.c
@@ -307,7 +307,7 @@ struct compiler compilers[] =
 		   %{v:-version} %{gg:-symout %g.sym} %{pg:-p} %{p}\
 		   %{pg:%{fomit-frame-pointer:%e-pg and -fomit-frame-pointer are incompatible}}\
 		   %{S:%W{o*}%{!o*:-o %b.s}}%{!S:-o %{|!pipe:%g.s}} |\n\
-              %{!S:i586-legacy-linux-gnulibc1-as %{R} %{j} %{J} %{h} %{d2} %a %{gg:-G %g.sym}\
+              %{!S:i486-legacy-linuxaout-as %{R} %{j} %{J} %{h} %{d2} %a %{gg:-G %g.sym}\
 		      %{c:%W{o*}%{!o*:-o %w%b.o}}%{!c:-o %d%w%b.o}\
                       %{!pipe:%g.s}\n }}}"},
   {".cc",
@@ -322,7 +322,7 @@ struct compiler compilers[] =
 		   %{v:-version} %{gg:-symout %g.sym} %{pg:-p} %{p}\
 		   %{pg:%{fomit-frame-pointer:%e-pg and -fomit-frame-pointer are incompatible}}\
 		   %{S:%W{o*}%{!o*:-o %b.s}}%{!S:-o %{|!pipe:%g.s}} |\n\
-              %{!S:i586-legacy-linux-gnulibc1-as %{R} %{j} %{J} %{h} %{d2} %a %{gg:-G %g.sym}\
+              %{!S:i486-legacy-linuxaout-as %{R} %{j} %{J} %{h} %{d2} %a %{gg:-G %g.sym}\
 		      %{c:%W{o*}%{!o*:-o %w%b.o}}%{!c:-o %d%w%b.o}\
                       %{!pipe:%g.s}\n }}}"},
   {".i",
@@ -330,10 +330,10 @@ struct compiler compilers[] =
 	%{g} %{O*} %{W*} %{w} %{pedantic} %{ansi} %{traditional}\
 	%{v:-version} %{gg:-symout %g.sym} %{pg:-p} %{p}\
 	%{S:%W{o*}%{!o*:-o %b.s}}%{!S:-o %{|!pipe:%g.s}} |\n\
-    %{!S:i586-legacy-linux-gnulibc1-as %{R} %{j} %{J} %{h} %{d2} %a %{gg:-G %g.sym}\
+    %{!S:i486-legacy-linuxaout-as %{R} %{j} %{J} %{h} %{d2} %a %{gg:-G %g.sym}\
             %{c:%W{o*}%{!o*:-o %w%b.o}}%{!c:-o %d%w%b.o} %{!pipe:%g.s}\n }"},
   {".s",
-   "%{!S:i586-legacy-linux-gnulibc1-as %{R} %{j} %{J} %{h} %{d2} %a \
+   "%{!S:i486-legacy-linuxaout-as %{R} %{j} %{J} %{h} %{d2} %a \
             %{c:%W{o*}%{!o*:-o %w%b.o}}%{!c:-o %d%w%b.o} %i\n }"},
   {".S",
    "cpp %{nostdinc} %{C} %{v} %{D*} %{U*} %{I*} %{M*} %{trigraphs} \
@@ -341,7 +341,7 @@ struct compiler compilers[] =
         %c %{O:-D__OPTIMIZE__} %{traditional} %{pedantic} %{P}\
 	%{Wcomment*} %{Wtrigraphs} %{Wall} %{w} %C\
         %i %{!M*:%{!E:%{!pipe:%g.s}}}%{E:%W{o*}}%{M*:%W{o*}} |\n\
-    %{!M*:%{!E:%{!S:i586-legacy-linux-gnulibc1-as %{R} %{j} %{J} %{h} %{d2} %a \
+    %{!M*:%{!E:%{!S:i486-legacy-linuxaout-as %{R} %{j} %{J} %{h} %{d2} %a \
                     %{c:%W{o*}%{!o*:-o %w%b.o}}%{!c:-o %d%w%b.o}\
 		    %{!pipe:%g.s}\n }}}"},
   /* Mark end of table */
@@ -349,9 +349,9 @@ struct compiler compilers[] =
 };
 
 /* Here is the spec for running the linker, after compiling all files.  */
-char *link_spec = "%{!c:%{!M*:%{!E:%{!S:i586-legacy-linux-gnulibc1-ld %{o*} %l\
+char *link_spec = "%{!c:%{!M*:%{!E:%{!S:i486-legacy-linuxaout-ld %{o*} %l\
  %{A} %{d} %{e*} %{N} %{n} %{r} %{s} %{S} %{T*} %{t} %{u*} %{X} %{x} %{z}\
- %{y*} %{!A:%{!nostdlib:%S}} -L /usr/i586-legacy-linux-gnulibc1/lib \
+ %{y*} %{!A:%{!nostdlib:%S}} -L /usr/i486-legacy-linuxaout/lib \
  %{L*} %o %{!nostdlib:%G gnulib%s %L gnulib%s %{!A:%E}}\n }}}}";
 
 /* Accumulate a command (program name and args), and run it.  */
