diff --git a/Makefile.in b/Makefile.in
index 41fdbf3..ecc30c7 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -56,7 +56,7 @@ INSTALL_DOSREL = install-dosrel-fake
 AS = as
 AR = ar
 AR_FLAGS = rc
-CC = cc
+CC = @CC@
 
 # Special variables passed down in EXTRA_GCC_FLAGS.  They are defined
 # here so that they can be overridden by Makefile fragments.
@@ -73,7 +73,7 @@ LIBCFLAGS_FOR_TARGET = $(CFLAGS_FOR_TARGET)
 PICFLAG = 
 PICFLAG_FOR_TARGET = 
 
-CXX = gcc
+CXX = @CXX@
 
 # Use -O2 to stress test the compiler.
 CXXFLAGS = -g -O2
@@ -271,6 +271,7 @@ BASE_FLAGS_TO_PASS = \
 	"CC_FOR_TARGET=$(CC_FOR_TARGET)" \
 	"CFLAGS=$(CFLAGS)" \
 	"CFLAGS_FOR_TARGET=$(CFLAGS_FOR_TARGET)" \
+	"DESTDIR=$(DESTDIR)" \
 	"CXX_FOR_BUILD=$(CXX_FOR_BUILD)" \
 	"CXXFLAGS=$(CXXFLAGS)" \
 	"CXXFLAGS_FOR_TARGET=$(CXXFLAGS_FOR_TARGET)" \
@@ -1288,8 +1289,8 @@ install-target: $(INSTALL_TARGET_MODULES)
 ### other supporting targets
 
 MAKEDIRS= \
-	$(prefix) \
-	$(exec_prefix)
+	$(DESTDIR)$(prefix) \
+	$(DESTDIR)$(exec_prefix)
 .PHONY: install-dirs
 install-dirs:
 	@for i in $(MAKEDIRS) ; do \
diff --git a/etc/Makefile.in b/etc/Makefile.in
index 699ada6..67a5acc 100644
--- a/etc/Makefile.in
+++ b/etc/Makefile.in
@@ -44,17 +44,17 @@ DVIFILES = configure.dvi standards.dvi cfg-paper.dvi
 all:
 
 install:  $(srcdir)/configure.man
-	$(INSTALL_DATA) $(srcdir)/configure.man $(man1dir)/configure.1
+	$(INSTALL_DATA) $(srcdir)/configure.man $(DESTDIR)$(man1dir)/configure.1
 
 uninstall:
-	cd $(infodir) && rm -f configure.info* standards.info* cfg-paper.info*
+	cd $(DESTDIR)$(infodir) && rm -f configure.info* standards.info* cfg-paper.info*
 
 info: $(INFOFILES)
 
 install-info: info
-	if test ! -f configure.info ; then cd $(srcdir); fi; \
+	if test ! -f configure.info ; then cd $(DESTDIR)$(srcdir); fi; \
 	for i in configure.info* standards.info* cfg-paper.info*; do \
-	  $(INSTALL_DATA) $$i $(infodir)/$$i; \
+	  $(INSTALL_DATA) $$i $(DESTDIR)$(infodir)/$$i; \
 	done
 
 dvi: $(DVIFILES)
diff --git a/libg++/Makefile.in b/libg++/Makefile.in
index 4f6d7e0..deb7a54 100644
--- a/libg++/Makefile.in
+++ b/libg++/Makefile.in
@@ -181,8 +181,8 @@ installcheck: check
 
 .PHONY: install
 install:
-	@if [ -f $(gxx_includedir)/unistd.h ] ; then echo; \
-	  echo '*** You seem to have files in $(gxx_includedir)/g++-include'; \
+	@if [ -f $(DESTDIR)$(gxx_includedir)/unistd.h ] ; then echo; \
+	  echo '*** You seem to have files in $(DESTDIR)$(gxx_includedir)/g++-include'; \
 	  echo '*** left over from an old release of libg++.  These must be removed.'; \
 	  echo '*** Please see $(srcdir)/README for more information.'; \
 	  echo; \
@@ -191,17 +191,17 @@ install:
 	fi
 	rootme=`pwd`/ ; export rootme ; \
 	for FILE in $(BUILD_LIBS) ; do \
-	  rm -f $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	  rm -f $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  if [ $$FILE = $(SHLINK) ] || [ $$FILE = $(MSHLINK) ]; then \
-	    ln -s -f $(SHLIB) $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    ln -s -f $(SHLIB) $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  elif [ $$FILE = $(SHLIB) ]; then \
-	    $(INSTALL_PROGRAM) $$FILE $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    $(INSTALL_PROGRAM) $$FILE $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	    : On the HP, shared libraries must be mode 555. ;\
-	    chmod 555 $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    chmod 555 $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  else \
-	    $(INSTALL_DATA) $$FILE $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
-	    $(RANLIB) $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
-	    chmod a-x $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    $(INSTALL_DATA) $$FILE $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    $(RANLIB) $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    chmod a-x $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  fi ; \
 	done
 	@for D in src genclass $(UTILS) ; do \
diff --git a/libg++/genclass/Makefile.in b/libg++/genclass/Makefile.in
index c6a04f5..d73fe87 100644
--- a/libg++/genclass/Makefile.in
+++ b/libg++/genclass/Makefile.in
@@ -20,4 +20,4 @@ gentest: $(srcdir)/gentest.sh $(srcdir)/expected.out
 
 .PHONY: install
 install:
-	$(INSTALL_PROGRAM) genclass $(bindir)/genclass
+	$(INSTALL_PROGRAM) genclass $(DESTDIR)$(bindir)/genclass
diff --git a/libg++/gperf/Makefile.in b/libg++/gperf/Makefile.in
index 7d1d1ef..07fe17a 100644
--- a/libg++/gperf/Makefile.in
+++ b/libg++/gperf/Makefile.in
@@ -32,7 +32,7 @@ src_all:
 install:
 	rootme=`pwd`/ ; export rootme ; cd src ; \
 		$(MAKE) $(FLAGS_TO_PASS) install
-	$(INSTALL_DATA) $(srcdir)/gperf.1 $(man1dir)/gperf.1
+	$(INSTALL_DATA) $(srcdir)/gperf.1 $(DESTDIR)$(man1dir)/gperf.1
 
 distrib: force
 	cd ..; rm -f gperf.tar.Z; tar cvf gperf.tar gperf; compress gperf.tar;\
diff --git a/libg++/gperf/src/Makefile.in b/libg++/gperf/src/Makefile.in
index 3ba9d70..e58d817 100644
--- a/libg++/gperf/src/Makefile.in
+++ b/libg++/gperf/src/Makefile.in
@@ -32,4 +32,4 @@ $(TARGETPROG): $(OBJECTS)
 
 .PHONY: install
 install: $(TARGETPROG)
-	$(INSTALL_PROGRAM) $(TARGETPROG) $(bindir)/$(TARGETPROG)
+	$(INSTALL_PROGRAM) $(TARGETPROG) $(DESTDIR)$(bindir)/$(TARGETPROG)
diff --git a/libg++/src/Makefile.in b/libg++/src/Makefile.in
index d675060..b198879 100644
--- a/libg++/src/Makefile.in
+++ b/libg++/src/Makefile.in
@@ -66,10 +66,10 @@ libgxx.list: stamp-picdir $(OBJS)
 install:
 	if [ -z "$(MULTISUBDIR)" ]; then \
 	  for FILE in `(cd ${srcdir}; echo *.h gen/*.ccP gen/*.hP)`; do \
-	    rm -f $(gxx_includedir)/$$FILE ; \
-	    $(INSTALL_DATA) ${srcdir}/$$FILE $(gxx_includedir)/$$FILE \
+	    rm -f $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+	    $(INSTALL_DATA) ${srcdir}/$$FILE $(DESTDIR)$(gxx_includedir)/$$FILE \
 	      || exit 1; \
-	    chmod a-x,a+r $(gxx_includedir)/$$FILE ; \
+	    chmod a-x,a+r $(DESTDIR)$(gxx_includedir)/$$FILE ; \
 	  done ; \
 	else true ; \
 	fi
diff --git a/libiberty/Makefile.in b/libiberty/Makefile.in
index 1594fae..7a43d4c 100644
--- a/libiberty/Makefile.in
+++ b/libiberty/Makefile.in
@@ -69,7 +69,7 @@ AR = ar
 AR_FLAGS = rc
 
 ERRORS_CC = $(CC)
-CC = cc
+CC = $(CC)
 CFLAGS = -g
 LIBCFLAGS = $(CFLAGS)
 MAKEINFO = makeinfo
@@ -106,6 +106,7 @@ FLAGS_TO_PASS = \
 	"CC=$(CC)" \
 	"CFLAGS=$(CFLAGS)" \
 	"LIBCFLAGS=$(LIBCFLAGS)" \
+	"DESTDIR=$(DESTDIR)" \
 	"EXTRA_OFILES=$(EXTRA_OFILES)" \
 	"HDEFINES=$(HDEFINES)" \
 	"LDFLAGS=$(LDFLAGS)" \
@@ -171,15 +172,15 @@ INSTALL_DEST = libdir
 install: install_to_$(INSTALL_DEST)
 
 install_to_libdir: all
-	$(INSTALL_DATA) $(TARGETLIB) $(libdir)/$(TARGETLIB).n
-	( cd $(libdir) ; $(RANLIB) $(libdir)/$(TARGETLIB).n )
-	mv -f $(libdir)/$(TARGETLIB).n $(libdir)$(MULTISUBDIR)/$(TARGETLIB)
+	$(INSTALL_DATA) $(TARGETLIB) $(DESTDIR)$(libdir)/$(TARGETLIB).n
+	( cd $(DESTDIR)$(libdir) ; $(RANLIB) $(DESTDIR)$(libdir)/$(TARGETLIB).n )
+	mv -f $(DESTDIR)$(libdir)/$(TARGETLIB).n $(DESTDIR)$(libdir)$(MULTISUBDIR)/$(TARGETLIB)
 	@$(MULTIDO) $(FLAGS_TO_PASS) multi-do DO=install
 
 install_to_tooldir: all
-	$(INSTALL_DATA) $(TARGETLIB) $(tooldir)/lib/$(TARGETLIB).n
-	( cd $(tooldir) ; $(RANLIB) $(tooldir)/lib/$(TARGETLIB).n )
-	mv -f $(tooldir)/lib/$(TARGETLIB).n $(tooldir)/lib$(MULTISUBDIR)/$(TARGETLIB)
+	$(INSTALL_DATA) $(TARGETLIB) $(DESTDIR)$(tooldir)/lib/$(TARGETLIB).n
+	( cd $(DESTDIR)$(tooldir) ; $(RANLIB) $(DESTDIR)$(tooldir)/lib/$(TARGETLIB).n )
+	mv -f $(DESTDIR)$(tooldir)/lib/$(TARGETLIB).n $(DESTDIR)$(tooldir)/lib$(MULTISUBDIR)/$(TARGETLIB)
 	@$(MULTIDO) $(FLAGS_TO_PASS) multi-do DO=install
 
 # The default configuration adds to libiberty all those functions that are
diff --git a/libio/Makefile.in b/libio/Makefile.in
index c78639a..0634842 100644
--- a/libio/Makefile.in
+++ b/libio/Makefile.in
@@ -104,15 +104,15 @@ install:
 	rootme=`pwd`/ ; export rootme ; \
 	if [ -z "$(MULTISUBDIR)" ]; then \
 	  if [ "$(_G_CONFIG_H)" != "" ]; then \
-	    rm -f $(tooldir)/include/_G_config.h ; \
-	    $(INSTALL_DATA) _G_config.h $(tooldir)/include/_G_config.h || exit 1; \
+	    rm -f $(DESTDIR)$(tooldir)/include/_G_config.h ; \
+	    $(INSTALL_DATA) _G_config.h $(DESTDIR)$(tooldir)/include/_G_config.h || exit 1; \
 	  else true; \
 	  fi ; \
 	  cd $(srcdir); \
 	  for FILE in $(USER_INCLUDES) ; do \
-	    rm -f $(gxx_includedir)/$$FILE ; \
-	    $(INSTALL_DATA) $$FILE $(gxx_includedir)/$$FILE ; \
-	    chmod a-x $(gxx_includedir)/$$FILE ; \
+	    rm -f $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+	    $(INSTALL_DATA) $$FILE $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+	    chmod a-x $(DESTDIR)$(gxx_includedir)/$$FILE ; \
 	  done ; \
 	else true; \
 	fi
diff --git a/libio/gen-params b/libio/gen-params
index b6e51b0..35b0f13 100755
--- a/libio/gen-params
+++ b/libio/gen-params
@@ -313,7 +313,6 @@ done
 cat <<!EOF!
 typedef ${clock_t-int /* default */} ${macro_prefix}clock_t;
 typedef ${dev_t-int /* default */} ${macro_prefix}dev_t;
-typedef ${fpos_t-long /* default */} ${macro_prefix}fpos_t;
 typedef ${gid_t-int /* default */} ${macro_prefix}gid_t;
 typedef ${ino_t-int /* default */} ${macro_prefix}ino_t;
 typedef ${mode_t-int /* default */} ${macro_prefix}mode_t;
@@ -324,7 +323,6 @@ typedef ${pid_t-int /* default */} ${macro_prefix}pid_t;
 #define __PTRDIFF_TYPE__ ${ptrdiff_t-long int /* default */}
 #endif
 typedef __PTRDIFF_TYPE__ ${macro_prefix}ptrdiff_t;
-typedef ${sigset_t-int /* default */} ${macro_prefix}sigset_t;
 #ifndef __SIZE_TYPE__
 #define __SIZE_TYPE__ ${size_t-unsigned long /* default */}
 #endif
diff --git a/libio/libio.h b/libio/libio.h
index 84a8217..b2b6367 100644
--- a/libio/libio.h
+++ b/libio/libio.h
@@ -28,8 +28,9 @@ the executable file might be covered by the GNU General Public License. */
 #define _IO_STDIO_H
 
 #include <_G_config.h>
-#define _IO_pos_t _G_fpos_t /* obsolete */
-#define _IO_fpos_t _G_fpos_t
+#define _G_fpos_t _G_off_t
+#define _IO_pos_t _G_off_t
+#define _IO_fpos_t _G_off_t
 #define _IO_size_t _G_size_t
 #define _IO_ssize_t _G_ssize_t
 #define _IO_off_t _G_off_t
diff --git a/libstdc++/Makefile.in b/libstdc++/Makefile.in
index a4bdc4f..f94f8a7 100644
--- a/libstdc++/Makefile.in
+++ b/libstdc++/Makefile.in
@@ -90,7 +90,7 @@ $(ARLIB): stdlist stl/stl.list
 	$(RANLIB) $(ARLIB)
 
 $(SHLIB): piclist
-	$(CXX) $(SHFLAGS) -shared -o $(SHLIB) `cat piclist` $(SHDEPS)
+	$(CC) $(SHFLAGS) -shared -o $(SHLIB) `cat piclist` $(SHDEPS)
 
 $(SHARLIB): $(SHLIB)
 	-rm -f t$(SHARLIB)
@@ -276,31 +276,31 @@ install:
 	if [ -z "$(MULTISUBDIR)" ]; then \
 	  cd $(srcdir); \
 	  for FILE in $(HEADERS) *.h std/*.*; do \
-	    rm -f $(gxx_includedir)/$$FILE ; \
-	    $(INSTALL_DATA) $$FILE $(gxx_includedir)/$$FILE ; \
-	    chmod a-x $(gxx_includedir)/$$FILE ; \
+	    rm -f $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+	    $(INSTALL_DATA) $$FILE $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+	    chmod a-x $(DESTDIR)$(gxx_includedir)/$$FILE ; \
 	  done ; \
 	  cd stl; \
 	  for FILE in *.h; do \
-	    rm -f $(gxx_includedir)/$$FILE ; \
-	    $(INSTALL_DATA) $$FILE $(gxx_includedir)/$$FILE ; \
-	    chmod a-x $(gxx_includedir)/$$FILE ; \
+	    rm -f $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+	    $(INSTALL_DATA) $$FILE $(DESTDIR)$(gxx_includedir)/$$FILE ; \
+	    chmod a-x $(DESTDIR)$(gxx_includedir)/$$FILE ; \
 	  done ; \
 	else true ; \
 	fi
 	rootme=`pwd`/ ; export rootme ; \
 	for FILE in $(LIBS) ; do \
-	  rm -f $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	  rm -f $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  if [ $$FILE = $(SHLINK) ] || [ $$FILE = $(MSHLINK) ]; then \
-	    ln -s -f $(SHLIB) $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    ln -s -f $(SHLIB) $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  elif [ $$FILE = $(SHLIB) ]; then \
-	    $(INSTALL_PROGRAM) $$FILE $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    $(INSTALL_PROGRAM) $$FILE $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	    : On the HP, shared libraries must be mode 555. ;\
-	    chmod 555 $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    chmod 555 $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  else \
-	    $(INSTALL_DATA) $$FILE $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
-	    $(RANLIB) $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
-	    chmod a-x $(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    $(INSTALL_DATA) $$FILE $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    $(RANLIB) $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
+	    chmod a-x $(DESTDIR)$(INSTALLDIR)$(MULTISUBDIR)/$$FILE ; \
 	  fi ; \
 	done
 	@rootme=`pwd`/ ; export rootme ; \
