diff --git a/libio/libio.h b/libio/libio.h
index b2b6367..84d9812 100644
--- a/libio/libio.h
+++ b/libio/libio.h
@@ -219,9 +219,11 @@ extern int __underflow __P((_IO_FILE*));
 extern int __uflow __P((_IO_FILE*));
 extern int __overflow __P((_IO_FILE*, int));
 
-#define _IO_getc(_fp) \
+#define _IO_getc_unlocked(_fp) \
        ((_fp)->_IO_read_ptr >= (_fp)->_IO_read_end ? __uflow(_fp) \
 	: *(unsigned char*)(_fp)->_IO_read_ptr++)
+extern int _IO_getc (_IO_FILE *__fp);
+
 #define _IO_peekc(_fp) \
        ((_fp)->_IO_read_ptr >= (_fp)->_IO_read_end \
 	  && __underflow(_fp) == EOF ? EOF \
