diff --git a/bfd/elflink.c b/bfd/elflink.c
index b41818ec..ec6fbd23 100644
--- a/bfd/elflink.c
+++ b/bfd/elflink.c
@@ -12549,9 +12549,6 @@ bfd_elf_final_link (bfd *abfd, struct bfd_link_info *info)
 		  if (info->error_textrel)
 		    info->callbacks->einfo
 		      (_("%P%X: read-only segment has dynamic relocations.\n"));
-		  else
-		    info->callbacks->einfo
-		      (_("%P: warning: creating a DT_TEXTREL in object.\n"));
 		  break;
 		}
 	    }
