diff --git a/gcc/testsuite/gcc.dg/graphite/id-pr48805.c b/gcc/testsuite/gcc.dg/graphite/id-pr48805.c
new file mode 100644
index 000000000..f13a828e1
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/graphite/id-pr48805.c
@@ -0,0 +1,20 @@
+void *A(const void *a);
+void C(void);
+
+static void B(unsigned short a[])
+{
+  long i, j, tmp;
+
+  for(i = 0; i < 3; i++) {
+    for(j = 1; j <= 3; j++)
+      tmp -= a[i - j];
+
+    a[i] = (unsigned short)(tmp >> 1);
+  }
+}
+void C(void)
+{
+  unsigned short a[3], b[3];
+  B(b + 1);
+  A(a);
+}
diff --git a/gcc/tree-scalar-evolution.c b/gcc/tree-scalar-evolution.c
index 1b68b36b3..591967cfe 100644
--- a/gcc/tree-scalar-evolution.c
+++ b/gcc/tree-scalar-evolution.c
@@ -2645,6 +2645,7 @@ instantiate_scev_r (basic_block instantiate_below,
 				   TREE_OPERAND (chrec, 0),
 				   fold_conversions, cache, size_expr);
 
+    case ADDR_EXPR:
     case SCEV_NOT_KNOWN:
       return chrec_dont_know;
 
