diff --git a/gcc/gimple-fold.c b/gcc/gimple-fold.c
index ad45dcb1a..736fe8da4 100644
--- a/gcc/gimple-fold.c
+++ b/gcc/gimple-fold.c
@@ -948,7 +948,7 @@ gimplify_and_update_call_from_tree (gimple_stmt_iterator *si_p, tree expr)
 	      unlink_stmt_vdef (stmt);
 	      release_defs (stmt);
 	    }
-	  gsi_remove (si_p, true);
+	  gsi_replace (si_p, gimple_build_nop (), true);
 	  return;
 	}
     }
diff --git a/gcc/testsuite/g++.dg/opt/pr55331.C b/gcc/testsuite/g++.dg/opt/pr55331.C
new file mode 100644
index 000000000..4717db809
--- /dev/null
+++ b/gcc/testsuite/g++.dg/opt/pr55331.C
@@ -0,0 +1,14 @@
+// PR tree-optimization/55331
+// { dg-do compile }
+// { dg-options "-O2 -fno-tree-fre" }
+
+struct A {};
+
+void
+foo (A *p, bool x)
+{
+  A a;
+  char *e = (char *) (&a + 1);
+  if (x)
+    __builtin_memmove (p, &a, e - (char *) &a);
+}
