diff --git a/gcc/gcse.c b/gcc/gcse.c
index 9656db1d6..c3683395d 100644
--- a/gcc/gcse.c
+++ b/gcc/gcse.c
@@ -4523,9 +4523,7 @@ hoist_code (void)
 		    {
 		      /* An occurence might've been already deleted
 			 while processing a dominator of BB.  */
-		      if (occr->deleted_p)
-			gcc_assert (MAX_HOIST_DEPTH > 1);
-		      else
+		      if (!occr->deleted_p)
 			{
 			  gcc_assert (NONDEBUG_INSN_P (occr->insn));
 			  hoistable++;
@@ -4557,10 +4555,7 @@ hoist_code (void)
 		  /* An occurence might've been already deleted
 		     while processing a dominator of BB.  */
 		  if (occr->deleted_p)
-		    {
-		      gcc_assert (MAX_HOIST_DEPTH > 1);
-		      continue;
-		    }
+		    continue;
 		  gcc_assert (NONDEBUG_INSN_P (occr->insn));
 
 		  max_distance = expr->max_distance;
