diff --git a/gcc/testsuite/gcc.dg/pr49899.c b/gcc/testsuite/gcc.dg/pr49899.c
new file mode 100644
index 000000000..b5b32591a
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/pr49899.c
@@ -0,0 +1,3 @@
+static int foo (void) { return 0; } /* { dg-error "weak declaration of 'foo' being applied to a already existing, static definition" } */
+int foo (void)  __attribute__((weak));
+
diff --git a/gcc/varasm.c b/gcc/varasm.c
index a020fd068..c78e11a41 100644
--- a/gcc/varasm.c
+++ b/gcc/varasm.c
@@ -5223,6 +5223,11 @@ merge_weak (tree newdecl, tree olddecl)
       gcc_assert (!TREE_USED (olddecl)
 	          || !TREE_SYMBOL_REFERENCED (DECL_ASSEMBLER_NAME (olddecl)));
 
+      /* PR 49899: You cannot convert a static function into a weak, public function.  */
+      if (! TREE_PUBLIC (olddecl) && TREE_PUBLIC (newdecl))
+	error ("weak declaration of %q+D being applied to a already "
+	       "existing, static definition", newdecl);
+
       if (TARGET_SUPPORTS_WEAK)
 	{
 	  /* We put the NEWDECL on the weak_decls list at some point.
