diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 3fea5f8ab..287ca1a96 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -7080,6 +7080,9 @@ lookup_template_class (tree d1,
 			      tsubst (ENUM_UNDERLYING_TYPE (template_type),
 				      arglist, complain, in_decl),
 			      SCOPED_ENUM_P (template_type), NULL);
+
+	      if (t == error_mark_node)
+		return t;
 	    }
 	  else
             {
diff --git a/gcc/testsuite/g++.dg/template/crash117.C b/gcc/testsuite/g++.dg/template/crash117.C
new file mode 100644
index 000000000..5aaaf7344
--- /dev/null
+++ b/gcc/testsuite/g++.dg/template/crash117.C
@@ -0,0 +1,9 @@
+// PR c++/58649
+
+template<typename> void foo()
+{
+  E();  // { dg-error "declaration|declared" }    
+  enum E {};
+}
+
+template void foo<int>();
