diff --git a/gcc/fortran/decl.c b/gcc/fortran/decl.c
index 10a48bbd0..67c888c26 100644
--- a/gcc/fortran/decl.c
+++ b/gcc/fortran/decl.c
@@ -5590,7 +5590,8 @@ match_procedure_interface (gfc_symbol **proc_if)
       /* Resolve interface if possible. That way, attr.procedure is only set
 	 if it is declared by a later procedure-declaration-stmt, which is
 	 invalid per F08:C1216 (cf. resolve_procedure_interface).  */
-      while ((*proc_if)->ts.interface)
+      while ((*proc_if)->ts.interface
+	     && *proc_if != (*proc_if)->ts.interface)
 	*proc_if = (*proc_if)->ts.interface;
 
       if ((*proc_if)->attr.flavor == FL_UNKNOWN
diff --git a/gcc/testsuite/gfortran.dg/pr66575.f90 b/gcc/testsuite/gfortran.dg/pr66575.f90
new file mode 100644
index 000000000..7a0a604ed
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/pr66575.f90
@@ -0,0 +1,6 @@
+! { dg-do compile }
+! Bug 66575 - Endless compilation on missing end interface 
+program p
+   procedure(g) :: g ! { dg-error "may not be used as its own interface" }
+   procedure(g) ! { dg-error "Syntax error in PROCEDURE statement" }
+end
