diff --git a/gcc/cp/call.c b/gcc/cp/call.c
index c2298106c..93e7c3cff 100644
--- a/gcc/cp/call.c
+++ b/gcc/cp/call.c
@@ -5817,7 +5817,7 @@ build_new_op_1 (location_t loc, enum tree_code code, int flags, tree arg1,
     case TRUTH_ORIF_EXPR:
     case TRUTH_AND_EXPR:
     case TRUTH_OR_EXPR:
-      if (complain & tf_warning)
+      if ((complain & tf_warning) && !processing_template_decl)
 	warn_logical_operator (loc, code, boolean_type_node,
 			       code_orig_arg1, arg1,
 			       code_orig_arg2, arg2);
diff --git a/gcc/testsuite/g++.dg/warn/Wlogical-op-4.C b/gcc/testsuite/g++.dg/warn/Wlogical-op-4.C
new file mode 100644
index 000000000..79169954e
--- /dev/null
+++ b/gcc/testsuite/g++.dg/warn/Wlogical-op-4.C
@@ -0,0 +1,22 @@
+// PR c++/107755
+// { dg-do compile }
+// { dg-options "-Wlogical-op" }
+
+struct Foo
+{
+  operator bool() const { return false; }
+};
+
+bool a;
+Foo b;
+
+template <typename ignored>
+static bool Bar()
+{
+  return (!false && (false ? a : b));
+}
+
+bool Baz()
+{
+  return Bar<void>();
+}
