diff --git a/gcc/expr.c b/gcc/expr.c
index fb306ae6a..60fd8d6e0 100644
--- a/gcc/expr.c
+++ b/gcc/expr.c
@@ -4993,9 +4993,14 @@ expand_assignment (tree to, tree from, bool nontemporal)
 		}
 	      else
 		{
-		  rtx from_rtx
-		    = simplify_gen_subreg (GET_MODE (to_rtx), result,
-					   TYPE_MODE (TREE_TYPE (from)), 0);
+		  rtx from_rtx;
+		  if (MEM_P (result))
+		    from_rtx = change_address (result, GET_MODE (to_rtx),
+					       NULL_RTX);
+		  else
+		    from_rtx
+		      = simplify_gen_subreg (GET_MODE (to_rtx), result,
+					     TYPE_MODE (TREE_TYPE (from)), 0);
 		  emit_move_insn (XEXP (to_rtx, 0),
 				  read_complex_part (from_rtx, false));
 		  emit_move_insn (XEXP (to_rtx, 1),
diff --git a/gcc/testsuite/gcc.c-torture/compile/pr89412.c b/gcc/testsuite/gcc.c-torture/compile/pr89412.c
new file mode 100644
index 000000000..0d6ddd5db
--- /dev/null
+++ b/gcc/testsuite/gcc.c-torture/compile/pr89412.c
@@ -0,0 +1,16 @@
+/* PR middle-end/89412 */
+
+struct S { double a, b; } d;
+int e;
+double f;
+
+void
+foo ()
+{
+  _Complex double h;
+  while (e)
+    {
+      f = h;
+      *(struct S *) &h = d;
+    }
+}
