diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index d09e05e2a..1a5db9e4c 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -10043,10 +10043,10 @@ grokdeclarator (const cp_declarator *declarator,
 
   if (declspecs->std_attributes)
     {
-      /* Apply the c++11 attributes to the type preceding them.  */
-      input_location = declspecs->locations[ds_std_attribute];
-      decl_attributes (&type, declspecs->std_attributes, 0);
-      input_location = saved_loc;
+      location_t attr_loc = declspecs->locations[ds_std_attribute];
+      if (warning_at (attr_loc, OPT_Wattributes, "attribute ignored"))
+	inform (attr_loc, "an attribute that appertains to a type-specifier "
+		"is ignored");
     }
 
   /* Determine the type of the entity declared by recurring on the
diff --git a/gcc/testsuite/g++.dg/cpp0x/alignas16.C b/gcc/testsuite/g++.dg/cpp0x/alignas16.C
new file mode 100644
index 000000000..7c3499297
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/alignas16.C
@@ -0,0 +1,9 @@
+// PR c++/85062
+// { dg-do compile { target c++11 } }
+
+template<typename... T> struct A
+{
+  int alignas(T...) i;		// { dg-warning "ignored" }
+};
+
+A<int> a;
