diff --git a/gcc/lra-constraints.c b/gcc/lra-constraints.c
index c5b43aa11..8cf8680c8 100644
--- a/gcc/lra-constraints.c
+++ b/gcc/lra-constraints.c
@@ -1929,7 +1929,6 @@ process_alt_operands (int only_alternative)
 		    len = 0;
 		    lra_assert (nop > m);
 
-		    this_alternative_matches = m;
 		    m_hregno = get_hard_regno (*curr_id->operand_loc[m]);
 		    /* We are supposed to match a previous operand.
 		       If we do, we win if that one did.  If we do
@@ -2031,6 +2030,7 @@ process_alt_operands (int only_alternative)
 		    else
 		      did_match = true;
 
+		    this_alternative_matches = m;
 		    /* This can be fixed with reloads if the operand
 		       we are supposed to match can be fixed with
 		       reloads. */
diff --git a/gcc/testsuite/gcc.target/i386/pr84985.c b/gcc/testsuite/gcc.target/i386/pr84985.c
new file mode 100644
index 000000000..e987e4113
--- /dev/null
+++ b/gcc/testsuite/gcc.target/i386/pr84985.c
@@ -0,0 +1,6 @@
+/* { dg-do compile } */
+/* { dg-options "-O0" } */
+int main() {
+  int a;
+  asm("" : "=d"(a) : "0"(a), "0ae"(&a));
+}
