diff --git a/gcc/c-family/c.opt b/gcc/c-family/c.opt
index 75e4aba18..27dd7f9c6 100644
--- a/gcc/c-family/c.opt
+++ b/gcc/c-family/c.opt
@@ -1475,7 +1475,7 @@ fsquangle
 C++ ObjC++ Ignore Warn(switch %qs is no longer supported)
 
 fsso-struct=
-C ObjC Joined Enum(sso_struct) Var(default_sso) Init(SSO_NATIVE)
+C ObjC Joined RejectNegative Enum(sso_struct) Var(default_sso) Init(SSO_NATIVE)
 -fsso-struct=[big-endian|little-endian]	Set the default scalar storage order.
 
 Enum
diff --git a/gcc/testsuite/gcc.dg/pr78957.c b/gcc/testsuite/gcc.dg/pr78957.c
new file mode 100644
index 000000000..305023e18
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/pr78957.c
@@ -0,0 +1,6 @@
+/* PR driver/78957 */
+/* { dg-do compile } */
+/* { dg-options "-fno-sso-struct=none" } */
+/* { dg-error "unrecognized command line option" "" { target *-*-* } 0 } */
+
+int i;
