diff --git a/gcc/gimplify.c b/gcc/gimplify.c
index 21e335f92..107a3e145 100644
--- a/gcc/gimplify.c
+++ b/gcc/gimplify.c
@@ -6415,7 +6415,7 @@ omp_notice_variable (struct gimplify_omp_ctx *ctx, tree decl, bool in_code)
 	  gcc_assert (DECL_P (t));
 	  n2 = splay_tree_lookup (ctx->variables, (splay_tree_key) t);
 	  if (n2)
-	    n2->value |= GOVD_SEEN;
+	    omp_notice_variable (ctx, t, true);
 	}
     }
 
diff --git a/gcc/testsuite/gfortran.dg/gomp/sharing-4.f90 b/gcc/testsuite/gfortran.dg/gomp/sharing-4.f90
new file mode 100644
index 000000000..05640df30
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/gomp/sharing-4.f90
@@ -0,0 +1,24 @@
+! { dg-do compile }
+! { dg-options "-fopenmp" }
+
+subroutine foo (v, n, r)
+  integer :: n
+  integer, intent(in) :: v(:)
+  integer, intent(out) :: r
+  integer :: i
+
+  r = 0
+
+!$omp parallel
+!$omp single
+
+  do i = 1, n
+!$omp task shared (v)
+    r = r + v(i)
+!$omp end task
+  enddo
+
+!$omp end single
+!$omp end parallel
+
+end
diff --git a/pr79342.dwo b/pr79342.dwo
new file mode 100644
index 000000000..1570bd013
Binary files /dev/null and b/pr79342.dwo differ
