diff --git a/gcc/cp/constraint.cc b/gcc/cp/constraint.cc
index 09ae301f0..4a47369bf 100644
--- a/gcc/cp/constraint.cc
+++ b/gcc/cp/constraint.cc
@@ -1930,7 +1930,7 @@ tsubst_constraint_info (tree t, tree args,
 tree
 tsubst_constraint (tree t, tree args, tsubst_flags_t complain, tree in_decl)
 {
-  if (t == NULL_TREE)
+  if (t == NULL_TREE || t == error_mark_node)
     return t;
   switch (TREE_CODE (t))
   {
diff --git a/gcc/testsuite/g++.dg/concepts/pr84330.C b/gcc/testsuite/g++.dg/concepts/pr84330.C
new file mode 100644
index 000000000..975ee8941
--- /dev/null
+++ b/gcc/testsuite/g++.dg/concepts/pr84330.C
@@ -0,0 +1,12 @@
+// PR c++/84330
+// { dg-options "-fconcepts" }
+
+struct A
+{
+  template<typename T> requires sizeof(T) >> 0 void foo(T);  // { dg-error "predicate constraint" }
+
+  void bar()
+  {
+    foo(0);  // { dg-error "no matching" }
+  }
+};
