diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 976d90480..40b2ebb57 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -5296,7 +5296,9 @@ maybe_deduce_size_from_array_init (tree decl, tree init)
 	      failure = 1;
 	}
 
-      if (!failure)
+      if (failure)
+	TREE_TYPE (decl) = error_mark_node;
+      else
 	{
 	  failure = cp_complete_array_type (&TREE_TYPE (decl), initializer,
 					    do_default);
diff --git a/gcc/testsuite/g++.dg/ext/desig10.C b/gcc/testsuite/g++.dg/ext/desig10.C
new file mode 100644
index 000000000..24057dd0a
--- /dev/null
+++ b/gcc/testsuite/g++.dg/ext/desig10.C
@@ -0,0 +1,4 @@
+// PR c++/84972
+// { dg-additional-options "-w" }
+
+char(a[])({.a = 0});  // { dg-error "designated initializer" }
