diff --git a/gcc/cp/decl.c b/gcc/cp/decl.c
index 410eae58b..860d9ff7e 100644
--- a/gcc/cp/decl.c
+++ b/gcc/cp/decl.c
@@ -5985,8 +5985,7 @@ reshape_init (tree type, tree init, tsubst_flags_t complain)
     {
       if (complain & tf_error)
 	error ("too many initializers for %qT", type);
-      else
-	return error_mark_node;
+      return error_mark_node;
     }
 
   return new_init;
diff --git a/gcc/testsuite/g++.dg/cpp0x/pr71577.C b/gcc/testsuite/g++.dg/cpp0x/pr71577.C
new file mode 100644
index 000000000..15794ad62
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/pr71577.C
@@ -0,0 +1,4 @@
+// PR c++/71577
+// { dg-do compile { target c++11 } }
+
+struct { int a; } s1, s2 = { s1, 0 };  // { dg-error "too many initializers" }
