diff --git a/gcc/cp/constexpr.c b/gcc/cp/constexpr.c
index dcf7e301c..bf6508ce9 100644
--- a/gcc/cp/constexpr.c
+++ b/gcc/cp/constexpr.c
@@ -3567,8 +3567,9 @@ cxx_eval_statement_list (const constexpr_ctx *ctx, tree t,
 {
   tree_stmt_iterator i;
   tree local_target;
-  /* In a statement-expression we want to return the last value.  */
-  tree r = NULL_TREE;
+  /* In a statement-expression we want to return the last value.
+     For empty statement expression return void_node.  */
+  tree r = void_node;
   if (!jump_target)
     {
       local_target = NULL_TREE;
diff --git a/gcc/testsuite/g++.dg/cpp0x/constexpr-79822.C b/gcc/testsuite/g++.dg/cpp0x/constexpr-79822.C
new file mode 100644
index 000000000..6a72702aa
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/constexpr-79822.C
@@ -0,0 +1,12 @@
+// PR c++/79822
+// { dg-do compile }
+// { dg-options "" }
+
+bool
+foo ()
+{
+  bool a = ({ }) && false;		// { dg-error "could not convert" }
+  bool b = ({ ; }) && false;		// { dg-error "could not convert" }
+  bool c = ({ (void) 1; }) && false;	// { dg-error "could not convert" }
+  return a && b && c;
+}
