diff --git a/gcc/testsuite/gcc.target/aarch64/aapcs64/aapcs64.exp b/gcc/testsuite/gcc.target/aarch64/aapcs64/aapcs64.exp
index ad92994fe..9215c7c4f 100644
--- a/gcc/testsuite/gcc.target/aarch64/aapcs64/aapcs64.exp
+++ b/gcc/testsuite/gcc.target/aarch64/aapcs64/aapcs64.exp
@@ -29,37 +29,6 @@ torture-init
 set-torture-options $C_TORTURE_OPTIONS
 set additional_flags "-W -Wall -Wno-abi"
 
-# Test parameter passing.
-foreach src [lsort [glob -nocomplain $srcdir/$subdir/test_*.c]] {
-    if {[runtest_file_p $runtests $src]} {
-	    c-torture-execute [list $src \
-				    $srcdir/$subdir/abitest.S] \
-				    $additional_flags
-    }
-}
-
-# Test unnamed argument retrieval via the va_arg macro.
-foreach src [lsort [glob -nocomplain $srcdir/$subdir/va_arg-*.c]] {
-    if {[runtest_file_p $runtests $src]} {
-	    c-torture-execute [list $src \
-				    $srcdir/$subdir/abitest.S] \
-				    $additional_flags
-    }
-}
-
-# Test function return value.
-#   Disable -fipa-ra to prevent the compiler from generating
-#   conflicting code.
-set additional_flags_for_func_ret $additional_flags
-append additional_flags_for_func_ret " -fno-ipa-ra"
-foreach src [lsort [glob -nocomplain $srcdir/$subdir/func-ret-*.c]] {
-    if {[runtest_file_p $runtests $src]} {
-	    c-torture-execute [list $src \
-				    $srcdir/$subdir/abitest.S] \
-				    $additional_flags_for_func_ret
-    }
-}
-
 # Test no internal compiler errors.
 foreach src [lsort [glob -nocomplain $srcdir/$subdir/ice_*.c]] {
     if {[runtest_file_p $runtests $src]} {
diff --git a/gcc/testsuite/gcc.target/aarch64/pr70398.c b/gcc/testsuite/gcc.target/aarch64/pr70398.c
index dbe5ad1b2..bb7122c0f 100644
--- a/gcc/testsuite/gcc.target/aarch64/pr70398.c
+++ b/gcc/testsuite/gcc.target/aarch64/pr70398.c
@@ -1,5 +1,5 @@
 /* { dg-do run } */
-/* { dg-options "-O -fno-tree-loop-optimize -fno-tree-ter -static" } */
+/* { dg-options "-O -fno-tree-loop-optimize -fno-tree-ter -static-libgcc" } */
 unsigned int in[8 * 8] =
   { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
     45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63 };
diff --git a/gcc/testsuite/gcc.target/aarch64/symbol-range-tiny.c b/gcc/testsuite/gcc.target/aarch64/symbol-range-tiny.c
deleted file mode 100644
index d7e46b059..000000000
--- a/gcc/testsuite/gcc.target/aarch64/symbol-range-tiny.c
+++ /dev/null
@@ -1,12 +0,0 @@
-/* { dg-do compile } */
-/* { dg-options "-O3 -save-temps -mcmodel=tiny" } */
-
-int fixed_regs[0x00200000];
-
-int
-foo()
-{
-  return fixed_regs[0x00080000];
-}
-
-/* { dg-final { scan-assembler-not "adr\tx\[0-9\]+, fixed_regs\\\+" } } */
diff --git a/gcc/testsuite/gcc.target/aarch64/symbol-range.c b/gcc/testsuite/gcc.target/aarch64/symbol-range.c
deleted file mode 100644
index 6574cf431..000000000
--- a/gcc/testsuite/gcc.target/aarch64/symbol-range.c
+++ /dev/null
@@ -1,12 +0,0 @@
-/* { dg-do compile } */
-/* { dg-options "-O3 -save-temps -mcmodel=small" } */
-
-int fixed_regs[0x200000000ULL];
-
-int
-foo()
-{
-  return fixed_regs[0x100000000ULL];
-}
-
-/* { dg-final { scan-assembler-not "adrp\tx\[0-9\]+, fixed_regs\\\+" } } */
