diff --git a/gcc/cp/pt.c b/gcc/cp/pt.c
index 60967299b..0cf0a3ef4 100644
--- a/gcc/cp/pt.c
+++ b/gcc/cp/pt.c
@@ -23960,7 +23960,7 @@ listify (tree arg)
 static tree
 listify_autos (tree type, tree auto_node)
 {
-  tree init_auto = listify (auto_node);
+  tree init_auto = listify (strip_top_quals (auto_node));
   tree argvec = make_tree_vec (1);
   TREE_VEC_ELT (argvec, 0) = init_auto;
   if (processing_template_decl)
diff --git a/gcc/testsuite/g++.dg/cpp0x/initlist-deduce2.C b/gcc/testsuite/g++.dg/cpp0x/initlist-deduce2.C
new file mode 100644
index 000000000..dbf68a40c
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/initlist-deduce2.C
@@ -0,0 +1,8 @@
+// Test for PR63149
+// { dg-do compile { target c++11 } }
+
+#include <initializer_list>
+
+const auto r = { 1, 2, 3 };
+using X = decltype(r);
+using X = const std::initializer_list<int>;
